{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation \n",
    "\n",
    "# Team Hafler. ASAP CRN Metadata validation\n",
    "\n",
    "28 Nov 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully imported\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> SANITY CHECK: verify reading from google doc works.\n",
    "\n",
    "```python\n",
    "CDEv1_ = read_CDE(\"v1\")\n",
    "CDEv2_ = read_CDE(\"v2.1\")\n",
    "CDEv3_ = read_CDE(\"v3.0\")\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean V1 Table\n",
    "write clean metadata tables according to CDE v1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hafler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/data/teams/hafler/sc_pmdbs/metadata')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## convert to seurat Object\n",
    "team = \"hafler\"\n",
    "collection = \"sc_pmdbs\"\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{collection}/metadata\"\n",
    "\n",
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\"]\n",
    "excel_path = metadata_path / \"og/ASAP_CDE_ALL_Team_Hafler_v1.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\",header=1).drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\",header=1).drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\",header=1).drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\",header=1).drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\",header=1).drop(columns=\"Field\")\n",
    "metadata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_version = \"v1\"\n",
    "\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_hafler_batch(sample_df):\n",
    "\n",
    "    # First batch: HSDG07HC HSDG10HC HSDG148PD HSDG199PD\n",
    "    # batch[batch.sample_id in ['hSDG07HC', 'hSDG10HC', 'hSDG148PD', 'hSDG199PD']]=1\n",
    "    Batch_1 = ['hSDG07', 'hSDG10', 'hSDG148', 'hSDG199'] \n",
    "    # Second batch: hsDG101HC hsDG13HC hsDG151PD hsDG197PD hsDG30HC hsDG99HC\n",
    "    Batch_2 = ['hSDG101', 'hSDG13', 'hSDG151', 'hSDG197', 'hSDG30', 'hSDG99']\n",
    "    # Third batch: hsDG142PD hsDG208PD\n",
    "    Batch_3 = ['hSDG142', 'hSDG208'] \n",
    "\n",
    "\n",
    "    batch_col = []\n",
    "    for row in sample_df.sample_id:\n",
    "        if row in Batch_1:\n",
    "            batch_col.append(\"Batch_1\")\n",
    "        elif row in Batch_2:\n",
    "            batch_col.append(\"Batch_2\")\n",
    "        elif row in Batch_3:\n",
    "            batch_col.append(\"Batch_3\")\n",
    "        else:\n",
    "            print(\"ERROR >>>>>>>> not no batch info\")\n",
    "            batch_col.append(\"\")\n",
    "\n",
    "\n",
    "    sample_df['batch'] = batch_col\n",
    "    return sample_df\n",
    "\n",
    "SAMPLE = add_hafler_batch(SAMPLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id source_sample_id subject_id replicate  replicate_count  \\\n",
       "0    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "1    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "2    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "3   hSDG101          hSDG101       HC03      Rep1                1   \n",
       "4   hSDG101          hSDG101       HC03      Rep1                1   \n",
       "\n",
       "   repeated_sample    batch tissue       brain_region  source_RIN  ...  \\\n",
       "0                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "1                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "2                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "3                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "4                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0                   NaN                                      NaN   \n",
       "1                   NaN                                      NaN   \n",
       "2                   NaN                                      NaN   \n",
       "3                   NaN                                      NaN   \n",
       "4                   NaN                                      NaN   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0                      NaN                     NaN                        NaN   \n",
       "1                      NaN                     NaN                        NaN   \n",
       "2                      NaN                     NaN                        NaN   \n",
       "3                      NaN                     NaN                        NaN   \n",
       "4                      NaN                     NaN                        NaN   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV2000 pm_PH  donor_id  \n",
       "0                    NaN             NaN    NaN   NaN       NaN  \n",
       "1                    NaN             NaN    NaN   NaN       NaN  \n",
       "2                    NaN             NaN    NaN   NaN       NaN  \n",
       "3                    NaN             NaN    NaN   NaN       NaN  \n",
       "4                    NaN             NaN    NaN   NaN       NaN  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix replicate & replicate_count\n",
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "SAMPLE['replicate'] = \"Rep1\"\n",
    "SAMPLE['replicate_count'] = 1\n",
    "\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>fastq</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id subject_id file_type  \\\n",
       "0     hSDG07           hSDG07       HC01     fastq   \n",
       "1     hSDG07           hSDG07       HC01     fastq   \n",
       "2     hSDG07           hSDG07       HC01     fastq   \n",
       "3    hSDG101          hSDG101       HC03     fastq   \n",
       "4    hSDG101          hSDG101       HC03     fastq   \n",
       "5    hSDG101          hSDG101       HC03     fastq   \n",
       "6     hSDG10           hSDG10       HC04     fastq   \n",
       "7     hSDG10           hSDG10       HC04     fastq   \n",
       "8     hSDG10           hSDG10       HC04     fastq   \n",
       "9     hSDG13           hSDG13       HC02     fastq   \n",
       "10    hSDG13           hSDG13       HC02     fastq   \n",
       "11    hSDG13           hSDG13       HC02     fastq   \n",
       "12   hSDG142          hSDG142       PD04     fastq   \n",
       "13   hSDG142          hSDG142       PD04     fastq   \n",
       "14   hSDG142          hSDG142       PD04     fastq   \n",
       "15   hSDG148          hSDG148       PD05     fastq   \n",
       "16   hSDG148          hSDG148       PD05     fastq   \n",
       "17   hSDG148          hSDG148       PD05     fastq   \n",
       "18   hSDG151          hSDG151       PD01     fastq   \n",
       "19   hSDG151          hSDG151       PD01     fastq   \n",
       "20   hSDG151          hSDG151       PD01     fastq   \n",
       "21   hSDG197          hSDG197       PD02     fastq   \n",
       "22   hSDG197          hSDG197       PD02     fastq   \n",
       "23   hSDG197          hSDG197       PD02     fastq   \n",
       "24   hSDG199          hSDG199       PD06     fastq   \n",
       "25   hSDG199          hSDG199       PD06     fastq   \n",
       "26   hSDG199          hSDG199       PD06     fastq   \n",
       "27   hSDG208          hSDG208       PD03     fastq   \n",
       "28   hSDG208          hSDG208       PD03     fastq   \n",
       "29   hSDG208          hSDG208       PD03     fastq   \n",
       "30    hSDG30           hSDG30       HC05     fastq   \n",
       "31    hSDG30           hSDG30       HC05     fastq   \n",
       "32    hSDG30           hSDG30       HC05     fastq   \n",
       "33    hSDG99           hSDG99       HC06     fastq   \n",
       "34    hSDG99           hSDG99       HC06     fastq   \n",
       "35    hSDG99           hSDG99       HC06     fastq   \n",
       "\n",
       "                                file_name  \n",
       "0    HSDG07HC_HHT_S6_L004_I1_001.fastq.gz  \n",
       "1    HSDG07HC_HHT_S6_L004_R1_001.fastq.gz  \n",
       "2    HSDG07HC_HHT_S6_L004_R2_001.fastq.gz  \n",
       "3   hsDG101HC_HHT_S2_L002_I1_001.fastq.gz  \n",
       "4   hsDG101HC_HHT_S2_L002_R1_001.fastq.gz  \n",
       "5   hsDG101HC_HHT_S2_L002_R2_001.fastq.gz  \n",
       "6    HSDG10HC_HHT_S7_L004_I1_001.fastq.gz  \n",
       "7    HSDG10HC_HHT_S7_L004_R1_001.fastq.gz  \n",
       "8    HSDG10HC_HHT_S7_L004_R2_001.fastq.gz  \n",
       "9    hsDG13HC_HHT_S1_L002_I1_001.fastq.gz  \n",
       "10   hsDG13HC_HHT_S1_L002_R1_001.fastq.gz  \n",
       "11   hsDG13HC_HHT_S1_L002_R2_001.fastq.gz  \n",
       "12  hsDG142PD_HHT_S1_L004_I1_001.fastq.gz  \n",
       "13  hsDG142PD_HHT_S1_L004_R1_001.fastq.gz  \n",
       "14  hsDG142PD_HHT_S1_L004_R2_001.fastq.gz  \n",
       "15  HSDG148PD_HHT_S8_L004_I1_001.fastq.gz  \n",
       "16  HSDG148PD_HHT_S8_L004_R1_001.fastq.gz  \n",
       "17  HSDG148PD_HHT_S8_L004_R2_001.fastq.gz  \n",
       "18  hsDG151PD_HHT_S3_L002_I1_001.fastq.gz  \n",
       "19  hsDG151PD_HHT_S3_L002_R1_001.fastq.gz  \n",
       "20  hsDG151PD_HHT_S3_L002_R2_001.fastq.gz  \n",
       "21  hsDG197PD_HHT_S4_L002_I1_001.fastq.gz  \n",
       "22  hsDG197PD_HHT_S4_L002_R1_001.fastq.gz  \n",
       "23  hsDG197PD_HHT_S4_L002_R2_001.fastq.gz  \n",
       "24  HSDG199PD_HHT_S9_L004_I1_001.fastq.gz  \n",
       "25  HSDG199PD_HHT_S9_L004_R1_001.fastq.gz  \n",
       "26  HSDG199PD_HHT_S9_L004_R2_001.fastq.gz  \n",
       "27  hsDG208PD_HHT_S2_L004_I1_001.fastq.gz  \n",
       "28  hsDG208PD_HHT_S2_L004_R1_001.fastq.gz  \n",
       "29  hsDG208PD_HHT_S2_L004_R2_001.fastq.gz  \n",
       "30   hsDG30HC_HHT_S5_L002_I1_001.fastq.gz  \n",
       "31   hsDG30HC_HHT_S5_L002_R1_001.fastq.gz  \n",
       "32   hsDG30HC_HHT_S5_L002_R2_001.fastq.gz  \n",
       "33   hsDG99HC_HHT_S6_L002_I1_001.fastq.gz  \n",
       "34   hsDG99HC_HHT_S6_L002_R1_001.fastq.gz  \n",
       "35   hsDG99HC_HHT_S6_L002_R2_001.fastq.gz  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE[['sample_id', 'source_sample_id', 'subject_id',\n",
    "       'file_type', 'file_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "STUDY = prep_table(STUDY, CDEv1)\n",
    "STUDY['metadata_version_date']\n",
    "STUDY.rename(columns={\"submittor_email\":\"submitter_email\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT = prep_table(SUBJECT, CDEv1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT['sex'] = SUBJECT['sex'].replace({'F':\"Female\", 'M':\"Male\"})\n",
    "SUBJECT['race'] = SUBJECT['race'].replace({'W':\"White\", 'B':\"Black or African American\"})\n",
    "\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace({'Normal control':\"Healthy Control\", \"Idiopathic Parkinson's disease\":\"Idiopathic PD\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# they \"race\" was reported under ethnicity, so code as \"Not Reported\"\n",
    "SUBJECT['ethnicity'] = 'Not Reported'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC06</td>\n",
       "      <td>hSDG99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PD01</td>\n",
       "      <td>hSDG151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PD02</td>\n",
       "      <td>hSDG197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PD03</td>\n",
       "      <td>hSDG208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PD04</td>\n",
       "      <td>hSDG142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PD05</td>\n",
       "      <td>hSDG148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PD06</td>\n",
       "      <td>hSDG199</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id\n",
       "0        HC01            hSDG07\n",
       "1        HC02            hSDG13\n",
       "2        HC03           hSDG101\n",
       "3        HC04            hSDG10\n",
       "4        HC05            hSDG30\n",
       "5        HC06            hSDG99\n",
       "6        PD01           hSDG151\n",
       "7        PD02           hSDG197\n",
       "8        PD03           hSDG208\n",
       "9        PD04           hSDG142\n",
       "10       PD05           hSDG148\n",
       "11       PD06           hSDG199"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT[['subject_id', 'source_subject_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = prep_table(SAMPLE,  CDEv1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# set time == 0 for all samples\n",
    "SAMPLE['time'] = 0\n",
    "\n",
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# need to join with subject to get \"sex\" and convert to ontology term\n",
    "SAMPLE_SUBJECT = SAMPLE.merge(SUBJECT, on='subject_id',  how='left')\n",
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['sex_ontology_term_id'] = SAMPLE_SUBJECT['sex'].replace({\"Male\":\"PATO:0000384 (male)\", \"Female\":\"PATO:0000383 (female)\" })\n",
    "\n",
    "# ignore development_stage_ontology_term_id, self_reported_ethnicity_ontology_term_id, assay_ontology_term_id, etc for now. (Check wiht Le)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix assay\n",
    "SAMPLE['assay'] = SAMPLE['assay'].replace({'v3.1 - Single Index, 10x Genomics ':\"v3.1 - Single Index\"})\n",
    "# fix assay\n",
    "SAMPLE['sequencing_length'] = SAMPLE['sequencing_length'].replace({'150bp x2':\"150\"})\n",
    "\n",
    "# fix DV200 \n",
    "SAMPLE.rename(columns={\"DV2000\":\"DV200\"}, inplace=True)\n",
    "#TODO: add file_description I1, R1, R2  \"Raw sequencing data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>file_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HSDG07HC_HHT_S6_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hsDG101HC_HHT_S2_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HSDG10HC_HHT_S7_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hsDG13HC_HHT_S1_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hsDG142PD_HHT_S1_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>HSDG148PD_HHT_S8_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hsDG151PD_HHT_S3_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hsDG197PD_HHT_S4_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>HSDG199PD_HHT_S9_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hsDG208PD_HHT_S2_L004_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hsDG30HC_HHT_S5_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hsDG99HC_HHT_S6_L002_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id subject_id source_sample_id  \\\n",
       "0     hSDG07       HC01           hSDG07   \n",
       "1     hSDG07       HC01           hSDG07   \n",
       "2     hSDG07       HC01           hSDG07   \n",
       "3    hSDG101       HC03          hSDG101   \n",
       "4    hSDG101       HC03          hSDG101   \n",
       "5    hSDG101       HC03          hSDG101   \n",
       "6     hSDG10       HC04           hSDG10   \n",
       "7     hSDG10       HC04           hSDG10   \n",
       "8     hSDG10       HC04           hSDG10   \n",
       "9     hSDG13       HC02           hSDG13   \n",
       "10    hSDG13       HC02           hSDG13   \n",
       "11    hSDG13       HC02           hSDG13   \n",
       "12   hSDG142       PD04          hSDG142   \n",
       "13   hSDG142       PD04          hSDG142   \n",
       "14   hSDG142       PD04          hSDG142   \n",
       "15   hSDG148       PD05          hSDG148   \n",
       "16   hSDG148       PD05          hSDG148   \n",
       "17   hSDG148       PD05          hSDG148   \n",
       "18   hSDG151       PD01          hSDG151   \n",
       "19   hSDG151       PD01          hSDG151   \n",
       "20   hSDG151       PD01          hSDG151   \n",
       "21   hSDG197       PD02          hSDG197   \n",
       "22   hSDG197       PD02          hSDG197   \n",
       "23   hSDG197       PD02          hSDG197   \n",
       "24   hSDG199       PD06          hSDG199   \n",
       "25   hSDG199       PD06          hSDG199   \n",
       "26   hSDG199       PD06          hSDG199   \n",
       "27   hSDG208       PD03          hSDG208   \n",
       "28   hSDG208       PD03          hSDG208   \n",
       "29   hSDG208       PD03          hSDG208   \n",
       "30    hSDG30       HC05           hSDG30   \n",
       "31    hSDG30       HC05           hSDG30   \n",
       "32    hSDG30       HC05           hSDG30   \n",
       "33    hSDG99       HC06           hSDG99   \n",
       "34    hSDG99       HC06           hSDG99   \n",
       "35    hSDG99       HC06           hSDG99   \n",
       "\n",
       "                                file_name  \n",
       "0    HSDG07HC_HHT_S6_L004_I1_001.fastq.gz  \n",
       "1    HSDG07HC_HHT_S6_L004_R1_001.fastq.gz  \n",
       "2    HSDG07HC_HHT_S6_L004_R2_001.fastq.gz  \n",
       "3   hsDG101HC_HHT_S2_L002_I1_001.fastq.gz  \n",
       "4   hsDG101HC_HHT_S2_L002_R1_001.fastq.gz  \n",
       "5   hsDG101HC_HHT_S2_L002_R2_001.fastq.gz  \n",
       "6    HSDG10HC_HHT_S7_L004_I1_001.fastq.gz  \n",
       "7    HSDG10HC_HHT_S7_L004_R1_001.fastq.gz  \n",
       "8    HSDG10HC_HHT_S7_L004_R2_001.fastq.gz  \n",
       "9    hsDG13HC_HHT_S1_L002_I1_001.fastq.gz  \n",
       "10   hsDG13HC_HHT_S1_L002_R1_001.fastq.gz  \n",
       "11   hsDG13HC_HHT_S1_L002_R2_001.fastq.gz  \n",
       "12  hsDG142PD_HHT_S1_L004_I1_001.fastq.gz  \n",
       "13  hsDG142PD_HHT_S1_L004_R1_001.fastq.gz  \n",
       "14  hsDG142PD_HHT_S1_L004_R2_001.fastq.gz  \n",
       "15  HSDG148PD_HHT_S8_L004_I1_001.fastq.gz  \n",
       "16  HSDG148PD_HHT_S8_L004_R1_001.fastq.gz  \n",
       "17  HSDG148PD_HHT_S8_L004_R2_001.fastq.gz  \n",
       "18  hsDG151PD_HHT_S3_L002_I1_001.fastq.gz  \n",
       "19  hsDG151PD_HHT_S3_L002_R1_001.fastq.gz  \n",
       "20  hsDG151PD_HHT_S3_L002_R2_001.fastq.gz  \n",
       "21  hsDG197PD_HHT_S4_L002_I1_001.fastq.gz  \n",
       "22  hsDG197PD_HHT_S4_L002_R1_001.fastq.gz  \n",
       "23  hsDG197PD_HHT_S4_L002_R2_001.fastq.gz  \n",
       "24  HSDG199PD_HHT_S9_L004_I1_001.fastq.gz  \n",
       "25  HSDG199PD_HHT_S9_L004_R1_001.fastq.gz  \n",
       "26  HSDG199PD_HHT_S9_L004_R2_001.fastq.gz  \n",
       "27  hsDG208PD_HHT_S2_L004_I1_001.fastq.gz  \n",
       "28  hsDG208PD_HHT_S2_L004_R1_001.fastq.gz  \n",
       "29  hsDG208PD_HHT_S2_L004_R2_001.fastq.gz  \n",
       "30   hsDG30HC_HHT_S5_L002_I1_001.fastq.gz  \n",
       "31   hsDG30HC_HHT_S5_L002_R1_001.fastq.gz  \n",
       "32   hsDG30HC_HHT_S5_L002_R2_001.fastq.gz  \n",
       "33   hsDG99HC_HHT_S6_L002_I1_001.fastq.gz  \n",
       "34   hsDG99HC_HHT_S6_L002_R1_001.fastq.gz  \n",
       "35   hsDG99HC_HHT_S6_L002_R2_001.fastq.gz  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE[['sample_id','subject_id','source_sample_id','file_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HC01': 'hSDG07',\n",
       " 'HC03': 'hSDG101',\n",
       " 'HC04': 'hSDG10',\n",
       " 'HC02': 'hSDG13',\n",
       " 'PD04': 'hSDG142',\n",
       " 'PD05': 'hSDG148',\n",
       " 'PD01': 'hSDG151',\n",
       " 'PD02': 'hSDG197',\n",
       " 'PD06': 'hSDG199',\n",
       " 'PD03': 'hSDG208',\n",
       " 'HC05': 'hSDG30',\n",
       " 'HC06': 'hSDG99'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the CLINPATH has sample_id as a subject_id, so we need to fix that\n",
    "subj_samp_mapper = dict(zip( SAMPLE['subject_id'], SAMPLE['sample_id']))\n",
    "subj_samp_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH = prep_table(CLINPATH, CDEv1)\n",
    "CLINPATH.rename(columns={\"path_thai\":\"path_thal\"}, inplace=True)\n",
    "CLINPATH.replace(\"Nan\", \"\", inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# redact \"Prefrontal Cortex\" from region_level_2 for now\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace({'Prefrontal Cortex':\"\"})\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# but remove commas\n",
    "CLINPATH[\"APOE_e4_status\"] = CLINPATH[\"APOE_e4_status\"].str.replace(\",\",\"\")\n",
    "\n",
    "# need to fix the path_autopsy_dx_main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC06</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PD01</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PD02</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PD03</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PD04</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PD05</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PD06</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id time_from_baseline GP2_id AMPPD_id\n",
       "0       HC01           hSDG07                                   \n",
       "1       HC02           hSDG13                                   \n",
       "2       HC03          hSDG101                                   \n",
       "3       HC04           hSDG10                                   \n",
       "4       HC05           hSDG30                                   \n",
       "5       HC06           hSDG99                                   \n",
       "6       PD01          hSDG151                                   \n",
       "7       PD02          hSDG197                                   \n",
       "8       PD03          hSDG208                                   \n",
       "9       PD04          hSDG142                                   \n",
       "10      PD05          hSDG148                                   \n",
       "11      PD06          hSDG199                                   "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH[['sample_id', 'source_sample_id', 'time_from_baseline', 'GP2_id',\n",
    "       'AMPPD_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_96310/2503552355.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  CLINPATH['sample_id'].replace(subj_samp_mapper, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id time_from_baseline GP2_id AMPPD_id\n",
       "0     hSDG07           hSDG07                                   \n",
       "1     hSDG13           hSDG13                                   \n",
       "2    hSDG101          hSDG101                                   \n",
       "3     hSDG10           hSDG10                                   \n",
       "4     hSDG30           hSDG30                                   \n",
       "5     hSDG99           hSDG99                                   \n",
       "6    hSDG151          hSDG151                                   \n",
       "7    hSDG197          hSDG197                                   \n",
       "8    hSDG208          hSDG208                                   \n",
       "9    hSDG142          hSDG142                                   \n",
       "10   hSDG148          hSDG148                                   \n",
       "11   hSDG199          hSDG199                                   "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['sample_id'].replace(subj_samp_mapper, inplace=True)\n",
    "CLINPATH[['sample_id', 'source_sample_id', 'time_from_baseline', 'GP2_id',\n",
    "       'AMPPD_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLINPATH = CLINPATH.rename(columns={\"source_sample_id\":\"source_subject_id\"})\n",
    "# CLINPATH[\"subject_id\"] = CLINPATH[\"source_subject_id\"].str.replace(\"hSDG\",\"hSDG0\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dump these extractions from the excel file as og"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_path = metadata_path / \"og\"\n",
    "STUDY.to_csv(og_path / f\"STUDY.csv\", index=False)\n",
    "SAMPLE.to_csv(og_path / f\"SAMPLE.csv\", index=False)\n",
    "SUBJECT.to_csv(og_path / f\"SUBJECT.csv\", index=False)\n",
    "PROTOCOL.to_csv(og_path / f\"PROTOCOL.csv\", index=False)\n",
    "CLINPATH.to_csv(og_path / f\"CLINPATH.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# validate v1\n",
    "\n",
    "QC these tables and make sure they are conformant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **17 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_RIN: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- file_description: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- preprocessing_references: 36/36 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- tissue_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_ontology_term_id: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- suspension_type: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 36/36 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 36/36 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 36/36 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values 💩'single nuclei, 8000'\n",
      "    - valid ➡️ int or NULL ('NA')\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in CLINPATH: dig_slide_avail, quant_path_avail**\n",
      "🚨⚠️❗ **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time_from_baseline: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- AMPPD_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 1/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_dx_main: 6/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **23 Fields with invalid entries:**\n",
      "- _*region_level_2*_:  invalid values 💩'Prefrontal cortex'\n",
      "    - valid ➡️ 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown'\n",
      "- _*hx_dementia_mci*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*hx_melanoma*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*education_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other'\n",
      "- _*smoking_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Current smoker', 'Former smoker', 'Never', 'Unknown'\n",
      "- _*APOE_e4_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '22', '23', '24', '33', '34', '44', 'Unknown'\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder'\n",
      "- _*path_braak_nft*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', 'I', 'II', 'III', 'IV', 'V', 'VI', 'I/II', 'III/IV', 'IV/V', 'V/VI'\n",
      "- _*path_braak_asyn*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_cerad*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Sparse', 'Moderate', 'Frequent'\n",
      "- _*path_thal*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*known_pathogenic_mutation*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Present', 'Unknown'\n",
      "- _*path_mckeith*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_infarcs*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*path_nia_ri*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*path_nia_aa_a*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3'\n",
      "- _*TDP43*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*arteriolosclerosis_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv1\n",
    "tables = CDE['Table'].unique()\n",
    "v1_path = metadata_path / \"v1\"\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(og_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table\n",
    "    df.to_csv(v1_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v1_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v1_path / f\"{table}.csv\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n"
     ]
    }
   ],
   "source": [
    "v2_path = metadata_path / \"v2\"\n",
    "\n",
    "v2_tables, aux_tables = v1_to_v2(v1_path, v2_path, CDEv1, CDEv2)\n",
    "\n",
    "# STUDYv2, PROTOCOLv2, SAMPLEv2, SUBJECTv2, CLINPATHv2, DATAv2 = v2_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>hx_dementia_mci</th>\n",
       "      <th>hx_melanoma</th>\n",
       "      <th>education_level</th>\n",
       "      <th>smoking_status</th>\n",
       "      <th>smoking_years</th>\n",
       "      <th>APOE_e4_status</th>\n",
       "      <th>cognitive_status</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC01</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>81.0</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC02</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>65.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC03</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>72.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC04</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>70.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC05</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>63.0</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HC06</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>69.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PD01</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>81.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PD02</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>72.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PD03</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>77.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PD04</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>79.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PD05</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>96.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PD06</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NKI/NYUGSOM</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>86.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id AMPPD_id GP2_id biobank_name organism     sex  \\\n",
       "0        HC01              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "1        HC02              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "2        HC03              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "3        HC04              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "4        HC05              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "5        HC06              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "6        PD01              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "7        PD02              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "8        PD03              None     None   None  NKI/NYUGSOM    Human  Female   \n",
       "9        PD04              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "10       PD05              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "11       PD06              None     None   None  NKI/NYUGSOM    Human    Male   \n",
       "\n",
       "    age_at_collection                       race     ethnicity  ...  \\\n",
       "0                81.0  Black or African American  Not Reported  ...   \n",
       "1                65.0                      White  Not Reported  ...   \n",
       "2                72.0                      White  Not Reported  ...   \n",
       "3                70.0                      White  Not Reported  ...   \n",
       "4                63.0  Black or African American  Not Reported  ...   \n",
       "5                69.0                      White  Not Reported  ...   \n",
       "6                81.0                      White  Not Reported  ...   \n",
       "7                72.0                      White  Not Reported  ...   \n",
       "8                77.0                      White  Not Reported  ...   \n",
       "9                79.0                      White  Not Reported  ...   \n",
       "10               96.0                      White  Not Reported  ...   \n",
       "11               86.0                      White  Not Reported  ...   \n",
       "\n",
       "   hx_dementia_mci hx_melanoma education_level smoking_status smoking_years  \\\n",
       "0             None        None            None           None          None   \n",
       "1             None        None            None           None          None   \n",
       "2             None        None            None           None          None   \n",
       "3             None        None            None           None          None   \n",
       "4             None        None            None           None          None   \n",
       "5             None        None            None           None          None   \n",
       "6             None        None            None           None          None   \n",
       "7             None        None            None           None          None   \n",
       "8             None        None            None           None          None   \n",
       "9             None        None            None           None          None   \n",
       "10            None        None            None           None          None   \n",
       "11            None        None            None           None          None   \n",
       "\n",
       "   APOE_e4_status cognitive_status time_from_baseline primary_diagnosis  \\\n",
       "0            None             None               None   Healthy Control   \n",
       "1            None             None               None   Healthy Control   \n",
       "2            None             None               None   Healthy Control   \n",
       "3            None             None               None   Healthy Control   \n",
       "4            None             None               None   Healthy Control   \n",
       "5            None             None               None   Healthy Control   \n",
       "6            None             None               None     Idiopathic PD   \n",
       "7            None             None               None     Idiopathic PD   \n",
       "8            None             None               None     Idiopathic PD   \n",
       "9            None             None               None     Idiopathic PD   \n",
       "10           None             None               None     Idiopathic PD   \n",
       "11           None             None               None     Idiopathic PD   \n",
       "\n",
       "   primary_diagnosis_text  \n",
       "0                      NA  \n",
       "1                      NA  \n",
       "2                      NA  \n",
       "3                      NA  \n",
       "4                      NA  \n",
       "5                      NA  \n",
       "6                      NA  \n",
       "7                      NA  \n",
       "8                      NA  \n",
       "9                      NA  \n",
       "10                     NA  \n",
       "11                     NA  \n",
       "\n",
       "[12 rows x 25 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2_tables['SUBJECT']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **15 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 1/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **6 Fields with invalid entries:**\n",
      "- _*family_history*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No', 'Unknown', 'Not Reported'\n",
      "- _*hx_dementia_mci*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*hx_melanoma*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*education_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other'\n",
      "- _*smoking_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Current smoker', 'Former smoker', 'Never', 'Unknown'\n",
      "- _*APOE_e4_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '22', '23', '24', '33', '34', '44', 'Unknown'\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **32 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_dx_main: 6/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **19 Fields with invalid entries:**\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder'\n",
      "- _*path_braak_nft*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', 'I', 'II', 'III', 'IV', 'V', 'VI', 'I/II', 'III/IV', 'IV/V', 'V/VI'\n",
      "- _*path_braak_asyn*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_cerad*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Sparse', 'Moderate', 'Frequent'\n",
      "- _*path_thal*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*known_pathogenic_mutation*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Present', 'Unknown'\n",
      "- _*path_mckeith*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_infarcs*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*path_nia_ri*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*path_nia_aa_a*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3'\n",
      "- _*TDP43*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*arteriolosclerosis_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "- _*dig_slide_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*quant_path_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- tissue_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_ontology_term_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- suspension_type: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **2 Fields with invalid entries:**\n",
      "- _*hemisphere*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Right', 'Left', 'Unknown'\n",
      "- _*input_cell_count*_:  invalid values 💩'single nuclei, 8000'\n",
      "    - valid ➡️ int or NULL ('NA')\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- file_description: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 36/36 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 36/36 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "for table,df in v2_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **16 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- condition_id: 12/12 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- tissue_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- assay_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **2 Fields with invalid entries:**\n",
      "- _*organism*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Human', 'Mouse', 'Dog', 'Fly', 'Other'\n",
      "- _*assay_type*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'scRNAseq', 'snRNAseq', 'bulkRNAseq'\n",
      "\n",
      "🚨⚠️❗ **Missing Required Fields in PMDBS: brain_region, region_level_1, region_level_2, region_level_3**\n",
      "🚨⚠️❗ **Missing Optional Fields in PMDBS: hemisphere**\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "🚨⚠️❗ **Extra field in PMDBS: tissue**\n",
      "🚨⚠️❗ **Extra field in PMDBS: technology**\n",
      "🚨⚠️❗ **Extra field in PMDBS: omic**\n",
      "🚨⚠️❗ **Extra field in PMDBS: RIN**\n",
      "🚨⚠️❗ **Extra field in PMDBS: molecular_source**\n",
      "🚨⚠️❗ **Extra field in PMDBS: input_cell_count**\n",
      "🚨⚠️❗ **Extra field in PMDBS: assay**\n",
      "🚨⚠️❗ **Extra field in PMDBS: sequencing_end**\n",
      "🚨⚠️❗ **Extra field in PMDBS: sequencing_length**\n",
      "🚨⚠️❗ **Extra field in PMDBS: sequencing_instrument**\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **46 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 1/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_dx_main: 6/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_nft: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 12/12 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 12/12 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **25 Fields with invalid entries:**\n",
      "- _*family_history*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No', 'Unknown', 'Not Reported'\n",
      "- _*hx_dementia_mci*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*hx_melanoma*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*education_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other'\n",
      "- _*smoking_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Current smoker', 'Former smoker', 'Never', 'Unknown'\n",
      "- _*APOE_e4_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '22', '23', '24', '33', '34', '44', 'Unknown'\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder'\n",
      "- _*path_braak_nft*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', 'I', 'II', 'III', 'IV', 'V', 'VI', 'I/II', 'III/IV', 'IV/V', 'V/VI'\n",
      "- _*path_braak_asyn*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_cerad*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Sparse', 'Moderate', 'Frequent'\n",
      "- _*path_thal*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*known_pathogenic_mutation*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Present', 'Unknown'\n",
      "- _*path_mckeith*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_infarcs*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*path_nia_ri*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*path_nia_aa_a*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3'\n",
      "- _*TDP43*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*arteriolosclerosis_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "- _*dig_slide_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*quant_path_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 2/2 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 2/2 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## transfer hafler metadata Hafler "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-hafler@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "# Lee\n",
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/hafler-credentials.json \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      1260  2023-11-30T21:03:06Z  gs://asap-raw-data-team-hafler/metadata/v2/CLINPATH.csv#1701378186816137  metageneration=1\n",
      "      5290  2023-11-30T21:03:07Z  gs://asap-raw-data-team-hafler/metadata/v2/DATA.csv#1701378187386113  metageneration=1\n",
      "      1761  2023-11-30T21:03:07Z  gs://asap-raw-data-team-hafler/metadata/v2/PROTOCOL.csv#1701378187767597  metageneration=1\n",
      "      3260  2023-11-30T21:03:07Z  gs://asap-raw-data-team-hafler/metadata/v2/SAMPLE.csv#1701378187201094  metageneration=1\n",
      "      1446  2023-11-30T21:03:07Z  gs://asap-raw-data-team-hafler/metadata/v2/STUDY.csv#1701378187591914  metageneration=1\n",
      "      1491  2023-11-30T21:03:07Z  gs://asap-raw-data-team-hafler/metadata/v2/SUBJECT.csv#1701378187002956  metageneration=1\n",
      "                                 gs://asap-raw-data-team-hafler/metadata/v2/v2_20231114/\n",
      "                                 gs://asap-raw-data-team-hafler/metadata/v2/v2_20231128/\n",
      "TOTAL: 6 objects, 14508 bytes (14.17 KiB)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!gsutil -u dnastack-asap-parkinsons ls -al \"gs://asap-raw-data-team-hafler/metadata/v2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/ergonyc/Projects/ASAP/meta-clean/hafler.ipynb Cell 40\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/ergonyc/Projects/ASAP/meta-clean/hafler.ipynb#X54sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m Path\u001b[39m.\u001b[39mcwd()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Path' is not defined"
     ]
    }
   ],
   "source": [
    "Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Hafler/v2_20231130/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/SUBJECT.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/SAMPLE.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/DATA.csv [Content-Type=text/csv]...\n",
      "- [4 files][ 11.0 KiB/ 11.0 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Hafler/v2_20231130/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/PROTOCOL.csv [Content-Type=text/csv]...\n",
      "- [6 files][ 14.2 KiB/ 14.2 KiB]                                                \n",
      "Operation completed over 6 objects/14.2 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "\n",
    "# !gsutil -u dnastack-asap-parkinsons cp -r ./clean/team-Hafler/v2_20231128  \"gs://asap-raw-data-team-hafler/metadata/v2\"\n",
    "!gsutil -u dnastack-asap-parkinsons cp -r \"./clean/team-Hafler/v2_20231130/*.csv\"  \"gs://asap-raw-data-team-hafler/metadata/v2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy to workflow-dev bucket\n",
    "\n",
    "First copy each set of metadata locally ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [admin-workflow-dev@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/wf-credentials.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Hafler/v2_20231130/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/SUBJECT.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/SAMPLE.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/DATA.csv [Content-Type=text/csv]...\n",
      "- [4 files][ 11.0 KiB/ 11.0 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Hafler/v2_20231130/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Hafler/v2_20231130/PROTOCOL.csv [Content-Type=text/csv]...\n",
      "\\ [6 files][ 14.2 KiB/ 14.2 KiB]                                                \n",
      "Operation completed over 6 objects/14.2 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil  cp -r \"./clean/team-Hafler/v2_20231130/*.csv\"  \"gs://asap-workflow-dev/metadata/v2/hafler\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "damn\n"
     ]
    }
   ],
   "source": [
    "print(\"damn\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check file md5s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.checksums import extract_md5_from_details, extract_md5_from_details2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-hafler@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/hafler-credentials.json  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "WARNING: You have requested checksumming but your crcmod installation isn't\n",
      "using the module's C extension, so checksumming will run very slowly. For help\n",
      "installing the extension, please see \"gsutil help crcmod\".\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# !gcloud storage hash \"gs://asap-raw-data-team-hardy/**/*.gz\"  --skip-crc32c --hex  --billing-project dnastack-asap-parkinsons > hardy_hexhash.log\n",
    "\n",
    "!gsutil -u dnastack-asap-parkinsons hash -h \"gs://asap-raw-data-team-hafler/**/*.gz\" > hafler_hexhash.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_56418/1240943610.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check1'] =checksum['file_MD5'].str.strip()\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_56418/1240943610.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n"
     ]
    }
   ],
   "source": [
    "bucket_files_md5 = extract_md5_from_details2(\"hafler_hexhash.log\")\n",
    "\n",
    "\n",
    "\n",
    "checksum = DATAv2[['file_name','file_MD5']]\n",
    "checksum['check1'] = checksum['file_MD5'].str.strip()\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "#empty means success!!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
