{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata compilation\n",
    "\n",
    "# Team Hardy. ASAP CRN Metadata scrubbing\n",
    "\n",
    "confirm Team Hardy scRNAseq metadata\n",
    "9 Oct 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, create_upload_medadata_package\n",
    "from checksums import get_md5_hashes, authenticate_with_service_account\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> SANITY CHECK: verify reading from google doc works.\n",
    "\n",
    "```python\n",
    "CDEv1_ = read_CDE(\"v1\")\n",
    "CDEv2_ = read_CDE(\"v2.1\")\n",
    "CDEv3_ = read_CDE(\"v3.0\")\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted as v1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Team Hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"hardy\"\n",
    "dataset_name = \"sn-rnaseq\"\n",
    "\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "og_path = metadata_path / \"og\"\n",
    "\n",
    "\n",
    "# v1_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'CLINPATH', 'SAMPLE']\n",
    "\n",
    "# dfs = {}\n",
    "# for table in v1_tables:\n",
    "#     dfs[table] =read_meta_table(og_path / f\"{table}.csv\")\n",
    "\n",
    "\n",
    "STUDY = read_meta_table(og_path / \"STUDY.csv\")\n",
    "PROTOCOL = read_meta_table(og_path / \"PROTOCOL.csv\")\n",
    "SUBJECT = read_meta_table(og_path / \"SUBJECT.csv\")\n",
    "CLINPATH = read_meta_table(og_path / \"CLINPATH.csv\")\n",
    "SAMPLE = read_meta_table(og_path / \"SAMPLE.csv\")\n",
    "\n",
    "\n",
    "metadata_version = \"v1\"\n",
    "\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>project_description</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Ryten Lab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>Mina Ryten</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PI_email</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>contributor_names</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>submitter_name</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>submitter_email</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASAP_grant_id</td>\n",
       "      <td>ASAP-000478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brain_regions</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DUA_version</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>metadata_version_date</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name                                              value\n",
       "0              project_name  Understanding mechanisms of Parkinson's diseas...\n",
       "1           project_dataset                                    Hardy snRNA-seq\n",
       "2       project_description  Genetic analysis has identified many risk gene...\n",
       "3            ASAP_team_name                                         TEAM-HARDY\n",
       "4             ASAP_lab_name                                          Ryten Lab\n",
       "5              PI_full_name                                         Mina Ryten\n",
       "6                  PI_email                               mina.ryten@ucl.ac.uk\n",
       "7         contributor_names  Aine Fairbrother-Browne, Jonathan Brenton, Mel...\n",
       "8            submitter_name                            Aine Fairbrother-Browne\n",
       "9           submitter_email               aine.fairbrother-browne.18@ucl.ac.uk\n",
       "10            ASAP_grant_id                                        ASAP-000478\n",
       "11     other_funding_source                                                 NA\n",
       "12          publication_DOI                                                 NA\n",
       "13         publication_PMID                                                 NA\n",
       "14  number_of_brain_samples                                                128\n",
       "15            brain_regions  Inferior Parietal Lobule (IPL), Anterior Cingu...\n",
       "16         types_of_samples  Late stage (Braak 5-6) PD and control post-mor...\n",
       "17                PI_ORCHID                                0000-0001-9520-6957\n",
       "18     PI_google_scholar_id  https://scholar.google.co.uk/citations?user=lt...\n",
       "19              DUA_version                                                 NA\n",
       "20    metadata_version_date                                 Version 1, 09/2023"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>project_description</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>PI_email</td>\n",
       "      <td>contributor_names</td>\n",
       "      <td>submitter_name</td>\n",
       "      <td>submitter_email</td>\n",
       "      <td>...</td>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>brain_regions</td>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>DUA_version</td>\n",
       "      <td>metadata_version_date</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  0                1   \\\n",
       "0                                       project_name  project_dataset   \n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "                                                  2               3   \\\n",
       "0                                project_description  ASAP_team_name   \n",
       "1  Genetic analysis has identified many risk gene...      TEAM-HARDY   \n",
       "\n",
       "              4             5                     6   \\\n",
       "0  ASAP_lab_name  PI_full_name              PI_email   \n",
       "1      Ryten Lab    Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                                  7                        8   \\\n",
       "0                                  contributor_names           submitter_name   \n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "                                     9   ...                    11  \\\n",
       "0                       submitter_email  ...  other_funding_source   \n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                    NA   \n",
       "\n",
       "                12                13                       14  \\\n",
       "0  publication_DOI  publication_PMID  number_of_brain_samples   \n",
       "1               NA                NA                      128   \n",
       "\n",
       "                                                  15  \\\n",
       "0                                      brain_regions   \n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "                                                  16                   17  \\\n",
       "0                                   types_of_samples            PI_ORCHID   \n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "                                                  18           19  \\\n",
       "0                               PI_google_scholar_id  DUA_version   \n",
       "1  https://scholar.google.co.uk/citations?user=lt...           NA   \n",
       "\n",
       "                      20  \n",
       "0  metadata_version_date  \n",
       "1     Version 1, 09/2023  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset  \\\n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "1  Genetic analysis has identified many risk gene...     TEAM-HARDY   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email  \\\n",
       "1     Ryten Lab   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "0                                  contributor_names           submitter_name  \\\n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "0                       submitter_email  ... other_funding_source  \\\n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                   NA   \n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples  \\\n",
       "1              NA               NA                     128   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID  \\\n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "0                               PI_google_scholar_id DUA_version  \\\n",
       "1  https://scholar.google.co.uk/citations?user=lt...          NA   \n",
       "\n",
       "0 metadata_version_date  \n",
       "1    Version 1, 09/2023  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NA</td>\n",
       "      <td>v1_20241025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset  \\\n",
       "0  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "0  Genetic analysis has identified many risk gene...     TEAM-HARDY   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email  \\\n",
       "0     Ryten Lab   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "0                                  contributor_names           submitter_name  \\\n",
       "0  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "0                       submitter_email  ... other_funding_source  \\\n",
       "0  aine.fairbrother-browne.18@ucl.ac.uk  ...                   NA   \n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples  \\\n",
       "0              NA               NA                     128   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "0  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID  \\\n",
       "0  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "0                               PI_google_scholar_id DUA_version  \\\n",
       "0  https://scholar.google.co.uk/citations?user=lt...          NA   \n",
       "\n",
       "0 metadata_version_date  \n",
       "0           v1_20241025  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "STUDY.reset_index(drop=True, inplace=True)\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_collection_summary</td>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell_extraction_summary</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib_prep_summary</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>data_processing_summary</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>github_url</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>protocols_io_DOI</td>\n",
       "      <td>Nuclear extraction protocol: 10.17504/protocol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>other_reference</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name  \\\n",
       "0  sample_collection_summary   \n",
       "1    cell_extraction_summary   \n",
       "2           lib_prep_summary   \n",
       "3    data_processing_summary   \n",
       "4                 github_url   \n",
       "5           protocols_io_DOI   \n",
       "6            other_reference   \n",
       "\n",
       "                                               value  \n",
       "0  This dataset contains cortical regions only, p...  \n",
       "1  From protocols.io: This protocol is used to is...  \n",
       "2  'Nuclei were extracted from homogenised post-m...  \n",
       "3  Cell ranger was used to convert raw sequencing...  \n",
       "4  Raw to fastq to mapped: https://github.com/RHR...  \n",
       "5  Nuclear extraction protocol: 10.17504/protocol...  \n",
       "6                                                 NA  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_collection_summary</td>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell_extraction_summary</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib_prep_summary</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>data_processing_summary</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>github_url</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name  \\\n",
       "0  sample_collection_summary   \n",
       "1    cell_extraction_summary   \n",
       "2           lib_prep_summary   \n",
       "3    data_processing_summary   \n",
       "4                 github_url   \n",
       "\n",
       "                                               value  \n",
       "0  This dataset contains cortical regions only, p...  \n",
       "1  From protocols.io: This protocol is used to is...  \n",
       "2  'Nuclei were extracted from homogenised post-m...  \n",
       "3  Cell ranger was used to convert raw sequencing...  \n",
       "4  Raw to fastq to mapped: https://github.com/RHR...  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # fix PROTOCOL formatting\n",
    "# tmp = pd.DataFrame()\n",
    "# tmp = PROTOCOL[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "# tmp.columns = tmp.iloc[0]\n",
    "# PROTOCOL = tmp.drop([0])\n",
    "PROTOCOL.reset_index(drop=True, inplace=True)\n",
    "PROTOCOL.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay</th>\n",
       "      <th>sequencing_end</th>\n",
       "      <th>sequencing_length</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3612</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3613</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>v3.1 - Dual Index</td>\n",
       "      <td>paired-end</td>\n",
       "      <td>190</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  assay sequencing_end sequencing_length  \\\n",
       "0     v3.1 - Dual Index     paired-end               190   \n",
       "1     v3.1 - Dual Index     paired-end               190   \n",
       "2     v3.1 - Dual Index     paired-end               190   \n",
       "3     v3.1 - Dual Index     paired-end               190   \n",
       "4     v3.1 - Dual Index     paired-end               190   \n",
       "...                 ...            ...               ...   \n",
       "3611  v3.1 - Dual Index     paired-end               190   \n",
       "3612  v3.1 - Dual Index     paired-end               190   \n",
       "3613  v3.1 - Dual Index     paired-end               190   \n",
       "3614  v3.1 - Dual Index     paired-end               190   \n",
       "3615  v3.1 - Dual Index     paired-end               190   \n",
       "\n",
       "      sequencing_instrument  \n",
       "0     Illumina NovaSeq 6000  \n",
       "1     Illumina NovaSeq 6000  \n",
       "2     Illumina NovaSeq 6000  \n",
       "3     Illumina NovaSeq 6000  \n",
       "4     Illumina NovaSeq 6000  \n",
       "...                     ...  \n",
       "3611  Illumina NovaSeq 6000  \n",
       "3612  Illumina NovaSeq 6000  \n",
       "3613  Illumina NovaSeq 6000  \n",
       "3614  Illumina NovaSeq 6000  \n",
       "3615  Illumina NovaSeq 6000  \n",
       "\n",
       "[3616 rows x 4 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE[['assay','sequencing_end', 'sequencing_length', 'sequencing_instrument']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3612</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3613</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id source_subject_id subject_id replicate replicate_count  \\\n",
       "0     babom_ACG             P2/14      babom        NA               1   \n",
       "1     babom_ACG             P2/14      babom        NA               1   \n",
       "2     babom_ACG             P2/14      babom        NA               1   \n",
       "3     babom_ACG             P2/14      babom        NA               1   \n",
       "4     babom_ACG             P2/14      babom        NA               1   \n",
       "...         ...               ...        ...       ...             ...   \n",
       "3611  zupam_IPL            P78/11      zupam      rep2               2   \n",
       "3612  zupam_IPL            P78/11      zupam      rep2               2   \n",
       "3613  zupam_IPL            P78/11      zupam      rep2               2   \n",
       "3614  zupam_IPL            P78/11      zupam      rep2               2   \n",
       "3615  zupam_IPL            P78/11      zupam      rep2               2   \n",
       "\n",
       "     repeated_sample batch tissue brain_region source_RIN  ...  \\\n",
       "0                  0    B1  brain          ACG         NA  ...   \n",
       "1                  0    B1  brain          ACG         NA  ...   \n",
       "2                  0    B1  brain          ACG         NA  ...   \n",
       "3                  0    B1  brain          ACG         NA  ...   \n",
       "4                  0    B1  brain          ACG         NA  ...   \n",
       "...              ...   ...    ...          ...        ...  ...   \n",
       "3611               1    B2  brain          IPL         NA  ...   \n",
       "3612               1    B2  brain          IPL         NA  ...   \n",
       "3613               1    B2  brain          IPL         NA  ...   \n",
       "3614               1    B2  brain          IPL         NA  ...   \n",
       "3615               1    B2  brain          IPL         NA  ...   \n",
       "\n",
       "       sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0     PATO:0000383 (female)                                       NA   \n",
       "1     PATO:0000383 (female)                                       NA   \n",
       "2     PATO:0000383 (female)                                       NA   \n",
       "3     PATO:0000383 (female)                                       NA   \n",
       "4     PATO:0000383 (female)                                       NA   \n",
       "...                     ...                                      ...   \n",
       "3611  PATO:0000383 (female)                                       NA   \n",
       "3612  PATO:0000383 (female)                                       NA   \n",
       "3613  PATO:0000383 (female)                                       NA   \n",
       "3614  PATO:0000383 (female)                                       NA   \n",
       "3615  PATO:0000383 (female)                                       NA   \n",
       "\n",
       "     disease_ontology_term_id tissue_ontology_term_id  \\\n",
       "0               MONDO:0005180          UBERON:0009835   \n",
       "1               MONDO:0005180          UBERON:0009835   \n",
       "2               MONDO:0005180          UBERON:0009835   \n",
       "3               MONDO:0005180          UBERON:0009835   \n",
       "4               MONDO:0005180          UBERON:0009835   \n",
       "...                       ...                     ...   \n",
       "3611            MONDO:0005180          UBERON:0006088   \n",
       "3612            MONDO:0005180          UBERON:0006088   \n",
       "3613            MONDO:0005180          UBERON:0006088   \n",
       "3614            MONDO:0005180          UBERON:0006088   \n",
       "3615            MONDO:0005180          UBERON:0006088   \n",
       "\n",
       "     cell_type_ontology_term_id assay_ontology_term_id suspension_type DV2000  \\\n",
       "0                            NA            EFO:0008913         nucleus     NA   \n",
       "1                            NA            EFO:0008913         nucleus     NA   \n",
       "2                            NA            EFO:0008913         nucleus     NA   \n",
       "3                            NA            EFO:0008913         nucleus     NA   \n",
       "4                            NA            EFO:0008913         nucleus     NA   \n",
       "...                         ...                    ...             ...    ...   \n",
       "3611                         NA            EFO:0008913         nucleus     NA   \n",
       "3612                         NA            EFO:0008913         nucleus     NA   \n",
       "3613                         NA            EFO:0008913         nucleus     NA   \n",
       "3614                         NA            EFO:0008913         nucleus     NA   \n",
       "3615                         NA            EFO:0008913         nucleus     NA   \n",
       "\n",
       "     pm_PH donor_id  \n",
       "0       NA       NA  \n",
       "1       NA       NA  \n",
       "2       NA       NA  \n",
       "3       NA       NA  \n",
       "4       NA       NA  \n",
       "...    ...      ...  \n",
       "3611    NA       NA  \n",
       "3612    NA       NA  \n",
       "3613    NA       NA  \n",
       "3614    NA       NA  \n",
       "3615    NA       NA  \n",
       "\n",
       "[3616 rows x 42 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "SAMPLE\n",
    "\n",
    "\n",
    "# add 'replicate' coding (nans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE['replicate'] = SAMPLE['replicate'].replace({NULL: \"Rep1\", \"rep1\": \"Rep1\", \"rep2\": \"Rep2\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NA</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NA</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_subject_id subject_id replicate replicate_count  \\\n",
       "0  babom_ACG             P2/14      babom      Rep1               1   \n",
       "1  babom_ACG             P2/14      babom      Rep1               1   \n",
       "2  babom_ACG             P2/14      babom      Rep1               1   \n",
       "3  babom_ACG             P2/14      babom      Rep1               1   \n",
       "4  babom_ACG             P2/14      babom      Rep1               1   \n",
       "\n",
       "  repeated_sample batch tissue brain_region source_RIN  ...  \\\n",
       "0               0    B1  brain          ACG         NA  ...   \n",
       "1               0    B1  brain          ACG         NA  ...   \n",
       "2               0    B1  brain          ACG         NA  ...   \n",
       "3               0    B1  brain          ACG         NA  ...   \n",
       "4               0    B1  brain          ACG         NA  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0  PATO:0000383 (female)                                       NA   \n",
       "1  PATO:0000383 (female)                                       NA   \n",
       "2  PATO:0000383 (female)                                       NA   \n",
       "3  PATO:0000383 (female)                                       NA   \n",
       "4  PATO:0000383 (female)                                       NA   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0            MONDO:0005180          UBERON:0009835                         NA   \n",
       "1            MONDO:0005180          UBERON:0009835                         NA   \n",
       "2            MONDO:0005180          UBERON:0009835                         NA   \n",
       "3            MONDO:0005180          UBERON:0009835                         NA   \n",
       "4            MONDO:0005180          UBERON:0009835                         NA   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV2000 pm_PH donor_id  \n",
       "0            EFO:0008913         nucleus     NA    NA       NA  \n",
       "1            EFO:0008913         nucleus     NA    NA       NA  \n",
       "2            EFO:0008913         nucleus     NA    NA       NA  \n",
       "3            EFO:0008913         nucleus     NA    NA       NA  \n",
       "4            EFO:0008913         nucleus     NA    NA       NA  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       6\n",
       "1       6\n",
       "2       6\n",
       "3       6\n",
       "4       0\n",
       "       ..\n",
       "133    NA\n",
       "134    NA\n",
       "135    NA\n",
       "136    NA\n",
       "137    NA\n",
       "Name: path_braak_asyn, Length: 138, dtype: object"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_braak_asyn']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_nft'] = CLINPATH['path_braak_nft'].replace({\"0\":\"0\", \n",
    "                                                                \"1\":\"I\", \n",
    "                                                                \"2\": \"II\", \n",
    "                                                                \"3\":\"III\", \n",
    "                                                                \"4\":\"IV\", \n",
    "                                                                \"5\":\"V\", \n",
    "                                                                \"6\":\"VI\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# code family_history as \"Not Reported\" (currently empty)\n",
    "CLINPATH['family_history'] = \"Not Reported\"\n",
    "\n",
    "\n",
    "\n",
    "# check APOE_e4_status ? currently empty\n",
    "\n",
    "# `path_autopsy_dx_main`  actually seems good parser might be wrong\n",
    "\n",
    "# code \"at least 4\" as \"4/5\" \n",
    "\n",
    "CLINPATH['path_thal'] = CLINPATH['path_thal'].replace({'At least 4':\"4/5\"})\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace({'Diffuse neocortical': \"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\", \n",
    "                                                        'Limbic transitional': \"Limbic (transitional)\" ,\n",
    "                                                        'Diffuse Neocortical':\"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CLINPATH['path_nia_aa_a'] = CLINPATH['path_nia_aa_a'].replace({\"0\":\"A0\", \n",
    "                                                                    \"1\":\"A1\", \n",
    "                                                                    \"2\": \"A2\", \n",
    "                                                                    \"3\":\"A3\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_b'] = CLINPATH['path_nia_aa_b'].replace({\"0\":\"B0\", \n",
    "                                                                \"1\":\"B1\", \n",
    "                                                                \"2\": \"B2\", \n",
    "                                                                \"3\":\"B3\"})\n",
    "\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_c'] = CLINPATH['path_nia_aa_c'].replace({\"0\":\"C0\", \n",
    "                                                                \"1\":\"C1\", \n",
    "                                                                \"2\": \"C2\", \n",
    "                                                                \"3\":\"C3\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "CLINPATH['path_ad_level'] = CLINPATH['path_ad_level'].replace({\"No evidence\": \"No evidence of Alzheimer\\'s disease neuropathological change\"})\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_autopsy_map = { \"Parkinson's disease with dementia\": \"Parkinson's disease with dementia\", \n",
    "       \"Parkinson's disease\": \"Parkinson's disease\",\n",
    "       'Control brain':\"Control, no misfolded protein or significant vascular pathology\", \n",
    "       'Pathological ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Control brain / Path ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Argyrophilic grain disease': \"Control, Argyrophilic grain disease\",\n",
    "       'Control brain, Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       'Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       \"Control brain, Alzheimer`s disease (intermediate level AD pathological change)\":\"Alzheimer's disease (intermediate level neuropathological change)\",\n",
    "       'Control brain / Path ageing, CAA':\"Control, Cerebrovascular disease (cerebral amyloid angiopathy)\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['path_autopsy_dx_main'] = CLINPATH['path_autopsy_dx_main'].replace(path_autopsy_map)\n",
    "# - _*age_at_diagnosis*_:  invalid values 💩'69', '61', 'Not applicable', '71', '60', '57', '79', '87', '67', '75', '50', '35', '44', '42', '58', '80', '72', '65', '48', '74', '62', '36', '77', '43', '56', '55', '37', '66', '64', '49', 'NA'\n",
    "\n",
    "CLINPATH['age_at_diagnosis'] = CLINPATH['age_at_diagnosis'].replace({\"Not applicable\": NULL})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 🚨⚠️❗ **Extra field in CLINPATH: source_subject_id**\n",
    "# rename source_subject_id to source_sample_id\n",
    "CLINPATH.rename(columns={\"source_subject_id\":\"source_sample_id\"}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace({\"Alzheimer?s disease\":\"Alzheimer's disease\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# - _*molecular_source*_:  invalid values 💩'nuclear RNA'\n",
    "#     - valid ➡️ 'Total RNA', 'PolyA RNA', 'Cytoplasmic RNA', 'Nuclear RNA', 'Genomic DNA', 'Protein', 'Other'\n",
    "# - _*input_cell_count*_:  invalid values 💩'16483.800000000003', '20418', '16714.643750000003', '17644.912977099237', '17353.600000000002', '17388.800000000003', '17952', '17776', '19712', '16948.800000000003', '17098', '18656', '17265.600000000002', '12223.199999999999', '14520', '18128', '17933.2', '12289.199999999999', '12051.599999999999', '13860', '13970', '18304', '17564.879999999997', '18480', '17916.800000000003', '19855.07096774194', '17368.707692307697', '16766', '17529.600000000002', '19360', '23100', '15180', '18100.800000000003', '19536', '18832', '13094.4', '18601.559999999998', '13194.72', '13226.4', '11893.199999999999', '16632', '20400', '21800', '17555.9', '15312', '12249.599999999999', '18426', '23738.000000000004', '17920', '18347.28', '18503.76', '16859', '19680', '18163.2', '20856', '15048', '14124', '17600', '12381.599999999999', '16566.800000000003', '19422', '21120', '17300.800000000003', '21472.100000000002', '45500.600000000006', '17001.600000000002', '17314', '10942.8', '9292.8', '20064', '18008.199999999997', '17476.800000000003', '12157.199999999999', '17688', '13570.656', '13572.24', '18249.48', '16684.800000000003', '13678.5', '17716.83', '18458', '25600', '23800', '13195.775999999998', '13190.099999999999', '18240', '17762', '19920', '18445.079999999998', '17738.6', '16772.800000000003', '17477.7', '19520', '19754', '17211.828571428574', '14388', '15582', '22400', '22000', '18260', '14552.999999999998', '13332', '19928.300000000003', '19256', '19008', '18560', '17760', '20617.2', '18053.879999999997', '27365.100000000002', '17885.2'\n",
    "#     - valid ➡️ int or NULL ('NA')\n",
    "# - _*sequencing_end*_:  invalid values 💩'paired-end'\n",
    "#     - valid ➡️ 'Single', 'Paired-end'\n",
    "# - _*sequencing_length*_:  invalid values 💩'190'\n",
    "#     - valid ➡️ '25', '50', '100', '150'\n",
    "# - _*technology*_:  invalid values 💩'sN'\n",
    "#     - valid ➡️ 'SN', 'SC', 'Spatial', 'Bulk', 'Seq', 'Array'\n",
    "# - _*adjustment*_:  invalid values 💩'raw'\n",
    "#     - valid ➡️ 'Raw', 'Processed'\n",
    "# - _*content*_:  invalid values 💩'reads'\n",
    "#     - valid ➡️ 'Counts', 'Probabilities', 'Genotypes', 'Dosages', 'Reads'\n",
    "\n",
    "SAMPLE = prep_table(SAMPLE, CDEv1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "v1_path = metadata_path / \"v1\"\n",
    "STUDY.to_csv(v1_path / f\"STUDY.csv\", index=False)\n",
    "SAMPLE.to_csv(v1_path / f\"SAMPLE.csv\", index=False)\n",
    "SUBJECT.to_csv(v1_path / f\"SUBJECT.csv\", index=False)\n",
    "PROTOCOL.to_csv(v1_path / f\"PROTOCOL.csv\", index=False)\n",
    "CLINPATH.to_csv(v1_path / f\"CLINPATH.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'f37738d04879c1cb8c380920964dad4d'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE[['sample_id','subject_id','file_name','file_MD5']]['file_MD5'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validate v1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 64/64 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 64/64 empty rows (REQUIRED)\n",
      "\n",
      "\t- duration_pmi: 1/64 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "🚨⚠️❗ **Missing Required Fields in SAMPLE: source_sample_id**\n",
      "🚨⚠️❗ **Missing Optional Fields in SAMPLE: DV200**\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_RIN: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 64/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- pm_PH: 3616/3616 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 3616/3616 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **2 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values 💩'16483.800000000003', '20418', '16714.643750000003', '17644.912977099237', '17353.600000000002', '17388.800000000003', '17952', '17776', '19712', '16948.800000000003', '17098', '18656', '17265.600000000002', '12223.199999999999', '14520', '18128', '17933.2', '12289.199999999999', '12051.599999999999', '13860', '13970', '18304', '17564.879999999997', '18480', '17916.800000000003', '19855.07096774194', '17368.707692307697', '16766', '17529.600000000002', '19360', '23100', '15180', '18100.800000000003', '19536', '18832', '13094.4', '18601.559999999998', '13194.72', '13226.4', '11893.199999999999', '16632', '20400', '21800', '17555.9', '15312', '12249.599999999999', '18426', '23738.000000000004', '17920', '18347.28', '18503.76', '16859', '19680', '18163.2', '20856', '15048', '14124', '17600', '12381.599999999999', '16566.800000000003', '19422', '21120', '17300.800000000003', '21472.100000000002', '45500.600000000006', '17001.600000000002', '17314', '10942.8', '9292.8', '20064', '18008.199999999997', '17476.800000000003', '12157.199999999999', '17688', '13570.656', '13572.24', '18249.48', '16684.800000000003', '13678.5', '17716.83', '18458', '25600', '23800', '13195.775999999998', '13190.099999999999', '18240', '17762', '19920', '18445.079999999998', '17738.6', '16772.800000000003', '17477.7', '19520', '19754', '17211.828571428574', '14388', '15582', '22400', '22000', '18260', '14552.999999999998', '13332', '19928.300000000003', '19256', '19008', '18560', '17760', '20617.2', '18053.879999999997', '27365.100000000002', '17885.2'\n",
      "    - valid ➡️ int or NULL ('NA')\n",
      "- _*sequencing_length*_:  invalid values 💩'190'\n",
      "    - valid ➡️ '25', '50', '100', '150', 'NA'\n",
      "🚨⚠️❗ **Extra field in SAMPLE: source_subject_id**\n",
      "🚨⚠️❗ **Extra field in SAMPLE: DV2000**\n",
      "\n",
      "🚨⚠️❗ **Missing Required Fields in PROTOCOL: sample_collection_summary, cell_extraction_summary, lib_prep_summary, data_processing_summary, github_url, protocols_io_DOI**\n",
      "🚨⚠️❗ **Missing Optional Fields in PROTOCOL: other_reference**\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "🚨⚠️❗ **Extra field in PROTOCOL: name**\n",
      "🚨⚠️❗ **Extra field in PROTOCOL: value**\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **44 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- GP2_id: 4/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- AMPPD_id: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 54/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_dx_main: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_death: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 24/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 64/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 10/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 10/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 10/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 10/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 10/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 12/138 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv1\n",
    "tables = CDE['Table'].unique()\n",
    "v1_path = metadata_path / \"v1\"\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(v1_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table\n",
    "    # df.to_csv(v1_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved SAMPLE_auxiliary.csv\n",
      "Saved PROTOCOL_auxiliary.csv\n"
     ]
    }
   ],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v1_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v1_path / f\"{table}.csv\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v1->v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n"
     ]
    }
   ],
   "source": [
    "v2_path = metadata_path / \"v2\"\n",
    "\n",
    "# rationalize the dataset id\n",
    "team_dataset_id = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n",
    "\n",
    "v2_tables, aux_tables = v1_to_v2(v1_path, v2_path, CDEv1, CDEv2, team_dataset_id=team_dataset_id)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>sn_rnaseq</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Https://github.com/RHReynolds/asap-pd-progress...</td>\n",
       "      <td>v2.1_20241025</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name  \\\n",
       "0     TEAM-HARDY     Ryten Lab   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Understanding mechanisms of Parkinson's diseas...       sn_rnaseq   \n",
       "\n",
       "   project_dataset                                project_description  \\\n",
       "0  Hardy snRNA-seq  Genetic analysis has identified many risk gene...   \n",
       "\n",
       "  PI_full_name              PI_email  \\\n",
       "0   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                   contributor_names           submitter_name  \\\n",
       "0  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "   ... publication_PMID number_of_brain_samples  \\\n",
       "0  ...               NA                     128   \n",
       "\n",
       "                                       brain_regions  \\\n",
       "0  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "                                    types_of_samples            PI_ORCHID  \\\n",
       "0  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "                                PI_google_scholar_id DUA_version  \\\n",
       "0  https://scholar.google.co.uk/citations?user=lt...          NA   \n",
       "\n",
       "                            preprocessing_references metadata_version_date  \\\n",
       "0  Https://github.com/RHReynolds/asap-pd-progress...         v2.1_20241025   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 24 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2_tables['STUDY']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **15 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 2/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- race: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 32/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 74/74 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **32 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- duration_pmi: 1/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_dx_main: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_death: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 17/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 74/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 37/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 10/74 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 10/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 10/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 10/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 10/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 74/74 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 11/74 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_sample_id: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 2/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_sample_id: 138/138 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **2 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values 💩'16483.800000000003', '20418', '16714.643750000003', '17644.912977099237', '17353.600000000002', '17388.800000000003', '17952', '17776', '19712', '16948.800000000003', '17098', '18656', '17265.600000000002', '12223.199999999999', '14520', '18128', '17933.2', '12289.199999999999', '12051.599999999999', '13860', '13970', '18304', '17564.879999999997', '18480', '17916.800000000003', '19855.07096774194', '17368.707692307697', '16766', '17529.600000000002', '19360', '23100', '15180', '18100.800000000003', '19536', '18832', '13094.4', '18601.559999999998', '13194.72', '13226.4', '11893.199999999999', '16632', '20400', '21800', '17555.9', '15312', '12249.599999999999', '18426', '23738.000000000004', '17920', '18347.28', '18503.76', '16859', '19680', '18163.2', '20856', '15048', '14124', '17600', '12381.599999999999', '16566.800000000003', '19422', '21120', '17300.800000000003', '21472.100000000002', '45500.600000000006', '17001.600000000002', '17314', '10942.8', '9292.8', '20064', '18008.199999999997', '17476.800000000003', '12157.199999999999', '17688', '13570.656', '13572.24', '18249.48', '16684.800000000003', '13678.5', '17716.83', '18458', '25600', '23800', '13195.775999999998', '13190.099999999999', '18240', '17762', '19920', '18445.079999999998', '17738.6', '16772.800000000003', '17477.7', '19520', '19754', '17211.828571428574', '14388', '15582', '22400', '22000', '18260', '14552.999999999998', '13332', '19928.300000000003', '19256', '19008', '18560', '17760', '20617.2', '18053.879999999997', '27365.100000000002', '17885.2'\n",
      "    - valid ➡️ int or NULL ('NA')\n",
      "- _*sequencing_length*_:  invalid values 💩'190'\n",
      "    - valid ➡️ '25', '50', '100', '150', 'NA'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 3616/3616 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 3616/3616 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "for table,df in v2_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v2->v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **6 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 64/64 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **12 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_sample_id: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- organism: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 138/138 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 2/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 138/138 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 138/138 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 3616/3616 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 3616/3616 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 3616/3616 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "No empty entries (NULL) found .\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values 💩'16483.800000000003', '20418', '16714.643750000003', '17644.912977099237', '17353.600000000002', '17388.800000000003', '17952', '17776', '19712', '16948.800000000003', '17098', '18656', '17265.600000000002', '12223.199999999999', '14520', '18128', '17933.2', '12289.199999999999', '12051.599999999999', '13860', '13970', '18304', '17564.879999999997', '18480', '17916.800000000003', '19855.07096774194', '17368.707692307697', '16766', '17529.600000000002', '19360', '23100', '15180', '18100.800000000003', '19536', '18832', '13094.4', '18601.559999999998', '13194.72', '13226.4', '11893.199999999999', '16632', '20400', '21800', '17555.9', '15312', '12249.599999999999', '18426', '23738.000000000004', '17920', '18347.28', '18503.76', '16859', '19680', '18163.2', '20856', '15048', '14124', '17600', '12381.599999999999', '16566.800000000003', '19422', '21120', '17300.800000000003', '21472.100000000002', '45500.600000000006', '17001.600000000002', '17314', '10942.8', '9292.8', '20064', '18008.199999999997', '17476.800000000003', '12157.199999999999', '17688', '13570.656', '13572.24', '18249.48', '16684.800000000003', '13678.5', '17716.83', '18458', '25600', '23800', '13195.775999999998', '13190.099999999999', '18240', '17762', '19920', '18445.079999999998', '17738.6', '16772.800000000003', '17477.7', '19520', '19754', '17211.828571428574', '14388', '15582', '22400', '22000', '18260', '14552.999999999998', '13332', '19928.300000000003', '19256', '19008', '18560', '17760', '20617.2', '18053.879999999997', '27365.100000000002', '17885.2'\n",
      "    - valid ➡️ int or NULL ('NA')\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **46 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- duration_pmi: 1/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_death: 20/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- AMPPD_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 47/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_dx_main: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 29/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_nft: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 49/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 20/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 21/94 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 4/4 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 4/4 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>sn_rnaseq</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>...</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>NA</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>Https://github.com/RHReynolds/asap-pd-progress...</td>\n",
       "      <td>v3.0_20241025</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name  \\\n",
       "0     TEAM-HARDY     Ryten Lab   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Understanding mechanisms of Parkinson's diseas...       sn_rnaseq   \n",
       "\n",
       "   project_dataset                                project_description  \\\n",
       "0  Hardy snRNA-seq  Genetic analysis has identified many risk gene...   \n",
       "\n",
       "  PI_full_name              PI_email  \\\n",
       "0   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                   contributor_names           submitter_name  \\\n",
       "0  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "   ... number_samples                                       sample_types  \\\n",
       "0  ...            128  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "                                    types_of_samples DUA_version  \\\n",
       "0  Late stage (Braak 5-6) PD and control post-mor...          NA   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "                                PI_google_scholar_id  \\\n",
       "0  https://scholar.google.co.uk/citations?user=lt...   \n",
       "\n",
       "                            preprocessing_references metadata_version_date  \\\n",
       "0  Https://github.com/RHReynolds/asap-pd-progress...         v3.0_20241025   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = v3_tables['STUDY']\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hardy\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/hardy-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-hardy@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-data-team-hardy/fastqs/**/*.gz\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "prefix = \"fastqs/**/*.gz\"\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_95602/3835353718.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check1'] = checksum['file_MD5']\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_95602/3835353718.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>check1</th>\n",
       "      <th>check2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom_ACG_S20_L008_R1_001.fastq.gz</td>\n",
       "      <td>f37738d04879c1cb8c380920964dad4d</td>\n",
       "      <td>f37738d04879c1cb8c380920964dad4d</td>\n",
       "      <td>f37738d04879c1cb8c380920964dad4d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>babom_ACG_S20_L008_I1_001.fastq.gz</td>\n",
       "      <td>f3053c59a33ce48cbea6a602dcc491c7</td>\n",
       "      <td>f3053c59a33ce48cbea6a602dcc491c7</td>\n",
       "      <td>f3053c59a33ce48cbea6a602dcc491c7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>babom_ACG_S20_L008_R2_001.fastq.gz</td>\n",
       "      <td>69fe97e968dc1cb782a3d995918b5bc8</td>\n",
       "      <td>69fe97e968dc1cb782a3d995918b5bc8</td>\n",
       "      <td>69fe97e968dc1cb782a3d995918b5bc8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>babom_ACG_S20_L008_I2_001.fastq.gz</td>\n",
       "      <td>4eeee300289a4fd18169dddbbfd8ad72</td>\n",
       "      <td>4eeee300289a4fd18169dddbbfd8ad72</td>\n",
       "      <td>4eeee300289a4fd18169dddbbfd8ad72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>babom_ACG_S20_L007_R1_001.fastq.gz</td>\n",
       "      <td>7b321fc0e55ffbd496a241dcb0746def</td>\n",
       "      <td>7b321fc0e55ffbd496a241dcb0746def</td>\n",
       "      <td>7b321fc0e55ffbd496a241dcb0746def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611</th>\n",
       "      <td>zupam_IPL_S1_L002_I2_001.fastq.gz</td>\n",
       "      <td>943b32d75d48c6c10e6128543d9ad200</td>\n",
       "      <td>943b32d75d48c6c10e6128543d9ad200</td>\n",
       "      <td>943b32d75d48c6c10e6128543d9ad200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3612</th>\n",
       "      <td>zupam_IPL_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>acb14c2a6300a13329a05f550be76362</td>\n",
       "      <td>acb14c2a6300a13329a05f550be76362</td>\n",
       "      <td>acb14c2a6300a13329a05f550be76362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3613</th>\n",
       "      <td>zupam_IPL_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>b7d9893357402609b9cac32eba0586f3</td>\n",
       "      <td>b7d9893357402609b9cac32eba0586f3</td>\n",
       "      <td>b7d9893357402609b9cac32eba0586f3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>zupam_IPL_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>cb43fbc20c0e38befcf21ed87b78ea0c</td>\n",
       "      <td>cb43fbc20c0e38befcf21ed87b78ea0c</td>\n",
       "      <td>cb43fbc20c0e38befcf21ed87b78ea0c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>zupam_IPL_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>343e3efec511d9fb44de0fdc55628aa1</td>\n",
       "      <td>343e3efec511d9fb44de0fdc55628aa1</td>\n",
       "      <td>343e3efec511d9fb44de0fdc55628aa1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               file_name                          file_MD5  \\\n",
       "0     babom_ACG_S20_L008_R1_001.fastq.gz  f37738d04879c1cb8c380920964dad4d   \n",
       "1     babom_ACG_S20_L008_I1_001.fastq.gz  f3053c59a33ce48cbea6a602dcc491c7   \n",
       "2     babom_ACG_S20_L008_R2_001.fastq.gz  69fe97e968dc1cb782a3d995918b5bc8   \n",
       "3     babom_ACG_S20_L008_I2_001.fastq.gz  4eeee300289a4fd18169dddbbfd8ad72   \n",
       "4     babom_ACG_S20_L007_R1_001.fastq.gz  7b321fc0e55ffbd496a241dcb0746def   \n",
       "...                                  ...                               ...   \n",
       "3611   zupam_IPL_S1_L002_I2_001.fastq.gz  943b32d75d48c6c10e6128543d9ad200   \n",
       "3612   zupam_IPL_S1_L001_R1_001.fastq.gz  acb14c2a6300a13329a05f550be76362   \n",
       "3613   zupam_IPL_S1_L001_I1_001.fastq.gz  b7d9893357402609b9cac32eba0586f3   \n",
       "3614   zupam_IPL_S1_L001_R2_001.fastq.gz  cb43fbc20c0e38befcf21ed87b78ea0c   \n",
       "3615   zupam_IPL_S1_L001_I2_001.fastq.gz  343e3efec511d9fb44de0fdc55628aa1   \n",
       "\n",
       "                                check1                            check2  \n",
       "0     f37738d04879c1cb8c380920964dad4d  f37738d04879c1cb8c380920964dad4d  \n",
       "1     f3053c59a33ce48cbea6a602dcc491c7  f3053c59a33ce48cbea6a602dcc491c7  \n",
       "2     69fe97e968dc1cb782a3d995918b5bc8  69fe97e968dc1cb782a3d995918b5bc8  \n",
       "3     4eeee300289a4fd18169dddbbfd8ad72  4eeee300289a4fd18169dddbbfd8ad72  \n",
       "4     7b321fc0e55ffbd496a241dcb0746def  7b321fc0e55ffbd496a241dcb0746def  \n",
       "...                                ...                               ...  \n",
       "3611  943b32d75d48c6c10e6128543d9ad200  943b32d75d48c6c10e6128543d9ad200  \n",
       "3612  acb14c2a6300a13329a05f550be76362  acb14c2a6300a13329a05f550be76362  \n",
       "3613  b7d9893357402609b9cac32eba0586f3  b7d9893357402609b9cac32eba0586f3  \n",
       "3614  cb43fbc20c0e38befcf21ed87b78ea0c  cb43fbc20c0e38befcf21ed87b78ea0c  \n",
       "3615  343e3efec511d9fb44de0fdc55628aa1  343e3efec511d9fb44de0fdc55628aa1  \n",
       "\n",
       "[3616 rows x 4 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "checksum = v3_tables['DATA'][['file_name','file_MD5']]\n",
    "checksum['check1'] = checksum['file_MD5']\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "checksum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------\n",
    "## Create metadata package\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_path = root_path / f\"{team}\"\n",
    "\n",
    "create_upload_medadata_package(export_path, v3_tables)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alzheimer's_disease</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field                           condition_id intervention_name  \\\n",
       "0                              idiopathic_pd      Case-Control   \n",
       "1      no_pd_nor_other_neurological_disorder      Case-Control   \n",
       "2                        alzheimer's_disease      Case-Control   \n",
       "3                other_neurological_disorder      Case-Control   \n",
       "\n",
       "Field intervention_id protocol_id intervention_aux_table  \n",
       "0                Case          NA                     NA  \n",
       "1             Control          NA                     NA  \n",
       "2               Other          NA                     NA  \n",
       "3               Other          NA                     NA  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# these are actually matches... probably unstripped line endings."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
