{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata compilation\n",
    "\n",
    "# Team Lee (Biederer). ASAP CRN Metadata construction\n",
    "\n",
    "\n",
    "This is a bulkRNAseq dataset which was originally transfered along with the scRNAseq dataset.  The metadata should be identical, save the details of the bulkRNAseq assay compared to the sc/snRNAseq previously platformed.\n",
    "\n",
    "\n",
    "\n",
    "15 Sept 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, create_upload_medadata_package\n",
    "from checksums import extract_md5_from_details2\n",
    "from checksums import get_md5_hashes, authenticate_with_service_account\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Team Lee"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "Infer the bulk metadata from the sn dataset v3.0 metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean v3 Table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "source = \"pmdbs\"\n",
    "team = \"lee\"\n",
    "dataset_name = \"bulk-rnaseq\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "\n",
    "v3_path = metadata_path / \"v3\"\n",
    "og_path = metadata_path / \"og\"  # this data is a copy of the lee/sn_pmdbs/metadata/v3\n",
    "\n",
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "in_tables = [table_name for table_name in v3_meta_tables if f\"{table_name}.csv\" in os.listdir(og_path)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tables = {}\n",
    "for table_name in in_tables:\n",
    "    df = read_meta_table(f\"{og_path}/{table_name}.csv\")\n",
    "    og_tables[table_name] = df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq'])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tables.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>sc_pmdbs</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>unsure</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "      <td>v1_20241016</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name  \\\n",
       "0       TEAM-LEE          Bras   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Is senescence a component of human PD and does...        sc_pmdbs   \n",
       "\n",
       "                                   project_dataset  \\\n",
       "0  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "                                 project_description PI_full_name  \\\n",
       "0  Characterize the neuropathological progression...   Jose, Bras   \n",
       "\n",
       "            PI_email                                  contributor_names  \\\n",
       "0  jose.bras@vai.org  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...   \n",
       "\n",
       "     submitter_name  ... number_samples  \\\n",
       "0  Kaitlyn E Westra  ...             75   \n",
       "\n",
       "                                        sample_types  \\\n",
       "0  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "                         types_of_samples DUA_version  \\\n",
       "0  human PD and control postmortem brains      unsure   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "  PI_google_scholar_id preprocessing_references metadata_version_date  \\\n",
       "0                   NA             NA(raw data)           v1_20241016   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tables['STUDY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- submitter_email: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 75/75 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 150/150 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 150/150 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **40 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 3/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 17/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 18/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 14/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 1/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 25/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- region_level_3: 75/75 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 3/3 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 3/3 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- RIN: 75/75 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "dfs = {}\n",
    "for table,df in og_tables.items():\n",
    "    if table == \"ASSAY_bulkRNAseq\":\n",
    "        table = \"ASSAY_RNAseq\"\n",
    "\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = dfs['STUDY']\n",
    "PROTOCOL = dfs['PROTOCOL']\n",
    "SUBJECT = dfs['SUBJECT']\n",
    "SAMPLE = dfs['SAMPLE']\n",
    "DATA = dfs['DATA']\n",
    "CLINPATH = dfs['CLINPATH']\n",
    "PMDBS = dfs['PMDBS']\n",
    "ASSAY_RNAseq = dfs['ASSAY_RNAseq']\n",
    "CONDITION = dfs['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rationalize the team_dataset_id\n",
    "STUDY['team_dataset_id'] = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n",
    "STUDY['metadata_tables'] = f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_version = \"v3.0\"\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\"\n",
    "\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>RIN</th>\n",
       "      <th>molecular_source</th>\n",
       "      <th>input_cell_count</th>\n",
       "      <th>assay</th>\n",
       "      <th>sequencing_end</th>\n",
       "      <th>sequencing_length</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>9106</td>\n",
       "      <td>v1</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>100</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP_HC_1225</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>6533</td>\n",
       "      <td>v1</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>100</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SN_HC_1225</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>989</td>\n",
       "      <td>v1</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>100</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>2675</td>\n",
       "      <td>v1</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>100</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP_HC_0602</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>2954</td>\n",
       "      <td>v1</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>100</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id tissue technology omic RIN molecular_source input_cell_count  \\\n",
       "0  MFG_HC_1225  Brain       Bulk  RNA  NA        PolyA RNA             9106   \n",
       "1  HIP_HC_1225  Brain       Bulk  RNA  NA        PolyA RNA             6533   \n",
       "2   SN_HC_1225  Brain       Bulk  RNA  NA        PolyA RNA              989   \n",
       "3  MFG_HC_0602  Brain       Bulk  RNA  NA        PolyA RNA             2675   \n",
       "4  HIP_HC_0602  Brain       Bulk  RNA  NA        PolyA RNA             2954   \n",
       "\n",
       "  assay sequencing_end sequencing_length  sequencing_instrument  \n",
       "0    v1     Paired-end               100  Illumina NovaSeq 6000  \n",
       "1    v1     Paired-end               100  Illumina NovaSeq 6000  \n",
       "2    v1     Paired-end               100  Illumina NovaSeq 6000  \n",
       "3    v1     Paired-end               100  Illumina NovaSeq 6000  \n",
       "4    v1     Paired-end               100  Illumina NovaSeq 6000  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix ASSAY_RNAseq  \n",
    "ASSAY_RNAseq['technology'] = \"Bulk\"\n",
    "ASSAY_RNAseq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "ASSAY:\n",
    ">> VAI’s Genomics Core (RRID:SCR_022913; they sequenced this data for us), and I can make some inferences after looking through it. From what I can see, this was likely sequenced on the NovaSeq 6000, with a run length of 2x100bp, using stranded total RNA, with a target genome coverage of 30M reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>9977258e598d6a52130c29c71aef6925</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>fe2cf93257801227b7072a4fb7d18792</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>110ca4864cf6938faca67567bebfb6cc</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>0dcc67217e43ab53bae0d0676f9bfe8b</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGPD0009_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>a2608d0bd192333b0076d7091c1c50ea</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id replicate replicate_count repeated_sample    batch file_type  \\\n",
       "0  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "1  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "2  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "3  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "4  MFG_PD_0009      rep1               1               0  BATCH_4     fastq   \n",
       "\n",
       "                           file_name     file_description  \\\n",
       "0  MFGHC1225_S9_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "1  MFGHC1225_S9_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "2  MFGHC0602_S2_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "3  MFGHC0602_S2_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "4  MFGPD0009_S3_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "\n",
       "                           file_MD5 adjustment content header annotation  \\\n",
       "0  9977258e598d6a52130c29c71aef6925        Raw   Reads     NA         NA   \n",
       "1  fe2cf93257801227b7072a4fb7d18792        Raw   Reads     NA         NA   \n",
       "2  110ca4864cf6938faca67567bebfb6cc        Raw   Reads     NA         NA   \n",
       "3  0dcc67217e43ab53bae0d0676f9bfe8b        Raw   Reads     NA         NA   \n",
       "4  a2608d0bd192333b0076d7091c1c50ea        Raw   Reads     NA         NA   \n",
       "\n",
       "  configuration_file  \n",
       "0       NA(raw data)  \n",
       "1       NA(raw data)  \n",
       "2       NA(raw data)  \n",
       "3       NA(raw data)  \n",
       "4       NA(raw data)  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From here we should be able to infer the samples and subjects to subset to\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>DV200</th>\n",
       "      <th>suspension_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>NA</td>\n",
       "      <td>Brain</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>NA</td>\n",
       "      <td>Brain</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SN_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>Brain</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>NA</td>\n",
       "      <td>Brain</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>NA</td>\n",
       "      <td>Brain</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0  MFG_HC_1225    HC_1225            12-25      rep1               1   \n",
       "1  HIP_HC_1225    HC_1225            12-25      rep1               1   \n",
       "2   SN_HC_1225    HC_1225            12-25      rep1               1   \n",
       "3  MFG_HC_0602    HC_0602            06-02      rep1               1   \n",
       "4  HIP_HC_0602    HC_0602            06-02      rep1               1   \n",
       "\n",
       "  repeated_sample    batch organism tissue assay_type  ...  \\\n",
       "0               0  BATCH_4       NA  Brain         NA  ...   \n",
       "1               0  BATCH_9       NA  Brain         NA  ...   \n",
       "2               0  BATCH_7       NA  Brain         NA  ...   \n",
       "3               0  BATCH_4       NA  Brain         NA  ...   \n",
       "4               0  BATCH_9       NA  Brain         NA  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id disease_ontology_term_id  \\\n",
       "0                                       NA             PATO:0000461   \n",
       "1                                       NA             PATO:0000461   \n",
       "2                                       NA             PATO:0000461   \n",
       "3                                       NA             PATO:0000461   \n",
       "4                                       NA             PATO:0000461   \n",
       "\n",
       "  tissue_ontology_term_id assay_ontology_term_id donor_id pm_PH  \\\n",
       "0          UBERON:0002702            EFO:0030004       NA    NA   \n",
       "1          UBERON:0002702            EFO:0030004       NA    NA   \n",
       "2          UBERON:0002702            EFO:0030004       NA    NA   \n",
       "3          UBERON:0002702            EFO:0030004       NA    NA   \n",
       "4          UBERON:0002702            EFO:0030004       NA    NA   \n",
       "\n",
       "  cell_type_ontology_term_id source_RIN DV200 suspension_type  \n",
       "0               NA(multiple)         NA    NA         nucleus  \n",
       "1               NA(multiple)         NA    NA         nucleus  \n",
       "2               NA(multiple)         NA    NA         nucleus  \n",
       "3               NA(multiple)         NA    NA         nucleus  \n",
       "4               NA(multiple)         NA    NA         nucleus  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>80.0</td>\n",
       "      <td>White</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>84.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Other neurological disorder</td>\n",
       "      <td>Mild Cognitive Impairment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>82.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>87.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id                          biobank_name   sex  \\\n",
       "0    HC_1225             12-25  Banner Sun Health Research Institute  Male   \n",
       "1    HC_0602             06-02  Banner Sun Health Research Institute  Male   \n",
       "2    PD_0009             00-09  Banner Sun Health Research Institute  Male   \n",
       "3    PD_1921             19-21  Banner Sun Health Research Institute  Male   \n",
       "4    PD_2058             20-58  Banner Sun Health Research Institute  Male   \n",
       "\n",
       "   age_at_collection   race                      primary_diagnosis  \\\n",
       "0               80.0  White  No PD nor other neurological disorder   \n",
       "1               84.0  White            Other neurological disorder   \n",
       "2               64.0  White                          Idiopathic PD   \n",
       "3               82.0  White                          Idiopathic PD   \n",
       "4               87.0  White                          Idiopathic PD   \n",
       "\n",
       "      primary_diagnosis_text  \n",
       "0                         NA  \n",
       "1  Mild Cognitive Impairment  \n",
       "2                         NA  \n",
       "3                         NA  \n",
       "4                         NA  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### infer the DATA and subject / sample from data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/lee-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-lee@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-data-team-lee/fastqs/bulk_MFG/*.gz\n"
     ]
    }
   ],
   "source": [
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "prefix = \"fastqs/bulk_MFG/*.gz\"\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>e2be134da750bc4cd0672dee9b507965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0602HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>653a5622dcc2b0e8aa3b43a5ef364deb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0602HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>39c58383dbad4b68645a0f601046e0ef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1225HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>cff31843f6024c7d1440378b3cbbfadd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1225HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>47fc7147767b66c6049c5e647495c49a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1308HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c57154c6ae4921eb67cb577cac60b216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1308HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>ef773e48069d43944a4b5dd94984d910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1312PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>8e8df75e784d2593bdec8d8d201f4ca6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1312PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>d0b026f2c29687552ba7918b51b26f24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1317PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>75f1cdf6ec725e841b8888dab186a9cc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1317PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>f4c37fb09238c0726807d7283d55d6ff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1344PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>75ff20098927202bf4df0df729420c25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1344PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>a0b3b195f8bc3b4636c0124c402d9372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1441PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>cdfb977580e07dec44c71effeb69c8e2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1441PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>a3f18b3207f92e01a5edd1b6f81cba7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1504PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>1bc1ca4f9322b07dfd908857c5682787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1504PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>9174ccfe1d55d0f6a6593ec34aa9e6f2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1858PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>9793c72859f7050c8c3b90e3d2ca9486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1858PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>b63adaf065139ebd5b4347d506046b0c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1862HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>5de54ea1089a08c35afc37c23dd2de88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1862HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>899cf283f0741d6f938db46a682c0f49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1864HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>8dab92f72aa77de35063f787445fa331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1864HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>e49bf32ebe17fddfb8875d1a7481d22c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1902PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>965c52eca79bdefdcf8405ac94c75601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1902PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>1bdebaae2687bcc501a9004b0b1be12e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1921PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>d1280dfafe0fcbac8802dfca641478b5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1921PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>539d6950e467a58ff47430bcdc2279b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1939HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>33353a7e2ee37d37accaa15d2d3cc85b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1939HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>7e0d665bb49cf1be466bfe864474d3f0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1973PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>5f7f7433986639d2a8bdf3e6bd11e04c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1973PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>a932f0df075ca9715c1ea7ed101b337a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2005PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>79d23db52992c3ba461bf6ea88500e21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>2005PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>199f85f9f4ea91b5095aa3d09d21a65b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>2038PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>3450e5f65c2557a24384193c458559c3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>2038PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>a915f747c1a567514b6c23945ef2408a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>2057HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>7caf225b976212493d28d0cd99ce1781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>2057HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>342cc4805abbaa85584daa4fbbfe8692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>2058PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>64109b483d579f8c65d783b04ecde664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>2058PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>0abb82aba78b894398b2423b3be90d77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>2061HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a751254028b7fe8fbaac1c070037bca9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>2061HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>359f7b2a4793900ac846a1b9a03ab6e3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>2062HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c2bb9a2b4f0007eaa8b7a8d0231c900a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>2062HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>b760ded55c542d83f59ba647a214c621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>2067HC_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>5043e22cefc8992305695d1979076175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>2067HC_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>6bec3861fc68e9e3d585560ea4186ee6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                filename                               md5\n",
       "0   0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854\n",
       "1   0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993\n",
       "2   0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203\n",
       "3   0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d\n",
       "4   0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0\n",
       "5   0413PD_MFG_bulk_L000_R2_001.fastq.gz  e2be134da750bc4cd0672dee9b507965\n",
       "6   0602HC_MFG_bulk_L000_R1_001.fastq.gz  653a5622dcc2b0e8aa3b43a5ef364deb\n",
       "7   0602HC_MFG_bulk_L000_R2_001.fastq.gz  39c58383dbad4b68645a0f601046e0ef\n",
       "8   1225HC_MFG_bulk_L000_R1_001.fastq.gz  cff31843f6024c7d1440378b3cbbfadd\n",
       "9   1225HC_MFG_bulk_L000_R2_001.fastq.gz  47fc7147767b66c6049c5e647495c49a\n",
       "10  1308HC_MFG_bulk_L000_R1_001.fastq.gz  c57154c6ae4921eb67cb577cac60b216\n",
       "11  1308HC_MFG_bulk_L000_R2_001.fastq.gz  ef773e48069d43944a4b5dd94984d910\n",
       "12  1312PD_MFG_bulk_L000_R1_001.fastq.gz  8e8df75e784d2593bdec8d8d201f4ca6\n",
       "13  1312PD_MFG_bulk_L000_R2_001.fastq.gz  d0b026f2c29687552ba7918b51b26f24\n",
       "14  1317PD_MFG_bulk_L000_R1_001.fastq.gz  75f1cdf6ec725e841b8888dab186a9cc\n",
       "15  1317PD_MFG_bulk_L000_R2_001.fastq.gz  f4c37fb09238c0726807d7283d55d6ff\n",
       "16  1344PD_MFG_bulk_L000_R1_001.fastq.gz  75ff20098927202bf4df0df729420c25\n",
       "17  1344PD_MFG_bulk_L000_R2_001.fastq.gz  a0b3b195f8bc3b4636c0124c402d9372\n",
       "18  1441PD_MFG_bulk_L000_R1_001.fastq.gz  cdfb977580e07dec44c71effeb69c8e2\n",
       "19  1441PD_MFG_bulk_L000_R2_001.fastq.gz  a3f18b3207f92e01a5edd1b6f81cba7a\n",
       "20  1504PD_MFG_bulk_L000_R1_001.fastq.gz  1bc1ca4f9322b07dfd908857c5682787\n",
       "21  1504PD_MFG_bulk_L000_R2_001.fastq.gz  9174ccfe1d55d0f6a6593ec34aa9e6f2\n",
       "22  1858PD_MFG_bulk_L000_R1_001.fastq.gz  9793c72859f7050c8c3b90e3d2ca9486\n",
       "23  1858PD_MFG_bulk_L000_R2_001.fastq.gz  b63adaf065139ebd5b4347d506046b0c\n",
       "24  1862HC_MFG_bulk_L000_R1_001.fastq.gz  5de54ea1089a08c35afc37c23dd2de88\n",
       "25  1862HC_MFG_bulk_L000_R2_001.fastq.gz  899cf283f0741d6f938db46a682c0f49\n",
       "26  1864HC_MFG_bulk_L000_R1_001.fastq.gz  8dab92f72aa77de35063f787445fa331\n",
       "27  1864HC_MFG_bulk_L000_R2_001.fastq.gz  e49bf32ebe17fddfb8875d1a7481d22c\n",
       "28  1902PD_MFG_bulk_L000_R1_001.fastq.gz  965c52eca79bdefdcf8405ac94c75601\n",
       "29  1902PD_MFG_bulk_L000_R2_001.fastq.gz  1bdebaae2687bcc501a9004b0b1be12e\n",
       "30  1921PD_MFG_bulk_L000_R1_001.fastq.gz  d1280dfafe0fcbac8802dfca641478b5\n",
       "31  1921PD_MFG_bulk_L000_R2_001.fastq.gz  539d6950e467a58ff47430bcdc2279b1\n",
       "32  1939HC_MFG_bulk_L000_R1_001.fastq.gz  33353a7e2ee37d37accaa15d2d3cc85b\n",
       "33  1939HC_MFG_bulk_L000_R2_001.fastq.gz  7e0d665bb49cf1be466bfe864474d3f0\n",
       "34  1973PD_MFG_bulk_L000_R1_001.fastq.gz  5f7f7433986639d2a8bdf3e6bd11e04c\n",
       "35  1973PD_MFG_bulk_L000_R2_001.fastq.gz  a932f0df075ca9715c1ea7ed101b337a\n",
       "36  2005PD_MFG_bulk_L000_R1_001.fastq.gz  79d23db52992c3ba461bf6ea88500e21\n",
       "37  2005PD_MFG_bulk_L000_R2_001.fastq.gz  199f85f9f4ea91b5095aa3d09d21a65b\n",
       "38  2038PD_MFG_bulk_L000_R1_001.fastq.gz  3450e5f65c2557a24384193c458559c3\n",
       "39  2038PD_MFG_bulk_L000_R2_001.fastq.gz  a915f747c1a567514b6c23945ef2408a\n",
       "40  2057HC_MFG_bulk_L000_R1_001.fastq.gz  7caf225b976212493d28d0cd99ce1781\n",
       "41  2057HC_MFG_bulk_L000_R2_001.fastq.gz  342cc4805abbaa85584daa4fbbfe8692\n",
       "42  2058PD_MFG_bulk_L000_R1_001.fastq.gz  64109b483d579f8c65d783b04ecde664\n",
       "43  2058PD_MFG_bulk_L000_R2_001.fastq.gz  0abb82aba78b894398b2423b3be90d77\n",
       "44  2061HC_MFG_bulk_L000_R1_001.fastq.gz  a751254028b7fe8fbaac1c070037bca9\n",
       "45  2061HC_MFG_bulk_L000_R2_001.fastq.gz  359f7b2a4793900ac846a1b9a03ab6e3\n",
       "46  2062HC_MFG_bulk_L000_R1_001.fastq.gz  c2bb9a2b4f0007eaa8b7a8d0231c900a\n",
       "47  2062HC_MFG_bulk_L000_R2_001.fastq.gz  b760ded55c542d83f59ba647a214c621\n",
       "48  2067HC_MFG_bulk_L000_R1_001.fastq.gz  5043e22cefc8992305695d1979076175\n",
       "49  2067HC_MFG_bulk_L000_R2_001.fastq.gz  6bec3861fc68e9e3d585560ea4186ee6"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(bucket_files_md5.items(), columns=[\"filename\", \"md5\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['id'] = df['filename'].str.split(\"_\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "      <td>0413PD</td>\n",
       "      <td>PD_0413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               filename                               md5  \\\n",
       "0  0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854   \n",
       "1  0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993   \n",
       "2  0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203   \n",
       "3  0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d   \n",
       "4  0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0   \n",
       "\n",
       "       id subject_id  \n",
       "0  0009PD    PD_0009  \n",
       "1  0009PD    PD_0009  \n",
       "2  0348PD    PD_0348  \n",
       "3  0348PD    PD_0348  \n",
       "4  0413PD    PD_0413  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['subject_id'] = df['id'].apply(lambda x: f\"{x[-2:]}_{x[:-2]}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "      <td>0413PD</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>MFG_PD_0413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               filename                               md5  \\\n",
       "0  0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854   \n",
       "1  0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993   \n",
       "2  0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203   \n",
       "3  0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d   \n",
       "4  0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0   \n",
       "\n",
       "       id subject_id    sample_id  \n",
       "0  0009PD    PD_0009  MFG_PD_0009  \n",
       "1  0009PD    PD_0009  MFG_PD_0009  \n",
       "2  0348PD    PD_0348  MFG_PD_0348  \n",
       "3  0348PD    PD_0348  MFG_PD_0348  \n",
       "4  0413PD    PD_0413  MFG_PD_0413  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['sample_id'] = \"MFG_\" + df['subject_id']\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['lane'] = df['filename'].str.split(\"_\").str[3]\n",
    "df['end'] = df['filename'].str.split(\"_\").str[4].str.rstrip(\".fastq.gz\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "BULK_FILES = df.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>L000</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>L000</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "      <td>L000</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "      <td>L000</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "      <td>0413PD</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>MFG_PD_0413</td>\n",
       "      <td>L000</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               filename                               md5  \\\n",
       "0  0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854   \n",
       "1  0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993   \n",
       "2  0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203   \n",
       "3  0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d   \n",
       "4  0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0   \n",
       "\n",
       "       id subject_id    sample_id  lane end  \n",
       "0  0009PD    PD_0009  MFG_PD_0009  L000  R1  \n",
       "1  0009PD    PD_0009  MFG_PD_0009  L000  R2  \n",
       "2  0348PD    PD_0348  MFG_PD_0348  L000  R1  \n",
       "3  0348PD    PD_0348  MFG_PD_0348  L000  R2  \n",
       "4  0413PD    PD_0413  MFG_PD_0413  L000  R1  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BULK_FILES.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "bulk_samples = set(BULK_FILES['sample_id'].unique())\n",
    "bulk_subjects = set(BULK_FILES['subject_id'].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>80.0</td>\n",
       "      <td>White</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>84.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Other neurological disorder</td>\n",
       "      <td>Mild Cognitive Impairment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>64.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>82.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Male</td>\n",
       "      <td>87.0</td>\n",
       "      <td>White</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id                          biobank_name   sex  \\\n",
       "0    HC_1225             12-25  Banner Sun Health Research Institute  Male   \n",
       "1    HC_0602             06-02  Banner Sun Health Research Institute  Male   \n",
       "2    PD_0009             00-09  Banner Sun Health Research Institute  Male   \n",
       "3    PD_1921             19-21  Banner Sun Health Research Institute  Male   \n",
       "4    PD_2058             20-58  Banner Sun Health Research Institute  Male   \n",
       "\n",
       "   age_at_collection   race                      primary_diagnosis  \\\n",
       "0               80.0  White  No PD nor other neurological disorder   \n",
       "1               84.0  White            Other neurological disorder   \n",
       "2               64.0  White                          Idiopathic PD   \n",
       "3               82.0  White                          Idiopathic PD   \n",
       "4               87.0  White                          Idiopathic PD   \n",
       "\n",
       "      primary_diagnosis_text  \n",
       "0                         NA  \n",
       "1  Mild Cognitive Impairment  \n",
       "2                         NA  \n",
       "3                         NA  \n",
       "4                         NA  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SUBJECT subset by subject\n",
    "SUBJECT_ = SUBJECT[SUBJECT['subject_id'].isin(bulk_subjects)].reset_index(drop=True)\n",
    "SUBJECT_.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# CLINPATH subset by subject\n",
    "CLINPATH_ = CLINPATH[CLINPATH['subject_id'].isin(bulk_subjects)].reset_index(drop=True)\n",
    "\n",
    "\n",
    "# PMDDBS subset by sample\n",
    "PMDBS_ = PMDBS[PMDBS['sample_id'].isin(bulk_samples)].reset_index(drop=True)\n",
    "\n",
    "# SAMPLE subset by sample\n",
    "SAMPLE_ = SAMPLE[SAMPLE['sample_id'].isin(bulk_samples)].reset_index(drop=True)\n",
    "\n",
    "# ASSAY_RNAseq subset by sample\n",
    "ASSAY_RNAseq_ = ASSAY_RNAseq[ASSAY_RNAseq['sample_id'].isin(bulk_samples)].reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MFG_HC_0602',\n",
       " 'MFG_HC_1225',\n",
       " 'MFG_HC_1308',\n",
       " 'MFG_HC_1862',\n",
       " 'MFG_HC_1864',\n",
       " 'MFG_HC_1939',\n",
       " 'MFG_HC_2057',\n",
       " 'MFG_HC_2061',\n",
       " 'MFG_HC_2062',\n",
       " 'MFG_HC_2067',\n",
       " 'MFG_PD_0009',\n",
       " 'MFG_PD_0348',\n",
       " 'MFG_PD_0413',\n",
       " 'MFG_PD_1312',\n",
       " 'MFG_PD_1317',\n",
       " 'MFG_PD_1344',\n",
       " 'MFG_PD_1441',\n",
       " 'MFG_PD_1504',\n",
       " 'MFG_PD_1858',\n",
       " 'MFG_PD_1902',\n",
       " 'MFG_PD_1921',\n",
       " 'MFG_PD_1973',\n",
       " 'MFG_PD_2005',\n",
       " 'MFG_PD_2038',\n",
       " 'MFG_PD_2058'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>9977258e598d6a52130c29c71aef6925</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>fe2cf93257801227b7072a4fb7d18792</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>110ca4864cf6938faca67567bebfb6cc</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>0dcc67217e43ab53bae0d0676f9bfe8b</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGPD0009_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>a2608d0bd192333b0076d7091c1c50ea</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id replicate replicate_count repeated_sample    batch file_type  \\\n",
       "0  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "1  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "2  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "3  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "4  MFG_PD_0009      rep1               1               0  BATCH_4     fastq   \n",
       "\n",
       "                           file_name     file_description  \\\n",
       "0  MFGHC1225_S9_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "1  MFGHC1225_S9_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "2  MFGHC0602_S2_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "3  MFGHC0602_S2_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "4  MFGPD0009_S3_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "\n",
       "                           file_MD5 adjustment content header annotation  \\\n",
       "0  9977258e598d6a52130c29c71aef6925        Raw   Reads     NA         NA   \n",
       "1  fe2cf93257801227b7072a4fb7d18792        Raw   Reads     NA         NA   \n",
       "2  110ca4864cf6938faca67567bebfb6cc        Raw   Reads     NA         NA   \n",
       "3  0dcc67217e43ab53bae0d0676f9bfe8b        Raw   Reads     NA         NA   \n",
       "4  a2608d0bd192333b0076d7091c1c50ea        Raw   Reads     NA         NA   \n",
       "\n",
       "  configuration_file  \n",
       "0       NA(raw data)  \n",
       "1       NA(raw data)  \n",
       "2       NA(raw data)  \n",
       "3       NA(raw data)  \n",
       "4       NA(raw data)  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# DATA suubset by sample\n",
    "\n",
    "DATA_ = DATA[DATA['sample_id'].isin(bulk_samples)].reset_index(drop=True)\n",
    "\n",
    "DATA_.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# now fix the files and md5\n",
    "# BULK_FILES.head()\n",
    "\n",
    "filenname_map = dict(zip(BULK_FILES['sample_id'], BULK_FILES['filename']))\n",
    "md5_map = dict(zip(BULK_FILES['sample_id'], BULK_FILES['md5']))\n",
    "\n",
    "DATA_['file_name'] = DATA_['sample_id'].map(filenname_map)\n",
    "DATA_['file_MD5'] = DATA_['sample_id'].map(md5_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'replicate', 'replicate_count', 'repeated_sample', 'batch',\n",
       "       'file_type', 'file_name', 'file_description', 'file_MD5', 'adjustment',\n",
       "       'content', 'header', 'annotation', 'configuration_file'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA_.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "define \"batch\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set batch to \"batch_x\" as we have confirmed that all samples were run in the same batch\n",
    "DATA_['batch'] = \"batch_x\"\n",
    "# set batch to batch_x\n",
    "SAMPLE_['batch'] = \"batch_x\"\n",
    "\n",
    "# # make each sample its own batch. \n",
    "# SAMPLE_['batch'] = [f\"batch_s{x}\" for x in SAMPLE_.index] \n",
    "# DATA_['batch'] = DATA_['sample_id'].map(dict(zip(SAMPLE_['sample_id'], SAMPLE_['batch'])))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            condition_id intervention_name intervention_id  \\\n",
       "0  no_pd_nor_other_neurological_disorder      Case-Control         Control   \n",
       "1            other_neurological_disorder      Case-Control           Other   \n",
       "2                          idiopathic_pd      Case-Control            Case   \n",
       "\n",
       "  protocol_id intervention_aux_table  \n",
       "0          NA                     NA  \n",
       "1          NA                     NA  \n",
       "2          NA                     NA  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pack the tables into a dictionary\n",
    "\n",
    "v3_tables = {\n",
    "    'STUDY': STUDY,\n",
    "    'PROTOCOL': PROTOCOL,\n",
    "    'SUBJECT': SUBJECT_,\n",
    "    'SAMPLE': SAMPLE_,\n",
    "    'DATA': DATA_,\n",
    "    'CLINPATH': CLINPATH_,\n",
    "    'PMDBS': PMDBS_,\n",
    "    'ASSAY_RNAseq': ASSAY_RNAseq_,\n",
    "    'CONDITION': CONDITION\n",
    "}\n",
    "CONDITION\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     no_pd_nor_other_neurological_disorder\n",
       "1               other_neurological_disorder\n",
       "2                             idiopathic_pd\n",
       "3                             idiopathic_pd\n",
       "4                             idiopathic_pd\n",
       "5                             idiopathic_pd\n",
       "6                             idiopathic_pd\n",
       "7     no_pd_nor_other_neurological_disorder\n",
       "8     no_pd_nor_other_neurological_disorder\n",
       "9     no_pd_nor_other_neurological_disorder\n",
       "10    no_pd_nor_other_neurological_disorder\n",
       "11    no_pd_nor_other_neurological_disorder\n",
       "12              other_neurological_disorder\n",
       "13    no_pd_nor_other_neurological_disorder\n",
       "14    no_pd_nor_other_neurological_disorder\n",
       "15                            idiopathic_pd\n",
       "16                            idiopathic_pd\n",
       "17                            idiopathic_pd\n",
       "18                            idiopathic_pd\n",
       "19                            idiopathic_pd\n",
       "20                            idiopathic_pd\n",
       "21                            idiopathic_pd\n",
       "22                            idiopathic_pd\n",
       "23                            idiopathic_pd\n",
       "24                            idiopathic_pd\n",
       "Name: condition_id, dtype: object"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE_['condition_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(og_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v3_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- submitter_email: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 25/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 50/50 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 50/50 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **40 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 3/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 17/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 18/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 14/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 1/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 25/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- region_level_3: 25/25 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- RIN: 25/25 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 3/3 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 3/3 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n",
    "We will skip this since we simply assumed the md5s were correct in constructing the metadata\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create metadata package\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            condition_id intervention_name intervention_id  \\\n",
       "0  no_pd_nor_other_neurological_disorder      Case-Control         Control   \n",
       "1            other_neurological_disorder      Case-Control           Other   \n",
       "2                          idiopathic_pd      Case-Control            Case   \n",
       "\n",
       "  protocol_id intervention_aux_table  \n",
       "0          NA                     NA  \n",
       "1          NA                     NA  \n",
       "2          NA                     NA  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Create metadata package\n",
    "\n",
    "export_path = root_path / f\"{team}\"\n",
    "\n",
    "create_upload_medadata_package(export_path, v3_tables)\n",
    "v3_tables['SAMPLE']['condition_id']\n",
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
