{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# Team Lee. ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "9 Oct 2024\n",
    "\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully imported\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, create_upload_medadata_package\n",
    "from checksums import get_md5_hashes, authenticate_with_service_account\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted as v1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Team Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"lee\"\n",
    "dataset_name = \"sn-rnaseq\"\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "\n",
    "og_path = metadata_path / \"og\"\n",
    "\n",
    "# v1_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'CLINPATH', 'SAMPLE']\n",
    "\n",
    "# STUDY = read_meta_table(og_path / \"STUDY.csv\")\n",
    "# PROTOCOL = read_meta_table(og_path / \"PROTOCOL.csv\")\n",
    "# SUBJECT = read_meta_table(og_path / \"SUBJECT.csv\")\n",
    "# CLINPATH = read_meta_table(og_path / \"CLINPATH.csv\")\n",
    "# SAMPLE = read_meta_table(og_path / \"SAMPLE.csv\")\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{og_path}/SUBJECT.tsv\", delimiter=\"\\t\")\n",
    "SAMPLE = pd.read_csv(f\"{og_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "CLINPATH = pd.read_csv(f\"{og_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{og_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{og_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "metadata_version = \"v1\"\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Team-Lee-Bras-Lab-Info</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>Data type</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Note</th>\n",
       "      <th>Required/Optional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name/Title</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unique and clear title.</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset name</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A Dataset name is required for each submission...</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Brief description of the goals and objectives ...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team e.g. \"Scherzer\"</td>\n",
       "      <td>Enum</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bras</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>ASAP Lab under the above team e.g. \"Dong\"</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0           Unnamed: 1  \\\n",
       "0  Is senescence a component of human PD and does...         project_name   \n",
       "1    Human snRNA-seq PD Senesence Jose Bras Team Lee      project_dataset   \n",
       "2  Characterize the neuropathological progression...  project_description   \n",
       "3                                           TEAM-LEE       ASAP_team_name   \n",
       "4                                               Bras        ASAP_lab_name   \n",
       "\n",
       "                                          Unnamed: 2 Team-Lee-Bras-Lab-Info  \\\n",
       "0                                 Project Name/Title                 String   \n",
       "1                                       Dataset name                 String   \n",
       "2  Brief description of the goals and objectives ...                 String   \n",
       "3                          ASAP Team e.g. \"Scherzer\"                   Enum   \n",
       "4          ASAP Lab under the above team e.g. \"Dong\"                 String   \n",
       "\n",
       "                                          Field  \\\n",
       "0                                           NaN   \n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]   \n",
       "4                                           NaN   \n",
       "\n",
       "                                         Description Data type  Validation  \\\n",
       "0                            Unique and clear title.  Required         NaN   \n",
       "1  A Dataset name is required for each submission...  Required         NaN   \n",
       "2                                                NaN  Required         NaN   \n",
       "3                                                NaN  Required         NaN   \n",
       "4                                                NaN  Required         NaN   \n",
       "\n",
       "   Note  Required/Optional  \n",
       "0   NaN                NaN  \n",
       "1   NaN                NaN  \n",
       "2   NaN                NaN  \n",
       "3   NaN                NaN  \n",
       "4   NaN                NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = pd.read_csv(og_path / \"STUDY.tsv\",delimiter=\"\\t\")\n",
    "STUDY.to_csv(og_path / \"STUDY_.csv\")\n",
    "STUDY = pd.read_csv(og_path / \"STUDY_.csv\")\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  \\\n",
       "1  Is senescence a component of human PD and does...   \n",
       "\n",
       "0                                  project_dataset  \\\n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "1  Characterize the neuropathological progression...       TEAM-LEE   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email  \\\n",
       "1          Bras   Jose, Bras  jose.bras@vai.org   \n",
       "\n",
       "0                                       submitter_id    submitter_name  \\\n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra   \n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI  \\\n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN   \n",
       "\n",
       "0 publication_PMID number_of_brain_samples  \\\n",
       "1              NaN                      75   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "1  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id  \\\n",
       "1  human PD and control postmortem brains       NaN                  NaN   \n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"Unnamed: 1\",\"Unnamed: 0\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "# STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>v1_20241025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  \\\n",
       "0  Is senescence a component of human PD and does...   \n",
       "\n",
       "0                                  project_dataset  \\\n",
       "0  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "0  Characterize the neuropathological progression...       TEAM-LEE   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email  \\\n",
       "0          Bras   Jose, Bras  jose.bras@vai.org   \n",
       "\n",
       "0                                  contributor_names    submitter_name  \\\n",
       "0  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra   \n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI  \\\n",
       "0  kaitlyn.westra@vai.org  ...                  NaN             NaN   \n",
       "\n",
       "0 publication_PMID number_of_brain_samples  \\\n",
       "0              NaN                      75   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "0  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id  \\\n",
       "0  human PD and control postmortem brains       NaN                  NaN   \n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "0      unsure           v1_20241025  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Testing the function with STUDY.csv and CDE.csv\n",
    "STUDY.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "# Need to rename submitter_id to contributor_names\n",
    "STUDY = STUDY.rename(columns={\"submitter_id\": \"contributor_names\"})\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "STUDY.reset_index(drop=True, inplace=True)\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect additional metadata from covar.csv .. i.e. batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "aux_metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "HIP_covar = pd.read_csv(f\"{aux_metadata_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{aux_metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "HIP_control = pd.read_csv(f\"{aux_metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{aux_metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{aux_metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{aux_metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{aux_metadata_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{aux_metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "SN_control = pd.read_csv(f\"{aux_metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hippocampus samples\n",
    "# HIP_cases[\"GROUPcv\"]=\"PD\"\n",
    "# HIP_control[\"GROUPcv\"]=\"HC\"\n",
    "\n",
    "HIP_meta = pd.concat([HIP_cases, HIP_control], axis=0, ignore_index=True)\n",
    "HIP_meta[\"GROUPcv\"]= HIP_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "HIP_meta['MERGE_ID'] = \"HIP_\" + HIP_meta['GROUPcv'] +\"_\" + HIP_meta['CaseID'].str.replace('-','')\n",
    "HIP_covar['MERGE_ID'] = HIP_covar['COUNT_ID']\n",
    "# the fastqs follow COUNT_ID insteald of SEQ_ID naming convention\n",
    "HIP_covar['SEQ_ID'] = HIP_covar['COUNT_ID']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# there's a bug in the meta table... skip for now\n",
    "HIP_TABLE = pd.merge(HIP_covar, HIP_meta, on='MERGE_ID', how='outer')\n",
    "\n",
    "# HIP_TABLE = HIP_covar\n",
    "HIP_TABLE['subdir']=\"HIP\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = HIP_TABLE[[\"MERGE_ID\",\"SEQ_ID\",\"GROUPcv\",\"subdir\",'PD']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "### medial frontal gyrus samples\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "MFG_meta[\"GROUPcv\"]= MFG_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\"\n",
    "\n",
    "\n",
    "\n",
    "# Substantia Nigra\n",
    "SN_meta = pd.concat([SN_cases, SN_control], axis=0, ignore_index=True)\n",
    "SN_meta[\"GROUPcv\"] = SN_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "SN_meta['MERGE_ID'] = \"SN_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "SN_covar['MERGE_ID'] = SN_covar['SAMPLE']\n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "SN_TABLE = pd.merge(SN_covar, SN_meta, on='MERGE_ID', how='outer')\n",
    "SN_TABLE['subdir']=\"SN\"\n",
    "\n",
    "\n",
    "### concatenate SN, MSG, and HIP tables into one 'all_samples' table\n",
    "all_samples = pd.concat([HIP_TABLE, MFG_TABLE, SN_TABLE], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL = SAMPLE.merge(all_samples, left_on='sample_id', right_on='MERGE_ID', how='left')\n",
    "\n",
    "\n",
    "SAMPLE_ALL.to_csv(og_path / \"alternate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['batch'] = SAMPLE_ALL['BATCH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "SAMPLE = prep_table(SAMPLE, CDEv1)\n",
    "# for field in string_enum_fields:\n",
    "#     if field in SAMPLE.columns:\n",
    "#         SAMPLE[field] = SAMPLE[field].astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the updated tables to the og folder with _ appended to the name\n",
    "SAMPLE.to_csv(og_path / \"SAMPLE_.csv\")\n",
    "SUBJECT.to_csv(og_path / \"SUBJECT_.csv\")\n",
    "CLINPATH.to_csv(og_path / \"CLINPATH_.csv\")\n",
    "STUDY.to_csv(og_path / \"STUDY_.csv\")\n",
    "PROTOCOL.to_csv(og_path / \"PROTOCOL_.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# reload the tables with read_meta_table\n",
    "SAMPLE = read_meta_table(og_path / \"SAMPLE_.csv\")\n",
    "SUBJECT = read_meta_table(og_path / \"SUBJECT_.csv\")\n",
    "CLINPATH = read_meta_table(og_path / \"CLINPATH_.csv\")\n",
    "STUDY = read_meta_table(og_path / \"STUDY_.csv\")\n",
    "PROTOCOL = read_meta_table(og_path / \"PROTOCOL_.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['No PD nor other neurological disorder',\n",
       "       'Other neurological disorder', 'Idiopathic PD'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['primary_diagnosis'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace('No PD nor other neurological disorder','No PD or other neurological disorder')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix file_name and file_MD5 which need to be exploded (do this last for simplicity. i.e. to keep one sample per row rather than one file per row)\n",
    "\n",
    "# Step 1: Split the values in the columns based on commas\n",
    "SAMPLE['file_name'] = SAMPLE['file_name'].str.split(',')\n",
    "SAMPLE['file_MD5(R1,R2)'] = SAMPLE['file_MD5(R1,R2)'].str.split(',')\n",
    "\n",
    "# Step 2: Explode both 'file_name' and 'file_MD5(R1,R2)' columns together\n",
    "SAMPLE = SAMPLE.explode(['file_name', 'file_MD5(R1,R2)'])\n",
    "\n",
    "# Step 3: Rename the \"file_MD5(R1,R2)\" column to \"file_MD5\"\n",
    "SAMPLE = SAMPLE.rename(columns={\"file_MD5(R1,R2)\": \"file_MD5\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix subject id in SAMPLE\n",
    "\n",
    "samp_id_mapper = dict(zip(SUBJECT['source_subject_id'], SUBJECT['subject_id']))\n",
    "SAMPLE['subject_id'] = SAMPLE['subject_id'].map(samp_id_mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate replicate_count  \\\n",
       "0  MFG_HC_1225            12-25    HC_1225      rep1               1   \n",
       "0  MFG_HC_1225            12-25    HC_1225      rep1               1   \n",
       "1  MFG_HC_0602            06-02    HC_0602      rep1               1   \n",
       "1  MFG_HC_0602            06-02    HC_0602      rep1               1   \n",
       "2  MFG_PD_0009            00-09    PD_0009      rep1               1   \n",
       "\n",
       "  repeated_sample tissue          brain_region source_RIN RIN  ...  \\\n",
       "0               0  Brain  Middle_Frontal_Gyrus         NA  NA  ...   \n",
       "0               0  Brain  Middle_Frontal_Gyrus         NA  NA  ...   \n",
       "1               0  Brain  Middle_Frontal_Gyrus         NA  NA  ...   \n",
       "1               0  Brain  Middle_Frontal_Gyrus         NA  NA  ...   \n",
       "2               0  Brain  Middle_Frontal_Gyrus         NA  NA  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id disease_ontology_term_id  \\\n",
       "0                                  Unknown             PATO:0000461   \n",
       "0                                  Unknown             PATO:0000461   \n",
       "1                                  Unknown             PATO:0000461   \n",
       "1                                  Unknown             PATO:0000461   \n",
       "2                                  Unknown            MONDO:0005180   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id  \\\n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "2          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "\n",
       "  suspension_type DV2000 pm_PH donor_id    batch  \n",
       "0         nucleus     NA    NA       NA  BATCH_4  \n",
       "0         nucleus     NA    NA       NA  BATCH_4  \n",
       "1         nucleus     NA    NA       NA  BATCH_4  \n",
       "1         nucleus     NA    NA       NA  BATCH_4  \n",
       "2         nucleus     NA    NA       NA  BATCH_4  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MFGHC1225_S9_L001_R1_001.fastq.gz',\n",
       " 'MFGHC1225_S9_L001_R2_001.fastq.gz',\n",
       " 'MFGHC0602_S2_L001_R1_001.fastq.gz',\n",
       " 'MFGHC0602_S2_L001_R2_001.fastq.gz',\n",
       " 'MFGPD0009_S3_L001_R1_001.fastq.gz',\n",
       " 'MFGPD0009_S3_L001_R2_001.fastq.gz',\n",
       " 'MFGPD1921_S9_L001_R1_001.fastq.gz',\n",
       " 'MFGPD1921_S9_L001_R2_001.fastq.gz',\n",
       " 'MFGPD2058_S5_L001_R1_001.fastq.gz',\n",
       " 'MFGPD2058_S5_L001_R2_001.fastq.gz',\n",
       " 'MFGPD1441_S9_L001_R1_001.fastq.gz',\n",
       " 'MFGPD1441_S9_L001_R2_001.fastq.gz',\n",
       " 'MFGPD1344_S7_L001_R1_001.fastq.gz',\n",
       " 'MFGPD1344_S7_L001_R2_001.fastq.gz',\n",
       " 'MFGHC1939_S15_L001_R1_001.fastq.gz',\n",
       " 'MFGHC1939_S15_L001_R2_001.fastq.gz',\n",
       " 'MFGHC1308_S8_L002_R1_001.fastq.gz',\n",
       " 'MFGHC1308_S8_L002_R2_001.fastq.gz',\n",
       " 'MFGHC1862_S5_L004_R1_001.fastq.gz',\n",
       " 'MFGHC1862_S5_L004_R2_001.fastq.gz',\n",
       " 'MFGHC1864_S1_L002_R1_001.fastq.gz',\n",
       " 'MFGHC1864_S1_L002_R2_001.fastq.gz',\n",
       " 'MFGHC2057_S7_L002_R1_001.fastq.gz',\n",
       " 'MFGHC2057_S7_L002_R2_001.fastq.gz',\n",
       " 'MFGHC2061_S7_L004_R1_001.fastq.gz',\n",
       " 'MFGHC2061_S7_L004_R2_001.fastq.gz',\n",
       " 'MFGHC2062_S2_L002_R1_001.fastq.gz',\n",
       " 'MFGHC2062_S2_L002_R2_001.fastq.gz',\n",
       " 'MFGHC2067_S6_L002_R1_001.fastq.gz',\n",
       " 'MFGHC2067_S6_L002_R2_001.fastq.gz',\n",
       " 'MFGPD0348_S10_L002_R1_001.fastq.gz',\n",
       " 'MFGPD0348_S10_L002_R2_001.fastq.gz',\n",
       " 'MFGPD0413_S3_L002_R1_001.fastq.gz',\n",
       " 'MFGPD0413_S3_L002_R2_001.fastq.gz',\n",
       " 'MFGPD1312_S9_L002_R1_001.fastq.gz',\n",
       " 'MFGPD1312_S9_L002_R2_001.fastq.gz',\n",
       " 'MFGPD1317_S5_L002_R1_001.fastq.gz',\n",
       " 'MFGPD1317_S5_L002_R2_001.fastq.gz',\n",
       " 'MFGPD1504_S9_L001_R1_001.fastq.gz',\n",
       " 'MFGPD1504_S9_L001_R2_001.fastq.gz',\n",
       " 'MFGPD1858_S4_L002_R1_001.fastq.gz',\n",
       " 'MFGPD1858_S4_L002_R2_001.fastq.gz',\n",
       " 'MFGPD1902_S1_L004_R1_001.fastq.gz',\n",
       " 'MFGPD1902_S1_L004_R2_001.fastq.gz',\n",
       " 'MFGPD1973_S2_L004_R1_001.fastq.gz',\n",
       " 'MFGPD1973_S2_L004_R2_001.fastq.gz',\n",
       " 'MFGPD2005_S9_L001_R1_001.fastq.gz',\n",
       " 'MFGPD2005_S9_L001_R2_001.fastq.gz',\n",
       " 'MFGPD2038_S4_L004_R1_001.fastq.gz',\n",
       " 'MFGPD2038_S4_L004_R2_001.fastq.gz',\n",
       " 'HIP_HC_1225_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_1225_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_0602_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_0602_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_0009_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_0009_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_1921_S11_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1921_S11_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_2058_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_2058_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_1441_S14_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1441_S14_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1344_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_1344_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_1939_S6_L001_R1_001.fastq.gz',\n",
       " 'HIP_HC_1939_S6_L001_R2_001.fastq.gz',\n",
       " 'HIP_HC_1308_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_1308_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_1862_S9_L001_R1_001.fastq.gz',\n",
       " 'HIP_HC_1862_S9_L001_R2_001.fastq.gz',\n",
       " 'HIP_HC_1864_S8_L001_R1_001.fastq.gz',\n",
       " 'HIP_HC_1864_S8_L001_R2_001.fastq.gz',\n",
       " 'HIP_HC_2057_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_2057_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_2061_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_2061_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_2062_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_HC_2062_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_HC_2067_S9_L001_R1_001.fastq.gz',\n",
       " 'HIP_HC_2067_S9_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_0348_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_0348_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_0413_S7_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_0413_S7_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1312_S9_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1312_S9_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1317_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_1317_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_1504_S9_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1504_S9_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1858_S10_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1858_S10_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1902_S12_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1902_S12_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_1973_S13_L001_R1_001.fastq.gz',\n",
       " 'HIP_PD_1973_S13_L001_R2_001.fastq.gz',\n",
       " 'HIP_PD_2005_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_2005_S0_L000_R2_001.fastq.gz',\n",
       " 'HIP_PD_2038_S0_L000_R1_001.fastq.gz',\n",
       " 'HIP_PD_2038_S0_L000_R2_001.fastq.gz',\n",
       " 'SN_1225_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1225_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_0602_HC_S9_L001_R1_001.fastq.gz',\n",
       " 'SN_0602_HC_S9_L001_R2_001.fastq.gz',\n",
       " 'SN_0009_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_0009_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1921_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1921_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_2058_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2058_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1441_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1441_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1344_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1344_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1939_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1939_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1308_HC_S1_L001_R1_001.fastq.gz',\n",
       " 'SN_1308_HC_S1_L001_R2_001.fastq.gz',\n",
       " 'SN_1862_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1862_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1864_HC_S9_L001_R1_001.fastq.gz',\n",
       " 'SN_1864_HC_S9_L001_R2_001.fastq.gz',\n",
       " 'SN_2057_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2057_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_2061_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2061_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_2062_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2062_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_2067_HC_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2067_HC_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_0348_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_0348_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_0413_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_0413_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1312_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1312_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1317_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1317_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1504_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1504_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1858_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1858_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_1902_PD_S1_L001_R1_001.fastq.gz',\n",
       " 'SN_1902_PD_S1_L001_R2_001.fastq.gz',\n",
       " 'SN_1973_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_1973_PD_S1_L000_R2_001.fastq.gz',\n",
       " 'SN_2005_PD_S1_L001_R1_001.fastq.gz',\n",
       " 'SN_2005_PD_S1_L001_R2_001.fastq.gz',\n",
       " 'SN_2038_PD_S1_L000_R1_001.fastq.gz',\n",
       " 'SN_2038_PD_S1_L000_R2_001.fastq.gz']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "SAMPLE['file_name'] = SAMPLE['file_name'].replace({'HIP_PD_0348_S0_L000_R2_001.fastq.gzgz':'HIP_PD_0348_S0_L000_R2_001.fastq.gz',\n",
    "                              'HIP_HC_1939_S6_L001_R2_001.fastq.gzgz':  'HIP_HC_1939_S6_L001_R2_001.fastq.gz'})\n",
    "SAMPLE.file_name.to_list()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['region_level_2'].unique()\n",
    "\n",
    "# change \"Hippocampus\" to \"CA1-CA4\"\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace('Hippocampus', 'CA1-CA4')\n",
    "\n",
    "# skip hx_melanoma and education level for now as there is not a \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# leave cognitive status as is, since there is no \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# potential \"path_braak_asyn\" coding \n",
    "braak_map = {'L. Olfactory Bulb-Only':\"1/2\", 'Lla. Brainstem Predominant':\"3\",\n",
    "       'Llb. Limbic Predominant':\"3/4\", 'LV. Neocortical':\"5\",\n",
    "       'Lll. Brainstem/Limbic':\"3/4\", '0. No Lewy bodies':\"0\"}\n",
    "# set to NaN for now since this is actualy path_mckeith coding\n",
    "\n",
    "CLINPATH['path_braak_asyn'] = \"\"\n",
    "\n",
    "mckeith_map = {'L. Olfactory Bulb-Only':\"Olfactory bulb only\", 'Lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'Llb. Limbic Predominant':\"Limbic (transitional)\", 'LV. Neocortical':\"Neocortical\",\n",
    "       'Lll. Brainstem/Limbic':\"Amygdala Predominant\", 'l. Olfactory Bulb-Only':\"Olfactory bulb only\", 'lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'llb. Limbic Predominant':\"Limbic (transitional)\", 'lV. Neocortical':\"Neocortical\",\n",
    "       'lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\"}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# - _*region_level_3*_:  invalid values 💩'unknown'\n",
    "#     - valid ➡️ 'Grey matter', 'White matter', 'Grey and white matter', 'Unknown', 'NA'\n",
    "# - _*path_mckeith*_:  invalid values 💩'l. Olfactory Bulb-Only', 'lla. Brainstem Predominant', 'llb. Limbic Predominant', 'lV. Neocortical', 'lll. Brainstem/Limbic'\n",
    "#     - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown', 'NA'\n",
    "# - _*path_nia_ri*_:  invalid values 💩'Criteria not met', 'Not AD', 'criteria not met'\n",
    "#     - valid ➡️ 'Low', 'Intermediate', 'High', 'None', 'NA'\n",
    "# - _*TDP43*_:  invalid values 💩'Na'\n",
    "#     - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
    "# - _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'Cerebral amyloid angiopathy, temporal and occipital lobe', 'Cerebral amyloid angiopathy, frontal lobe'\n",
    "#     - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
    "# - _*path_ad_level*_:  invalid values 💩'Microscopic changes of Alzheimer's disease, insufficient for diagnosis', 'Microscopic lesions of Alzheimer's disease, insufficient for diagnosis'\n",
    "#     - valid ➡️ 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown', 'NA'\n",
    "CLINPATH['region_level_3'] = CLINPATH['region_level_3'].replace('unknown', 'Unknown')\n",
    "# CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(braak_map)\n",
    "CLINPATH['path_nia_ri'] = CLINPATH['path_nia_ri'].replace({'Criteria not met': 'None', 'Not AD': 'None', 'criteria not met': 'None'})\n",
    "CLINPATH['TDP43'] = CLINPATH['TDP43'].replace('Na', 'NA')\n",
    "CLINPATH['amyloid_angiopathy_severity_scale'] = CLINPATH['amyloid_angiopathy_severity_scale'].replace({'Cerebral amyloid angiopathy, temporal and occipital lobe': 'Severe', 'Cerebral amyloid angiopathy, frontal lobe': 'Severe'})\n",
    "CLINPATH['path_ad_level'] = CLINPATH['path_ad_level'].replace({'Microscopic changes of Alzheimer\\'s disease, insufficient for diagnosis': 'Low level Alzheimer\\'s disease neuropathological change', 'Microscopic lesions of Alzheimer\\'s disease, insufficient for diagnosis': 'Unknown'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL_CP = SAMPLE_ALL.merge(CLINPATH, on='sample_id', how='outer')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/data/teams/lee/sn-rnaseq/metadata/v1')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1_path = metadata_path / \"v1\"\n",
    "v1_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL_CP.to_csv((v1_path / f\"auxiluary_metadata.csv\"))\n",
    "SAMPLE.to_csv((v1_path / f\"SAMPLE.csv\"))\n",
    "SUBJECT.to_csv((v1_path / f\"SUBJECT.csv\"))\n",
    "CLINPATH.to_csv((v1_path / f\"CLINPATH.csv\"))\n",
    "STUDY.to_csv((v1_path / f\"STUDY.csv\"))\n",
    "PROTOCOL.to_csv((v1_path / f\"PROTOCOL.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validate v1\n",
    "\n",
    "QC these tables and make sure they are conformant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "🚨⚠️❗ **Missing Required Fields in STUDY: submitter_email**\n",
      "🚨⚠️❗ **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "🚨⚠️❗ **Extra field in STUDY: submittor_email**\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in SAMPLE: DV200**\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_RIN: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- pm_PH: 150/150 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 150/150 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "🚨⚠️❗ **Extra field in SAMPLE: DV2000**\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in CLINPATH: dig_slide_avail, quant_path_avail**\n",
      "🚨⚠️❗ **38 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- GP2_id: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- AMPPD_id: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 9/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 51/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 54/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 72/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 72/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 66/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 66/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 42/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 69/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 3/75 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv1\n",
    "tables = CDE['Table'].unique()\n",
    "v1_path = metadata_path / \"v1\"\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(v1_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table\n",
    "    # df.to_csv(v1_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### save extras as auxillary tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved STUDY_auxiliary.csv\n",
      "Saved SAMPLE_auxiliary.csv\n"
     ]
    }
   ],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v1_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v1_path / f\"{table}.csv\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['No PD nor other neurological disorder',\n",
       "       'Other neurological disorder', 'Idiopathic PD'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['SUBJECT']['primary_diagnosis'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v1->v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n"
     ]
    }
   ],
   "source": [
    "v2_path = metadata_path / \"v2\"\n",
    "team_dataset_id = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n",
    "\n",
    "v2_tables, aux_tables = v1_to_v2(v1_path, v2_path, CDEv1, CDEv2, team_dataset_id=team_dataset_id)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- submitter_email: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **12 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- last_diagnosis: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 3/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **29 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_second_dx: 17/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_asyn: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 18/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 25/25 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 14/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 1/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 25/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- region_level_3: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_sample_id: 75/75 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 150/150 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 150/150 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "for table,df in v2_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v2->v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "f\"{v3_meta_tables}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- submitter_email: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 75/75 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 75/75 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 150/150 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 150/150 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 75/75 empty rows (OPTIONAL)\n",
      "\n",
      "\t- region_level_3: 75/75 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- RIN: 75/75 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **40 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 3/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 17/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 18/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 14/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 1/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 25/25 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 3/3 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 3/3 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>sn_rnaseq</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>unsure</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "      <td>v3.0_20241025</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name  \\\n",
       "0       TEAM-LEE          Bras   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Is senescence a component of human PD and does...       sn_rnaseq   \n",
       "\n",
       "                                   project_dataset  \\\n",
       "0  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "                                 project_description PI_full_name  \\\n",
       "0  Characterize the neuropathological progression...   Jose, Bras   \n",
       "\n",
       "            PI_email                                  contributor_names  \\\n",
       "0  jose.bras@vai.org  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...   \n",
       "\n",
       "     submitter_name  ... number_samples  \\\n",
       "0  Kaitlyn E Westra  ...             75   \n",
       "\n",
       "                                        sample_types  \\\n",
       "0  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "                         types_of_samples DUA_version  \\\n",
       "0  human PD and control postmortem brains      unsure   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "  PI_google_scholar_id preprocessing_references metadata_version_date  \\\n",
       "0                   NA             NA(raw data)         v3.0_20241025   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = v3_tables['STUDY']\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lee\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/lee-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-lee@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-data-team-lee/fastqs/**/*.gz\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "prefix = \"fastqs/**/*.gz\"\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_37810/2789596456.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_37810/2789596456.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check1'] = checksum['file_MD5']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "checksum = v3_tables['DATA'][['file_name','file_MD5']]\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "checksum['check1'] = checksum['file_MD5']\n",
    "\n",
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create metadata package\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_path = root_path / f\"{team}\"\n",
    "\n",
    "create_upload_medadata_package(export_path, v3_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     no_pd_nor_other_neurological_disorder\n",
       "1     no_pd_nor_other_neurological_disorder\n",
       "2     no_pd_nor_other_neurological_disorder\n",
       "3               other_neurological_disorder\n",
       "4               other_neurological_disorder\n",
       "                      ...                  \n",
       "70                            idiopathic_pd\n",
       "71                            idiopathic_pd\n",
       "72                            idiopathic_pd\n",
       "73                            idiopathic_pd\n",
       "74                            idiopathic_pd\n",
       "Name: condition_id, Length: 75, dtype: object"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['SAMPLE']['condition_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field                           condition_id intervention_name  \\\n",
       "0      no_pd_nor_other_neurological_disorder      Case-Control   \n",
       "1                other_neurological_disorder      Case-Control   \n",
       "2                              idiopathic_pd      Case-Control   \n",
       "\n",
       "Field intervention_id protocol_id intervention_aux_table  \n",
       "0             Control          NA                     NA  \n",
       "1               Other          NA                     NA  \n",
       "2                Case          NA                     NA  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
