{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata compilation\n",
    "\n",
    "# Team Lee (Biederer). ASAP CRN Metadata construction\n",
    "\n",
    "\n",
    "This is a bulkRNAseq dataset which was originally transfered along with the scRNAseq dataset.  The metadata should be identical, save the details of the bulkRNAseq assay compared to the sc/snRNAseq previously platformed.\n",
    "\n",
    "\n",
    "\n",
    "15 Sept 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "WARNING: You have requested checksumming but your crcmod installation isn't\n",
      "using the module's C extension, so checksumming will run very slowly. For help\n",
      "installing the extension, please see \"gsutil help crcmod\".\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons hash -h \"gs://asap-raw-data-team-lee/fastqs/bulk_MFG/*.gz\" > lee_hexhash.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helpers (eventually load these from validate.py + io_helpers.py, etc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_CDE(metadata_version:str=\"v3.0-beta\", local_path:str|bool|Path=False):\n",
    "    \"\"\"\n",
    "    Load CDE from local csv and cache it, return a dataframe and dictionary of dtypes\n",
    "    \"\"\"\n",
    "    # Construct the path to CSD.csv\n",
    "    GOOGLE_SHEET_ID = \"1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc\"\n",
    "\n",
    "    if metadata_version == \"v1\":\n",
    "        sheet_name = \"ASAP_CDE_v1\"\n",
    "    elif metadata_version == \"v2\":\n",
    "        sheet_name = \"ASAP_CDE_v2\"\n",
    "    elif metadata_version == \"v2.1\":\n",
    "        sheet_name = \"ASAP_CDE_v2.1\"\n",
    "    elif metadata_version in [\"v3.0\",\"v3.0-beta\"]:\n",
    "        sheet_name = \"ASAP_CDE_v3.0.0-beta\"\n",
    "    else:\n",
    "        sheet_name = \"ASAP_CDE_v2.1\"\n",
    "\n",
    "\n",
    "    if metadata_version in [\"v1\",\"v2\",\"v2.1\",\"v3.0-beta\"]:\n",
    "        print(f\"metadata_version: {sheet_name}\")\n",
    "    else:\n",
    "        print(f\"Unsupported metadata_version: {sheet_name}\")\n",
    "        return 0,0\n",
    "    \n",
    "    cde_url = f\"https://docs.google.com/spreadsheets/d/{GOOGLE_SHEET_ID}/gviz/tq?tqx=out:csv&sheet={sheet_name}\"\n",
    "    if local_path:\n",
    "        cde_url = Path(local_path) / f\"{sheet_name}.csv\"\n",
    "        print(cde_url)\n",
    "\n",
    "    \n",
    "    try:\n",
    "        CDE_df = pd.read_csv(cde_url)\n",
    "        read_source = \"url\" if not local_path else \"local file\"\n",
    "        print(f\"read {read_source}\")\n",
    "    except:\n",
    "        CDE_df = pd.read_csv(f\"{sheet_name}.csv\")\n",
    "        print(\"read local file\")\n",
    "\n",
    "    # drop rows with no table name (i.e. ASAP_ids)\n",
    "    CDE_df.dropna(subset=['Table'], inplace=True)\n",
    "\n",
    "    return CDE_df\n",
    "\n",
    "\n",
    "# Function to parse the file to extract MD5 and filenames\n",
    "def extract_md5_from_details(md5_file):\n",
    "    md5s = {}\n",
    "    with open(md5_file, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        current_file = None\n",
    "        for line in lines:\n",
    "            if line.startswith(\"gs://\"):\n",
    "                current_file = line.strip().rstrip(\":\")\n",
    "                current_file = current_file.split(\"/\")[-1]\n",
    "            if \"Hash (md5)\" in line:\n",
    "                md5s[current_file] = line.split(\":\")[1].strip()\n",
    "    return md5s\n",
    "\n",
    "\n",
    "# Function to parse the file to extract MD5 and filenames\n",
    "def extract_md5_from_details2(md5_file):\n",
    "    md5s = {}\n",
    "    with open(md5_file, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        current_file = None\n",
    "        for line in lines:\n",
    "            if line.startswith(\"Hashes [hex]\"):\n",
    "                current_file = line.strip().rstrip(\":\")\n",
    "                current_file = current_file.split(\"/\")[-1]\n",
    "            if \"Hash (md5)\" in line:\n",
    "                md5s[current_file] = line.split(\":\")[1].strip()\n",
    "    return md5s\n",
    "\n",
    "\n",
    "\n",
    "# Function to parse the file to extract crc32c and filenames\n",
    "def extract_crc32c_from_details2(md5_file):\n",
    "    crcs = {}\n",
    "    with open(md5_file, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        current_file = None\n",
    "        for line in lines:\n",
    "            if line.startswith(\"Hashes [hex]\"):\n",
    "                current_file = line.strip().rstrip(\":\")\n",
    "                current_file = current_file.split(\"/\")[-1]\n",
    "            if \"Hash (crc32c)\" in line:\n",
    "                crcs[current_file] = line.split(\":\")[1].strip()\n",
    "    return crcs\n",
    "\n",
    "\n",
    "\n",
    "# Function to parse the file to extract crc32c and filenames\n",
    "def extract_hashes_from_gcloudstorage(source_hash):\n",
    "\n",
    "    crcs = {}\n",
    "    md5s = {}\n",
    "\n",
    "    with open(source_hash, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        current_file = None\n",
    "        for line in lines:\n",
    "            \n",
    "            if line.startswith(\"crc32c_hash:\"):\n",
    "                curr_crc =  line.split(\":\")[1].strip()\n",
    "\n",
    "            elif line.startswith(\"md5_hash:\"):\n",
    "                curr_md5 =  line.split(\":\")[1].strip()\n",
    "\n",
    "            elif line.startswith(\"url:\"):\n",
    "                current_file = line.split(\"/\")[-1].strip()\n",
    "                crcs[current_file] = curr_crc\n",
    "                md5s[current_file] = curr_md5\n",
    "            # else:\n",
    "            #     print(f'cruff:{line.strip()}')\n",
    "\n",
    "\n",
    "    return crcs, md5s\n",
    "\n",
    "\n",
    "\n",
    "# Function to parse the file to extract crc32c and filenames\n",
    "def extract_hashes_from_gsutil(source_hash):\n",
    "\n",
    "    crcs = {}\n",
    "    md5s = {}\n",
    "\n",
    "    with open(source_hash, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        current_file = None\n",
    "        for line in lines:\n",
    "            if line.startswith(\"Hashes [hex]\"):\n",
    "                current_file = line.strip().rstrip(\":\")\n",
    "                current_file = current_file.split(\"/\")[-1]\n",
    "            if \"Hash (crc32c)\" in line:\n",
    "                crcs[current_file] = line.split(\":\")[1].strip()\n",
    "            if \"Hash (md5)\" in line:\n",
    "                md5s[current_file] = line.split(\":\")[1].strip()\n",
    "\n",
    "    return crcs, md5s\n",
    "\n",
    "NULL = \"NA\"\n",
    "\n",
    "def read_file(data_file):\n",
    "    \"\"\"\n",
    "    TODO: depricate dtypes\n",
    "    \"\"\"\n",
    "    encoding = 'latin1'\n",
    "\n",
    "    print(f\"reading {data_file} txt/csv, encoding={encoding}\")\n",
    "    df = pd.read_csv(data_file, dtype=\"str\", encoding=encoding, index_col=0)        \n",
    "\n",
    "    for col in df.select_dtypes(include='object').columns:\n",
    "        df[col] = df[col].str.encode('latin1', errors='replace').str.decode('utf-8', errors='replace')\n",
    "\n",
    "    df.replace({\"\":NULL, pd.NA:NULL}, inplace=True)\n",
    "\n",
    "    return df\n",
    "    \n",
    "\n",
    "bucket_files_md5 = extract_md5_from_details2(\"lee_hexhash.log\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0009PD_MFG_bulk_L000_R1_001.fastq.gz': 'c9d64ec5b02de6b09a45dd7ac6610854',\n",
       " '0009PD_MFG_bulk_L000_R2_001.fastq.gz': '43a19a5fafd6892a0726660a560c4993',\n",
       " '0348PD_MFG_bulk_L000_R1_001.fastq.gz': '0ad336467cb40bfa3f6e7f582ce53203',\n",
       " '0348PD_MFG_bulk_L000_R2_001.fastq.gz': '8b6c00b59f2f0e08c053cd437a17584d',\n",
       " '0413PD_MFG_bulk_L000_R1_001.fastq.gz': 'a1e499ed8fd73cc2ae75835792ea23d0',\n",
       " '0413PD_MFG_bulk_L000_R2_001.fastq.gz': 'e2be134da750bc4cd0672dee9b507965',\n",
       " '0602HC_MFG_bulk_L000_R1_001.fastq.gz': '653a5622dcc2b0e8aa3b43a5ef364deb',\n",
       " '0602HC_MFG_bulk_L000_R2_001.fastq.gz': '39c58383dbad4b68645a0f601046e0ef',\n",
       " '1225HC_MFG_bulk_L000_R1_001.fastq.gz': 'cff31843f6024c7d1440378b3cbbfadd',\n",
       " '1225HC_MFG_bulk_L000_R2_001.fastq.gz': '47fc7147767b66c6049c5e647495c49a',\n",
       " '1308HC_MFG_bulk_L000_R1_001.fastq.gz': 'c57154c6ae4921eb67cb577cac60b216',\n",
       " '1308HC_MFG_bulk_L000_R2_001.fastq.gz': 'ef773e48069d43944a4b5dd94984d910',\n",
       " '1312PD_MFG_bulk_L000_R1_001.fastq.gz': '8e8df75e784d2593bdec8d8d201f4ca6',\n",
       " '1312PD_MFG_bulk_L000_R2_001.fastq.gz': 'd0b026f2c29687552ba7918b51b26f24',\n",
       " '1317PD_MFG_bulk_L000_R1_001.fastq.gz': '75f1cdf6ec725e841b8888dab186a9cc',\n",
       " '1317PD_MFG_bulk_L000_R2_001.fastq.gz': 'f4c37fb09238c0726807d7283d55d6ff',\n",
       " '1344PD_MFG_bulk_L000_R1_001.fastq.gz': '75ff20098927202bf4df0df729420c25',\n",
       " '1344PD_MFG_bulk_L000_R2_001.fastq.gz': 'a0b3b195f8bc3b4636c0124c402d9372',\n",
       " '1441PD_MFG_bulk_L000_R1_001.fastq.gz': 'cdfb977580e07dec44c71effeb69c8e2',\n",
       " '1441PD_MFG_bulk_L000_R2_001.fastq.gz': 'a3f18b3207f92e01a5edd1b6f81cba7a',\n",
       " '1504PD_MFG_bulk_L000_R1_001.fastq.gz': '1bc1ca4f9322b07dfd908857c5682787',\n",
       " '1504PD_MFG_bulk_L000_R2_001.fastq.gz': '9174ccfe1d55d0f6a6593ec34aa9e6f2',\n",
       " '1858PD_MFG_bulk_L000_R1_001.fastq.gz': '9793c72859f7050c8c3b90e3d2ca9486',\n",
       " '1858PD_MFG_bulk_L000_R2_001.fastq.gz': 'b63adaf065139ebd5b4347d506046b0c',\n",
       " '1862HC_MFG_bulk_L000_R1_001.fastq.gz': '5de54ea1089a08c35afc37c23dd2de88',\n",
       " '1862HC_MFG_bulk_L000_R2_001.fastq.gz': '899cf283f0741d6f938db46a682c0f49',\n",
       " '1864HC_MFG_bulk_L000_R1_001.fastq.gz': '8dab92f72aa77de35063f787445fa331',\n",
       " '1864HC_MFG_bulk_L000_R2_001.fastq.gz': 'e49bf32ebe17fddfb8875d1a7481d22c',\n",
       " '1902PD_MFG_bulk_L000_R1_001.fastq.gz': '965c52eca79bdefdcf8405ac94c75601',\n",
       " '1902PD_MFG_bulk_L000_R2_001.fastq.gz': '1bdebaae2687bcc501a9004b0b1be12e',\n",
       " '1921PD_MFG_bulk_L000_R1_001.fastq.gz': 'd1280dfafe0fcbac8802dfca641478b5',\n",
       " '1921PD_MFG_bulk_L000_R2_001.fastq.gz': '539d6950e467a58ff47430bcdc2279b1',\n",
       " '1939HC_MFG_bulk_L000_R1_001.fastq.gz': '33353a7e2ee37d37accaa15d2d3cc85b',\n",
       " '1939HC_MFG_bulk_L000_R2_001.fastq.gz': '7e0d665bb49cf1be466bfe864474d3f0',\n",
       " '1973PD_MFG_bulk_L000_R1_001.fastq.gz': '5f7f7433986639d2a8bdf3e6bd11e04c',\n",
       " '1973PD_MFG_bulk_L000_R2_001.fastq.gz': 'a932f0df075ca9715c1ea7ed101b337a',\n",
       " '2005PD_MFG_bulk_L000_R1_001.fastq.gz': '79d23db52992c3ba461bf6ea88500e21',\n",
       " '2005PD_MFG_bulk_L000_R2_001.fastq.gz': '199f85f9f4ea91b5095aa3d09d21a65b',\n",
       " '2038PD_MFG_bulk_L000_R1_001.fastq.gz': '3450e5f65c2557a24384193c458559c3',\n",
       " '2038PD_MFG_bulk_L000_R2_001.fastq.gz': 'a915f747c1a567514b6c23945ef2408a',\n",
       " '2057HC_MFG_bulk_L000_R1_001.fastq.gz': '7caf225b976212493d28d0cd99ce1781',\n",
       " '2057HC_MFG_bulk_L000_R2_001.fastq.gz': '342cc4805abbaa85584daa4fbbfe8692',\n",
       " '2058PD_MFG_bulk_L000_R1_001.fastq.gz': '64109b483d579f8c65d783b04ecde664',\n",
       " '2058PD_MFG_bulk_L000_R2_001.fastq.gz': '0abb82aba78b894398b2423b3be90d77',\n",
       " '2061HC_MFG_bulk_L000_R1_001.fastq.gz': 'a751254028b7fe8fbaac1c070037bca9',\n",
       " '2061HC_MFG_bulk_L000_R2_001.fastq.gz': '359f7b2a4793900ac846a1b9a03ab6e3',\n",
       " '2062HC_MFG_bulk_L000_R1_001.fastq.gz': 'c2bb9a2b4f0007eaa8b7a8d0231c900a',\n",
       " '2062HC_MFG_bulk_L000_R2_001.fastq.gz': 'b760ded55c542d83f59ba647a214c621',\n",
       " '2067HC_MFG_bulk_L000_R1_001.fastq.gz': '5043e22cefc8992305695d1979076175',\n",
       " '2067HC_MFG_bulk_L000_R2_001.fastq.gz': '6bec3861fc68e9e3d585560ea4186ee6'}"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bucket_files_md5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(bucket_files_md5.items(), columns=[\"filename\", \"md5\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['id'] = df['filename'].str.split(\"_\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "      <td>0413PD</td>\n",
       "      <td>PD_0413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               filename                               md5  \\\n",
       "0  0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854   \n",
       "1  0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993   \n",
       "2  0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203   \n",
       "3  0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d   \n",
       "4  0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0   \n",
       "\n",
       "       id subject_id  \n",
       "0  0009PD    PD_0009  \n",
       "1  0009PD    PD_0009  \n",
       "2  0348PD    PD_0348  \n",
       "3  0348PD    PD_0348  \n",
       "4  0413PD    PD_0413  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['subject_id'] = df['id'].apply(lambda x: f\"{x[-2:]}_{x[:-2]}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>c9d64ec5b02de6b09a45dd7ac6610854</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0009PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>43a19a5fafd6892a0726660a560c4993</td>\n",
       "      <td>0009PD</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>0ad336467cb40bfa3f6e7f582ce53203</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0348PD_MFG_bulk_L000_R2_001.fastq.gz</td>\n",
       "      <td>8b6c00b59f2f0e08c053cd437a17584d</td>\n",
       "      <td>0348PD</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>MFG_PD_0348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0413PD_MFG_bulk_L000_R1_001.fastq.gz</td>\n",
       "      <td>a1e499ed8fd73cc2ae75835792ea23d0</td>\n",
       "      <td>0413PD</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>MFG_PD_0413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               filename                               md5  \\\n",
       "0  0009PD_MFG_bulk_L000_R1_001.fastq.gz  c9d64ec5b02de6b09a45dd7ac6610854   \n",
       "1  0009PD_MFG_bulk_L000_R2_001.fastq.gz  43a19a5fafd6892a0726660a560c4993   \n",
       "2  0348PD_MFG_bulk_L000_R1_001.fastq.gz  0ad336467cb40bfa3f6e7f582ce53203   \n",
       "3  0348PD_MFG_bulk_L000_R2_001.fastq.gz  8b6c00b59f2f0e08c053cd437a17584d   \n",
       "4  0413PD_MFG_bulk_L000_R1_001.fastq.gz  a1e499ed8fd73cc2ae75835792ea23d0   \n",
       "\n",
       "       id subject_id    sample_id  \n",
       "0  0009PD    PD_0009  MFG_PD_0009  \n",
       "1  0009PD    PD_0009  MFG_PD_0009  \n",
       "2  0348PD    PD_0348  MFG_PD_0348  \n",
       "3  0348PD    PD_0348  MFG_PD_0348  \n",
       "4  0413PD    PD_0413  MFG_PD_0413  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['sample_id'] = \"MFG_\" + df['subject_id']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean V1 Table\n",
    "write clean metadata tables according to CDE v1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Team Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/SUBJECT.csv txt/csv, encoding=latin1\n",
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/SAMPLE.csv txt/csv, encoding=latin1\n",
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/CLINPATH.csv txt/csv, encoding=latin1\n",
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/STUDY.csv txt/csv, encoding=latin1\n",
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/PROTOCOL.csv txt/csv, encoding=latin1\n",
      "reading /Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2/DATA.csv txt/csv, encoding=latin1\n"
     ]
    }
   ],
   "source": [
    "## convert \n",
    "metadata_path = Path.home() / (\"Projects/ASAP/meta-clean\") / \"clean/team-Lee/v2\"\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\", index_col=0)\n",
    "SUBJECT = read_file(f\"{metadata_path}/SUBJECT.csv\")\n",
    "# SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\", index_col=0)\n",
    "SAMPLE = read_file(f\"{metadata_path}/SAMPLE.csv\")\n",
    "# CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\", index_col=0)\n",
    "CLINPATH = read_file(f\"{metadata_path}/CLINPATH.csv\")\n",
    "\n",
    "# STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\", index_col=0)\n",
    "STUDY = read_file(f\"{metadata_path}/STUDY.csv\")\n",
    "# PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\", index_col=0)\n",
    "PROTOCOL = read_file(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "# DATA = pd.read_csv(f\"{metadata_path}/DATA.csv\", index_col=0)\n",
    "DATA = read_file(f\"{metadata_path}/DATA.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bool(Path.cwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v3.0.0-beta\n",
      "/Users/ergonyc/Projects/ASAP/harmonized-wf-dev/data/teams/lee/ASAP_CDE_v3.0.0-beta.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "metadata_version = \"v3.0-beta\"\n",
    "CDE_df = read_CDE(metadata_version, local_path=Path.cwd())\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "      <th>V0</th>\n",
       "      <th>comment</th>\n",
       "      <th>denormalized</th>\n",
       "      <th>dataset relavent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>team_dataset_id</td>\n",
       "      <td>The \"project_name\" is often too verbose for pr...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table            Field                                        Description  \\\n",
       "2  STUDY   ASAP_team_name  ASAP Team Name:   Name of the ASAP CRN Team. i...   \n",
       "3  STUDY    ASAP_lab_name  Lab Name. :   Lab name that is submitting data...   \n",
       "4  STUDY     project_name  Project Name:   A Title of the overall project...   \n",
       "5  STUDY  team_dataset_id  The \"project_name\" is often too verbose for pr...   \n",
       "6  STUDY  project_dataset  Dataset Name:   A unique name is required for ...   \n",
       "\n",
       "  DataType  Required                                         Validation   V0  \\\n",
       "2     Enum  Required  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...  NaN   \n",
       "3   String  Required                                                NaN  NaN   \n",
       "4   String  Required                                                NaN  NaN   \n",
       "5   String  Required                                                NaN  NaN   \n",
       "6   String  Required                                                NaN  NaN   \n",
       "\n",
       "  comment  denormalized dataset relavent  \n",
       "2     NaN           NaN              NaN  \n",
       "3     NaN           NaN              NaN  \n",
       "4     NaN           NaN              NaN  \n",
       "5     NaN           NaN              NaN  \n",
       "6     NaN           NaN              NaN  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>9977258e598d6a52130c29c71aef6925</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC1225_S9_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>fe2cf93257801227b7072a4fb7d18792</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>110ca4864cf6938faca67567bebfb6cc</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGHC0602_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>0dcc67217e43ab53bae0d0676f9bfe8b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>fastq</td>\n",
       "      <td>MFGPD0009_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>Raw sequencing data</td>\n",
       "      <td>a2608d0bd192333b0076d7091c1c50ea</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA(raw data)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id replicate replicate_count repeated_sample    batch file_type  \\\n",
       "0  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "1  MFG_HC_1225      rep1               1               0  BATCH_4     fastq   \n",
       "2  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "3  MFG_HC_0602      rep1               1               0  BATCH_4     fastq   \n",
       "4  MFG_PD_0009      rep1               1               0  BATCH_4     fastq   \n",
       "\n",
       "                           file_name     file_description  \\\n",
       "0  MFGHC1225_S9_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "1  MFGHC1225_S9_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "2  MFGHC0602_S2_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "3  MFGHC0602_S2_L001_R2_001.fastq.gz  Raw sequencing data   \n",
       "4  MFGPD0009_S3_L001_R1_001.fastq.gz  Raw sequencing data   \n",
       "\n",
       "                           file_MD5 technology omic adjustment content time  \\\n",
       "0  9977258e598d6a52130c29c71aef6925         SN  RNA        Raw   Reads    0   \n",
       "1  fe2cf93257801227b7072a4fb7d18792         SN  RNA        Raw   Reads    0   \n",
       "2  110ca4864cf6938faca67567bebfb6cc         SN  RNA        Raw   Reads    0   \n",
       "3  0dcc67217e43ab53bae0d0676f9bfe8b         SN  RNA        Raw   Reads    0   \n",
       "4  a2608d0bd192333b0076d7091c1c50ea         SN  RNA        Raw   Reads    0   \n",
       "\n",
       "  header annotation configuration_file  \n",
       "0     NA         NA       NA(raw data)  \n",
       "1     NA         NA       NA(raw data)  \n",
       "2     NA         NA       NA(raw data)  \n",
       "3     NA         NA       NA(raw data)  \n",
       "4     NA         NA       NA(raw data)  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SN_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_7</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia_Nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0  MFG_HC_1225    HC_1225            12-25      rep1               1   \n",
       "1  HIP_HC_1225    HC_1225            12-25      rep1               1   \n",
       "2   SN_HC_1225    HC_1225            12-25      rep1               1   \n",
       "3  MFG_HC_0602    HC_0602            06-02      rep1               1   \n",
       "4  HIP_HC_0602    HC_0602            06-02      rep1               1   \n",
       "\n",
       "  repeated_sample    batch tissue          brain_region hemisphere  ...  \\\n",
       "0               0  BATCH_4  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "1               0  BATCH_9  Brain           Hippocampus    Unknown  ...   \n",
       "2               0  BATCH_7  Brain     Substantia_Nigra     Unknown  ...   \n",
       "3               0  BATCH_4  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "4               0  BATCH_9  Brain           Hippocampus    Unknown  ...   \n",
       "\n",
       "  sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0  PATO:0000384 (male)                                  Unknown   \n",
       "1  PATO:0000384 (male)                                  Unknown   \n",
       "2  PATO:0000384 (male)                                  Unknown   \n",
       "3  PATO:0000384 (male)                                  Unknown   \n",
       "4  PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0             PATO:0000461          UBERON:0002702               NA(multiple)   \n",
       "1             PATO:0000461          UBERON:0002702               NA(multiple)   \n",
       "2             PATO:0000461          UBERON:0002702               NA(multiple)   \n",
       "3             PATO:0000461          UBERON:0002702               NA(multiple)   \n",
       "4             PATO:0000461          UBERON:0002702               NA(multiple)   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV200 pm_PH donor_id  \n",
       "0            EFO:0030004         nucleus    NA    NA       NA  \n",
       "1            EFO:0030004         nucleus    NA    NA       NA  \n",
       "2            EFO:0030004         nucleus    NA    NA       NA  \n",
       "3            EFO:0030004         nucleus    NA    NA       NA  \n",
       "4            EFO:0030004         nucleus    NA    NA       NA  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge df and DATA on sample_id\n",
    "DATA_ = df.merge(DATA, left_on=\"sample_id\", right_on=\"sample_id\", how=\"left\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((50, 5), (150, 17), (50, 21))"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicates on md5\n",
    "DATA_.drop_duplicates(subset=\"md5\", inplace=True)\n",
    "\n",
    "df.shape, DATA.shape, DATA_.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['filename', 'md5', 'id', 'subject_id', 'sample_id', 'replicate',\n",
       "       'replicate_count', 'repeated_sample', 'batch', 'file_type', 'file_name',\n",
       "       'file_description', 'file_MD5', 'technology', 'omic', 'adjustment',\n",
       "       'content', 'time', 'header', 'annotation', 'configuration_file'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA_.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "copy 'filename' -> 'file_name'\n",
    "\n",
    "copy  'md5' -> 'file_md5'\n",
    "\n",
    "change 'technology' -> 'bulk'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_['file_MD5'] = DATA_['md5']\n",
    "DATA_['file_name'] = DATA_['filename']\n",
    "DATA_['technology'] = 'bulk'\n",
    "\n",
    "# set batch to \"batch_x\"\n",
    "DATA_['batch'] = \"batch_x\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next take the unique sample_id and subset SAMPLE, and CLINPATH to only include the bulk samples.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['MFG_PD_0009', 'MFG_PD_0348', 'MFG_PD_0413', 'MFG_HC_0602',\n",
       "        'MFG_HC_1225', 'MFG_HC_1308', 'MFG_PD_1312', 'MFG_PD_1317',\n",
       "        'MFG_PD_1344', 'MFG_PD_1441', 'MFG_PD_1504', 'MFG_PD_1858',\n",
       "        'MFG_HC_1862', 'MFG_HC_1864', 'MFG_PD_1902', 'MFG_PD_1921',\n",
       "        'MFG_HC_1939', 'MFG_PD_1973', 'MFG_PD_2005', 'MFG_PD_2038',\n",
       "        'MFG_HC_2057', 'MFG_PD_2058', 'MFG_HC_2061', 'MFG_HC_2062',\n",
       "        'MFG_HC_2067'], dtype=object),\n",
       " array(['PD_0009', 'PD_0348', 'PD_0413', 'HC_0602', 'HC_1225', 'HC_1308',\n",
       "        'PD_1312', 'PD_1317', 'PD_1344', 'PD_1441', 'PD_1504', 'PD_1858',\n",
       "        'HC_1862', 'HC_1864', 'PD_1902', 'PD_1921', 'HC_1939', 'PD_1973',\n",
       "        'PD_2005', 'PD_2038', 'HC_2057', 'PD_2058', 'HC_2061', 'HC_2062',\n",
       "        'HC_2067'], dtype=object))"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_ids = DATA_['sample_id'].unique()\n",
    "subject_ids = DATA_['subject_id'].unique()\n",
    "sample_ids, subject_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ = SAMPLE[SAMPLE['sample_id'].isin(sample_ids)].copy()\n",
    "CLINPATH_ = CLINPATH[CLINPATH['subject_id'].isin(subject_ids)].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT_ = SUBJECT[SUBJECT['subject_id'].isin(subject_ids)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reindex SAMPLE_ and CLINPATH_\n",
    "SAMPLE_.reset_index(inplace=True, drop=True)\n",
    "CLINPATH_.reset_index(inplace=True, drop=True)\n",
    "\n",
    "\n",
    "# set batch to batch_x\n",
    "SAMPLE_['batch'] = \"batch_x\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MFG_PD_1441</td>\n",
       "      <td>PD_1441</td>\n",
       "      <td>14-41</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MFG_PD_1344</td>\n",
       "      <td>PD_1344</td>\n",
       "      <td>13-44</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MFG_HC_1939</td>\n",
       "      <td>HC_1939</td>\n",
       "      <td>19-39</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MFG_HC_1308</td>\n",
       "      <td>HC_1308</td>\n",
       "      <td>13-08</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MFG_HC_1862</td>\n",
       "      <td>HC_1862</td>\n",
       "      <td>18-62</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MFG_HC_1864</td>\n",
       "      <td>HC_1864</td>\n",
       "      <td>18-64</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MFG_HC_2057</td>\n",
       "      <td>HC_2057</td>\n",
       "      <td>20-57</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MFG_HC_2061</td>\n",
       "      <td>HC_2061</td>\n",
       "      <td>20-61</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MFG_HC_2062</td>\n",
       "      <td>HC_2062</td>\n",
       "      <td>20-62</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MFG_HC_2067</td>\n",
       "      <td>HC_2067</td>\n",
       "      <td>20-67</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MFG_PD_0348</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>03-48</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MFG_PD_0413</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>04-13</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MFG_PD_1312</td>\n",
       "      <td>PD_1312</td>\n",
       "      <td>13-12</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFG_PD_1317</td>\n",
       "      <td>PD_1317</td>\n",
       "      <td>13-17</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MFG_PD_1504</td>\n",
       "      <td>PD_1504</td>\n",
       "      <td>15-04</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MFG_PD_1858</td>\n",
       "      <td>PD_1858</td>\n",
       "      <td>18-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MFG_PD_1902</td>\n",
       "      <td>PD_1902</td>\n",
       "      <td>19-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MFG_PD_1973</td>\n",
       "      <td>PD_1973</td>\n",
       "      <td>19-73</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MFG_PD_2005</td>\n",
       "      <td>PD_2005</td>\n",
       "      <td>20-05</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MFG_PD_2038</td>\n",
       "      <td>PD_2038</td>\n",
       "      <td>20-38</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0   MFG_HC_1225    HC_1225            12-25      rep1               1   \n",
       "1   MFG_HC_0602    HC_0602            06-02      rep1               1   \n",
       "2   MFG_PD_0009    PD_0009            00-09      rep1               1   \n",
       "3   MFG_PD_1921    PD_1921            19-21      rep1               1   \n",
       "4   MFG_PD_2058    PD_2058            20-58      rep1               1   \n",
       "5   MFG_PD_1441    PD_1441            14-41      rep1               1   \n",
       "6   MFG_PD_1344    PD_1344            13-44      rep1               1   \n",
       "7   MFG_HC_1939    HC_1939            19-39      rep1               1   \n",
       "8   MFG_HC_1308    HC_1308            13-08      rep1               1   \n",
       "9   MFG_HC_1862    HC_1862            18-62      rep1               1   \n",
       "10  MFG_HC_1864    HC_1864            18-64      rep1               1   \n",
       "11  MFG_HC_2057    HC_2057            20-57      rep1               1   \n",
       "12  MFG_HC_2061    HC_2061            20-61      rep1               1   \n",
       "13  MFG_HC_2062    HC_2062            20-62      rep1               1   \n",
       "14  MFG_HC_2067    HC_2067            20-67      rep1               1   \n",
       "15  MFG_PD_0348    PD_0348            03-48      rep1               1   \n",
       "16  MFG_PD_0413    PD_0413            04-13      rep1               1   \n",
       "17  MFG_PD_1312    PD_1312            13-12      rep1               1   \n",
       "18  MFG_PD_1317    PD_1317            13-17      rep1               1   \n",
       "19  MFG_PD_1504    PD_1504            15-04      rep1               1   \n",
       "20  MFG_PD_1858    PD_1858            18-58      rep1               1   \n",
       "21  MFG_PD_1902    PD_1902            19-02      rep1               1   \n",
       "22  MFG_PD_1973    PD_1973            19-73      rep1               1   \n",
       "23  MFG_PD_2005    PD_2005            20-05      rep1               1   \n",
       "24  MFG_PD_2038    PD_2038            20-38      rep1               1   \n",
       "\n",
       "   repeated_sample    batch tissue          brain_region hemisphere  ...  \\\n",
       "0                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "1                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "2                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "3                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "4                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "5                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "6                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "7                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "8                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "9                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "10               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "11               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "12               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "13               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "14               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "15               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "16               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "17               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "18               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "19               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "20               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "21               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "22               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "23               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "24               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "\n",
       "     sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0     PATO:0000384 (male)                                  Unknown   \n",
       "1     PATO:0000384 (male)                                  Unknown   \n",
       "2     PATO:0000384 (male)                                  Unknown   \n",
       "3     PATO:0000384 (male)                                  Unknown   \n",
       "4     PATO:0000384 (male)                                  Unknown   \n",
       "5   PATO:0000383 (female)                                  Unknown   \n",
       "6   PATO:0000383 (female)                                  Unknown   \n",
       "7   PATO:0000383 (female)                                  Unknown   \n",
       "8     PATO:0000384 (male)                                  Unknown   \n",
       "9     PATO:0000384 (male)                                  Unknown   \n",
       "10    PATO:0000384 (male)                                  Unknown   \n",
       "11    PATO:0000384 (male)                                  Unknown   \n",
       "12  PATO:0000383 (female)                                  Unknown   \n",
       "13    PATO:0000384 (male)                                  Unknown   \n",
       "14    PATO:0000384 (male)                                  Unknown   \n",
       "15    PATO:0000384 (male)                                  Unknown   \n",
       "16    PATO:0000384 (male)                                  Unknown   \n",
       "17    PATO:0000384 (male)                                  Unknown   \n",
       "18    PATO:0000384 (male)                                  Unknown   \n",
       "19  PATO:0000383 (female)                                  Unknown   \n",
       "20    PATO:0000384 (male)                                  Unknown   \n",
       "21    PATO:0000384 (male)                                  Unknown   \n",
       "22    PATO:0000384 (male)                                  Unknown   \n",
       "23    PATO:0000384 (male)                                  Unknown   \n",
       "24    PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id  \\\n",
       "0              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "2             MONDO:0005180          UBERON:0002702   \n",
       "3             MONDO:0005180          UBERON:0002702   \n",
       "4             MONDO:0005180          UBERON:0002702   \n",
       "5             MONDO:0005180          UBERON:0002702   \n",
       "6             MONDO:0005180          UBERON:0002702   \n",
       "7              PATO:0000461          UBERON:0002702   \n",
       "8              PATO:0000461          UBERON:0002702   \n",
       "9              PATO:0000461          UBERON:0002702   \n",
       "10             PATO:0000461          UBERON:0002702   \n",
       "11             PATO:0000461          UBERON:0002702   \n",
       "12             PATO:0000461          UBERON:0002702   \n",
       "13             PATO:0000461          UBERON:0002702   \n",
       "14             PATO:0000461          UBERON:0002702   \n",
       "15            MONDO:0005180          UBERON:0002702   \n",
       "16            MONDO:0005180          UBERON:0002702   \n",
       "17            MONDO:0005180          UBERON:0002702   \n",
       "18            MONDO:0005180          UBERON:0002702   \n",
       "19            MONDO:0005180          UBERON:0002702   \n",
       "20            MONDO:0005180          UBERON:0002702   \n",
       "21            MONDO:0005180          UBERON:0002702   \n",
       "22            MONDO:0005180          UBERON:0002702   \n",
       "23            MONDO:0005180          UBERON:0002702   \n",
       "24            MONDO:0005180          UBERON:0002702   \n",
       "\n",
       "   cell_type_ontology_term_id assay_ontology_term_id suspension_type DV200  \\\n",
       "0                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "1                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "2                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "3                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "4                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "5                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "6                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "7                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "8                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "9                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "10               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "11               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "12               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "13               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "14               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "15               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "16               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "17               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "18               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "19               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "20               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "21               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "22               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "23               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "24               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "\n",
       "   pm_PH donor_id  \n",
       "0     NA       NA  \n",
       "1     NA       NA  \n",
       "2     NA       NA  \n",
       "3     NA       NA  \n",
       "4     NA       NA  \n",
       "5     NA       NA  \n",
       "6     NA       NA  \n",
       "7     NA       NA  \n",
       "8     NA       NA  \n",
       "9     NA       NA  \n",
       "10    NA       NA  \n",
       "11    NA       NA  \n",
       "12    NA       NA  \n",
       "13    NA       NA  \n",
       "14    NA       NA  \n",
       "15    NA       NA  \n",
       "16    NA       NA  \n",
       "17    NA       NA  \n",
       "18    NA       NA  \n",
       "19    NA       NA  \n",
       "20    NA       NA  \n",
       "21    NA       NA  \n",
       "22    NA       NA  \n",
       "23    NA       NA  \n",
       "24    NA       NA  \n",
       "\n",
       "[25 rows x 33 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "validating sample_id\n",
      "validating replicate\n",
      "validating replicate_count\n",
      "validating repeated_sample\n",
      "validating batch\n",
      "validating file_type\n",
      "validating file_name\n",
      "validating file_description\n",
      "validating file_MD5\n",
      "validating adjustment\n",
      "validating content\n",
      "validating header\n",
      "validating annotation\n",
      "validating configuration_file\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 50/50 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 50/50 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# wrape this in try/except to make suing the ReportCollector portable\n",
    "# probably an abstract base class would be better\n",
    "\n",
    "class DummyStreamlit:\n",
    "    @staticmethod\n",
    "    def markdown(self,msg):\n",
    "        pass\n",
    "    def error(self,msg):\n",
    "        pass\n",
    "    def header(self,msg):\n",
    "        pass        \n",
    "    def subheader(self,msg):\n",
    "        pass    \n",
    "    def divider(self):\n",
    "        pass\n",
    "st = DummyStreamlit()\n",
    "\n",
    "\n",
    "def get_log(log_file):\n",
    "    \"\"\" grab logged information from the log file.\"\"\"\n",
    "    with open(log_file, 'r') as f:\n",
    "        report_content = f.read()\n",
    "    return report_content\n",
    "\n",
    "def columnize( itemlist ):\n",
    "    NEWLINE_DASH = ' \\n- '\n",
    "    if len(itemlist) > 1:\n",
    "        return f\"- {itemlist[0]}{NEWLINE_DASH.join(itemlist[1:])}\"\n",
    "    else:\n",
    "        return f\"- {itemlist[0]}\"\n",
    "    \n",
    "def read_meta_table(table_path):\n",
    "    # read the whole table\n",
    "    try:\n",
    "        table_df = pd.read_csv(table_path,dtype=str)\n",
    "    except UnicodeDecodeError:\n",
    "        table_df = pd.read_csv(table_path, encoding='latin1',dtype=str)\n",
    "\n",
    "    # drop the first column if it is just the index\n",
    "    if table_df.columns[0] == \"Unnamed: 0\":\n",
    "        table_df = table_df.drop(columns=[\"Unnamed: 0\"])\n",
    "\n",
    "    return table_df\n",
    "\n",
    "\n",
    "class ReportCollector:\n",
    "    \"\"\"\n",
    "    Class to collect and log messages, errors, and markdown to a log file and/or streamlit\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, destination=\"both\"):\n",
    "        self.entries = []\n",
    "        self.filename = None\n",
    "\n",
    "        if destination in [\"both\", \"streamlit\"]:\n",
    "            self.publish_to_streamlit = True\n",
    "        else:\n",
    "            self.publish_to_streamlit = False\n",
    "\n",
    "\n",
    "    def add_markdown(self, msg):\n",
    "        self.entries.append((\"markdown\", msg))\n",
    "        if self.publish_to_streamlit:\n",
    "            st.markdown(msg)\n",
    "\n",
    "\n",
    "    def add_error(self, msg):\n",
    "        self.entries.append((\"error\", msg))\n",
    "        if self.publish_to_streamlit:\n",
    "            st.error(msg)\n",
    "\n",
    "    def add_header(self, msg):\n",
    "        self.entries.append((\"header\", msg))\n",
    "        if self.publish_to_streamlit:    \n",
    "            st.header(msg)\n",
    "\n",
    "    def add_subheader(self, msg):\n",
    "        self.entries.append((\"subheader\", msg))\n",
    "        if self.publish_to_streamlit:    \n",
    "            st.subheader(msg)\n",
    "\n",
    "    def add_divider(self):\n",
    "        self.entries.append((\"divider\", None))\n",
    "        if self.publish_to_streamlit:    \n",
    "            st.divider()\n",
    "\n",
    "    \n",
    "    def write_to_file(self, filename):\n",
    "        self.filename = filename\n",
    "        with open(filename, 'w') as f:\n",
    "            report_content = self.get_log()\n",
    "            f.write(report_content)\n",
    "    \n",
    "\n",
    "    def get_log(self):\n",
    "        \"\"\" grab logged information from the log file.\"\"\"\n",
    "        report_content = []\n",
    "        for msg_type, msg in self.entries:\n",
    "            if msg_type == \"markdown\":\n",
    "                report_content += msg + '\\n'\n",
    "            elif msg_type == \"error\":\n",
    "                report_content += f\"🚨⚠️❗ **{msg}**\\n\"\n",
    "            elif msg_type == \"header\":\n",
    "                report_content += f\"# {msg}\\n\"\n",
    "            elif msg_type == \"subheader\":\n",
    "                report_content += f\"## {msg}\\n\"\n",
    "            elif msg_type == \"divider\":\n",
    "                report_content += 60*'-' + '\\n'\n",
    "        \n",
    "        return \"\".join(report_content)\n",
    "\n",
    "    def reset(self):\n",
    "        self.entries = []\n",
    "        self.filename = None\n",
    "\n",
    "    def print_log(self):\n",
    "        print(self.get_log())\n",
    "\n",
    "\n",
    "def validate_table(df: pd.DataFrame, table_name: str, specific_cde_df: pd.DataFrame, out: ReportCollector ):\n",
    "    \"\"\"\n",
    "    Validate the table against the specific table entries from the CDE\n",
    "    \"\"\"\n",
    "    df.replace({\"\":NULL, pd.NA:NULL}, inplace=True)\n",
    "    def my_str(x):\n",
    "        return f\"'{str(x)}'\"\n",
    "    missing_required = []\n",
    "    missing_optional = []\n",
    "    null_fields = []\n",
    "    invalid_entries = []\n",
    "    total_rows = df.shape[0]\n",
    "    df_out = pd.DataFrame()\n",
    "    for field in specific_cde_df[\"Field\"]:\n",
    "        entry_idx = specific_cde_df[\"Field\"]==field\n",
    "        print(f\"validating {field}\")\n",
    "        \n",
    "        opt_req = \"REQUIRED\" if specific_cde_df.loc[entry_idx, \"Required\"].item()==\"Required\" else \"OPTIONAL\"\n",
    "\n",
    "        if field not in df.columns:\n",
    "            if opt_req == \"REQUIRED\":\n",
    "                missing_required.append(field)\n",
    "            else:\n",
    "                missing_optional.append(field)\n",
    "\n",
    "            # print(f\"missing {opt_req} column {field}\")\n",
    "            df_out[field] = [NULL]*total_rows\n",
    "\n",
    "        else:\n",
    "            datatype = specific_cde_df.loc[entry_idx,\"DataType\"]\n",
    "            if datatype.item() == \"Integer\":\n",
    "                # recode \"Unknown\" as NULL\n",
    "                df.replace({\"Unknown\":NULL, \"unknown\":NULL}, inplace=True)\n",
    "                df[field].apply(lambda x: int(x) if x!=NULL else x )\n",
    "                # test that all are integer or NULL, flag NULL entries\n",
    "            elif datatype.item() == \"Float\":\n",
    "                # recode \"Unknown\" as NULL\n",
    "                df.replace({\"Unknown\":NULL, \"unknown\":NULL}, inplace=True)\n",
    "                df[field].apply(lambda x: float(x) if x!=NULL else x )\n",
    "                # test that all are float or NULL, flag NULL entries\n",
    "            elif datatype.item() == \"Enum\":\n",
    "\n",
    "                valid_values = eval(specific_cde_df.loc[entry_idx,\"Validation\"].item())\n",
    "                entries = df[field]\n",
    "                valid_entries = entries.apply(lambda x: x in valid_values)\n",
    "                invalid_values = entries[~valid_entries].unique()\n",
    "                n_invalid = invalid_values.shape[0]\n",
    "                if n_invalid > 0:\n",
    "                    valstr = ', '.join(map(my_str, valid_values))\n",
    "                    invalstr = ', '.join(map(my_str,invalid_values))\n",
    "                    invalid_entries.append((opt_req, field, n_invalid, valstr, invalstr))\n",
    "            else: #dtype == String\n",
    "                pass\n",
    "            \n",
    "            n_null = (df[field]==NULL).sum()\n",
    "            if n_null > 0:            \n",
    "                null_fields.append((opt_req, field, n_null))\n",
    "            \n",
    "            df_out[field] = df[field]\n",
    "            \n",
    "\n",
    "    # now compose report...\n",
    "    if len(missing_required) > 0:\n",
    "        out.add_error(f\"Missing Required Fields in {table_name}: {', '.join(missing_required)}\")\n",
    "    else:\n",
    "        out.add_markdown(f\"All required fields are present in *{table_name}* table.\")\n",
    "\n",
    "    if len(missing_optional) > 0:\n",
    "        out.add_error(f\"Missing Optional Fields in {table_name}: {', '.join(missing_optional)}\")\n",
    "    \n",
    "\n",
    "    if len(null_fields) > 0:\n",
    "        # print(f\"{opt_req} {field} has {n_null}/{df.shape[0]} NULL entries \")\n",
    "        out.add_error(f\"{len(null_fields)} Fields with empty (NULL) values:\")\n",
    "        for opt_req, field, count in null_fields:\n",
    "            out.add_markdown(f\"\\n\\t- {field}: {count}/{total_rows} empty rows ({opt_req})\")\n",
    "    else:\n",
    "        out.add_markdown(f\"No empty entries (NULL) found .\")\n",
    "\n",
    "\n",
    "    if len(invalid_entries) > 0:\n",
    "        out.add_error(f\"{len(invalid_entries)} Fields with invalid entries:\")\n",
    "        for opt_req, field, count, valstr, invalstr in invalid_entries:\n",
    "            str_out = f\"- _*{field}*_:  invalid values 💩{invalstr}\\n\"\n",
    "            str_out += f\"    - valid ➡️ {valstr}\"\n",
    "            out.add_markdown(str_out)\n",
    "    else:\n",
    "        out.add_markdown(f\"No invalid entries found in Enum fields.\")\n",
    "\n",
    "\n",
    "    return df, df_out, out\n",
    "\n",
    "\n",
    "report = ReportCollector(destination=\"log\")\n",
    "\n",
    "df = DATA_.copy()\n",
    "table_choice = \"DATA\"\n",
    "# perform the valadation\n",
    "\n",
    "# specific_cde_df = CDE_df[CDE_df['Table'] == table_choice]\n",
    "specific_cde_df = CDE_df[CDE_df['Table'].str.startswith(table_choice)]\n",
    "\n",
    "retval = validate_table(df, table_choice, specific_cde_df, report)\n",
    "\n",
    "report.print_log()\n",
    "\n",
    "DATA_export, DATA_out, report = retval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "validating sample_id\n",
      "validating subject_id\n",
      "validating source_sample_id\n",
      "validating replicate\n",
      "validating replicate_count\n",
      "validating repeated_sample\n",
      "validating batch\n",
      "validating condition\n",
      "validating tissue\n",
      "validating time\n",
      "validating alternate_id\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in SAMPLE: condition, time, alternate_id**\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"log\")\n",
    "\n",
    "df = SAMPLE_.copy()\n",
    "table_choice = \"SAMPLE\"\n",
    "# perform the valadation\n",
    "\n",
    "# specific_cde_df = CDE_df[CDE_df['Table'] == table_choice]\n",
    "specific_cde_df = CDE_df[CDE_df['Table'].str.startswith(table_choice)]\n",
    "\n",
    "retval = validate_table(df, table_choice, specific_cde_df, report)\n",
    "\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_export, SAMPLE_out, report = retval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((25, 33), (25, 33), (75, 33), (25, 11))"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE_export.shape, SAMPLE_.shape, SAMPLE.shape, SAMPLE_out.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "validating subject_id\n",
      "validating source_subject_id\n",
      "validating duration_pmi\n",
      "validating age_at_death\n",
      "validating path_autopsy_dx_main\n",
      "validating path_autopsy_second_dx\n",
      "validating path_autopsy_third_dx\n",
      "validating path_autopsy_fourth_dx\n",
      "validating path_autopsy_fifth_dx\n",
      "validating path_autopsy_sixth_dx\n",
      "validating path_autopsy_seventh_dx\n",
      "validating path_autopsy_eight_dx\n",
      "validating path_year_death\n",
      "validating cause_death\n",
      "validating other_cause_death_1\n",
      "validating other_cause_death_2\n",
      "validating brain_weight\n",
      "validating path_braak_nft\n",
      "validating path_braak_asyn\n",
      "validating path_cerad\n",
      "validating path_thal\n",
      "validating known_pathogenic_mutation\n",
      "validating PD_pathogenic_mutation\n",
      "validating path_mckeith\n",
      "validating sn_neuronal_loss\n",
      "validating path_infarcs\n",
      "validating path_nia_ri\n",
      "validating path_nia_aa_a\n",
      "validating path_nia_aa_b\n",
      "validating path_nia_aa_c\n",
      "validating TDP43\n",
      "validating arteriolosclerosis_severity_scale\n",
      "validating amyloid_angiopathy_severity_scale\n",
      "validating path_ad_level\n",
      "validating dig_slide_avail\n",
      "validating quant_path_avail\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **27 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_second_dx: 17/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 18/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 24/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 22/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 12/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 25/25 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **16 Fields with invalid entries:**\n",
      "- _*path_braak_asyn*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_cerad*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Sparse', 'Moderate', 'Frequent'\n",
      "- _*path_thal*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*known_pathogenic_mutation*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Present', 'Unknown'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_infarcs*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*path_nia_ri*_:  invalid values 💩'Criteria not met', 'Not AD'\n",
      "    - valid ➡️ 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*path_nia_aa_a*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3'\n",
      "- _*TDP43*_:  invalid values 💩'NA', 'Na'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*arteriolosclerosis_severity_scale*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'NA', 'Cerebral amyloid angiopathy, temporal and occipital lobe', 'Cerebral amyloid angiopathy, frontal lobe'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values 💩'Microscopic changes of Alzheimer's disease, insufficient for diagnosis', 'Microscopic lesions of Alzheimer's disease, insufficient for diagnosis'\n",
      "    - valid ➡️ 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "- _*dig_slide_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*quant_path_avail*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"log\")\n",
    "\n",
    "df = CLINPATH_.copy()\n",
    "table_choice = \"CLINPATH\"\n",
    "# perform the valadation\n",
    "\n",
    "# specific_cde_df = CDE_df[CDE_df['Table'] == table_choice]\n",
    "specific_cde_df = CDE_df[CDE_df['Table'].str.startswith(table_choice)]\n",
    "\n",
    "retval = validate_table(df, table_choice, specific_cde_df, report)\n",
    "\n",
    "report.print_log()\n",
    "\n",
    "CLINPATH_export, CLINPATH_out, report = retval\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "df.replace({\"\":NULL, pd.NA:NULL, np.NaN:NULL}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NA'"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NA'"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT_['age_at_onset'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "validating subject_id\n",
      "validating source_subject_id\n",
      "validating biobank_name\n",
      "validating organism\n",
      "validating sex\n",
      "validating age_at_collection\n",
      "validating race\n",
      "validating primary_diagnosis\n",
      "validating primary_diagnosis_text\n",
      "validating AMPPD_id\n",
      "validating GP2_id\n",
      "validating ethnicity\n",
      "validating family_history\n",
      "validating last_diagnosis\n",
      "validating age_at_onset\n",
      "validating age_at_diagnosis\n",
      "validating first_motor_symptom\n",
      "validating hx_dementia_mci\n",
      "validating hx_melanoma\n",
      "validating education_level\n",
      "validating smoking_status\n",
      "validating smoking_years\n",
      "validating APOE_e4_status\n",
      "validating cognitive_status\n",
      "validating time_from_baseline\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **12 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- AMPPD_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 25/25 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 3/25 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **4 Fields with invalid entries:**\n",
      "- _*hx_melanoma*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*education_level*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other'\n",
      "- _*smoking_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Current smoker', 'Former smoker', 'Never', 'Unknown'\n",
      "- _*cognitive_status*_:  invalid values 💩'NA'\n",
      "    - valid ➡️ 'Normal', 'MCI', 'Dementia'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"log\")\n",
    "\n",
    "df = SUBJECT_.copy()\n",
    "table_choice = \"SUBJECT\"\n",
    "# perform the valadation\n",
    "\n",
    "# specific_cde_df = CDE_df[CDE_df['Table'] == table_choice]\n",
    "specific_cde_df = CDE_df[CDE_df['Table'].str.startswith(table_choice)]\n",
    "\n",
    "retval = validate_table(df, table_choice, specific_cde_df, report)\n",
    "\n",
    "report.print_log()\n",
    "\n",
    "SUBJECT_export, SUBJECT_out, report = retval\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now compose a PROTOCOL, and STUDY\n",
    "\n",
    "\n",
    "ASSAY:\n",
    ">> VAI’s Genomics Core (RRID:SCR_022913; they sequenced this data for us), and I can make some inferences after looking through it. From what I can see, this was likely sequenced on the NovaSeq 6000, with a run length of 2x100bp, using stranded total RNA, with a target genome coverage of 30M reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The samples were collected by the Banner Sun H...</td>\n",
       "      <td>See protocol.</td>\n",
       "      <td>Library preps were generated via the 5' 10x Ge...</td>\n",
       "      <td>We use cellranger count (6.0.1) to align and g...</td>\n",
       "      <td>private repo at https://github.com/TheGBLab/ASAP</td>\n",
       "      <td>https://dx.doi.org/10.17504/protocols.io.bzmgp43w</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sample_collection_summary cell_extraction_summary  \\\n",
       "0  The samples were collected by the Banner Sun H...           See protocol.   \n",
       "\n",
       "                                    lib_prep_summary  \\\n",
       "0  Library preps were generated via the 5' 10x Ge...   \n",
       "\n",
       "                             data_processing_summary  \\\n",
       "0  We use cellranger count (6.0.1) to align and g...   \n",
       "\n",
       "                                         github_url  \\\n",
       "0  private repo at https://github.com/TheGBLab/ASAP   \n",
       "\n",
       "                                    protocols_io_DOI other_reference  \n",
       "0  https://dx.doi.org/10.17504/protocols.io.bzmgp43w              NA  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY['brain_regions'] = \"middle frontal gyrus\"\n",
    "STUDY['team_dataset_id'] = \"Human_bulkRNA_seq_PD_Senesence_Jose_Bras_Team_Lee\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create ASSAY table:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY['project_dataset']= 'Human bulkRNA-seq PD Senesence Jose Bras Team Lee'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "### medial frontal gyrus samples\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "MFG_meta[\"GROUPcv\"]= MFG_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adapt ASSAY_bulkRNAseq table from team Hardy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = Path.home() / \"Projects/ASAP/meta-clean/data/teams/hardy/bulkRNAseq/metadata\"\n",
    "\n",
    "ASSAY_comp = pd.read_csv(metadata_path / \"ASSAY_bulkRNAseq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>RIN</th>\n",
       "      <th>molecular_source</th>\n",
       "      <th>assay</th>\n",
       "      <th>sequencing_end</th>\n",
       "      <th>sequencing_length</th>\n",
       "      <th>sequencing_instrument</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>fragment_library_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hardy_001</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>4.1</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Left</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hardy_002</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>3.1</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Left</td>\n",
       "      <td>Temporal lobe</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hardy_003</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>3.2</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Left</td>\n",
       "      <td>Parietal lobe</td>\n",
       "      <td>Inferior parietal lobule</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hardy_004</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>3.9</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Left</td>\n",
       "      <td>Cingulate gyrus</td>\n",
       "      <td>Anterior cingulate gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hardy_005</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>4.3</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Hardy_284</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>6.8</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Cingulate gyrus</td>\n",
       "      <td>Anterior cingulate gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Hardy_285</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>6.4</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>Hardy_286</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>6.2</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Temporal lobe</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>Hardy_287</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>5.6</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Parietal lobe</td>\n",
       "      <td>Inferior parietal lobule</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>Hardy_288</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Bulk</td>\n",
       "      <td>RNA</td>\n",
       "      <td>6.8</td>\n",
       "      <td>PolyA RNA</td>\n",
       "      <td>Other</td>\n",
       "      <td>Paired-end</td>\n",
       "      <td>150</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>Right</td>\n",
       "      <td>Cingulate gyrus</td>\n",
       "      <td>Anterior cingulate gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>ISR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id tissue technology omic  RIN molecular_source  assay  \\\n",
       "0    Hardy_001  Brain       Bulk  RNA  4.1        PolyA RNA  Other   \n",
       "1    Hardy_002  Brain       Bulk  RNA  3.1        PolyA RNA  Other   \n",
       "2    Hardy_003  Brain       Bulk  RNA  3.2        PolyA RNA  Other   \n",
       "3    Hardy_004  Brain       Bulk  RNA  3.9        PolyA RNA  Other   \n",
       "4    Hardy_005  Brain       Bulk  RNA  4.3        PolyA RNA  Other   \n",
       "..         ...    ...        ...  ...  ...              ...    ...   \n",
       "283  Hardy_284  Brain       Bulk  RNA  6.8        PolyA RNA  Other   \n",
       "284  Hardy_285  Brain       Bulk  RNA  6.4        PolyA RNA  Other   \n",
       "285  Hardy_286  Brain       Bulk  RNA  6.2        PolyA RNA  Other   \n",
       "286  Hardy_287  Brain       Bulk  RNA  5.6        PolyA RNA  Other   \n",
       "287  Hardy_288  Brain       Bulk  RNA  6.8        PolyA RNA  Other   \n",
       "\n",
       "    sequencing_end  sequencing_length  sequencing_instrument hemisphere  \\\n",
       "0       Paired-end                150  Illumina NovaSeq 6000       Left   \n",
       "1       Paired-end                150  Illumina NovaSeq 6000       Left   \n",
       "2       Paired-end                150  Illumina NovaSeq 6000       Left   \n",
       "3       Paired-end                150  Illumina NovaSeq 6000       Left   \n",
       "4       Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "..             ...                ...                    ...        ...   \n",
       "283     Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "284     Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "285     Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "286     Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "287     Paired-end                150  Illumina NovaSeq 6000      Right   \n",
       "\n",
       "      region_level_1            region_level_2 region_level_3  \\\n",
       "0       Frontal lobe      Middle frontal gyrus    Grey matter   \n",
       "1      Temporal lobe     Middle temporal gyrus    Grey matter   \n",
       "2      Parietal lobe  Inferior parietal lobule    Grey matter   \n",
       "3    Cingulate gyrus  Anterior cingulate gyrus    Grey matter   \n",
       "4       Frontal lobe      Middle frontal gyrus    Grey matter   \n",
       "..               ...                       ...            ...   \n",
       "283  Cingulate gyrus  Anterior cingulate gyrus    Grey matter   \n",
       "284     Frontal lobe      Middle frontal gyrus    Grey matter   \n",
       "285    Temporal lobe     Middle temporal gyrus    Grey matter   \n",
       "286    Parietal lobe  Inferior parietal lobule    Grey matter   \n",
       "287  Cingulate gyrus  Anterior cingulate gyrus    Grey matter   \n",
       "\n",
       "    fragment_library_types  \n",
       "0                      ISR  \n",
       "1                      ISR  \n",
       "2                      ISR  \n",
       "3                      ISR  \n",
       "4                      ISR  \n",
       "..                     ...  \n",
       "283                    ISR  \n",
       "284                    ISR  \n",
       "285                    ISR  \n",
       "286                    ISR  \n",
       "287                    ISR  \n",
       "\n",
       "[288 rows x 15 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASSAY_comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "ASSAY = SAMPLE_ \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *ASSAY_bulkRNAseq* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MFG_PD_1441</td>\n",
       "      <td>PD_1441</td>\n",
       "      <td>14-41</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MFG_PD_1344</td>\n",
       "      <td>PD_1344</td>\n",
       "      <td>13-44</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MFG_HC_1939</td>\n",
       "      <td>HC_1939</td>\n",
       "      <td>19-39</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MFG_HC_1308</td>\n",
       "      <td>HC_1308</td>\n",
       "      <td>13-08</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MFG_HC_1862</td>\n",
       "      <td>HC_1862</td>\n",
       "      <td>18-62</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MFG_HC_1864</td>\n",
       "      <td>HC_1864</td>\n",
       "      <td>18-64</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MFG_HC_2057</td>\n",
       "      <td>HC_2057</td>\n",
       "      <td>20-57</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MFG_HC_2061</td>\n",
       "      <td>HC_2061</td>\n",
       "      <td>20-61</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MFG_HC_2062</td>\n",
       "      <td>HC_2062</td>\n",
       "      <td>20-62</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MFG_HC_2067</td>\n",
       "      <td>HC_2067</td>\n",
       "      <td>20-67</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MFG_PD_0348</td>\n",
       "      <td>PD_0348</td>\n",
       "      <td>03-48</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MFG_PD_0413</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>04-13</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MFG_PD_1312</td>\n",
       "      <td>PD_1312</td>\n",
       "      <td>13-12</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFG_PD_1317</td>\n",
       "      <td>PD_1317</td>\n",
       "      <td>13-17</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MFG_PD_1504</td>\n",
       "      <td>PD_1504</td>\n",
       "      <td>15-04</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MFG_PD_1858</td>\n",
       "      <td>PD_1858</td>\n",
       "      <td>18-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MFG_PD_1902</td>\n",
       "      <td>PD_1902</td>\n",
       "      <td>19-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MFG_PD_1973</td>\n",
       "      <td>PD_1973</td>\n",
       "      <td>19-73</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MFG_PD_2005</td>\n",
       "      <td>PD_2005</td>\n",
       "      <td>20-05</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MFG_PD_2038</td>\n",
       "      <td>PD_2038</td>\n",
       "      <td>20-38</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch_x</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0   MFG_HC_1225    HC_1225            12-25      rep1               1   \n",
       "1   MFG_HC_0602    HC_0602            06-02      rep1               1   \n",
       "2   MFG_PD_0009    PD_0009            00-09      rep1               1   \n",
       "3   MFG_PD_1921    PD_1921            19-21      rep1               1   \n",
       "4   MFG_PD_2058    PD_2058            20-58      rep1               1   \n",
       "5   MFG_PD_1441    PD_1441            14-41      rep1               1   \n",
       "6   MFG_PD_1344    PD_1344            13-44      rep1               1   \n",
       "7   MFG_HC_1939    HC_1939            19-39      rep1               1   \n",
       "8   MFG_HC_1308    HC_1308            13-08      rep1               1   \n",
       "9   MFG_HC_1862    HC_1862            18-62      rep1               1   \n",
       "10  MFG_HC_1864    HC_1864            18-64      rep1               1   \n",
       "11  MFG_HC_2057    HC_2057            20-57      rep1               1   \n",
       "12  MFG_HC_2061    HC_2061            20-61      rep1               1   \n",
       "13  MFG_HC_2062    HC_2062            20-62      rep1               1   \n",
       "14  MFG_HC_2067    HC_2067            20-67      rep1               1   \n",
       "15  MFG_PD_0348    PD_0348            03-48      rep1               1   \n",
       "16  MFG_PD_0413    PD_0413            04-13      rep1               1   \n",
       "17  MFG_PD_1312    PD_1312            13-12      rep1               1   \n",
       "18  MFG_PD_1317    PD_1317            13-17      rep1               1   \n",
       "19  MFG_PD_1504    PD_1504            15-04      rep1               1   \n",
       "20  MFG_PD_1858    PD_1858            18-58      rep1               1   \n",
       "21  MFG_PD_1902    PD_1902            19-02      rep1               1   \n",
       "22  MFG_PD_1973    PD_1973            19-73      rep1               1   \n",
       "23  MFG_PD_2005    PD_2005            20-05      rep1               1   \n",
       "24  MFG_PD_2038    PD_2038            20-38      rep1               1   \n",
       "\n",
       "   repeated_sample    batch tissue          brain_region hemisphere  ...  \\\n",
       "0                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "1                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "2                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "3                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "4                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "5                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "6                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "7                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "8                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "9                0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "10               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "11               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "12               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "13               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "14               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "15               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "16               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "17               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "18               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "19               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "20               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "21               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "22               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "23               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "24               0  batch_x  Brain  Middle_Frontal_Gyrus    Unknown  ...   \n",
       "\n",
       "     sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0     PATO:0000384 (male)                                  Unknown   \n",
       "1     PATO:0000384 (male)                                  Unknown   \n",
       "2     PATO:0000384 (male)                                  Unknown   \n",
       "3     PATO:0000384 (male)                                  Unknown   \n",
       "4     PATO:0000384 (male)                                  Unknown   \n",
       "5   PATO:0000383 (female)                                  Unknown   \n",
       "6   PATO:0000383 (female)                                  Unknown   \n",
       "7   PATO:0000383 (female)                                  Unknown   \n",
       "8     PATO:0000384 (male)                                  Unknown   \n",
       "9     PATO:0000384 (male)                                  Unknown   \n",
       "10    PATO:0000384 (male)                                  Unknown   \n",
       "11    PATO:0000384 (male)                                  Unknown   \n",
       "12  PATO:0000383 (female)                                  Unknown   \n",
       "13    PATO:0000384 (male)                                  Unknown   \n",
       "14    PATO:0000384 (male)                                  Unknown   \n",
       "15    PATO:0000384 (male)                                  Unknown   \n",
       "16    PATO:0000384 (male)                                  Unknown   \n",
       "17    PATO:0000384 (male)                                  Unknown   \n",
       "18    PATO:0000384 (male)                                  Unknown   \n",
       "19  PATO:0000383 (female)                                  Unknown   \n",
       "20    PATO:0000384 (male)                                  Unknown   \n",
       "21    PATO:0000384 (male)                                  Unknown   \n",
       "22    PATO:0000384 (male)                                  Unknown   \n",
       "23    PATO:0000384 (male)                                  Unknown   \n",
       "24    PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id  \\\n",
       "0              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "2             MONDO:0005180          UBERON:0002702   \n",
       "3             MONDO:0005180          UBERON:0002702   \n",
       "4             MONDO:0005180          UBERON:0002702   \n",
       "5             MONDO:0005180          UBERON:0002702   \n",
       "6             MONDO:0005180          UBERON:0002702   \n",
       "7              PATO:0000461          UBERON:0002702   \n",
       "8              PATO:0000461          UBERON:0002702   \n",
       "9              PATO:0000461          UBERON:0002702   \n",
       "10             PATO:0000461          UBERON:0002702   \n",
       "11             PATO:0000461          UBERON:0002702   \n",
       "12             PATO:0000461          UBERON:0002702   \n",
       "13             PATO:0000461          UBERON:0002702   \n",
       "14             PATO:0000461          UBERON:0002702   \n",
       "15            MONDO:0005180          UBERON:0002702   \n",
       "16            MONDO:0005180          UBERON:0002702   \n",
       "17            MONDO:0005180          UBERON:0002702   \n",
       "18            MONDO:0005180          UBERON:0002702   \n",
       "19            MONDO:0005180          UBERON:0002702   \n",
       "20            MONDO:0005180          UBERON:0002702   \n",
       "21            MONDO:0005180          UBERON:0002702   \n",
       "22            MONDO:0005180          UBERON:0002702   \n",
       "23            MONDO:0005180          UBERON:0002702   \n",
       "24            MONDO:0005180          UBERON:0002702   \n",
       "\n",
       "   cell_type_ontology_term_id assay_ontology_term_id suspension_type DV200  \\\n",
       "0                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "1                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "2                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "3                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "4                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "5                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "6                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "7                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "8                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "9                NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "10               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "11               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "12               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "13               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "14               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "15               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "16               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "17               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "18               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "19               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "20               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "21               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "22               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "23               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "24               NA(multiple)            EFO:0030004         nucleus    NA   \n",
       "\n",
       "   pm_PH donor_id  \n",
       "0     NA       NA  \n",
       "1     NA       NA  \n",
       "2     NA       NA  \n",
       "3     NA       NA  \n",
       "4     NA       NA  \n",
       "5     NA       NA  \n",
       "6     NA       NA  \n",
       "7     NA       NA  \n",
       "8     NA       NA  \n",
       "9     NA       NA  \n",
       "10    NA       NA  \n",
       "11    NA       NA  \n",
       "12    NA       NA  \n",
       "13    NA       NA  \n",
       "14    NA       NA  \n",
       "15    NA       NA  \n",
       "16    NA       NA  \n",
       "17    NA       NA  \n",
       "18    NA       NA  \n",
       "19    NA       NA  \n",
       "20    NA       NA  \n",
       "21    NA       NA  \n",
       "22    NA       NA  \n",
       "23    NA       NA  \n",
       "24    NA       NA  \n",
       "\n",
       "[25 rows x 33 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"log\")\n",
    "\n",
    "df = ASSAY.copy()\n",
    "table_choice = \"ASSAY_bulkRNAseq\"\n",
    "# perform the valadation\n",
    "\n",
    "# specific_cde_df = CDE_df[CDE_df['Table'] == table_choice]\n",
    "specific_cde_df = CDE_df[CDE_df['Table'].str.startswith(table_choice)]\n",
    "\n",
    "retval = validate_table(df, table_choice, specific_cde_df, report)\n",
    "\n",
    "report.print_log()\n",
    "\n",
    "ASSAY_export, ASSAY_out, report = retval\n",
    "\n",
    "ASSAY_export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_path = Path.cwd() / \"data/lee/bulk_rna_seq/metadata\"\n",
    "\n",
    "if not export_path.exists():\n",
    "    export_path.mkdir(parents=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(export_path / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_path / \"PROTOCOL.csv\")\n",
    "CLINPATH_export.to_csv(export_path / \"CLINPATH.csv\")\n",
    "SAMPLE_export.to_csv(export_path / \"SAMPLE.csv\")\n",
    "SUBJECT_export.to_csv(export_path / \"SUBJECT.csv\")\n",
    "\n",
    "DATA_export.to_csv(export_path / \"DATA.csv\")\n",
    "\n",
    "MFG_TABLE.to_csv(export_path / \"aux_meta.csv\")\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "ASSAY_export.to_csv(export_path / \"ASSAY_bulkRNAseq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/harmonized-wf-dev/data/teams/lee/data/lee/bulk_rna_seq/metadata/ASSAY_bulkRNAseq.csv')"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "export_path / \"ASSAY_bulkRNAseq.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update the table to v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>HC_0602</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>PD_0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count  \\\n",
       "0  MFG_HC_1225            12-25    HC_1225      rep1                1   \n",
       "0  MFG_HC_1225            12-25    HC_1225      rep1                1   \n",
       "1  MFG_HC_0602            06-02    HC_0602      rep1                1   \n",
       "1  MFG_HC_0602            06-02    HC_0602      rep1                1   \n",
       "2  MFG_PD_0009            00-09    PD_0009      rep1                1   \n",
       "\n",
       "   repeated_sample    batch tissue          brain_region  source_RIN  ...  \\\n",
       "0                0  BATCH_4  Brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "0                0  BATCH_4  Brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "1                0  BATCH_4  Brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "1                0  BATCH_4  Brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "2                0  BATCH_4  Brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0   PATO:0000384 (male)                                  Unknown   \n",
       "0   PATO:0000384 (male)                                  Unknown   \n",
       "1   PATO:0000384 (male)                                  Unknown   \n",
       "1   PATO:0000384 (male)                                  Unknown   \n",
       "2   PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id  \\\n",
       "0              PATO:0000461          UBERON:0002702   \n",
       "0              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "2             MONDO:0005180          UBERON:0002702   \n",
       "\n",
       "  cell_type_ontology_term_id assay_ontology_term_id suspension_type DV200  \\\n",
       "0               NA(multiple)            EFO:0030004         nucleus  <NA>   \n",
       "0               NA(multiple)            EFO:0030004         nucleus  <NA>   \n",
       "1               NA(multiple)            EFO:0030004         nucleus  <NA>   \n",
       "1               NA(multiple)            EFO:0030004         nucleus  <NA>   \n",
       "2               NA(multiple)            EFO:0030004         nucleus  <NA>   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN           \n",
       "0   NaN           \n",
       "1   NaN           \n",
       "1   NaN           \n",
       "2   NaN           \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "from update_schema import update_tables_to_CDEv2\n",
    "\n",
    "tables_path = Path.cwd() / \"clean/team-Lee\"\n",
    "\n",
    "\n",
    "CDEv1 = pd.read_csv( Path.cwd() / \"ASAP_CDE_v1.csv\" )\n",
    "CDEv2 = pd.read_csv( Path.cwd() / \"ASAP_CDE_v2.csv\" )\n",
    "\n",
    "\n",
    "STUDYv2, PROTOCOLv2, SAMPLEv2, SUBJECTv2, CLINPATHv2, DATAv2 = update_tables_to_CDEv2(tables_path, CDEv1, CDEv2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "export_root = Path.cwd() / \"clean/team-Lee/v2\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "STUDYv2.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOLv2.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLEv2.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECTv2.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATHv2.to_csv(export_root / \"CLINPATH.csv\")\n",
    "DATAv2.to_csv(export_root / \"DATA.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Lee/v2')"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "export_root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transfer cleaned metadata to raw buckets \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-lee@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "# Lee\n",
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/lee-credentials.json \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      6243  2023-11-29T22:26:06Z  gs://asap-raw-data-team-lee/metadata/v2/CLINPATH.csv#1701296766077305  metageneration=1\n",
      "     24218  2023-11-29T22:26:06Z  gs://asap-raw-data-team-lee/metadata/v2/DATA.csv#1701296766689376  metageneration=1\n",
      "       968  2023-11-29T22:26:07Z  gs://asap-raw-data-team-lee/metadata/v2/PROTOCOL.csv#1701296767085192  metageneration=1\n",
      "     23092  2023-11-29T22:26:06Z  gs://asap-raw-data-team-lee/metadata/v2/SAMPLE.csv#1701296766486332  metageneration=1\n",
      "      1054  2023-11-29T22:26:06Z  gs://asap-raw-data-team-lee/metadata/v2/STUDY.csv#1701296766878081  metageneration=1\n",
      "      4277  2023-11-29T22:26:06Z  gs://asap-raw-data-team-lee/metadata/v2/SUBJECT.csv#1701296766288192  metageneration=1\n",
      "TOTAL: 6 objects, 59852 bytes (58.45 KiB)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!gsutil -u dnastack-asap-parkinsons ls -al \"gs://asap-raw-data-team-lee/metadata/v2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/CLINPATH.csv#1701213902124849...\n",
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/DATA.csv#1701213902800745...\n",
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/PROTOCOL.csv#1701213903243233...\n",
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/SAMPLE.csv#1701213902582568...\n",
      "/ [4 objects]                                                                   \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m rm ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/STUDY.csv#1701213903050409...\n",
      "Removing gs://asap-raw-data-team-lee/metadata/v2/v2_20231128/SUBJECT.csv#1701213902358354...\n",
      "/ [6 objects]                                                                   \n",
      "Operation completed over 6 objects.                                              \n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons rm -r \"gs://asap-raw-data-team-lee/metadata/v2/v2_20231128\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/meta-clean')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Lee/v2_20231130/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/SUBJECT.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/SAMPLE.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/DATA.csv [Content-Type=text/csv]... \n",
      "- [4 files][ 56.5 KiB/ 56.5 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Lee/v2_20231130/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/PROTOCOL.csv [Content-Type=text/csv]...\n",
      "\\ [6 files][ 58.4 KiB/ 58.4 KiB]                                                \n",
      "Operation completed over 6 objects/58.4 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "\n",
    "!gsutil -u dnastack-asap-parkinsons cp -r \"./clean/team-Lee/v2_20231130/*.csv\"  \"gs://asap-raw-data-team-lee/metadata/v2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Lee/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/SUBJECT.csv [Content-Type=text/csv]...          \n",
      "Copying file://./clean/team-Lee/auxiluary_metadata.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/SAMPLE.csv [Content-Type=text/csv]...           \n",
      "\\ [4 files][187.4 KiB/187.4 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Lee/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/auxilarry_metadata.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/PROTOCOL.csv [Content-Type=text/csv]...         \n",
      "Copying file://./clean/team-Lee/CDE.csv [Content-Type=text/csv]...              \n",
      "| [8 files][328.0 KiB/328.0 KiB]                                                \n",
      "Operation completed over 8 objects/328.0 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons cp -r \"./clean/team-Lee/*.csv\"  \"gs://asap-raw-data-team-lee/metadata/clean/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     37741  2023-11-28T21:58:33Z  gs://asap-raw-data-team-lee/metadata/clean/CDE.csv#1701208713502213  metageneration=1\n",
      "     24241  2023-11-28T21:58:31Z  gs://asap-raw-data-team-lee/metadata/clean/CLINPATH.csv#1701208711158586  metageneration=1\n",
      "       968  2023-11-28T21:58:33Z  gs://asap-raw-data-team-lee/metadata/clean/PROTOCOL.csv#1701208713057031  metageneration=1\n",
      "     60412  2023-11-28T21:58:32Z  gs://asap-raw-data-team-lee/metadata/clean/SAMPLE.csv#1701208712230186  metageneration=1\n",
      "       969  2023-11-28T21:58:32Z  gs://asap-raw-data-team-lee/metadata/clean/STUDY.csv#1701208712490381  metageneration=1\n",
      "      3078  2023-11-28T21:58:31Z  gs://asap-raw-data-team-lee/metadata/clean/SUBJECT.csv#1701208711431682  metageneration=1\n",
      "    104336  2023-11-28T21:58:32Z  gs://asap-raw-data-team-lee/metadata/clean/auxilarry_metadata.csv#1701208712852367  metageneration=1\n",
      "    109904  2023-09-22T10:45:50Z  gs://asap-raw-data-team-lee/metadata/clean/auxiliary_metadata.csv#1695379550068707  metageneration=1\n",
      "    104153  2023-11-28T21:58:31Z  gs://asap-raw-data-team-lee/metadata/clean/auxiluary_metadata.csv#1701208711886220  metageneration=1\n",
      "TOTAL: 9 objects, 445802 bytes (435.35 KiB)\n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons ls -al \"gs://asap-raw-data-team-lee/metadata/clean\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy to workflow-dev bucket\n",
    "\n",
    "First copy each set of metadata locally ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [admin-workflow-dev@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/wf-credentials.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      6243  2023-11-28T22:21:42Z  gs://asap-workflow-dev/metadata/v2/lee/CLINPATH.csv#1701210102669978  metageneration=1\n",
      "     24218  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/DATA.csv#1701210103715354  metageneration=1\n",
      "       968  2023-11-28T22:21:44Z  gs://asap-workflow-dev/metadata/v2/lee/PROTOCOL.csv#1701210104255218  metageneration=1\n",
      "     23092  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/SAMPLE.csv#1701210103292269  metageneration=1\n",
      "      1054  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/STUDY.csv#1701210103992608  metageneration=1\n",
      "      4277  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/SUBJECT.csv#1701210103010639  metageneration=1\n",
      "                                 gs://asap-workflow-dev/metadata/v2/lee/v2_20231128/\n",
      "TOTAL: 6 objects, 59852 bytes (58.45 KiB)\n"
     ]
    }
   ],
   "source": [
    "# !gsutil  ls -al \"gs://asap-workflow-dev/CDE\"\n",
    "\n",
    "!gsutil  ls -al \"gs://asap-workflow-dev/metadata/v2/lee\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Lee/v2_20231130/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/SUBJECT.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/SAMPLE.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/DATA.csv [Content-Type=text/csv]... \n",
      "- [4 files][ 56.5 KiB/ 56.5 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Lee/v2_20231130/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Lee/v2_20231130/PROTOCOL.csv [Content-Type=text/csv]...\n",
      "\\ [6 files][ 58.4 KiB/ 58.4 KiB]                                                \n",
      "Operation completed over 6 objects/58.4 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil  cp -r \"./clean/team-Lee/v2_20231130/*.csv\" \"gs://asap-workflow-dev/metadata/v2/lee\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      6243  2023-11-28T22:21:42Z  gs://asap-workflow-dev/metadata/v2/lee/CLINPATH.csv#1701210102669978  metageneration=1\n",
      "     24218  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/DATA.csv#1701210103715354  metageneration=1\n",
      "       968  2023-11-28T22:21:44Z  gs://asap-workflow-dev/metadata/v2/lee/PROTOCOL.csv#1701210104255218  metageneration=1\n",
      "     23092  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/SAMPLE.csv#1701210103292269  metageneration=1\n",
      "      1054  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/STUDY.csv#1701210103992608  metageneration=1\n",
      "      4277  2023-11-28T22:21:43Z  gs://asap-workflow-dev/metadata/v2/lee/SUBJECT.csv#1701210103010639  metageneration=1\n",
      "TOTAL: 6 objects, 59852 bytes (58.45 KiB)\n"
     ]
    }
   ],
   "source": [
    "!gsutil ls -al \"gs://asap-workflow-dev/metadata/v2/lee\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check file md5s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.checksums import extract_md5_from_details, extract_md5_from_details2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-lee@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/lee-credentials.json  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "WARNING: You have requested checksumming but your crcmod installation isn't\n",
      "using the module's C extension, so checksumming will run very slowly. For help\n",
      "installing the extension, please see \"gsutil help crcmod\".\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# !gcloud storage hash \"gs://asap-raw-data-team-lee/**/*.gz\"  --skip-crc32c --hex  --billing-project dnastack-asap-parkinsons > hardy_hexhash.log\n",
    "\n",
    "!gsutil -u dnastack-asap-parkinsons hash -h \"gs://asap-raw-data-team-lee/**/*.gz\" > lee_hexhash.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_56157/1302936753.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check'] = checksum['file_name'].map(bucket_files_md5)\n"
     ]
    }
   ],
   "source": [
    "bucket_files_md5 = extract_md5_from_details2(\"lee_hexhash.log\")\n",
    "\n",
    "\n",
    "\n",
    "checksum = DATAv2[['file_name','file_MD5']]\n",
    "checksum['check1'] = checksum['file_MD5'].str.strip()\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "#empty means success!!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>check</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFGHC1225_S9_L001_R1_001.fastq.gz</td>\n",
       "      <td>9977258e598d6a52130c29c71aef6925</td>\n",
       "      <td>9977258e598d6a52130c29c71aef6925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFGHC1225_S9_L001_R2_001.fastq.gz</td>\n",
       "      <td>fe2cf93257801227b7072a4fb7d18792</td>\n",
       "      <td>fe2cf93257801227b7072a4fb7d18792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFGHC0602_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>110ca4864cf6938faca67567bebfb6cc</td>\n",
       "      <td>110ca4864cf6938faca67567bebfb6cc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFGHC0602_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>0dcc67217e43ab53bae0d0676f9bfe8b</td>\n",
       "      <td>0dcc67217e43ab53bae0d0676f9bfe8b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFGPD0009_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>a2608d0bd192333b0076d7091c1c50ea</td>\n",
       "      <td>a2608d0bd192333b0076d7091c1c50ea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>SN_1973_PD_S1_L000_R2_001.fastq.gz</td>\n",
       "      <td>53f6c6b4a00299fb41f32b34509835fa</td>\n",
       "      <td>53f6c6b4a00299fb41f32b34509835fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>SN_2005_PD_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>5bde8fdd8ab28e2c00e3ae327fecc80b</td>\n",
       "      <td>5bde8fdd8ab28e2c00e3ae327fecc80b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>SN_2005_PD_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>14364dc52760fbd3ad08b5ca582d849b</td>\n",
       "      <td>14364dc52760fbd3ad08b5ca582d849b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SN_2038_PD_S1_L000_R1_001.fastq.gz</td>\n",
       "      <td>3215d4e4a68a85546183e820b74cac1f</td>\n",
       "      <td>3215d4e4a68a85546183e820b74cac1f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SN_2038_PD_S1_L000_R2_001.fastq.gz</td>\n",
       "      <td>aeb21109e488f2cd78f077b3feb4a180</td>\n",
       "      <td>aeb21109e488f2cd78f077b3feb4a180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              file_name                          file_MD5  \\\n",
       "0     MFGHC1225_S9_L001_R1_001.fastq.gz  9977258e598d6a52130c29c71aef6925   \n",
       "1     MFGHC1225_S9_L001_R2_001.fastq.gz  fe2cf93257801227b7072a4fb7d18792   \n",
       "2     MFGHC0602_S2_L001_R1_001.fastq.gz  110ca4864cf6938faca67567bebfb6cc   \n",
       "3     MFGHC0602_S2_L001_R2_001.fastq.gz  0dcc67217e43ab53bae0d0676f9bfe8b   \n",
       "4     MFGPD0009_S3_L001_R1_001.fastq.gz  a2608d0bd192333b0076d7091c1c50ea   \n",
       "..                                  ...                               ...   \n",
       "145  SN_1973_PD_S1_L000_R2_001.fastq.gz  53f6c6b4a00299fb41f32b34509835fa   \n",
       "146  SN_2005_PD_S1_L001_R1_001.fastq.gz  5bde8fdd8ab28e2c00e3ae327fecc80b   \n",
       "147  SN_2005_PD_S1_L001_R2_001.fastq.gz  14364dc52760fbd3ad08b5ca582d849b   \n",
       "148  SN_2038_PD_S1_L000_R1_001.fastq.gz  3215d4e4a68a85546183e820b74cac1f   \n",
       "149  SN_2038_PD_S1_L000_R2_001.fastq.gz  aeb21109e488f2cd78f077b3feb4a180   \n",
       "\n",
       "                                check  \n",
       "0    9977258e598d6a52130c29c71aef6925  \n",
       "1    fe2cf93257801227b7072a4fb7d18792  \n",
       "2    110ca4864cf6938faca67567bebfb6cc  \n",
       "3    0dcc67217e43ab53bae0d0676f9bfe8b  \n",
       "4    a2608d0bd192333b0076d7091c1c50ea  \n",
       "..                                ...  \n",
       "145  53f6c6b4a00299fb41f32b34509835fa  \n",
       "146  5bde8fdd8ab28e2c00e3ae327fecc80b  \n",
       "147  14364dc52760fbd3ad08b5ca582d849b  \n",
       "148  3215d4e4a68a85546183e820b74cac1f  \n",
       "149  aeb21109e488f2cd78f077b3feb4a180  \n",
       "\n",
       "[150 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sl11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
