{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation \n",
    "# Team Scherzer. ASAP CRN Metadata validation \n",
    "\n",
    "10 Oct 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table, create_metadata_package\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, intervention_typer\n",
    "from checksums import extract_md5_from_details2, get_md5_hashes \n",
    "from bucket_util import authenticate_with_service_account, gsutil_ls, gsutil_cp, gsutil_mv \n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted as v2.1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Team Scherzer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"scherzer\"\n",
    "dataset_name = \"sn-rnaseq-mtg\"\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "\n",
    "og_path = metadata_path / \"og\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\",\"DATA\"]\n",
    "\n",
    "excel_path = og_path / \"ASAP_CDE-ScherzerTeam.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\")#,header=1)#.drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\")#,header=1)#.drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\")#,header=1)#.drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\")#,header=1)#.drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\")#,header=1)#.drop(columns=\"Field\")\n",
    "DATA = pd.read_excel(excel_path,sheet_name=\"DATA\")#,header=1)#.drop(columns=\"Field\")\n",
    "\n",
    "metadata_version = \"v2\"\n",
    "\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = prep_table(STUDY,CDEv2)\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "STUDY.rename({'submittor_email':'submitter_email'}, axis=1, inplace=True)\n",
    "STUDY['ASAP_team_name'] = \"TEAM-SCHERZER\"\n",
    "# STUDY['team_dataset_id'] = \"sc_pmdbs\" # \"MTG_snRNAseq\" #\"Scherzer_MTG_snRNAseq\"\n",
    "STUDY['team_dataset_id'] = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-SCHERZER</td>\n",
       "      <td>Scherzer, Dong, and Levin</td>\n",
       "      <td>Parkinson5D: deconstructing proximal disease m...</td>\n",
       "      <td>sn_rnaseq_mtg</td>\n",
       "      <td>PD5D_MTG_snRNAseq</td>\n",
       "      <td>Here we will develop a molecular atlas of Park...</td>\n",
       "      <td>Dr.Clemens Scherzer</td>\n",
       "      <td>Clemens.scherzer@yale.edu</td>\n",
       "      <td>Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...</td>\n",
       "      <td>Dr.Xianjun Dong</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>94</td>\n",
       "      <td>MTG</td>\n",
       "      <td>PD and control postmortem brains</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>ASAP Access &amp; Use Policy</td>\n",
       "      <td>Nan</td>\n",
       "      <td>v2_20241102</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name              ASAP_lab_name  \\\n",
       "0  TEAM-SCHERZER  Scherzer, Dong, and Levin   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Parkinson5D: deconstructing proximal disease m...   sn_rnaseq_mtg   \n",
       "\n",
       "     project_dataset                                project_description  \\\n",
       "0  PD5D_MTG_snRNAseq  Here we will develop a molecular atlas of Park...   \n",
       "\n",
       "          PI_full_name                   PI_email  \\\n",
       "0  Dr.Clemens Scherzer  Clemens.scherzer@yale.edu   \n",
       "\n",
       "                                   contributor_names   submitter_name  ...  \\\n",
       "0  Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...  Dr.Xianjun Dong  ...   \n",
       "\n",
       "  publication_PMID number_of_brain_samples brain_regions  \\\n",
       "0              Nan                      94           MTG   \n",
       "\n",
       "                   types_of_samples PI_ORCHID  PI_google_scholar_id  \\\n",
       "0  PD and control postmortem brains       Nan                   Nan   \n",
       "\n",
       "                DUA_version preprocessing_references metadata_version_date  \\\n",
       "0  ASAP Access & Use Policy                      Nan           v2_20241102   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 24 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY['alternate_dataset_id'] = NULL\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['00-09', '03-29', '03-39', '03-41', '03-47', '03-48', '04-15',\n",
       "       '04-52', '04-64', '05-36', '06-02', '06-15', '06-35', '06-44',\n",
       "       '06-51', '06-55', '06-62', '07-04', '07-37', '07-46', '08-55',\n",
       "       '09-34', '09-52', '10-22', '10-76', '11-28', '11-44', '11-60',\n",
       "       '12-04', '12-06', '12-21', '12-66', '13-05', '13-08', '13-17',\n",
       "       '13-39', '13-40', '13-51', '13-61', '14-12', '14-24', '14-30',\n",
       "       '15-04', '15-06', '15-18', '15-35', '15-46', '15-54', '15-67',\n",
       "       '15-78', '15-85', '16-14', '16-44', '17-19', '17-22', '17-26',\n",
       "       '17-30', '17-37', '17-41', '17-47', '17-49', '17-52', '17-56',\n",
       "       '17-62', '18-09', '18-12', '18-14', '18-05', '18-17', '18-22',\n",
       "       '18-27', '18-36', '18-39', '18-42', '18-48', '18-49', '18-55',\n",
       "       '18-62', '18-65', '18-72', '18-78', '19-02', '19-10', '19-34',\n",
       "       '19-39', '19-57', '19-59', '19-60', '19-74', '20-03', '20-15',\n",
       "       '99-44', '99-47', '99-66'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix the excel saving source_subject_id as a date\n",
    "SUBJECT = prep_table(SUBJECT,CDEv2)\n",
    "\n",
    "recoder = {'2023-03-29 00:00:00': '03-29',\n",
    "       '2023-04-15 00:00:00': '04-15', \n",
    "       '2023-06-02 00:00:00': '06-02',\n",
    "        '2023-06-15 00:00:00': '06-15',\n",
    "        '2023-07-04 00:00:00': '07-04', \n",
    "       '2023-10-22 00:00:00': '10-22',\n",
    "       '2023-11-28 00:00:00': '11-28',\n",
    "        '2023-12-04 00:00:00': '12-04',\n",
    "       '2023-12-21 00:00:00': '12-21'}\n",
    "\n",
    "SUBJECT['source_subject_id'] = SUBJECT['source_subject_id'].replace(recoder)\n",
    "\n",
    "SUBJECT['source_subject_id'].unique()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>hx_dementia_mci</th>\n",
       "      <th>hx_melanoma</th>\n",
       "      <th>education_level</th>\n",
       "      <th>smoking_status</th>\n",
       "      <th>smoking_years</th>\n",
       "      <th>APOE_e4_status</th>\n",
       "      <th>cognitive_status</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>64</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>30</td>\n",
       "      <td>33</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN0329</td>\n",
       "      <td>03-29</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>79</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NA</td>\n",
       "      <td>14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>23</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN0339</td>\n",
       "      <td>03-39</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>86</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>12</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>50</td>\n",
       "      <td>33</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN0341</td>\n",
       "      <td>03-41</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>89</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>16</td>\n",
       "      <td>Never</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN0347</td>\n",
       "      <td>03-47</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>95</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>12</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>33</td>\n",
       "      <td>MCI</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>BN2003</td>\n",
       "      <td>20-03</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>72</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NA</td>\n",
       "      <td>16</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>33</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>BN2015</td>\n",
       "      <td>20-15</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>90</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>20</td>\n",
       "      <td>Never</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>34</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Prodromal motor PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>BN9944</td>\n",
       "      <td>99-44</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>69</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>15</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>34</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>BN9947</td>\n",
       "      <td>99-47</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>84</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>20</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>33</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Prodromal motor PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>BN9966</td>\n",
       "      <td>99-66</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>74</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>16</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>33</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Prodromal motor PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id AMPPD_id GP2_id biobank_name organism     sex  \\\n",
       "0      BN0009             00-09      Nan    Nan        BSHRI    Human    Male   \n",
       "1      BN0329             03-29      Nan    Nan        BSHRI    Human    Male   \n",
       "2      BN0339             03-39      Nan    Nan        BSHRI    Human    Male   \n",
       "3      BN0341             03-41      Nan    Nan        BSHRI    Human    Male   \n",
       "4      BN0347             03-47      Nan    Nan        BSHRI    Human  Female   \n",
       "..        ...               ...      ...    ...          ...      ...     ...   \n",
       "89     BN2003             20-03      Nan    Nan        BSHRI    Human    Male   \n",
       "90     BN2015             20-15      Nan    Nan        BSHRI    Human    Male   \n",
       "91     BN9944             99-44      Nan    Nan        BSHRI    Human    Male   \n",
       "92     BN9947             99-47      Nan    Nan        BSHRI    Human    Male   \n",
       "93     BN9966             99-66      Nan    Nan        BSHRI    Human    Male   \n",
       "\n",
       "    age_at_collection   race ethnicity  ... hx_dementia_mci hx_melanoma  \\\n",
       "0                  64  White   Unknown  ...             Yes          NA   \n",
       "1                  79  White   Unknown  ...             Yes          NA   \n",
       "2                  86  White   Unknown  ...              No          NA   \n",
       "3                  89  White   Unknown  ...              No          NA   \n",
       "4                  95  White   Unknown  ...              No          NA   \n",
       "..                ...    ...       ...  ...             ...         ...   \n",
       "89                 72  White   Unknown  ...             Yes          NA   \n",
       "90                 90  White   Unknown  ...              No          NA   \n",
       "91                 69  White   Unknown  ...              No          NA   \n",
       "92                 84  White   Unknown  ...              No          NA   \n",
       "93                 74  White   Unknown  ...              No          NA   \n",
       "\n",
       "   education_level smoking_status smoking_years APOE_e4_status  \\\n",
       "0               NA  Former smoker            30             33   \n",
       "1               14        Unknown       Unknown             23   \n",
       "2               12  Former smoker            50             33   \n",
       "3               16          Never             0             34   \n",
       "4               12  Former smoker       Unknown             33   \n",
       "..             ...            ...           ...            ...   \n",
       "89              16  Former smoker       Unknown             33   \n",
       "90              20          Never       Unknown             34   \n",
       "91              15  Former smoker       Unknown             34   \n",
       "92              20  Former smoker       Unknown             33   \n",
       "93              16  Former smoker       Unknown             33   \n",
       "\n",
       "   cognitive_status time_from_baseline   primary_diagnosis  \\\n",
       "0          Dementia            Unknown       Idiopathic PD   \n",
       "1          Dementia            Unknown       Idiopathic PD   \n",
       "2            Normal            Unknown     Healthy Control   \n",
       "3            Normal            Unknown     Healthy Control   \n",
       "4               MCI            Unknown     Healthy Control   \n",
       "..              ...                ...                 ...   \n",
       "89         Dementia            Unknown       Idiopathic PD   \n",
       "90           Normal            Unknown  Prodromal motor PD   \n",
       "91           Normal            Unknown     Healthy Control   \n",
       "92           Normal            Unknown  Prodromal motor PD   \n",
       "93           Normal            Unknown  Prodromal motor PD   \n",
       "\n",
       "   primary_diagnosis_text  \n",
       "0                     NaN  \n",
       "1                     NaN  \n",
       "2                     NaN  \n",
       "3                     NaN  \n",
       "4                     NaN  \n",
       "..                    ...  \n",
       "89                    NaN  \n",
       "90                    NaN  \n",
       "91                    NaN  \n",
       "92                    NaN  \n",
       "93                    NaN  \n",
       "\n",
       "[94 rows x 25 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "SUBJECT['hx_melanoma'] = SUBJECT['hx_melanoma'].replace({'Present':'Yes', 'Unknown': NULL})\n",
    "SUBJECT['cognitive_status'] = SUBJECT['cognitive_status'].replace({'Unknown': NULL })\n",
    "\n",
    "# SUBJECT.replace(\"Unknown\", NULL, inplace=True)\n",
    "\n",
    "# eductation levels \n",
    "SUBJECT['education_level'] = SUBJECT['education_level'].replace({'Unknown': NULL })\n",
    "\n",
    "# map \"Unknown\" to NULL\n",
    "\n",
    "# map ['14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'] ???\n",
    "SUBJECT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = prep_table(SAMPLE,CDEv2)\n",
    "\n",
    "SAMPLE['region_level_1'] = SAMPLE['region_level_1'].replace({'Temporal Lobe':'Temporal lobe'})\n",
    "SAMPLE['molecular_source'] = SAMPLE['molecular_source'].replace({'PolyA_RNA':'PolyA RNA'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "subj_sex_mapper = dict(zip(SUBJECT['subject_id'], SUBJECT['sex']))\n",
    "SAMPLE['sex_ontology_term_id'] = SAMPLE['subject_id'].map(subj_sex_mapper).replace({\"Male\":\"PATO:0000384 (male)\", \"Female\":\"PATO:0000383 (female)\" })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# autopsy coding is too too idosyncratic.  leave \n",
    "# the path_nia_aa_a are coded as path_nia_ri... swap them \n",
    "\n",
    "CLINPATH = prep_table(CLINPATH,CDEv2)\n",
    "\n",
    "\n",
    "# 'Low', 'Intermediate', 'High', 'None'\n",
    "CLINPATH['path_nia_ri'] = CLINPATH['path_nia_aa_a'].replace({'Low ADNC': 'Low', \n",
    "                                                            'Not AD': 'None', \n",
    "                                                            'Intermediate ADNC': 'Intermediate'})\n",
    "CLINPATH['path_nia_aa_a'] = NULL\n",
    "\n",
    "\n",
    "# path_thal\n",
    "CLINPATH['path_thal']= CLINPATH['path_thal'].replace({'Phase 0 (A0)':'0', \n",
    "                                'Phase 1 (A1)':'1', \n",
    "                                'Phase 2 (A1)':'2', \n",
    "                                'Phase 3 (A2)':'3', \n",
    "                                'Phase 4 (A3)':'4',\n",
    "                                'Phase 5 (A3)':'5', \n",
    "                                'Unknown':NULL, \n",
    "                                'Missing/unknown':NULL})\n",
    "\n",
    "CLINPATH['path_infarcs'] = CLINPATH['path_infarcs'].replace('Unknown', NULL)\n",
    "\n",
    "\n",
    "# no idea what No, undx, yes maps to.  leave\n",
    "CLINPATH['TDP43'] = CLINPATH['TDP43'].replace('Unknown', NULL)\n",
    "\n",
    "# sn_neuronal_loss has alternate 0-3 coding.  leave\n",
    "\n",
    "# add 'replicate' coding (nans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>path_autopsy_dx_main</th>\n",
       "      <th>path_autopsy_second_dx</th>\n",
       "      <th>path_autopsy_third_dx</th>\n",
       "      <th>path_autopsy_fourth_dx</th>\n",
       "      <th>path_autopsy_fifth_dx</th>\n",
       "      <th>path_autopsy_sixth_dx</th>\n",
       "      <th>path_autopsy_seventh_dx</th>\n",
       "      <th>...</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>dig_slide_avail</th>\n",
       "      <th>quant_path_avail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>4.00</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Charcot-Marie-Tooth disease (history)</td>\n",
       "      <td>GBA L444P/WT, L444P mutation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN0329</td>\n",
       "      <td>03-29</td>\n",
       "      <td>4.50</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Seizure disorder (history)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN0339</td>\n",
       "      <td>03-39</td>\n",
       "      <td>2.75</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>CAA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mild</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN0341</td>\n",
       "      <td>03-41</td>\n",
       "      <td>2.50</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>CWMR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>No</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Mild</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN0347</td>\n",
       "      <td>03-47</td>\n",
       "      <td>3.50</td>\n",
       "      <td>Control (MCI)</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Argyrophilic grains, mesial temporal lobe</td>\n",
       "      <td>Infarct(s)</td>\n",
       "      <td>CWMR</td>\n",
       "      <td>Several microscopic foci of cerebellar cortica...</td>\n",
       "      <td>Etat crible, putamen and caudate nucleus</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mild</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>BN2003</td>\n",
       "      <td>20-03</td>\n",
       "      <td>2.65</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Focal non-specific glial tauopathy, cortex of ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>No</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>BN2015</td>\n",
       "      <td>20-15</td>\n",
       "      <td>5.40</td>\n",
       "      <td>Control (history)</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Incidental Lewy body disease</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>No</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>BN9944</td>\n",
       "      <td>99-44</td>\n",
       "      <td>2.16</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Alzheimer Type II astrocytosis consistent with...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>BN9947</td>\n",
       "      <td>99-47</td>\n",
       "      <td>2.50</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Alzheimer type II astrocytosis</td>\n",
       "      <td>Inc LBs</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>BN9966</td>\n",
       "      <td>99-66</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Control</td>\n",
       "      <td>Incidental LBs</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Argyrophilic grains in mesial temporal lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Mild</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id  duration_pmi path_autopsy_dx_main  \\\n",
       "0      BN0009             00-09          4.00               PD/Dem   \n",
       "1      BN0329             03-29          4.50               PD/Dem   \n",
       "2      BN0339             03-39          2.75              Control   \n",
       "3      BN0341             03-41          2.50              Control   \n",
       "4      BN0347             03-47          3.50        Control (MCI)   \n",
       "..        ...               ...           ...                  ...   \n",
       "89     BN2003             20-03          2.65               PD/Dem   \n",
       "90     BN2015             20-15          5.40    Control (history)   \n",
       "91     BN9944             99-44          2.16              Control   \n",
       "92     BN9947             99-47          2.50              Control   \n",
       "93     BN9966             99-66          2.00              Control   \n",
       "\n",
       "                               path_autopsy_second_dx  \\\n",
       "0               Charcot-Marie-Tooth disease (history)   \n",
       "1                          Seizure disorder (history)   \n",
       "2                  Non-diagnostic Alzheimer's changes   \n",
       "3                  Non-diagnostic Alzheimer's changes   \n",
       "4                  Non-diagnostic Alzheimer's changes   \n",
       "..                                                ...   \n",
       "89  Microscopic changes of Alzheimer's disease, in...   \n",
       "90  Microscopic changes of Alzheimer's disease, in...   \n",
       "91                 Non-diagnostic Alzheimer's changes   \n",
       "92                 Non-diagnostic Alzheimer's changes   \n",
       "93                                     Incidental LBs   \n",
       "\n",
       "                                path_autopsy_third_dx  \\\n",
       "0                        GBA L444P/WT, L444P mutation   \n",
       "1                                                None   \n",
       "2                                                 CAA   \n",
       "3                                                CWMR   \n",
       "4           Argyrophilic grains, mesial temporal lobe   \n",
       "..                                                ...   \n",
       "89  Focal non-specific glial tauopathy, cortex of ...   \n",
       "90                       Incidental Lewy body disease   \n",
       "91  Alzheimer Type II astrocytosis consistent with...   \n",
       "92                     Alzheimer type II astrocytosis   \n",
       "93                 Non-diagnostic Alzheimer's changes   \n",
       "\n",
       "                         path_autopsy_fourth_dx path_autopsy_fifth_dx  \\\n",
       "0                                          None                  None   \n",
       "1                                          None                  None   \n",
       "2                                          None                  None   \n",
       "3                                          None                  None   \n",
       "4                                    Infarct(s)                  CWMR   \n",
       "..                                          ...                   ...   \n",
       "89                                         None                  None   \n",
       "90                                         None                  None   \n",
       "91                                         None                  None   \n",
       "92                                      Inc LBs                  None   \n",
       "93  Argyrophilic grains in mesial temporal lobe                  None   \n",
       "\n",
       "                                path_autopsy_sixth_dx  \\\n",
       "0                                                None   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4   Several microscopic foci of cerebellar cortica...   \n",
       "..                                                ...   \n",
       "89                                               None   \n",
       "90                                               None   \n",
       "91                                               None   \n",
       "92                                               None   \n",
       "93                                               None   \n",
       "\n",
       "                     path_autopsy_seventh_dx  ...   path_nia_ri  \\\n",
       "0                                       None  ...           Nan   \n",
       "1                                       None  ...           Nan   \n",
       "2                                       None  ...           Nan   \n",
       "3                                       None  ...           Nan   \n",
       "4   Etat crible, putamen and caudate nucleus  ...           Nan   \n",
       "..                                       ...  ...           ...   \n",
       "89                                      None  ...  Intermediate   \n",
       "90                                      None  ...           Nan   \n",
       "91                                      None  ...           Nan   \n",
       "92                                      None  ...           Nan   \n",
       "93                                      None  ...           Nan   \n",
       "\n",
       "    path_nia_aa_a path_nia_aa_b path_nia_aa_c TDP43  \\\n",
       "0              NA           Nan           Nan    NA   \n",
       "1              NA           Nan           Nan    NA   \n",
       "2              NA           Nan           Nan    NA   \n",
       "3              NA           Nan           Nan    No   \n",
       "4              NA           Nan           Nan    NA   \n",
       "..            ...           ...           ...   ...   \n",
       "89             NA           Nan           Nan    No   \n",
       "90             NA           Nan           Nan    No   \n",
       "91             NA           Nan           Nan    NA   \n",
       "92             NA           Nan           Nan    NA   \n",
       "93             NA           Nan           Nan    NA   \n",
       "\n",
       "   arteriolosclerosis_severity_scale  amyloid_angiopathy_severity_scale  \\\n",
       "0                                Nan                                Nan   \n",
       "1                                Nan                                Nan   \n",
       "2                               Mild                           Moderate   \n",
       "3                           Moderate                               Mild   \n",
       "4                               Mild                                Nan   \n",
       "..                               ...                                ...   \n",
       "89                           Unknown                            Unknown   \n",
       "90                           Unknown                            Unknown   \n",
       "91                           Unknown                                Nan   \n",
       "92                           Unknown                                Nan   \n",
       "93                           Unknown                               Mild   \n",
       "\n",
       "   path_ad_level dig_slide_avail quant_path_avail  \n",
       "0        Unknown             Nan              Nan  \n",
       "1        Unknown             Nan              Nan  \n",
       "2        Unknown             Nan              Nan  \n",
       "3        Unknown             Nan              Nan  \n",
       "4        Unknown             Nan              Nan  \n",
       "..           ...             ...              ...  \n",
       "89       Unknown             Nan              Nan  \n",
       "90       Unknown             Nan              Nan  \n",
       "91       Unknown             Nan              Nan  \n",
       "92       Unknown             Nan              Nan  \n",
       "93       Unknown             Nan              Nan  \n",
       "\n",
       "[94 rows x 36 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "mckeith_map = {'l. Olfactory Bulb-Only':\"Olfactory bulb only\", 'lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'llb. Limbic Predominant':\"Limbic (transitional)\", 'lV. Neocortical':\"Neocortical\",\n",
    "       'lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\", \"Unknown\":NULL}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n",
    "CLINPATH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chat GPT help\n",
    "\n",
    "```python\n",
    "\n",
    "mapping = {\n",
    "    'PD/Dem': ['Parkinson\\'s disease with dementia', 'Dementia with Lewy bodies'],\n",
    "    'Control': ['Control, no misfolded protein or significant vascular pathology'],\n",
    "    'Control (MCI)': ['Control, Low level AD neuropathological change', \n",
    "                       'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)'],\n",
    "    'PD/ND': ['Parkinson\\'s disease'],\n",
    "    'Control, Non-motoric': ['Control, no misfolded protein or significant vascular pathology'],\n",
    "    'PDD': ['Parkinson\\'s disease with dementia', 'Dementia with Lewy bodies'],\n",
    "    'MCI': ['Alzheimer\\'s disease (intermediate level neuropathological change)', \n",
    "            'Control, Low level AD neuropathological change'],\n",
    "    'PD (MCI)': ['Parkinson\\'s disease with dementia', 'Parkinson\\'s disease'],\n",
    "    'Control (history)': ['Control, no misfolded protein or significant vascular pathology'],\n",
    "    'Control (diseased)': ['Control, Cerebrovascular disease (atherosclerosis)', \n",
    "                           'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', \n",
    "                           'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', \n",
    "                           'Other neurological disorder'],\n",
    "    'MCI (clinical history)': ['Control, Low level AD neuropathological change', \n",
    "                               'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)'],\n",
    "    'Control (clinical history)': ['Control, no misfolded protein or significant vascular pathology']\n",
    "}\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path_autopsy_main_mapper = {\n",
    "#     'PD/Dem': 'Parkinson\\'s disease with dementia',\n",
    "#     'Control': 'Control, no misfolded protein or significant vascular pathology',\n",
    "#     'Control (MCI)': 'Control, Low level AD neuropathological change',\n",
    "#     'PD/ND': 'Parkinson\\'s disease',\n",
    "#     'Control, Non-motoric': 'Control, no misfolded protein or significant vascular pathology',\n",
    "#     'PDD': 'Parkinson\\'s disease with dementia',\n",
    "#     'MCI': 'Alzheimer\\'s disease (intermediate level neuropathological change)',\n",
    "#     'PD (MCI)': 'Parkinson\\'s disease with dementia',\n",
    "#     'Control (history)': 'Control, no misfolded protein or significant vascular pathology',\n",
    "#     'Control (diseased)': 'Control, Cerebrovascular disease (atherosclerosis)',\n",
    "#     'MCI (clinical history)': 'Control, Low level AD neuropathological change',\n",
    "#     'Control (clinical history)': 'Control, no misfolded protein or significant vascular pathology'\n",
    "# }\n",
    "\n",
    "# CLINPATH['path_autopsy_dx_main'] = CLINPATH['path_autopsy_dx_main'].replace(path_autopsy_main_mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name_source</th>\n",
       "      <th>file_name</th>\n",
       "      <th>local_path</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN0009_S2_L001_I1_001.fastq.gz</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>Index 1 Lane 1</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN0009_S2_L001_I2_001.fastq.gz</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>Index 2 Lane 1</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN0009_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>Read 1 Lane 1</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN0009_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>Read 2 Lane 1</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN0009_S2_L002_I1_001.fastq.gz</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>Index 1 Lane 2</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN1762</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch31</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN1762_S3_L003_R2_001.fastq.gz</td>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...</td>\n",
       "      <td>Read 2 Lane 3</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN1762</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch31</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN1762_S3_L004_I1_001.fastq.gz</td>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...</td>\n",
       "      <td>Index 1 Lane 4</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN1762</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch31</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN1762_S3_L004_I2_001.fastq.gz</td>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...</td>\n",
       "      <td>Index 2 Lane 4</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN1762</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch31</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN1762_S3_L004_R1_001.fastq.gz</td>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...</td>\n",
       "      <td>Read 1 Lane 4</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN1762</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch31</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BN1762_S3_L004_R2_001.fastq.gz</td>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...</td>\n",
       "      <td>Read 2 Lane 4</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1552 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sample_id source_sample_id replicate  replicate_count  \\\n",
       "0     PD_BN0009_MTG_snRNAseq_rep1           BN0009      Rep1                1   \n",
       "1     PD_BN0009_MTG_snRNAseq_rep1           BN0009      Rep1                1   \n",
       "2     PD_BN0009_MTG_snRNAseq_rep1           BN0009      Rep1                1   \n",
       "3     PD_BN0009_MTG_snRNAseq_rep1           BN0009      Rep1                1   \n",
       "4     PD_BN0009_MTG_snRNAseq_rep1           BN0009      Rep1                1   \n",
       "...                           ...              ...       ...              ...   \n",
       "1547  HC_BN1762_MTG_snRNAseq_rep1           BN1762      Rep1                1   \n",
       "1548  HC_BN1762_MTG_snRNAseq_rep1           BN1762      Rep1                1   \n",
       "1549  HC_BN1762_MTG_snRNAseq_rep1           BN1762      Rep1                1   \n",
       "1550  HC_BN1762_MTG_snRNAseq_rep1           BN1762      Rep1                1   \n",
       "1551  HC_BN1762_MTG_snRNAseq_rep1           BN1762      Rep1                1   \n",
       "\n",
       "      repeated_sample    batch file_type                file_name_source  \\\n",
       "0                   0   batch1     fastq  BN0009_S2_L001_I1_001.fastq.gz   \n",
       "1                   0   batch1     fastq  BN0009_S2_L001_I2_001.fastq.gz   \n",
       "2                   0   batch1     fastq  BN0009_S2_L001_R1_001.fastq.gz   \n",
       "3                   0   batch1     fastq  BN0009_S2_L001_R2_001.fastq.gz   \n",
       "4                   0   batch1     fastq  BN0009_S2_L002_I1_001.fastq.gz   \n",
       "...               ...      ...       ...                             ...   \n",
       "1547                0  batch31     fastq  BN1762_S3_L003_R2_001.fastq.gz   \n",
       "1548                0  batch31     fastq  BN1762_S3_L004_I1_001.fastq.gz   \n",
       "1549                0  batch31     fastq  BN1762_S3_L004_I2_001.fastq.gz   \n",
       "1550                0  batch31     fastq  BN1762_S3_L004_R1_001.fastq.gz   \n",
       "1551                0  batch31     fastq  BN1762_S3_L004_R2_001.fastq.gz   \n",
       "\n",
       "                                             file_name  \\\n",
       "0     PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz   \n",
       "1     PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz   \n",
       "2     PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz   \n",
       "3     PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz   \n",
       "4     PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz   \n",
       "...                                                ...   \n",
       "1547  HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz   \n",
       "1548  HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz   \n",
       "1549  HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz   \n",
       "1550  HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz   \n",
       "1551  HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz   \n",
       "\n",
       "                                             local_path file_description  \\\n",
       "0     /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...   Index 1 Lane 1   \n",
       "1     /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...   Index 2 Lane 1   \n",
       "2     /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...    Read 1 Lane 1   \n",
       "3     /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...    Read 2 Lane 1   \n",
       "4     /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...   Index 1 Lane 2   \n",
       "...                                                 ...              ...   \n",
       "1547  /data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...    Read 2 Lane 3   \n",
       "1548  /data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...   Index 1 Lane 4   \n",
       "1549  /data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...   Index 2 Lane 4   \n",
       "1550  /data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...    Read 1 Lane 4   \n",
       "1551  /data/neurogen/ASAP/scRNAseq/data/batch31_ASAP...    Read 2 Lane 4   \n",
       "\n",
       "                              file_MD5 technology omic adjustment content  \\\n",
       "0     bc037ea92d5359f3da1ba6da4f974e1c         SN  RNA        Raw   Reads   \n",
       "1     0c8ef844b76212e82428af03d341048b         SN  RNA        Raw   Reads   \n",
       "2     c6a41f29849152df2ba2391dc559149b         SN  RNA        Raw   Reads   \n",
       "3     4d610fdf5a72c8206b2f1505d8fcecd7         SN  RNA        Raw   Reads   \n",
       "4     e34917b3ec01cd09ea8b52c82db1dc83         SN  RNA        Raw   Reads   \n",
       "...                                ...        ...  ...        ...     ...   \n",
       "1547  887e3d000587b2774ce3cd7022078577         SN  RNA        Raw   Reads   \n",
       "1548  44ac0c6d4f7f3b4502eead39564c8723         SN  RNA        Raw   Reads   \n",
       "1549  dc114df490e1857072462ca090799005         SN  RNA        Raw   Reads   \n",
       "1550  e24eddb1ac0b1f4d2bdd32b5f69dae27         SN  RNA        Raw   Reads   \n",
       "1551  1d316a45079ccb5213fee401535beb02         SN  RNA        Raw   Reads   \n",
       "\n",
       "      time  header  annotation  configuration_file  \n",
       "0      NaN     NaN         NaN                 NaN  \n",
       "1      NaN     NaN         NaN                 NaN  \n",
       "2      NaN     NaN         NaN                 NaN  \n",
       "3      NaN     NaN         NaN                 NaN  \n",
       "4      NaN     NaN         NaN                 NaN  \n",
       "...    ...     ...         ...                 ...  \n",
       "1547   NaN     NaN         NaN                 NaN  \n",
       "1548   NaN     NaN         NaN                 NaN  \n",
       "1549   NaN     NaN         NaN                 NaN  \n",
       "1550   NaN     NaN         NaN                 NaN  \n",
       "1551   NaN     NaN         NaN                 NaN  \n",
       "\n",
       "[1552 rows x 20 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA['replicate'] = DATA['replicate'].apply(lambda x: x[0].upper()+x[1:] if x is not NULL else x)\n",
    "DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "v2_path = metadata_path / \"v2\"\n",
    "STUDY.to_csv(v2_path / f\"STUDY.csv\", index=False)\n",
    "SAMPLE.to_csv(v2_path / f\"SAMPLE.csv\", index=False)\n",
    "SUBJECT.to_csv(v2_path / f\"SUBJECT.csv\", index=False)\n",
    "PROTOCOL.to_csv(v2_path / f\"PROTOCOL.csv\", index=False)\n",
    "CLINPATH.to_csv(v2_path / f\"CLINPATH.csv\", index=False)\n",
    "DATA.to_csv(v2_path / f\"DATA.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **15 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 86/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 49/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 59/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 63/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 93/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 7/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 64/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 11/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 94/94 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*education_level*_:  invalid values 💩'14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other', 'NA'\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **29 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_dx_main: 1/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 36/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 69/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 81/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 91/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 44/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 21/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 66/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 32/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 70/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 88/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 94/94 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **4 Fields with invalid entries:**\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'Control (clinical history)'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder', 'NA'\n",
      "- _*path_mckeith*_:  invalid values 💩'Llb. Limbic Predominant', 'Lll. Brainstem/Limbic', 'Lla. Brainstem Predominant', 'LV. Neocortical', 'L. Olfactory Bulb-Only'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown', 'NA'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'3.0', '0.0', '1.0', '2.0'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
      "- _*TDP43*_:  invalid values 💩'No', 'Undx', 'Yes'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in SAMPLE: alternate_sample_id**\n",
      "🚨⚠️❗ **9 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- development_stage_ontology_term_id: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- tissue_ontology_term_id: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- suspension_type: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 97/97 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*sequencing_length*_:  invalid values 💩'120'\n",
      "    - valid ➡️ '25', '50', '100', '150', 'NA'\n",
      "🚨⚠️❗ **Extra field in SAMPLE: cellranger_outs_folder**\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time: 1552/1552 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 1552/1552 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 1552/1552 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 1552/1552 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "🚨⚠️❗ **Extra field in DATA: source_sample_id**\n",
      "🚨⚠️❗ **Extra field in DATA: file_name_source**\n",
      "🚨⚠️❗ **Extra field in DATA: local_path**\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "tables = CDE['Table'].unique()\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(v2_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table\n",
    "    # df.to_csv(v1_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### save extras as auxillary tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved SAMPLE_auxiliary.csv\n",
      "Saved DATA_auxiliary.csv\n"
     ]
    }
   ],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "v2_path = metadata_path / \"v2\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v2_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v2_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v2->v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 94/94 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 97/97 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- tissue_ontology_term_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 97/97 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 97/97 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 1552/1552 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 1552/1552 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 1552/1552 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 86/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 49/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 63/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 93/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 7/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 64/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_dx_main: 1/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 36/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 69/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 81/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 91/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 44/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 21/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 66/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 32/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 70/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 88/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 94/94 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **5 Fields with invalid entries:**\n",
      "- _*education_level*_:  invalid values 💩'14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other', 'NA'\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'Control (clinical history)'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder', 'NA'\n",
      "- _*path_mckeith*_:  invalid values 💩'Llb. Limbic Predominant', 'Lll. Brainstem/Limbic', 'Lla. Brainstem Predominant', 'LV. Neocortical', 'L. Olfactory Bulb-Only'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown', 'NA'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'3.0', '0.0', '1.0', '2.0'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
      "- _*TDP43*_:  invalid values 💩'No', 'Undx', 'Yes'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 3/3 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 3/3 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>prodromal_motor_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field        condition_id intervention_name intervention_id protocol_id  \\\n",
       "0           idiopathic_pd      Case-Control            Case          NA   \n",
       "1         healthy_control      Case-Control         Control          NA   \n",
       "2      prodromal_motor_pd      Case-Control           Other          NA   \n",
       "\n",
       "Field intervention_aux_table  \n",
       "0                         NA  \n",
       "1                         NA  \n",
       "2                         NA  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-SCHERZER</td>\n",
       "      <td>Scherzer, Dong, and Levin</td>\n",
       "      <td>Parkinson5D: deconstructing proximal disease m...</td>\n",
       "      <td>sn_rnaseq_mtg</td>\n",
       "      <td>PD5D_MTG_snRNAseq</td>\n",
       "      <td>Here we will develop a molecular atlas of Park...</td>\n",
       "      <td>Dr.Clemens Scherzer</td>\n",
       "      <td>Clemens.scherzer@yale.edu</td>\n",
       "      <td>Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...</td>\n",
       "      <td>Dr.Xianjun Dong</td>\n",
       "      <td>...</td>\n",
       "      <td>94</td>\n",
       "      <td>MTG</td>\n",
       "      <td>PD and control postmortem brains</td>\n",
       "      <td>ASAP Access &amp; Use Policy</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>v3.0_20241102</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name              ASAP_lab_name  \\\n",
       "0  TEAM-SCHERZER  Scherzer, Dong, and Levin   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  Parkinson5D: deconstructing proximal disease m...   sn_rnaseq_mtg   \n",
       "\n",
       "     project_dataset                                project_description  \\\n",
       "0  PD5D_MTG_snRNAseq  Here we will develop a molecular atlas of Park...   \n",
       "\n",
       "          PI_full_name                   PI_email  \\\n",
       "0  Dr.Clemens Scherzer  Clemens.scherzer@yale.edu   \n",
       "\n",
       "                                   contributor_names   submitter_name  ...  \\\n",
       "0  Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...  Dr.Xianjun Dong  ...   \n",
       "\n",
       "  number_samples sample_types                  types_of_samples  \\\n",
       "0             94          MTG  PD and control postmortem brains   \n",
       "\n",
       "                DUA_version  \\\n",
       "0  ASAP Access & Use Policy   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "  PI_google_scholar_id preprocessing_references metadata_version_date  \\\n",
       "0                   NA                       NA         v3.0_20241102   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = v3_tables['STUDY']\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n",
    "Team Scherzer uploaded their data originally with gcloud tools which by default don't create md5s.  We have prefiously checked the crc hashes instead.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scherzer\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/scherzer-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-scherzer@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "\n",
    "# # make sure to get ALL the fastq files in the bucket\n",
    "# prefix = \"fastqs/*.gz\"\n",
    "# bucket_files_md5 = get_md5_hashes( bucket, prefix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # def check_md5_sums()\n",
    "\n",
    "\n",
    "# checksum = v3_tables['DATA'][['file_name','file_MD5']]\n",
    "# checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "# checksum['check1'] = checksum['file_MD5']\n",
    "# checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "# #empty means success!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## prep metadata in raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/upload\n",
    "- 2. copy metadata/upload to dataset upload (upload subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil command succeeded: gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "metadata_subdir = \"metadata\"\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "dataset_name_ = \"sn-rnaseq-mfc\"\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name_}\"\n",
    "\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons mv -r gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload/upload\n",
      "gsutil command succeeded: \n"
     ]
    }
   ],
   "source": [
    "metadata_subdir2 = \"metadata/upload\"\n",
    "bucket = current_files[0].split(\"/\")[2]\n",
    "\n",
    "for file in current_files:\n",
    "    if file == \"\":\n",
    "        continue\n",
    "    file_nm = Path(file).name\n",
    "    is_dir = not file_nm.__contains__(\".\")\n",
    "\n",
    "    source = f\"gs://{bucket}/{metadata_subdir}/{file_nm}\"\n",
    "\n",
    "    destination = f\"gs://{bucket}/{metadata_subdir2}/{file_nm}\"\n",
    "    gsutil_mv(source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload\n",
      "gsutil command succeeded: gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload/upload/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "metadata_subdir = \"metadata/upload\"\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archive the uploaded metadata locally\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_source = f\"gs://{bucket}/{metadata_subdir}\"\n",
    "destination = f\"{metadata_path}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons cp -r gs://asap-raw-team-scherzer-pmdbs-sn-rnaseq-mfc/metadata/upload /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata\n",
      "gsutil command succeeded: \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "is_dir = True\n",
    "gsutil_cp(file_source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------\n",
    "## Create metadata package\n",
    "\n",
    "This will copy the final updated to v3.0 metadata to `asap-could-processing-resources`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(PosixPath('/Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata'),\n",
       " PosixPath('/Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/scherzer-pmdbs-sn-rnaseq-mtg'))"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_source = metadata_path\n",
    "\n",
    "\n",
    "source = \"pmdbs\"\n",
    "archive_root = Path.home() / \"Projects/ASAP/asap-crn-metadata/datasets\"\n",
    "dataset_path = archive_root / f\"{team}-{source}-{dataset_name}\"\n",
    "# bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "metadata_source, dataset_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping empty folder /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata/v1\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata/v2 to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/scherzer-pmdbs-sn-rnaseq-mtg/metadata/v2\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata/v3 to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/scherzer-pmdbs-sn-rnaseq-mtg/metadata/v3\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata/og to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/scherzer-pmdbs-sn-rnaseq-mtg/metadata/og\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/scherzer/sn-rnaseq-mtg/metadata/upload to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/scherzer-pmdbs-sn-rnaseq-mtg/metadata/upload\n"
     ]
    }
   ],
   "source": [
    "fnms = create_metadata_package(metadata_source, dataset_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "generate ASAP IDs + transfering back to raw data bucket via `asap-crn-metadata` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## transfer metadata to raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/archive\n",
    "- 2. copy package to metadata/ . i.e. /og/*.csv, /v??/*.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from utils.checksums import extract_md5_from_details, extract_md5_from_details2,  extract_crc32c_from_details2, extract_hashes_from_gcloudstorage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_hash = Path.home() / \"Projects/ASAP/team-scherzer/scherzer_hash.log\"\n",
    "\n",
    "source_file_crc = extract_crc32c_from_details2(source_hash)\n",
    "source_file_md5 = extract_md5_from_details2(source_hash)\n",
    "\n",
    "source_hash.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_file_crc,source_file_md5 = extract_hashes_from_gcloudstorage(source_hash)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## crc32c hashes \n",
    "\n",
    "due to gcloud defaults we have to use crc hashes to validate the fastqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_files_md5 = extract_md5_from_details2(\"scherzer_hexhash.log\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_files_crc = extract_crc32c_from_details2(\"scherzer_hexhash.log\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_src_crc'] = checksum['file_name'].map(source_file_crc)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_bucket_crc'] = checksum['file_name'].map(bucket_files_crc)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['meta_MD5'] = checksum['file_MD5'].str.strip()\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_src_md5'] = checksum['file_name'].map(source_file_md5)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_bucket_md5'] = checksum['file_name'].map(bucket_files_md5)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "checksum = DATA[['file_name','file_MD5']]\n",
    "\n",
    "\n",
    "checksum['check_src_crc'] = checksum['file_name'].map(source_file_crc)\n",
    "checksum['check_bucket_crc'] = checksum['file_name'].map(bucket_files_crc)\n",
    "\n",
    "\n",
    "checksum['meta_MD5'] = checksum['file_MD5'].str.strip()\n",
    "checksum['check_src_md5'] = checksum['file_name'].map(source_file_md5)\n",
    "checksum['check_bucket_md5'] = checksum['file_name'].map(bucket_files_md5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>check_src_crc</th>\n",
       "      <th>check_bucket_crc</th>\n",
       "      <th>meta_MD5</th>\n",
       "      <th>check_src_md5</th>\n",
       "      <th>check_bucket_md5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>c99915b4</td>\n",
       "      <td>c99915b4</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>f0af5e35</td>\n",
       "      <td>f0af5e35</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>151d9d97</td>\n",
       "      <td>151d9d97</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>58bc244e</td>\n",
       "      <td>58bc244e</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>600d0703</td>\n",
       "      <td>600d0703</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>93e1775e</td>\n",
       "      <td>93e1775e</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>a361f850</td>\n",
       "      <td>a361f850</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>b617330f</td>\n",
       "      <td>b617330f</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>e49ff1eb</td>\n",
       "      <td>e49ff1eb</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>55d83b85</td>\n",
       "      <td>55d83b85</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1552 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             file_name  \\\n",
       "0     PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz   \n",
       "1     PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz   \n",
       "2     PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz   \n",
       "3     PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz   \n",
       "4     PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz   \n",
       "...                                                ...   \n",
       "1547  HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz   \n",
       "1548  HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz   \n",
       "1549  HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz   \n",
       "1550  HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz   \n",
       "1551  HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz   \n",
       "\n",
       "                              file_MD5 check_src_crc check_bucket_crc  \\\n",
       "0     bc037ea92d5359f3da1ba6da4f974e1c      c99915b4         c99915b4   \n",
       "1     0c8ef844b76212e82428af03d341048b      f0af5e35         f0af5e35   \n",
       "2     c6a41f29849152df2ba2391dc559149b      151d9d97         151d9d97   \n",
       "3     4d610fdf5a72c8206b2f1505d8fcecd7      58bc244e         58bc244e   \n",
       "4     e34917b3ec01cd09ea8b52c82db1dc83      600d0703         600d0703   \n",
       "...                                ...           ...              ...   \n",
       "1547  887e3d000587b2774ce3cd7022078577      93e1775e         93e1775e   \n",
       "1548  44ac0c6d4f7f3b4502eead39564c8723      a361f850         a361f850   \n",
       "1549  dc114df490e1857072462ca090799005      b617330f         b617330f   \n",
       "1550  e24eddb1ac0b1f4d2bdd32b5f69dae27      e49ff1eb         e49ff1eb   \n",
       "1551  1d316a45079ccb5213fee401535beb02      55d83b85         55d83b85   \n",
       "\n",
       "                              meta_MD5                     check_src_md5  \\\n",
       "0     bc037ea92d5359f3da1ba6da4f974e1c  bc037ea92d5359f3da1ba6da4f974e1c   \n",
       "1     0c8ef844b76212e82428af03d341048b  0c8ef844b76212e82428af03d341048b   \n",
       "2     c6a41f29849152df2ba2391dc559149b  c6a41f29849152df2ba2391dc559149b   \n",
       "3     4d610fdf5a72c8206b2f1505d8fcecd7  4d610fdf5a72c8206b2f1505d8fcecd7   \n",
       "4     e34917b3ec01cd09ea8b52c82db1dc83  e34917b3ec01cd09ea8b52c82db1dc83   \n",
       "...                                ...                               ...   \n",
       "1547  887e3d000587b2774ce3cd7022078577  887e3d000587b2774ce3cd7022078577   \n",
       "1548  44ac0c6d4f7f3b4502eead39564c8723  44ac0c6d4f7f3b4502eead39564c8723   \n",
       "1549  dc114df490e1857072462ca090799005  dc114df490e1857072462ca090799005   \n",
       "1550  e24eddb1ac0b1f4d2bdd32b5f69dae27  e24eddb1ac0b1f4d2bdd32b5f69dae27   \n",
       "1551  1d316a45079ccb5213fee401535beb02  1d316a45079ccb5213fee401535beb02   \n",
       "\n",
       "     check_bucket_md5  \n",
       "0                 NaN  \n",
       "1                 NaN  \n",
       "2                 NaN  \n",
       "3                 NaN  \n",
       "4                 NaN  \n",
       "...               ...  \n",
       "1547              NaN  \n",
       "1548              NaN  \n",
       "1549              NaN  \n",
       "1550              NaN  \n",
       "1551              NaN  \n",
       "\n",
       "[1552 rows x 7 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
