{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation \n",
    "# Team Scherzer. ASAP CRN Metadata validation \n",
    "\n",
    "10 Oct 2024\n",
    "Andy Henrie\n",
    "\n",
    "## scrnaseq_hybsel\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully imported\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, create_upload_medadata_package\n",
    "from checksums import get_md5_hashes, authenticate_with_service_account\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted as v2.1 (actually early v3.0.0 beta)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Team Scherzer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"scherzer\"\n",
    "dataset_name = \"sn-rnaseq-mtg-hybsel\"\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "og_path = metadata_path / \"og\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\",\"DATA\"]\n",
    "\n",
    "excel_path = og_path / \"ASAP CDE v3.0.0-beta_sks_new.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\")#,header=1)#.drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\")#,header=1)#.drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\")#,header=1)#.drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\")#,header=1)#.drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\")#,header=1)#.drop(columns=\"Field\")\n",
    "DATA = pd.read_excel(excel_path,sheet_name=\"DATA\")#,header=1)#.drop(columns=\"Field\")\n",
    "\n",
    "metadata_version = \"v2\"\n",
    "\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create csvs \n",
    "STUDY = STUDY.rename({'submittor_email':'submitter_email'}, axis=1)\n",
    "\n",
    "STUDY.to_csv(og_path / f\"STUDY.csv\", index=False)\n",
    "SAMPLE.to_csv(og_path / f\"SAMPLE.csv\", index=False)\n",
    "SUBJECT.to_csv(og_path / f\"SUBJECT.csv\", index=False)\n",
    "PROTOCOL.to_csv(og_path / f\"PROTOCOL.csv\", index=False)\n",
    "CLINPATH.to_csv(og_path / f\"CLINPATH.csv\", index=False)\n",
    "DATA.to_csv(og_path / f\"DATA.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "ðŸš¨âš ï¸â— **Missing Optional Fields in STUDY: PI_ORCHID**\n",
      "ðŸš¨âš ï¸â— **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- metadata_version_date: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **1 Fields with invalid entries:**\n",
      "- _*ASAP_team_name*_:  invalid values ðŸ’©'Team Scherzer'\n",
      "    - valid âž¡ï¸ 'TEAM-LEE', 'TEAM-HAFLER', 'TEAM-HARDY', 'TEAM-JAKOBSSON', 'TEAM-SCHERZER', 'TEAM-SULZER', 'TEAM-VOET', 'TEAM-WOOD', 'NA'\n",
      "ðŸš¨âš ï¸â— **Extra field in STUDY: PI_ORCID**\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "ðŸš¨âš ï¸â— **6 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- sample_collection_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_extraction_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- lib_prep_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- data_processing_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "ðŸš¨âš ï¸â— **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- ethnicity: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 86/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 49/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 59/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 63/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 93/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 7/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 64/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 11/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **3 Fields with invalid entries:**\n",
      "- _*hx_dementia_mci*_:  invalid values ðŸ’©'yes', 'no'\n",
      "    - valid âž¡ï¸ 'Yes', 'No', 'NA'\n",
      "- _*hx_melanoma*_:  invalid values ðŸ’©'Present'\n",
      "    - valid âž¡ï¸ 'Yes', 'No', 'NA'\n",
      "- _*education_level*_:  invalid values ðŸ’©'14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'\n",
      "    - valid âž¡ï¸ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other', 'NA'\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "ðŸš¨âš ï¸â— **29 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_dx_main: 1/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 36/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 69/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 81/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 91/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 44/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 17/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 43/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 32/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 70/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 88/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **7 Fields with invalid entries:**\n",
      "- _*path_autopsy_dx_main*_:  invalid values ðŸ’©'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'control (clinical history)'\n",
      "    - valid âž¡ï¸ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder', 'NA'\n",
      "- _*path_cerad*_:  invalid values ðŸ’©'moderate', 'sparse', 'frequent'\n",
      "    - valid âž¡ï¸ 'None', 'Sparse', 'Moderate', 'Frequent', 'NA'\n",
      "- _*path_thal*_:  invalid values ðŸ’©'Phase 0 (A0)', 'Phase 1 (A1)', 'Phase 5 (A3)', 'Phase 3 (A2)', 'Phase 2 (A1)', 'Missing/unknown', 'Phase 4 (A3)'\n",
      "    - valid âž¡ï¸ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5', 'NA'\n",
      "- _*path_mckeith*_:  invalid values ðŸ’©'llb. Limbic Predominant', 'lll. Brainstem/Limbic', '0. No Lewy bodies', 'lla. Brainstem Predominant', 'lV. Neocortical', 'l. Olfactory Bulb-Only'\n",
      "    - valid âž¡ï¸ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown', 'NA'\n",
      "- _*sn_neuronal_loss*_:  invalid values ðŸ’©'3.0', '0.0', '1.0', '2.0'\n",
      "    - valid âž¡ï¸ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
      "- _*path_nia_aa_a*_:  invalid values ðŸ’©'Low ADNC', 'Not AD', 'Intermediate ADNC'\n",
      "    - valid âž¡ï¸ 'A0', 'A1', 'A2', 'A3', 'NA'\n",
      "- _*TDP43*_:  invalid values ðŸ’©'no', 'undx', 'yes'\n",
      "    - valid âž¡ï¸ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "ðŸš¨âš ï¸â— **6 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- suspension_type: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_sample_id: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **4 Fields with invalid entries:**\n",
      "- _*region_level_1*_:  invalid values ðŸ’©'Temporal Lobe'\n",
      "    - valid âž¡ï¸ 'Frontal lobe', 'Temporal lobe', 'Parietal lobe', 'Occipital lobe', 'Cingulate gyrus', 'Striatum', 'Thalamus', 'Midbrain', 'Pons', 'Medulla', 'Cerebellum', 'Intestine', 'Unknown', 'NA'\n",
      "- _*molecular_source*_:  invalid values ðŸ’©'PolyA_RNA'\n",
      "    - valid âž¡ï¸ 'Total RNA', 'PolyA RNA', 'Cytoplasmic RNA', 'Nuclear RNA', 'Genomic DNA', 'Protein', 'Other', 'NA'\n",
      "- _*sequencing_length*_:  invalid values ðŸ’©'200'\n",
      "    - valid âž¡ï¸ '25', '50', '100', '150', 'NA'\n",
      "- _*sex_ontology_term_id*_:  invalid values ðŸ’©'PATO:0000384', 'PATO:0000383'\n",
      "    - valid âž¡ï¸ 'PATO:0000384 (male)', 'PATO:0000383 (female)', 'Unknown', 'NA'\n",
      "ðŸš¨âš ï¸â— **Extra field in SAMPLE: hybrid_selection**\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "ðŸš¨âš ï¸â— **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time: 808/808 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 808/808 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 808/808 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 808/808 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "tables = CDE['Table'].unique()\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(og_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = prep_table(dfs['STUDY'],CDEv2)\n",
    "STUDY['team_dataset_id'] = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "STUDY['ASAP_team_name'] = \"TEAM-SCHERZER\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT = prep_table(dfs['SUBJECT'],CDEv2)\n",
    "SUBJECT['hx_melanoma'] = SUBJECT['hx_melanoma'].replace({\"Present\":\"Yes\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>path_autopsy_dx_main</th>\n",
       "      <th>path_autopsy_second_dx</th>\n",
       "      <th>path_autopsy_third_dx</th>\n",
       "      <th>path_autopsy_fourth_dx</th>\n",
       "      <th>path_autopsy_fifth_dx</th>\n",
       "      <th>path_autopsy_sixth_dx</th>\n",
       "      <th>path_autopsy_seventh_dx</th>\n",
       "      <th>...</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>dig_slide_avail</th>\n",
       "      <th>quant_path_avail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>4.00</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Charcot-Marie-Tooth disease (history)</td>\n",
       "      <td>GBA L444P/WT, L444P mutation</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN0329</td>\n",
       "      <td>03-29</td>\n",
       "      <td>4.50</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Seizure disorder (history)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN0339</td>\n",
       "      <td>03-39</td>\n",
       "      <td>2.75</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>CAA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mild</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN0341</td>\n",
       "      <td>03-41</td>\n",
       "      <td>2.50</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>CWMR</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Mild</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN0347</td>\n",
       "      <td>03-47</td>\n",
       "      <td>3.50</td>\n",
       "      <td>Control (MCI)</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Argyrophilic grains, mesial temporal lobe</td>\n",
       "      <td>Infarct(s)</td>\n",
       "      <td>CWMR</td>\n",
       "      <td>Several microscopic foci of cerebellar cortica...</td>\n",
       "      <td>Etat crible, putamen and caudate nucleus</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mild</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>BN2003</td>\n",
       "      <td>20-03</td>\n",
       "      <td>2.65</td>\n",
       "      <td>PD/Dem</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Focal non-specific glial tauopathy, cortex of ...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>BN2015</td>\n",
       "      <td>20-15</td>\n",
       "      <td>5.40</td>\n",
       "      <td>Control (history)</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Incidental Lewy body disease</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>BN9944</td>\n",
       "      <td>99-44</td>\n",
       "      <td>2.16</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Alzheimer Type II astrocytosis consistent with...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>BN9947</td>\n",
       "      <td>99-47</td>\n",
       "      <td>2.50</td>\n",
       "      <td>Control</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Alzheimer type II astrocytosis</td>\n",
       "      <td>Inc LBs</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>BN9966</td>\n",
       "      <td>99-66</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Control</td>\n",
       "      <td>Incidental LBs</td>\n",
       "      <td>Non-diagnostic Alzheimer's changes</td>\n",
       "      <td>Argyrophilic grains in mesial temporal lobe</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mild</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id  duration_pmi path_autopsy_dx_main  \\\n",
       "0      BN0009             00-09          4.00               PD/Dem   \n",
       "1      BN0329             03-29          4.50               PD/Dem   \n",
       "2      BN0339             03-39          2.75              Control   \n",
       "3      BN0341             03-41          2.50              Control   \n",
       "4      BN0347             03-47          3.50        Control (MCI)   \n",
       "..        ...               ...           ...                  ...   \n",
       "89     BN2003             20-03          2.65               PD/Dem   \n",
       "90     BN2015             20-15          5.40    Control (history)   \n",
       "91     BN9944             99-44          2.16              Control   \n",
       "92     BN9947             99-47          2.50              Control   \n",
       "93     BN9966             99-66          2.00              Control   \n",
       "\n",
       "                               path_autopsy_second_dx  \\\n",
       "0               Charcot-Marie-Tooth disease (history)   \n",
       "1                          Seizure disorder (history)   \n",
       "2                  Non-diagnostic Alzheimer's changes   \n",
       "3                  Non-diagnostic Alzheimer's changes   \n",
       "4                  Non-diagnostic Alzheimer's changes   \n",
       "..                                                ...   \n",
       "89  Microscopic changes of Alzheimer's disease, in...   \n",
       "90  Microscopic changes of Alzheimer's disease, in...   \n",
       "91                 Non-diagnostic Alzheimer's changes   \n",
       "92                 Non-diagnostic Alzheimer's changes   \n",
       "93                                     Incidental LBs   \n",
       "\n",
       "                                path_autopsy_third_dx  \\\n",
       "0                        GBA L444P/WT, L444P mutation   \n",
       "1                                                  NA   \n",
       "2                                                 CAA   \n",
       "3                                                CWMR   \n",
       "4           Argyrophilic grains, mesial temporal lobe   \n",
       "..                                                ...   \n",
       "89  Focal non-specific glial tauopathy, cortex of ...   \n",
       "90                       Incidental Lewy body disease   \n",
       "91  Alzheimer Type II astrocytosis consistent with...   \n",
       "92                     Alzheimer type II astrocytosis   \n",
       "93                 Non-diagnostic Alzheimer's changes   \n",
       "\n",
       "                         path_autopsy_fourth_dx path_autopsy_fifth_dx  \\\n",
       "0                                            NA                    NA   \n",
       "1                                            NA                    NA   \n",
       "2                                            NA                    NA   \n",
       "3                                            NA                    NA   \n",
       "4                                    Infarct(s)                  CWMR   \n",
       "..                                          ...                   ...   \n",
       "89                                           NA                    NA   \n",
       "90                                           NA                    NA   \n",
       "91                                           NA                    NA   \n",
       "92                                      Inc LBs                    NA   \n",
       "93  Argyrophilic grains in mesial temporal lobe                    NA   \n",
       "\n",
       "                                path_autopsy_sixth_dx  \\\n",
       "0                                                  NA   \n",
       "1                                                  NA   \n",
       "2                                                  NA   \n",
       "3                                                  NA   \n",
       "4   Several microscopic foci of cerebellar cortica...   \n",
       "..                                                ...   \n",
       "89                                                 NA   \n",
       "90                                                 NA   \n",
       "91                                                 NA   \n",
       "92                                                 NA   \n",
       "93                                                 NA   \n",
       "\n",
       "                     path_autopsy_seventh_dx  ...   path_nia_ri path_nia_aa_a  \\\n",
       "0                                         NA  ...            NA            NA   \n",
       "1                                         NA  ...            NA            NA   \n",
       "2                                         NA  ...            NA            NA   \n",
       "3                                         NA  ...            NA            NA   \n",
       "4   Etat crible, putamen and caudate nucleus  ...            NA            NA   \n",
       "..                                       ...  ...           ...           ...   \n",
       "89                                        NA  ...  Intermediate            NA   \n",
       "90                                        NA  ...            NA            NA   \n",
       "91                                        NA  ...            NA            NA   \n",
       "92                                        NA  ...            NA            NA   \n",
       "93                                        NA  ...            NA            NA   \n",
       "\n",
       "   path_nia_aa_b path_nia_aa_c TDP43 arteriolosclerosis_severity_scale  \\\n",
       "0             NA            NA    NA                                NA   \n",
       "1             NA            NA    NA                                NA   \n",
       "2             NA            NA    NA                              Mild   \n",
       "3             NA            NA    No                          Moderate   \n",
       "4             NA            NA    NA                              Mild   \n",
       "..           ...           ...   ...                               ...   \n",
       "89            NA            NA    No                                NA   \n",
       "90            NA            NA    No                                NA   \n",
       "91            NA            NA    NA                                NA   \n",
       "92            NA            NA    NA                                NA   \n",
       "93            NA            NA    NA                                NA   \n",
       "\n",
       "    amyloid_angiopathy_severity_scale path_ad_level dig_slide_avail  \\\n",
       "0                                  NA            NA              NA   \n",
       "1                                  NA            NA              NA   \n",
       "2                            Moderate            NA              NA   \n",
       "3                                Mild            NA              NA   \n",
       "4                                  NA            NA              NA   \n",
       "..                                ...           ...             ...   \n",
       "89                                 NA            NA              NA   \n",
       "90                                 NA            NA              NA   \n",
       "91                                 NA            NA              NA   \n",
       "92                                 NA            NA              NA   \n",
       "93                               Mild            NA              NA   \n",
       "\n",
       "   quant_path_avail  \n",
       "0                NA  \n",
       "1                NA  \n",
       "2                NA  \n",
       "3                NA  \n",
       "4                NA  \n",
       "..              ...  \n",
       "89               NA  \n",
       "90               NA  \n",
       "91               NA  \n",
       "92               NA  \n",
       "93               NA  \n",
       "\n",
       "[94 rows x 36 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH = prep_table(dfs['CLINPATH'],CDEv2)\n",
    "\n",
    "# autopsy coding is too too idosyncratic.  leave \n",
    "# the path_nia_aa_a are coded as path_nia_ri... swap them \n",
    "\n",
    "CLINPATH = prep_table(CLINPATH,CDEv2)\n",
    "\n",
    "\n",
    "# 'Low', 'Intermediate', 'High', 'None'\n",
    "CLINPATH['path_nia_ri'] = CLINPATH['path_nia_aa_a'].replace({'Low ADNC': 'Low', \n",
    "                                                            'Not AD': 'None', \n",
    "                                                            'Intermediate ADNC': 'Intermediate'})\n",
    "CLINPATH['path_nia_aa_a'] = NULL\n",
    "\n",
    "\n",
    "# path_thal\n",
    "CLINPATH['path_thal']= CLINPATH['path_thal'].replace({'Phase 0 (A0)':'0', \n",
    "                                'Phase 1 (A1)':'1', \n",
    "                                'Phase 2 (A1)':'2', \n",
    "                                'Phase 3 (A2)':'3', \n",
    "                                'Phase 4 (A3)':'4',\n",
    "                                'Phase 5 (A3)':'5', \n",
    "                                'Unknown':NULL, \n",
    "                                'Missing/unknown':NULL})\n",
    "\n",
    "CLINPATH['path_infarcs'] = CLINPATH['path_infarcs'].replace('Unknown', NULL)\n",
    "\n",
    "\n",
    "# no idea what No, undx, yes maps to.  leave\n",
    "CLINPATH['TDP43'] = CLINPATH['TDP43'].replace({'Unknown': NULL,'Undx':NULL})\n",
    "\n",
    "# sn_neuronal_loss has alternate 0-3 coding.  leave\n",
    "\n",
    "# add 'replicate' coding (nans)\n",
    "mckeith_map = {'l. Olfactory Bulb-Only':\"Olfactory bulb only\", 'lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'lb. Limbic Predominant':\"Limbic (transitional)\", 'lV. Neocortical':\"Neocortical\",\n",
    "       'lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\", \"Unknown\":NULL,\n",
    "       'L. Olfactory Bulb-Only':\"Olfactory bulb only\", 'Lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'Lb. Limbic Predominant':\"Limbic (transitional)\", 'LV. Neocortical':\"Neocortical\",\n",
    "       'Lll. Brainstem/Limbic':\"Amygdala Predominant\", 'Llb. Limbic Predominant': 'Limbic (amygdala) predominant'\n",
    "       }\n",
    "    \n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n",
    "CLINPATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *DATA* table.\n",
      "ðŸš¨âš ï¸â— **29 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_dx_main: 1/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_autopsy_second_dx: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 36/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 69/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 81/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 91/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 44/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 21/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 43/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 75/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 70/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 88/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **3 Fields with invalid entries:**\n",
      "- _*path_autopsy_dx_main*_:  invalid values ðŸ’©'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'Control (clinical history)'\n",
      "    - valid âž¡ï¸ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder', 'NA'\n",
      "- _*sn_neuronal_loss*_:  invalid values ðŸ’©'3.0', '0.0', '1.0', '2.0'\n",
      "    - valid âž¡ï¸ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
      "- _*TDP43*_:  invalid values ðŸ’©'No', 'Yes'\n",
      "    - valid âž¡ï¸ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "schema = CDE[CDE['Table'] == \"CLINPATH\"]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(CLINPATH.copy(), table, schema, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = prep_table(dfs['SAMPLE'],CDEv2)\n",
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "SAMPLE[\"cell_type_ontology_term_id\"] = \"CL:0002319\" # recode to fix encoding error\n",
    "SAMPLE[\"molecular_source\"] = \"PolyA RNA\" # fix encoding with underscore\n",
    "SAMPLE[\"development_stage_ontology_term_id\"] = \"HsapDv:0000258\"\n",
    "hybrid_selection = SAMPLE['hybrid_selection'][0]\n",
    "SAMPLE[\"alternate_sample_id\"] = hybrid_selection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *DATA* table.\n",
      "ðŸš¨âš ï¸â— **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- suspension_type: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **3 Fields with invalid entries:**\n",
      "- _*region_level_1*_:  invalid values ðŸ’©'Temporal Lobe'\n",
      "    - valid âž¡ï¸ 'Frontal lobe', 'Temporal lobe', 'Parietal lobe', 'Occipital lobe', 'Cingulate gyrus', 'Striatum', 'Thalamus', 'Midbrain', 'Pons', 'Medulla', 'Cerebellum', 'Intestine', 'Unknown', 'NA'\n",
      "- _*sequencing_length*_:  invalid values ðŸ’©'200'\n",
      "    - valid âž¡ï¸ '25', '50', '100', '150', 'NA'\n",
      "- _*sex_ontology_term_id*_:  invalid values ðŸ’©'PATO:0000384', 'PATO:0000383'\n",
      "    - valid âž¡ï¸ 'PATO:0000384 (male)', 'PATO:0000383 (female)', 'Unknown', 'NA'\n",
      "ðŸš¨âš ï¸â— **Extra field in DATA: hybrid_selection**\n",
      "\n"
     ]
    }
   ],
   "source": [
    "schema = CDE[CDE['Table'] == \"SAMPLE\"]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(SAMPLE.copy(), table, schema, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch30</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_I1_001.f...</td>\n",
       "      <td>Index 1 Lane 1</td>\n",
       "      <td>d2220622e5e3e83a7537e82b4184bd64</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch30</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_I2_001.f...</td>\n",
       "      <td>Index 2 Lane 1</td>\n",
       "      <td>ea30b06f7583bc46c6a8d636f42eac25</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch30</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_R1_001.f...</td>\n",
       "      <td>Read 1 Lane 1</td>\n",
       "      <td>4de7de820f840117b32038cecf2efdb1</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch30</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_R2_001.f...</td>\n",
       "      <td>Read 2 Lane 1</td>\n",
       "      <td>a15ecacea5b8b3328535b08fcbe1dd0e</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch30</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L002_I1_001.f...</td>\n",
       "      <td>Index 1 Lane 2</td>\n",
       "      <td>490f9acf9d4651ce3896f286227a7262</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L001_R2_001.f...</td>\n",
       "      <td>Read 2 Lane 1</td>\n",
       "      <td>1cc4397535a2638c1e8db96563cd83a0</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_I1_001.f...</td>\n",
       "      <td>Index 1 Lane 2</td>\n",
       "      <td>7bca81bb1b5d7d6b9da4750170dfe1e6</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_I2_001.f...</td>\n",
       "      <td>Index 2 Lane 2</td>\n",
       "      <td>f8f366baec1c7a7e0b2a8eda4d3837f4</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_R1_001.f...</td>\n",
       "      <td>Read 1 Lane 2</td>\n",
       "      <td>1bf3c8c30cb8b6d4d127084bd694b2fa</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch7</td>\n",
       "      <td>fastq</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_R2_001.f...</td>\n",
       "      <td>Read 2 Lane 2</td>\n",
       "      <td>7d18185e750616f056db05320f8f90f0</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_id replicate replicate_count repeated_sample    batch  \\\n",
       "0    BN1204_hybsel      rep1               1               0  batch30   \n",
       "1    BN1204_hybsel      rep1               1               0  batch30   \n",
       "2    BN1204_hybsel      rep1               1               0  batch30   \n",
       "3    BN1204_hybsel      rep1               1               0  batch30   \n",
       "4    BN1204_hybsel      rep1               1               0  batch30   \n",
       "..             ...       ...             ...             ...      ...   \n",
       "803  BN0952_hybsel      rep1               1               0   batch7   \n",
       "804  BN0952_hybsel      rep1               1               0   batch7   \n",
       "805  BN0952_hybsel      rep1               1               0   batch7   \n",
       "806  BN0952_hybsel      rep1               1               0   batch7   \n",
       "807  BN0952_hybsel      rep1               1               0   batch7   \n",
       "\n",
       "    file_type                                          file_name  \\\n",
       "0       fastq  novaseq_230914a_BN1204_BN0615_S6_L001_I1_001.f...   \n",
       "1       fastq  novaseq_230914a_BN1204_BN0615_S6_L001_I2_001.f...   \n",
       "2       fastq  novaseq_230914a_BN1204_BN0615_S6_L001_R1_001.f...   \n",
       "3       fastq  novaseq_230914a_BN1204_BN0615_S6_L001_R2_001.f...   \n",
       "4       fastq  novaseq_230914a_BN1204_BN0615_S6_L002_I1_001.f...   \n",
       "..        ...                                                ...   \n",
       "803     fastq  novaseq_230914c_BN1317_BN0952_S7_L001_R2_001.f...   \n",
       "804     fastq  novaseq_230914c_BN1317_BN0952_S7_L002_I1_001.f...   \n",
       "805     fastq  novaseq_230914c_BN1317_BN0952_S7_L002_I2_001.f...   \n",
       "806     fastq  novaseq_230914c_BN1317_BN0952_S7_L002_R1_001.f...   \n",
       "807     fastq  novaseq_230914c_BN1317_BN0952_S7_L002_R2_001.f...   \n",
       "\n",
       "    file_description                          file_MD5 technology omic  \\\n",
       "0     Index 1 Lane 1  d2220622e5e3e83a7537e82b4184bd64         SN  RNA   \n",
       "1     Index 2 Lane 1  ea30b06f7583bc46c6a8d636f42eac25         SN  RNA   \n",
       "2      Read 1 Lane 1  4de7de820f840117b32038cecf2efdb1         SN  RNA   \n",
       "3      Read 2 Lane 1  a15ecacea5b8b3328535b08fcbe1dd0e         SN  RNA   \n",
       "4     Index 1 Lane 2  490f9acf9d4651ce3896f286227a7262         SN  RNA   \n",
       "..               ...                               ...        ...  ...   \n",
       "803    Read 2 Lane 1  1cc4397535a2638c1e8db96563cd83a0         SN  RNA   \n",
       "804   Index 1 Lane 2  7bca81bb1b5d7d6b9da4750170dfe1e6         SN  RNA   \n",
       "805   Index 2 Lane 2  f8f366baec1c7a7e0b2a8eda4d3837f4         SN  RNA   \n",
       "806    Read 1 Lane 2  1bf3c8c30cb8b6d4d127084bd694b2fa         SN  RNA   \n",
       "807    Read 2 Lane 2  7d18185e750616f056db05320f8f90f0         SN  RNA   \n",
       "\n",
       "    adjustment content time header annotation configuration_file  \n",
       "0          Raw   Reads   NA     NA         NA                 NA  \n",
       "1          Raw   Reads   NA     NA         NA                 NA  \n",
       "2          Raw   Reads   NA     NA         NA                 NA  \n",
       "3          Raw   Reads   NA     NA         NA                 NA  \n",
       "4          Raw   Reads   NA     NA         NA                 NA  \n",
       "..         ...     ...  ...    ...        ...                ...  \n",
       "803        Raw   Reads   NA     NA         NA                 NA  \n",
       "804        Raw   Reads   NA     NA         NA                 NA  \n",
       "805        Raw   Reads   NA     NA         NA                 NA  \n",
       "806        Raw   Reads   NA     NA         NA                 NA  \n",
       "807        Raw   Reads   NA     NA         NA                 NA  \n",
       "\n",
       "[808 rows x 17 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA = prep_table(dfs['DATA'],CDEv2)\n",
    "DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Dx.doi.org/10.17504/protocols.io.j8nlk4k55g5r/v1</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_collection_summary cell_extraction_summary lib_prep_summary  \\\n",
       "0                        NA                      NA               NA   \n",
       "\n",
       "  data_processing_summary github_url  \\\n",
       "0                      NA         NA   \n",
       "\n",
       "                                   protocols_io_DOI other_reference  \n",
       "0  Dx.doi.org/10.17504/protocols.io.j8nlk4k55g5r/v1              NA  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROTOCOL = prep_table(dfs['PROTOCOL'],CDEv2)\n",
    "PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs['STUDY'] = STUDY\n",
    "dfs['SUBJECT'] = SUBJECT\n",
    "dfs['CLINPATH'] = CLINPATH\n",
    "dfs['SAMPLE'] = SAMPLE\n",
    "dfs['DATA'] = DATA \n",
    "dfs['PROTOCOL'] = PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "### save extras as auxillary tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved STUDY_auxiliary.csv\n",
      "Saved SAMPLE_auxiliary.csv\n"
     ]
    }
   ],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "v2_path = metadata_path / \"v2\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v2_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v2_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'subject_id', 'source_sample_id', 'replicate',\n",
       "       'replicate_count', 'repeated_sample', 'batch', 'tissue', 'brain_region',\n",
       "       'hemisphere', 'region_level_1', 'region_level_2', 'region_level_3',\n",
       "       'RIN', 'source_RIN', 'molecular_source', 'input_cell_count', 'assay',\n",
       "       'sequencing_end', 'sequencing_length', 'sequencing_instrument',\n",
       "       'organism_ontology_term_id', 'development_stage_ontology_term_id',\n",
       "       'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
       "       'disease_ontology_term_id', 'tissue_ontology_term_id',\n",
       "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
       "       'suspension_type', 'DV200', 'pm_PH', 'donor_id', 'alternate_sample_id',\n",
       "       'hybrid_selection'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs[\"SAMPLE\"].columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v2->v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "metadata_version = \"v3.0\"\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\"\n",
    "\n",
    "\n",
    "f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "ðŸš¨âš ï¸â— **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "ðŸš¨âš ï¸â— **6 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- sample_collection_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_extraction_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- lib_prep_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- data_processing_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "ðŸš¨âš ï¸â— **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 94/94 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "ðŸš¨âš ï¸â— **8 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- organism: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 94/94 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **1 Fields with invalid entries:**\n",
      "- _*sex_ontology_term_id*_:  invalid values ðŸ’©'PATO:0000384', 'PATO:0000383'\n",
      "    - valid âž¡ï¸ 'PATO:0000384 (male)', 'PATO:0000383 (female)', 'Unknown', 'NA'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "ðŸš¨âš ï¸â— **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 808/808 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 808/808 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 808/808 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "No empty entries (NULL) found .\n",
      "ðŸš¨âš ï¸â— **1 Fields with invalid entries:**\n",
      "- _*region_level_1*_:  invalid values ðŸ’©'Temporal Lobe'\n",
      "    - valid âž¡ï¸ 'Frontal lobe', 'Temporal lobe', 'Parietal lobe', 'Occipital lobe', 'Cingulate gyrus', 'Striatum', 'Thalamus', 'Midbrain', 'Pons', 'Medulla', 'Cerebellum', 'Intestine', 'Unknown', 'NA'\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "ðŸš¨âš ï¸â— **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 86/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 49/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 63/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 93/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 7/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 64/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_dx_main: 1/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 11/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 36/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 59/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 69/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 81/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 91/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 44/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 21/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 68/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 6/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 3/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 66/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 75/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 70/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 88/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- dig_slide_avail: 94/94 empty rows (OPTIONAL)\n",
      "\n",
      "\t- quant_path_avail: 94/94 empty rows (OPTIONAL)\n",
      "ðŸš¨âš ï¸â— **4 Fields with invalid entries:**\n",
      "- _*education_level*_:  invalid values ðŸ’©'14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'\n",
      "    - valid âž¡ï¸ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other', 'NA'\n",
      "- _*path_autopsy_dx_main*_:  invalid values ðŸ’©'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'Control (clinical history)'\n",
      "    - valid âž¡ï¸ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder', 'NA'\n",
      "- _*sn_neuronal_loss*_:  invalid values ðŸ’©'3.0', '0.0', '1.0', '2.0'\n",
      "    - valid âž¡ï¸ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown', 'NA'\n",
      "- _*TDP43*_:  invalid values ðŸ’©'No', 'Yes'\n",
      "    - valid âž¡ï¸ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown', 'NA'\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "ðŸš¨âš ï¸â— **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 3/3 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 3/3 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>prodromal_motor_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field        condition_id intervention_name intervention_id protocol_id  \\\n",
       "0           idiopathic_pd      Case-Control            Case          NA   \n",
       "1         healthy_control      Case-Control         Control          NA   \n",
       "2      prodromal_motor_pd      Case-Control           Other          NA   \n",
       "\n",
       "Field intervention_aux_table  \n",
       "0                         NA  \n",
       "1                         NA  \n",
       "2                         NA  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-SCHERZER</td>\n",
       "      <td>Scherzer, Dong, and Levin</td>\n",
       "      <td>Parkinson5D: deconstructing proximal disease m...</td>\n",
       "      <td>sn_rnaseq_mtg_hybsel</td>\n",
       "      <td>PD5D_MTG_snRNAseq_hybsel</td>\n",
       "      <td>Here we will develop a molecular atlas of Park...</td>\n",
       "      <td>Dr. Joshua Levin</td>\n",
       "      <td>Jlevin@broadinstitute.org</td>\n",
       "      <td>Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...</td>\n",
       "      <td>Dr. Sean Simmons</td>\n",
       "      <td>...</td>\n",
       "      <td>94</td>\n",
       "      <td>MTG</td>\n",
       "      <td>PD and control postmortem brains</td>\n",
       "      <td>ASAP Access &amp; Use Policy</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>v3.0_20241025</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name              ASAP_lab_name  \\\n",
       "0  TEAM-SCHERZER  Scherzer, Dong, and Levin   \n",
       "\n",
       "                                        project_name       team_dataset_id  \\\n",
       "0  Parkinson5D: deconstructing proximal disease m...  sn_rnaseq_mtg_hybsel   \n",
       "\n",
       "            project_dataset  \\\n",
       "0  PD5D_MTG_snRNAseq_hybsel   \n",
       "\n",
       "                                 project_description      PI_full_name  \\\n",
       "0  Here we will develop a molecular atlas of Park...  Dr. Joshua Levin   \n",
       "\n",
       "                    PI_email  \\\n",
       "0  Jlevin@broadinstitute.org   \n",
       "\n",
       "                                   contributor_names    submitter_name  ...  \\\n",
       "0  Clemens, Scherzer; Xianjun, Dong; Joshua, Levi...  Dr. Sean Simmons  ...   \n",
       "\n",
       "  number_samples sample_types                  types_of_samples  \\\n",
       "0             94          MTG  PD and control postmortem brains   \n",
       "\n",
       "                DUA_version  \\\n",
       "0  ASAP Access & Use Policy   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "  PI_google_scholar_id preprocessing_references metadata_version_date  \\\n",
       "0                   NA                       NA         v3.0_20241025   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = v3_tables['STUDY']\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY[\"metadata_tables\"].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scherzer\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/scherzer-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-scherzer@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-data-team-scherzer/snRNA_hybsel/*.gz\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "\n",
    "# make sure to get ALL the fastq files in the bucket\n",
    "prefix = \"snRNA_hybsel/*.gz\"\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>check2</th>\n",
       "      <th>check1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_I1_001.f...</td>\n",
       "      <td>d2220622e5e3e83a7537e82b4184bd64</td>\n",
       "      <td>d2220622e5e3e83a7537e82b4184bd64</td>\n",
       "      <td>d2220622e5e3e83a7537e82b4184bd64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_I2_001.f...</td>\n",
       "      <td>ea30b06f7583bc46c6a8d636f42eac25</td>\n",
       "      <td>ea30b06f7583bc46c6a8d636f42eac25</td>\n",
       "      <td>ea30b06f7583bc46c6a8d636f42eac25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_R1_001.f...</td>\n",
       "      <td>4de7de820f840117b32038cecf2efdb1</td>\n",
       "      <td>4de7de820f840117b32038cecf2efdb1</td>\n",
       "      <td>4de7de820f840117b32038cecf2efdb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L001_R2_001.f...</td>\n",
       "      <td>a15ecacea5b8b3328535b08fcbe1dd0e</td>\n",
       "      <td>a15ecacea5b8b3328535b08fcbe1dd0e</td>\n",
       "      <td>a15ecacea5b8b3328535b08fcbe1dd0e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN1204_hybsel</td>\n",
       "      <td>novaseq_230914a_BN1204_BN0615_S6_L002_I1_001.f...</td>\n",
       "      <td>490f9acf9d4651ce3896f286227a7262</td>\n",
       "      <td>490f9acf9d4651ce3896f286227a7262</td>\n",
       "      <td>490f9acf9d4651ce3896f286227a7262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L001_R2_001.f...</td>\n",
       "      <td>1cc4397535a2638c1e8db96563cd83a0</td>\n",
       "      <td>1cc4397535a2638c1e8db96563cd83a0</td>\n",
       "      <td>1cc4397535a2638c1e8db96563cd83a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_I1_001.f...</td>\n",
       "      <td>7bca81bb1b5d7d6b9da4750170dfe1e6</td>\n",
       "      <td>7bca81bb1b5d7d6b9da4750170dfe1e6</td>\n",
       "      <td>7bca81bb1b5d7d6b9da4750170dfe1e6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_I2_001.f...</td>\n",
       "      <td>f8f366baec1c7a7e0b2a8eda4d3837f4</td>\n",
       "      <td>f8f366baec1c7a7e0b2a8eda4d3837f4</td>\n",
       "      <td>f8f366baec1c7a7e0b2a8eda4d3837f4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_R1_001.f...</td>\n",
       "      <td>1bf3c8c30cb8b6d4d127084bd694b2fa</td>\n",
       "      <td>1bf3c8c30cb8b6d4d127084bd694b2fa</td>\n",
       "      <td>1bf3c8c30cb8b6d4d127084bd694b2fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>BN0952_hybsel</td>\n",
       "      <td>novaseq_230914c_BN1317_BN0952_S7_L002_R2_001.f...</td>\n",
       "      <td>7d18185e750616f056db05320f8f90f0</td>\n",
       "      <td>7d18185e750616f056db05320f8f90f0</td>\n",
       "      <td>7d18185e750616f056db05320f8f90f0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_id                                          file_name  \\\n",
       "0    BN1204_hybsel  novaseq_230914a_BN1204_BN0615_S6_L001_I1_001.f...   \n",
       "1    BN1204_hybsel  novaseq_230914a_BN1204_BN0615_S6_L001_I2_001.f...   \n",
       "2    BN1204_hybsel  novaseq_230914a_BN1204_BN0615_S6_L001_R1_001.f...   \n",
       "3    BN1204_hybsel  novaseq_230914a_BN1204_BN0615_S6_L001_R2_001.f...   \n",
       "4    BN1204_hybsel  novaseq_230914a_BN1204_BN0615_S6_L002_I1_001.f...   \n",
       "..             ...                                                ...   \n",
       "803  BN0952_hybsel  novaseq_230914c_BN1317_BN0952_S7_L001_R2_001.f...   \n",
       "804  BN0952_hybsel  novaseq_230914c_BN1317_BN0952_S7_L002_I1_001.f...   \n",
       "805  BN0952_hybsel  novaseq_230914c_BN1317_BN0952_S7_L002_I2_001.f...   \n",
       "806  BN0952_hybsel  novaseq_230914c_BN1317_BN0952_S7_L002_R1_001.f...   \n",
       "807  BN0952_hybsel  novaseq_230914c_BN1317_BN0952_S7_L002_R2_001.f...   \n",
       "\n",
       "                             file_MD5                            check2  \\\n",
       "0    d2220622e5e3e83a7537e82b4184bd64  d2220622e5e3e83a7537e82b4184bd64   \n",
       "1    ea30b06f7583bc46c6a8d636f42eac25  ea30b06f7583bc46c6a8d636f42eac25   \n",
       "2    4de7de820f840117b32038cecf2efdb1  4de7de820f840117b32038cecf2efdb1   \n",
       "3    a15ecacea5b8b3328535b08fcbe1dd0e  a15ecacea5b8b3328535b08fcbe1dd0e   \n",
       "4    490f9acf9d4651ce3896f286227a7262  490f9acf9d4651ce3896f286227a7262   \n",
       "..                                ...                               ...   \n",
       "803  1cc4397535a2638c1e8db96563cd83a0  1cc4397535a2638c1e8db96563cd83a0   \n",
       "804  7bca81bb1b5d7d6b9da4750170dfe1e6  7bca81bb1b5d7d6b9da4750170dfe1e6   \n",
       "805  f8f366baec1c7a7e0b2a8eda4d3837f4  f8f366baec1c7a7e0b2a8eda4d3837f4   \n",
       "806  1bf3c8c30cb8b6d4d127084bd694b2fa  1bf3c8c30cb8b6d4d127084bd694b2fa   \n",
       "807  7d18185e750616f056db05320f8f90f0  7d18185e750616f056db05320f8f90f0   \n",
       "\n",
       "                               check1  \n",
       "0    d2220622e5e3e83a7537e82b4184bd64  \n",
       "1    ea30b06f7583bc46c6a8d636f42eac25  \n",
       "2    4de7de820f840117b32038cecf2efdb1  \n",
       "3    a15ecacea5b8b3328535b08fcbe1dd0e  \n",
       "4    490f9acf9d4651ce3896f286227a7262  \n",
       "..                                ...  \n",
       "803  1cc4397535a2638c1e8db96563cd83a0  \n",
       "804  7bca81bb1b5d7d6b9da4750170dfe1e6  \n",
       "805  f8f366baec1c7a7e0b2a8eda4d3837f4  \n",
       "806  1bf3c8c30cb8b6d4d127084bd694b2fa  \n",
       "807  7d18185e750616f056db05320f8f90f0  \n",
       "\n",
       "[808 rows x 5 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_47191/3054458882.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_47191/3054458882.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check1'] = checksum['file_MD5']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def check_md5_sums()\n",
    "\n",
    "\n",
    "checksum = v3_tables['DATA'][['sample_id','file_name','file_MD5']]\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "checksum['check1'] = checksum['file_MD5']\n",
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "#empty means success!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------\n",
    "## Create metadata package\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_path = root_path / f\"{team}\"\n",
    "\n",
    "create_upload_medadata_package(export_path, v3_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          idiopathic_pd\n",
       "1        healthy_control\n",
       "2        healthy_control\n",
       "3     prodromal_motor_pd\n",
       "4          idiopathic_pd\n",
       "             ...        \n",
       "89       healthy_control\n",
       "90       healthy_control\n",
       "91         idiopathic_pd\n",
       "92         idiopathic_pd\n",
       "93       healthy_control\n",
       "Name: condition_id, Length: 94, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['SAMPLE']['condition_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>prodromal_motor_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field        condition_id intervention_name intervention_id protocol_id  \\\n",
       "0           idiopathic_pd      Case-Control            Case          NA   \n",
       "1         healthy_control      Case-Control         Control          NA   \n",
       "2      prodromal_motor_pd      Case-Control           Other          NA   \n",
       "\n",
       "Field intervention_aux_table  \n",
       "0                         NA  \n",
       "1                         NA  \n",
       "2                         NA  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_tables['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
