{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 2\n",
    "\n",
    "# Team Sulzer. ASAP CRN Metadata validation - wave 2\n",
    "\n",
    "13 Oct 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully imported\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table, create_metadata_package\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, intervention_typer\n",
    "from checksums import extract_md5_from_details2, get_md5_hashes \n",
    "from bucket_util import authenticate_with_service_account, gsutil_ls, gsutil_cp, gsutil_mv \n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted as v2.1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Team Sulzer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"sulzer\"\n",
    "dataset_name = \"sn-rnaseq\"\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata\"\n",
    "og_path = metadata_path / \"og\"\n",
    "\n",
    "\n",
    "metadata_version = \"v2\"\n",
    "\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in STUDY: submitter_email**\n",
      "üö®‚ö†Ô∏è‚ùó **Missing Optional Fields in STUDY: alternate_dataset_id**\n",
      "üö®‚ö†Ô∏è‚ùó **8 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- ASAP_grant_id: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- metadata_version_date: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Extra field in STUDY: submittor_email**\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "üö®‚ö†Ô∏è‚ùó **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **18 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- race: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 23/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 46/46 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **2 Fields with invalid entries:**\n",
      "- _*sex*_:  invalid values üí©'M', 'F'\n",
      "    - valid ‚û°Ô∏è 'Male', 'Female', 'Intersex', 'Unnown', 'NA'\n",
      "- _*age_at_collection*_:  invalid values üí©'77', '89+ ', '79', '88', '83', '78', '87', '75', '86', '85', '68', '73', '80', '72', '71', '50', '74', '81', '70', '66', '62'\n",
      "    - valid ‚û°Ô∏è float or NULL ('NA')\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **28 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_second_dx: 44/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 30/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 1/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 19/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 46/46 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Missing Optional Fields in SAMPLE: alternate_sample_id**\n",
      "üö®‚ö†Ô∏è‚ùó **12 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- region_level_2: 18/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- input_cell_count: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_ontology_term_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 46/46 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **4 Fields with invalid entries:**\n",
      "- _*region_level_2*_:  invalid values üí©'Substantia Nigra'\n",
      "    - valid ‚û°Ô∏è 'Prefrontal cortex', 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown', 'NA'\n",
      "- _*molecular_source*_:  invalid values üí©'RNA'\n",
      "    - valid ‚û°Ô∏è 'Total RNA', 'PolyA RNA', 'Cytoplasmic RNA', 'Nuclear RNA', 'Genomic DNA', 'Protein', 'Other', 'NA'\n",
      "- _*input_cell_count*_:  invalid values üí©'915.1258 ng', '1411.8474 ng', '1608.6002 ng', '704.4034 ng', '1444.6272 ng', '896.3852 ng', '2392.286 ng', '714.5692 ng', '923.9646 ng', '866.4158 ng', '1420.8972 ng', '2169.106 ng', '1069.3564 ng', '1192.589 ng', '1008.9174 ng', '983.9328 ng', '1266.1952 ng', '945.216 ng', 'NA'\n",
      "    - valid ‚û°Ô∏è int or NULL ('NA')\n",
      "- _*organism_ontology_term_id*_:  invalid values üí©'NCBITaxon:9607', 'NCBITaxon:9608', 'NCBITaxon:9609', 'NCBITaxon:9610', 'NCBITaxon:9611', 'NCBITaxon:9612', 'NCBITaxon:9613', 'NCBITaxon:9614', 'NCBITaxon:9615', 'NCBITaxon:9616', 'NCBITaxon:9617', 'NCBITaxon:9618', 'NCBITaxon:9619', 'NCBITaxon:9620', 'NCBITaxon:9621', 'NCBITaxon:9622', 'NCBITaxon:9623', 'NCBITaxon:9624', 'NCBITaxon:9625', 'NCBITaxon:9626', 'NCBITaxon:9627', 'NCBITaxon:9628', 'NCBITaxon:9629', 'NCBITaxon:9630', 'NCBITaxon:9631', 'NCBITaxon:9632', 'NCBITaxon:9633', 'NCBITaxon:9634', 'NCBITaxon:9635', 'NCBITaxon:9636', 'NCBITaxon:9637', 'NCBITaxon:9638', 'NCBITaxon:9639', 'NCBITaxon:9640', 'NCBITaxon:9641', 'NCBITaxon:9642', 'NCBITaxon:9643', 'NCBITaxon:9644', 'NCBITaxon:9645', 'NCBITaxon:9646', 'NCBITaxon:9647', 'NCBITaxon:9648', 'NCBITaxon:9649', 'NCBITaxon:9650', 'NCBITaxon:9651'\n",
      "    - valid ‚û°Ô∏è 'NCBITaxon:9606', 'NA'\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "üö®‚ö†Ô∏è‚ùó **7 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- file_name: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- file_description: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- file_MD5: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 46/46 empty rows (REQUIRED)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*technology*_:  invalid values üí©'sN'\n",
      "    - valid ‚û°Ô∏è 'SN', 'SC', 'Spatial', 'Bulk', 'Seq', 'Array', 'NA'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "tables = CDE['Table'].unique()\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(og_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table\n",
    "    # df.to_csv(v1_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT = read_meta_table(og_path / \"SUBJECT.csv\")\n",
    "CLINPATH = read_meta_table(og_path / \"CLINPATH.csv\")\n",
    "STUDY = read_meta_table(og_path / \"STUDY.csv\")\n",
    "PROTOCOL = read_meta_table(og_path / \"PROTOCOL.csv\")\n",
    "SAMPLE = read_meta_table(og_path / \"SAMPLE.csv\")\n",
    "DATA = read_meta_table(og_path / \"DATA.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "STUDY.rename(columns={\"submittor_email\":\"submitter_email\"}, inplace=True)\n",
    "STUDY['alternate_dataset_id'] = NULL\n",
    "STUDY['team_dataset_id'] = dataset_name.replace(\" \", \"_\").replace(\"-\", \"_\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>hx_dementia_mci</th>\n",
       "      <th>hx_melanoma</th>\n",
       "      <th>education_level</th>\n",
       "      <th>smoking_status</th>\n",
       "      <th>smoking_years</th>\n",
       "      <th>APOE_e4_status</th>\n",
       "      <th>cognitive_status</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OJ9</td>\n",
       "      <td>T-4530</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>New York Brain Bank</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>77</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OJ8</td>\n",
       "      <td>T-4105</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>New York Brain Bank</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>89+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OJ7</td>\n",
       "      <td>T-5259</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>New York Brain Bank</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>79</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OJ6</td>\n",
       "      <td>T-5385</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>New York Brain Bank</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>88</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OJ5</td>\n",
       "      <td>T-2465</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>New York Brain Bank</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>83</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id AMPPD_id GP2_id         biobank_name organism  \\\n",
       "0        OJ9            T-4530       NA     NA  New York Brain Bank    Human   \n",
       "1        OJ8            T-4105       NA     NA  New York Brain Bank    Human   \n",
       "2        OJ7            T-5259       NA     NA  New York Brain Bank    Human   \n",
       "3        OJ6            T-5385       NA     NA  New York Brain Bank    Human   \n",
       "4        OJ5            T-2465       NA     NA  New York Brain Bank    Human   \n",
       "\n",
       "      sex age_at_collection                       race  \\\n",
       "0    Male                77                    Unknown   \n",
       "1  Female              89+                     Unknown   \n",
       "2    Male                79                      White   \n",
       "3    Male                88                      White   \n",
       "4  Female                83  Black or African American   \n",
       "\n",
       "                ethnicity  ... hx_dementia_mci hx_melanoma education_level  \\\n",
       "0                 Unknown  ...         Unknown     Unknown         Unknown   \n",
       "1      Hispanic or Latino  ...         Unknown     Unknown         Unknown   \n",
       "2  Not Hispanic or Latino  ...         Unknown     Unknown         Unknown   \n",
       "3  Not Hispanic or Latino  ...         Unknown     Unknown         Unknown   \n",
       "4  Not Hispanic or Latino  ...         Unknown     Unknown         Unknown   \n",
       "\n",
       "  smoking_status smoking_years APOE_e4_status cognitive_status  \\\n",
       "0        Unknown       Unknown        Unknown          Unknown   \n",
       "1        Unknown       Unknown        Unknown          Unknown   \n",
       "2        Unknown       Unknown        Unknown          Unknown   \n",
       "3        Unknown       Unknown        Unknown          Unknown   \n",
       "4        Unknown       Unknown        Unknown          Unknown   \n",
       "\n",
       "  time_from_baseline primary_diagnosis primary_diagnosis_text  \n",
       "0            Unknown           Unknown                Unknown  \n",
       "1            Unknown           Unknown                Unknown  \n",
       "2            Unknown           Unknown                Unknown  \n",
       "3            Unknown           Unknown                Unknown  \n",
       "4            Unknown           Unknown                Unknown  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "SUBJECT = prep_table(SUBJECT, CDE)\n",
    "SUBJECT['sex'] = SUBJECT['sex'].replace({\"M\":\"Male\", \"F\":\"Female\"})\n",
    "SUBJECT.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46,)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['subject_id'].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['T-4530', 'T-4105', 'T-5259', 'T-5385', 'T-2465', 'T-3799',\n",
       "       'T-4424', 'T-5133', 'T-5227', 'T-5547', 'T-5359', 'T-4116',\n",
       "       'T-4839', 'T-2007', 'T-4938', 'T-5498', 'T-2791', 'T-4695',\n",
       "       'T-4934', 'T-3870', 'T-124', 'T-81', 'T-4515', 'T-4235', 'T-4815',\n",
       "       'T-4857', 'T-3998', 'T-4462', 'T-4108', 'T-4174', 'T3860',\n",
       "       'T-4162', 'T-4210', 'T-233', 'T1976', 'T-2017', 'T-308', 'T-1952',\n",
       "       'T-777', 'T-1619', 'T290', 'T-551', 'T-257', 'T-343', 'T-362'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['source_subject_id'].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "SUBJECT['hx_dementia_mci'] = SUBJECT['hx_dementia_mci'].replace(\"Unknown\", NULL)\n",
    "\n",
    "SUBJECT['hx_melanoma'] = SUBJECT['hx_melanoma'].replace(\"Unknown\", NULL)\n",
    "SUBJECT['education_level'] = SUBJECT['education_level'].replace(\"Unknown\", NULL)\n",
    "SUBJECT['cognitive_status'] = SUBJECT['cognitive_status'].replace(\"Unknown\", NULL)\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace(\"Unknown\", NULL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T-362</td>\n",
       "      <td>OF001</td>\n",
       "      <td>T-362</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T-308</td>\n",
       "      <td>OF009</td>\n",
       "      <td>T-308</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T3860</td>\n",
       "      <td>OF016</td>\n",
       "      <td>T3860</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T290</td>\n",
       "      <td>OF005</td>\n",
       "      <td>T290</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T-233</td>\n",
       "      <td>OF013</td>\n",
       "      <td>T-233</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0     T-362      OF001            T-362      Rep1               1   \n",
       "1     T-308      OF009            T-308      Rep1               1   \n",
       "2     T3860      OF016            T3860      Rep1               1   \n",
       "3      T290      OF005             T290      Rep1               1   \n",
       "4     T-233      OF013            T-233      Rep1               1   \n",
       "\n",
       "  repeated_sample batch tissue      brain_region hemisphere  ...  \\\n",
       "0               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "1               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "2               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "3               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "4               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0  PATO:0000383 (female)                                  Unknown   \n",
       "1    PATO:0000384 (male)                                  Unknown   \n",
       "2    PATO:0000384 (male)                                  Unknown   \n",
       "3    PATO:0000384 (male)                                  Unknown   \n",
       "4    PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0                       NA          UBERON:0003027                         NA   \n",
       "1                       NA          UBERON:0003027                         NA   \n",
       "2                       NA          UBERON:0003027                         NA   \n",
       "3                       NA          UBERON:0003027                         NA   \n",
       "4                       NA          UBERON:0003027                         NA   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV200 pm_PH donor_id  \n",
       "0                     NA         Nucleus    NA    NA       NA  \n",
       "1                     NA         Nucleus    NA    NA       NA  \n",
       "2                     NA         Nucleus    NA    NA       NA  \n",
       "3                     NA         Nucleus    NA    NA       NA  \n",
       "4                     NA         Nucleus    NA    NA       NA  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE = prep_table(SAMPLE, CDE)\n",
    "\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "SAMPLE['molecular_source'] = SAMPLE['molecular_source'].replace(\"RNA\", \"PolyA RNA\")\n",
    "SAMPLE['region_level_2'] = SAMPLE['region_level_2'].replace('Substantia Nigra', 'Substantia nigra',)\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T-362</td>\n",
       "      <td>OF001</td>\n",
       "      <td>T-362</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T-308</td>\n",
       "      <td>OF009</td>\n",
       "      <td>T-308</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T-3860</td>\n",
       "      <td>OF016</td>\n",
       "      <td>T-3860</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T-290</td>\n",
       "      <td>OF005</td>\n",
       "      <td>T-290</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T-233</td>\n",
       "      <td>OF013</td>\n",
       "      <td>T-233</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id subject_id source_sample_id replicate replicate_count  \\\n",
       "0     T-362      OF001            T-362      Rep1               1   \n",
       "1     T-308      OF009            T-308      Rep1               1   \n",
       "2    T-3860      OF016           T-3860      Rep1               1   \n",
       "3     T-290      OF005            T-290      Rep1               1   \n",
       "4     T-233      OF013            T-233      Rep1               1   \n",
       "\n",
       "  repeated_sample batch tissue      brain_region hemisphere  ...  \\\n",
       "0               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "1               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "2               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "3               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "4               0     1  Brain  Cingulate Cortex         NA  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0  PATO:0000383 (female)                                  Unknown   \n",
       "1    PATO:0000384 (male)                                  Unknown   \n",
       "2    PATO:0000384 (male)                                  Unknown   \n",
       "3    PATO:0000384 (male)                                  Unknown   \n",
       "4    PATO:0000384 (male)                                  Unknown   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0                       NA          UBERON:0003027                         NA   \n",
       "1                       NA          UBERON:0003027                         NA   \n",
       "2                       NA          UBERON:0003027                         NA   \n",
       "3                       NA          UBERON:0003027                         NA   \n",
       "4                       NA          UBERON:0003027                         NA   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV200 pm_PH donor_id  \n",
       "0                     NA         Nucleus    NA    NA       NA  \n",
       "1                     NA         Nucleus    NA    NA       NA  \n",
       "2                     NA         Nucleus    NA    NA       NA  \n",
       "3                     NA         Nucleus    NA    NA       NA  \n",
       "4                     NA         Nucleus    NA    NA       NA  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fix_samples = {'T3860':'T-3860', 'T290':'T-290','T1976':'T-1976'}\n",
    "SAMPLE['sample_id'] = SAMPLE['sample_id'].replace(fix_samples)\n",
    "SAMPLE['source_sample_id'] = SAMPLE['source_sample_id'].replace(fix_samples)\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_mapper = {\"Cingulate Cortex\":\"cing1\", \"Substantia Nigra\": \"sn1\"}\n",
    "SAMPLE['batch']=SAMPLE['brain_region'].map(batch_mapper)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üö®‚ö†Ô∏è‚ùó EXTREME CURATION HERE üö®‚ö†Ô∏è‚ùó\n",
    "The codeing by Team-Sulzer is inconsistent.  However I am able to infer the following:\n",
    "\n",
    "TODO:  Define `sample_id`s from `subject_id`, and create `sample_id` as `source_subject_id`\n",
    "\n",
    "\n",
    "`source_subject_id`s seem to be unique and correct.  Although the CLINPATH.duration_pmi are different for `source_subject_id`=\"T-2465\", and some demographic info is missing for the cingulate sample, these seem to be two samples from the same subject.  \n",
    "\n",
    "Hence:  We will assume `source_subject_id` is ground truth.   Redefine `sample_id` from the current `subject_id` (which matches file naming conventsions), and  create a _new_ `subject_id` from the `source_subject_id` by prepending \"SUL_\".  Finally, `source_sample_id` will be set to pd.NA as is currently just a copy of `source_subject_id`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fix SAMPLE table first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simple copy subject_id to sample_id, sample_id == source_sample_id\n",
    "SAMPLE['sample_id'] = SAMPLE['subject_id']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>alternate_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OF001</td>\n",
       "      <td>SUL_T_362</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OF009</td>\n",
       "      <td>SUL_T_308</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OF016</td>\n",
       "      <td>SUL_T_3860</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OF005</td>\n",
       "      <td>SUL_T_290</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OF013</td>\n",
       "      <td>SUL_T_233</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id  subject_id source_sample_id replicate replicate_count  \\\n",
       "0     OF001   SUL_T_362               NA      Rep1               1   \n",
       "1     OF009   SUL_T_308               NA      Rep1               1   \n",
       "2     OF016  SUL_T_3860               NA      Rep1               1   \n",
       "3     OF005   SUL_T_290               NA      Rep1               1   \n",
       "4     OF013   SUL_T_233               NA      Rep1               1   \n",
       "\n",
       "  repeated_sample  batch tissue      brain_region hemisphere  ...  \\\n",
       "0               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "1               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "2               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "3               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "4               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id disease_ontology_term_id  \\\n",
       "0                                  Unknown                       NA   \n",
       "1                                  Unknown                       NA   \n",
       "2                                  Unknown                       NA   \n",
       "3                                  Unknown                       NA   \n",
       "4                                  Unknown                       NA   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id  \\\n",
       "0          UBERON:0003027                         NA                     NA   \n",
       "1          UBERON:0003027                         NA                     NA   \n",
       "2          UBERON:0003027                         NA                     NA   \n",
       "3          UBERON:0003027                         NA                     NA   \n",
       "4          UBERON:0003027                         NA                     NA   \n",
       "\n",
       "  suspension_type DV200 pm_PH donor_id alternate_sample_id  \n",
       "0         Nucleus    NA    NA       NA                  NA  \n",
       "1         Nucleus    NA    NA       NA                  NA  \n",
       "2         Nucleus    NA    NA       NA                  NA  \n",
       "3         Nucleus    NA    NA       NA                  NA  \n",
       "4         Nucleus    NA    NA       NA                  NA  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fix_samples = {'T3860':'T-3860', 'T290':'T-290','T1976':'T-1976'}\n",
    "SAMPLE['source_sample_id'] = SAMPLE['source_sample_id'].replace(fix_samples)\n",
    "# then make a subject_id by copying souce_sample_id\n",
    "SAMPLE['subject_id'] = \"SUL_\" + SAMPLE['source_sample_id'].str.replace(\"-\",\"_\")\n",
    "\n",
    "# # finally set source_sample_id to pd.NA\n",
    "SAMPLE['source_sample_id'] = NULL\n",
    "\n",
    "SAMPLE['alternate_sample_id'] = NULL\n",
    "\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### now SUBJECT table "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_4530', 'SUL_T_4105', 'SUL_T_5259', 'SUL_T_5385',\n",
       "       'SUL_T_2465', 'SUL_T_3799', 'SUL_T_4424', 'SUL_T_5133',\n",
       "       'SUL_T_5227', 'SUL_T_5547', 'SUL_T_5359', 'SUL_T_4116',\n",
       "       'SUL_T_4839', 'SUL_T_2007', 'SUL_T_4938', 'SUL_T_5498',\n",
       "       'SUL_T_2791', 'SUL_T_4695', 'SUL_T_4934', 'SUL_T_3870',\n",
       "       'SUL_T_124', 'SUL_T_81', 'SUL_T_4515', 'SUL_T_4235', 'SUL_T_4815',\n",
       "       'SUL_T_4857', 'SUL_T_3998', 'SUL_T_4462', 'SUL_T_4108',\n",
       "       'SUL_T_4174', 'SUL_T_3860', 'SUL_T_4162', 'SUL_T_4210',\n",
       "       'SUL_T_233', 'SUL_T_1976', 'SUL_T_2017', 'SUL_T_308', 'SUL_T_1952',\n",
       "       'SUL_T_777', 'SUL_T_1619', 'SUL_T_290', 'SUL_T_551', 'SUL_T_257',\n",
       "       'SUL_T_343', 'SUL_T_362'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fix_samples = {'T3860':'T-3860', 'T290':'T-290','T1976':'T-1976'}\n",
    "SUBJECT['source_subject_id'] = SUBJECT['source_subject_id'].replace(fix_samples)\n",
    "\n",
    "SUBJECT['subject_id'] = \"SUL_\" + SUBJECT['source_subject_id'].str.replace(\"-\",\"_\")\n",
    "SUBJECT['subject_id'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### and CLINPATH table is already coded as it should be\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['T-4530', 'T-4105', 'T-5259', 'T-5385', 'T-2465', 'T-3799',\n",
       "       'T-4424', 'T-5133', 'T-5227', 'T-5547', 'T-5359', 'T-4116',\n",
       "       'T-4839', 'T-2007', 'T-4938', 'T-5498', 'T-2791', 'T-4695',\n",
       "       'T-4934', 'T-3870', 'T-124', 'T-81', 'T-4515', 'T-4235', 'T-4815',\n",
       "       'T-4857', 'T-3998', 'T-4462', 'T-4108', 'T-4174', 'T-3860',\n",
       "       'T-4162', 'T-4210', 'T-233', 'T-1976', 'T-2017', 'T-308', 'T-1952',\n",
       "       'T-777', 'T-1619', 'T-290', 'T-551', 'T-257', 'T-343', 'T-362'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "fix_samples = {'T3860':'T-3860', 'T290':'T-290','T1976':'T-1976'}\n",
    "CLINPATH['source_subject_id'] = CLINPATH['source_subject_id'].replace(fix_samples)\n",
    "CLINPATH['source_subject_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>path_autopsy_dx_main</th>\n",
       "      <th>path_autopsy_second_dx</th>\n",
       "      <th>path_autopsy_third_dx</th>\n",
       "      <th>path_autopsy_fourth_dx</th>\n",
       "      <th>path_autopsy_fifth_dx</th>\n",
       "      <th>path_autopsy_sixth_dx</th>\n",
       "      <th>path_autopsy_seventh_dx</th>\n",
       "      <th>...</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>dig_slide_avail</th>\n",
       "      <th>quant_path_avail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SUL_T_4530</td>\n",
       "      <td>T-4530</td>\n",
       "      <td>17.06666667</td>\n",
       "      <td>Parkinson's disease with dementia</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SUL_T_4105</td>\n",
       "      <td>T-4105</td>\n",
       "      <td>20</td>\n",
       "      <td>Alzheimer's disease (intermediate level neurop...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SUL_T_5259</td>\n",
       "      <td>T-5259</td>\n",
       "      <td>10.5</td>\n",
       "      <td>Parkinson's disease with dementia</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SUL_T_5385</td>\n",
       "      <td>T-5385</td>\n",
       "      <td>14.05</td>\n",
       "      <td>Control, Primary age-related tauopathy (PART)</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SUL_T_2465</td>\n",
       "      <td>T-2465</td>\n",
       "      <td>9.25</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_id source_subject_id duration_pmi  \\\n",
       "0  SUL_T_4530            T-4530  17.06666667   \n",
       "1  SUL_T_4105            T-4105           20   \n",
       "2  SUL_T_5259            T-5259         10.5   \n",
       "3  SUL_T_5385            T-5385        14.05   \n",
       "4  SUL_T_2465            T-2465         9.25   \n",
       "\n",
       "                                path_autopsy_dx_main path_autopsy_second_dx  \\\n",
       "0                  Parkinson's disease with dementia                     NA   \n",
       "1  Alzheimer's disease (intermediate level neurop...                     NA   \n",
       "2                  Parkinson's disease with dementia                     NA   \n",
       "3      Control, Primary age-related tauopathy (PART)                     NA   \n",
       "4     Control, Low level AD neuropathological change                     NA   \n",
       "\n",
       "  path_autopsy_third_dx path_autopsy_fourth_dx path_autopsy_fifth_dx  \\\n",
       "0                    NA                     NA                    NA   \n",
       "1                    NA                     NA                    NA   \n",
       "2                    NA                     NA                    NA   \n",
       "3                    NA                     NA                    NA   \n",
       "4                    NA                     NA                    NA   \n",
       "\n",
       "  path_autopsy_sixth_dx path_autopsy_seventh_dx  ... path_nia_ri  \\\n",
       "0                    NA                      NA  ...          NA   \n",
       "1                    NA                      NA  ...          NA   \n",
       "2                    NA                      NA  ...          NA   \n",
       "3                    NA                      NA  ...          NA   \n",
       "4                    NA                      NA  ...          NA   \n",
       "\n",
       "  path_nia_aa_a path_nia_aa_b path_nia_aa_c TDP43  \\\n",
       "0            NA            NA            NA    NA   \n",
       "1            NA            NA            NA    NA   \n",
       "2            NA            NA            NA    NA   \n",
       "3            NA            NA            NA    NA   \n",
       "4            NA            NA            NA    NA   \n",
       "\n",
       "  arteriolosclerosis_severity_scale amyloid_angiopathy_severity_scale  \\\n",
       "0                                NA                                NA   \n",
       "1                                NA                                NA   \n",
       "2                                NA                                NA   \n",
       "3                                NA                                NA   \n",
       "4                                NA                                NA   \n",
       "\n",
       "  path_ad_level dig_slide_avail quant_path_avail  \n",
       "0            NA              No               No  \n",
       "1            NA              No               No  \n",
       "2            NA              No               No  \n",
       "3            NA              No               No  \n",
       "4            NA              No               No  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['subject_id'] = \"SUL_\" + CLINPATH['source_subject_id'].str.replace(\"-\",\"_\")\n",
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DATA table\n",
    "\n",
    "The DATA.csv was not filled out.  So we will construct it by joining the manifest from the source of data transfer for Team Sulzer.\n",
    "\n",
    "### read_file_list\n",
    "Helper function to read sulzer manifest "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cingulate_files = Path.home() / \"Projects/ASAP/team-sulzer/cingulate_manifest.txt\"\n",
    "nigra_files = Path.home() / \"Projects/ASAP/team-sulzer/nigra_manifest.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp2 = pd.read_csv(\"sulzer_fastqs.txt\", delimiter=\" \", header=None, names=['file_name'])\n",
    "def read_gz_file_list(file_nm):\n",
    "    # Read the text file\n",
    "    with open(file_nm, 'r') as file:\n",
    "        lines = [line.strip() for line in file]\n",
    "\n",
    "    # Create a DataFrame\n",
    "    df = pd.DataFrame(lines, columns=['file_name'])\n",
    "\n",
    "    df['parts'] = df['file_name'].apply(lambda x: x.lstrip(\"gs://\").split(\"/\"))\n",
    "    df['source'] = df['parts'].apply(lambda x: x[1].lstrip(\"fastqs_\") )\n",
    "    df['name'] = df['parts'].apply(lambda x: x[-1])\n",
    "\n",
    "    df['name_parts'] = df['name'].apply(lambda x: x.rstrip('.fastq.gz').split(\"_\"))\n",
    "    df['sample_id'] = df['name_parts'].apply(lambda x: x[0])\n",
    "    df['lane_id'] = df['name_parts'].apply(lambda x: x[2])\n",
    "    df['file_type'] = df['name_parts'].apply(lambda x: x[3])\n",
    "    df.drop(columns=['parts','name_parts'], inplace=True)\n",
    "    return df\n",
    "\n",
    "# tmp2 = pd.read_csv(\"sulzer_fastqs.txt\", delimiter=\" \", header=None, names=['file_name'])\n",
    "def read_file_list_sulzer(file_nm):\n",
    "    # Read the text file\n",
    "    keep_ends = (\"fastq.gz\", \"bam\", \"tar\")\n",
    "    with open(file_nm, 'r') as file:\n",
    "        lines = [line.strip() for line in file if line.strip().endswith(keep_ends)]\n",
    "\n",
    "    # Create a DataFrame\n",
    "    df = pd.DataFrame(lines, columns=['file_name'])\n",
    "\n",
    "    df['parts'] = df['file_name'].apply(lambda x: x.split(\"/\"))\n",
    "    # df['source'] = df['parts'].apply(lambda x: x[5].lstrip(\"fastqs_\") )\n",
    "    df['file_name'] = df['parts'].apply(lambda x: x[-1])\n",
    "    df['sample_id'] = df['parts'].apply(lambda x: x[6] if x[-1].endswith((\"fastq.gz\",\"tar\")) else x[7])\n",
    "    df['file_type'] = df['parts'].apply(lambda x: x[7] if x[7].strip() in [\"fastq\",\"analysis\"] else \"bam\")\n",
    "    # df['tmp1'] = df['parts'].apply(lambda x: x[7] if len(x) > 7 else pd.NA)\n",
    "\n",
    "    # df['name_parts'] = df['name'].apply(lambda x: x.rstrip('.fastq.gz').split(\"_\"))\n",
    "    # df['sample_id'] = df['name_parts'].apply(lambda x: x[0])\n",
    "    # df['lane_id'] = df['name_parts'].apply(lambda x: x[2])\n",
    "    # df['file_type'] = df['name_parts'].apply(lambda x: x[3])\n",
    "    df.drop(columns=['parts'], inplace=True)\n",
    "    return df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "c_fastqs = read_file_list_sulzer(cingulate_files)\n",
    "c_fastqs['src'] = \"cingulate\"\n",
    "n_fastqs = read_file_list_sulzer(nigra_files)\n",
    "n_fastqs['src'] = \"nigra\"\n",
    "\n",
    "from_source = pd.concat([c_fastqs, n_fastqs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>file_type</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OJ1_cellranger_count_outs.tar</td>\n",
       "      <td>OJ1</td>\n",
       "      <td>analysis</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OJ1_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>OJ1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OJ1_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>OJ1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OJ1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>OJ1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OJ1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>OJ1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>OJ9_S1_L003_R2_001.fastq.gz</td>\n",
       "      <td>OJ9</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>OJ9_S1_L004_I1_001.fastq.gz</td>\n",
       "      <td>OJ9</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>OJ9_S1_L004_I2_001.fastq.gz</td>\n",
       "      <td>OJ9</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>OJ9_S1_L004_R1_001.fastq.gz</td>\n",
       "      <td>OJ9</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>OJ9_S1_L004_R2_001.fastq.gz</td>\n",
       "      <td>OJ9</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>459 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         file_name sample_id file_type    src\n",
       "0    OJ1_cellranger_count_outs.tar       OJ1  analysis  nigra\n",
       "1      OJ1_S1_L001_I1_001.fastq.gz       OJ1     fastq  nigra\n",
       "2      OJ1_S1_L001_I2_001.fastq.gz       OJ1     fastq  nigra\n",
       "3      OJ1_S1_L001_R1_001.fastq.gz       OJ1     fastq  nigra\n",
       "4      OJ1_S1_L001_R2_001.fastq.gz       OJ1     fastq  nigra\n",
       "..                             ...       ...       ...    ...\n",
       "454    OJ9_S1_L003_R2_001.fastq.gz       OJ9     fastq  nigra\n",
       "455    OJ9_S1_L004_I1_001.fastq.gz       OJ9     fastq  nigra\n",
       "456    OJ9_S1_L004_I2_001.fastq.gz       OJ9     fastq  nigra\n",
       "457    OJ9_S1_L004_R1_001.fastq.gz       OJ9     fastq  nigra\n",
       "458    OJ9_S1_L004_R2_001.fastq.gz       OJ9     fastq  nigra\n",
       "\n",
       "[459 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_fastqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OF009</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OF016</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OF005</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OF013</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id replicate replicate_count repeated_sample  batch\n",
       "0     OF001      Rep1               1               0  cing1\n",
       "1     OF009      Rep1               1               0  cing1\n",
       "2     OF016      Rep1               1               0  cing1\n",
       "3     OF005      Rep1               1               0  cing1\n",
       "4     OF013      Rep1               1               0  cing1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## now make DATA from scratch to ready to join into\n",
    "# Start with sample_id\n",
    "# DATA_ = read_meta_table(metadata_path / \"DATA.csv\")\n",
    "\n",
    "DATA = SAMPLE[[\"sample_id\", \"replicate\", \"replicate_count\",\"repeated_sample\",\"batch\"]].copy()\n",
    "\n",
    "DATA.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_type</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>OJ26_S1_L003_R2_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>OJ26_S1_L004_I1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>OJ26_S1_L004_I2_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>OJ26_S1_L004_R1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>OJ26_S1_L004_R2_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>nigra</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id replicate replicate_count repeated_sample batch  \\\n",
       "707      OJ26      Rep1               1               0   sn1   \n",
       "708      OJ26      Rep1               1               0   sn1   \n",
       "709      OJ26      Rep1               1               0   sn1   \n",
       "710      OJ26      Rep1               1               0   sn1   \n",
       "711      OJ26      Rep1               1               0   sn1   \n",
       "\n",
       "                        file_name file_type    src  \n",
       "707  OJ26_S1_L003_R2_001.fastq.gz     fastq  nigra  \n",
       "708  OJ26_S1_L004_I1_001.fastq.gz     fastq  nigra  \n",
       "709  OJ26_S1_L004_I2_001.fastq.gz     fastq  nigra  \n",
       "710  OJ26_S1_L004_R1_001.fastq.gz     fastq  nigra  \n",
       "711  OJ26_S1_L004_R2_001.fastq.gz     fastq  nigra  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA = pd.merge(DATA, from_source, on='sample_id', how='left')\n",
    "DATA.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA['omic'] = 'RNA'\n",
    "DATA['adjustments'] = 'Raw'\n",
    "DATA['technology'] = \"SN\"\n",
    "DATA['file_description']= DATA['file_type'].map({\"bam\": \"Binary aligned sequence\",\n",
    "                                                \"fastq\": \"gzip fastq\",\n",
    "                                                \"analysis\": \"cellranger counts tar archive\"})\n",
    "DATA['file_description']= DATA['file_type'].map({\"bam\": \"Reads\",\n",
    "                                                \"fastq\": \"Reads\",\n",
    "                                                \"analysis\": \"Counts\"})\n",
    "\n",
    "# now fix the file_type\n",
    "DATA['file_type'] = DATA['file_type'].replace({'fastq':'fastq', \"bam\":'Per sample raw file', \"analysis\":'Per sample processed file'})\n",
    "DATA['content'] = 'Counts'\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get \"md5s\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "cingulate_md5s = Path.home() / \"Projects/ASAP/team-sulzer/cingulate_fastq_md5.log\"\n",
    "nigra_md5s = Path.home() / \"Projects/ASAP/team-sulzer/nigra_fastq_md5.log\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from checksums import extract_hashes_from_gsutil\n",
    "\n",
    "# cingulate\n",
    "cingulate_hashs = extract_hashes_from_gsutil(cingulate_md5s)\n",
    "\n",
    "# nigra\n",
    "nigra_hashs = extract_hashes_from_gsutil(nigra_md5s)\n",
    "\n",
    "# combine cingulate and nigra\n",
    "source_file_crc = cingulate_hashs[0] | nigra_hashs[0]\n",
    "source_file_md5 = cingulate_hashs[1] | nigra_hashs[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA['file_MD5'] = DATA['file_name'].map(source_file_md5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_type</th>\n",
       "      <th>src</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustments</th>\n",
       "      <th>technology</th>\n",
       "      <th>file_description</th>\n",
       "      <th>content</th>\n",
       "      <th>file_MD5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>possorted_genome_bam.bam</td>\n",
       "      <td>Per sample raw file</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_cellranger_count_outs.tar</td>\n",
       "      <td>Per sample processed file</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Counts</td>\n",
       "      <td>Counts</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>74a74c8929b01a6da7e79b83f91a6607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>644d95728b50f1ff9db7becfa5b5c394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>129dfb6fe6115966e2cf2af5e3c82bdf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id replicate replicate_count repeated_sample  batch  \\\n",
       "0     OF001      Rep1               1               0  cing1   \n",
       "1     OF001      Rep1               1               0  cing1   \n",
       "2     OF001      Rep1               1               0  cing1   \n",
       "3     OF001      Rep1               1               0  cing1   \n",
       "4     OF001      Rep1               1               0  cing1   \n",
       "\n",
       "                         file_name                  file_type        src omic  \\\n",
       "0         possorted_genome_bam.bam        Per sample raw file  cingulate  RNA   \n",
       "1  OF001_cellranger_count_outs.tar  Per sample processed file  cingulate  RNA   \n",
       "2    OF001_S1_L001_I1_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "3    OF001_S1_L001_I2_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "4    OF001_S1_L001_R1_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "\n",
       "  adjustments technology file_description content  \\\n",
       "0         Raw         SN            Reads  Counts   \n",
       "1         Raw         SN           Counts  Counts   \n",
       "2         Raw         SN            Reads  Counts   \n",
       "3         Raw         SN            Reads  Counts   \n",
       "4         Raw         SN            Reads  Counts   \n",
       "\n",
       "                           file_MD5  \n",
       "0                               NaN  \n",
       "1                               NaN  \n",
       "2  74a74c8929b01a6da7e79b83f91a6607  \n",
       "3  644d95728b50f1ff9db7becfa5b5c394  \n",
       "4  129dfb6fe6115966e2cf2af5e3c82bdf  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "DATA.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# No data found for `OJ8`!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# load manifests and md5s\n",
    "\n",
    "Separate manifests for cingulate and nigra to make it easier for Team Sulzer to collect.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(46, 34)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(45, 34)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE[SAMPLE['sample_id']=='OJ8']\n",
    "empty_sample = SAMPLE['sample_id']=='OJ8'\n",
    "print(SAMPLE.shape)\n",
    "SAMPLE = SAMPLE[~empty_sample]\n",
    "SAMPLE.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(712, 14)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(711, 14)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA[DATA['sample_id']=='OJ8']\n",
    "empty_sample = DATA['sample_id']=='OJ8'\n",
    "print(DATA.shape)\n",
    "DATA = DATA[~empty_sample]\n",
    "DATA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_type</th>\n",
       "      <th>src</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>technology</th>\n",
       "      <th>file_description</th>\n",
       "      <th>content</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>possorted_genome_bam.bam</td>\n",
       "      <td>Per sample raw file</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_cellranger_count_outs.tar</td>\n",
       "      <td>Per sample processed file</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Counts</td>\n",
       "      <td>Counts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_I1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>74a74c8929b01a6da7e79b83f91a6607</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_I2_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>644d95728b50f1ff9db7becfa5b5c394</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OF001</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>OF001_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fastq</td>\n",
       "      <td>cingulate</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>SN</td>\n",
       "      <td>Reads</td>\n",
       "      <td>Counts</td>\n",
       "      <td>129dfb6fe6115966e2cf2af5e3c82bdf</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id replicate replicate_count repeated_sample  batch  \\\n",
       "0     OF001      Rep1               1               0  cing1   \n",
       "1     OF001      Rep1               1               0  cing1   \n",
       "2     OF001      Rep1               1               0  cing1   \n",
       "3     OF001      Rep1               1               0  cing1   \n",
       "4     OF001      Rep1               1               0  cing1   \n",
       "\n",
       "                         file_name                  file_type        src omic  \\\n",
       "0         possorted_genome_bam.bam        Per sample raw file  cingulate  RNA   \n",
       "1  OF001_cellranger_count_outs.tar  Per sample processed file  cingulate  RNA   \n",
       "2    OF001_S1_L001_I1_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "3    OF001_S1_L001_I2_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "4    OF001_S1_L001_R1_001.fastq.gz                      fastq  cingulate  RNA   \n",
       "\n",
       "  adjustment technology file_description content  \\\n",
       "0        Raw         SN            Reads  Counts   \n",
       "1        Raw         SN           Counts  Counts   \n",
       "2        Raw         SN            Reads  Counts   \n",
       "3        Raw         SN            Reads  Counts   \n",
       "4        Raw         SN            Reads  Counts   \n",
       "\n",
       "                           file_MD5 time header annotation configuration_file  \n",
       "0                               NaN   NA     NA         NA                 NA  \n",
       "1                               NaN   NA     NA         NA                 NA  \n",
       "2  74a74c8929b01a6da7e79b83f91a6607   NA     NA         NA                 NA  \n",
       "3  644d95728b50f1ff9db7becfa5b5c394   NA     NA         NA                 NA  \n",
       "4  129dfb6fe6115966e2cf2af5e3c82bdf   NA     NA         NA                 NA  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA.rename(columns={'adjustments':'adjustment'}, inplace=True)\n",
    "DATA['time'] = NULL\n",
    "DATA['header'] = NULL\n",
    "DATA['annotation'] = NULL\n",
    "DATA['configuration_file'] = NULL\n",
    "\n",
    "DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_4530', 'SUL_T_4105', 'SUL_T_5259', 'SUL_T_5385',\n",
       "       'SUL_T_2465', 'SUL_T_3799', 'SUL_T_4424', 'SUL_T_5133',\n",
       "       'SUL_T_5227', 'SUL_T_5547', 'SUL_T_5359', 'SUL_T_4116',\n",
       "       'SUL_T_4839', 'SUL_T_2007', 'SUL_T_4938', 'SUL_T_5498',\n",
       "       'SUL_T_2791', 'SUL_T_4695', 'SUL_T_4934', 'SUL_T_3870',\n",
       "       'SUL_T_124', 'SUL_T_81', 'SUL_T_4515', 'SUL_T_4235', 'SUL_T_4815',\n",
       "       'SUL_T_4857', 'SUL_T_3998', 'SUL_T_4462', 'SUL_T_4108',\n",
       "       'SUL_T_4174', 'SUL_T_3860', 'SUL_T_4162', 'SUL_T_4210',\n",
       "       'SUL_T_233', 'SUL_T_1976', 'SUL_T_2017', 'SUL_T_308', 'SUL_T_1952',\n",
       "       'SUL_T_777', 'SUL_T_1619', 'SUL_T_290', 'SUL_T_551', 'SUL_T_257',\n",
       "       'SUL_T_343', 'SUL_T_362'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['subject_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_4530', 'SUL_T_4105', 'SUL_T_5259', 'SUL_T_5385',\n",
       "       'SUL_T_2465', 'SUL_T_3799', 'SUL_T_4424', 'SUL_T_5133',\n",
       "       'SUL_T_5227', 'SUL_T_5547', 'SUL_T_5359', 'SUL_T_4116',\n",
       "       'SUL_T_4839', 'SUL_T_2007', 'SUL_T_4938', 'SUL_T_5498',\n",
       "       'SUL_T_2791', 'SUL_T_4695', 'SUL_T_4934', 'SUL_T_3870',\n",
       "       'SUL_T_124', 'SUL_T_81', 'SUL_T_4515', 'SUL_T_4235', 'SUL_T_4815',\n",
       "       'SUL_T_4857', 'SUL_T_3998', 'SUL_T_4462', 'SUL_T_4108',\n",
       "       'SUL_T_4174', 'SUL_T_3860', 'SUL_T_4162', 'SUL_T_4210',\n",
       "       'SUL_T_233', 'SUL_T_1976', 'SUL_T_2017', 'SUL_T_308', 'SUL_T_1952',\n",
       "       'SUL_T_777', 'SUL_T_1619', 'SUL_T_290', 'SUL_T_551', 'SUL_T_257',\n",
       "       'SUL_T_343', 'SUL_T_362'], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['subject_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_4530', 'SUL_T_4105', 'SUL_T_5259', 'SUL_T_5385',\n",
       "       'SUL_T_2465', 'SUL_T_3799', 'SUL_T_4424', 'SUL_T_5133',\n",
       "       'SUL_T_5227', 'SUL_T_5547', 'SUL_T_5359', 'SUL_T_4116',\n",
       "       'SUL_T_4839', 'SUL_T_2007', 'SUL_T_4938', 'SUL_T_5498',\n",
       "       'SUL_T_2791', 'SUL_T_4695', 'SUL_T_4934', 'SUL_T_3870',\n",
       "       'SUL_T_124', 'SUL_T_81', 'SUL_T_4515', 'SUL_T_4235', 'SUL_T_4815',\n",
       "       'SUL_T_4857', 'SUL_T_3998', 'SUL_T_4462', 'SUL_T_4108',\n",
       "       'SUL_T_4174', 'SUL_T_3860', 'SUL_T_4162', 'SUL_T_4210',\n",
       "       'SUL_T_233', 'SUL_T_1976', 'SUL_T_2017', 'SUL_T_308', 'SUL_T_1952',\n",
       "       'SUL_T_777', 'SUL_T_1619', 'SUL_T_290', 'SUL_T_551', 'SUL_T_257',\n",
       "       'SUL_T_343', 'SUL_T_362'], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['subject_id'].unique()\n",
    "# CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>alternate_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OF001</td>\n",
       "      <td>SUL_T_362</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OF009</td>\n",
       "      <td>SUL_T_308</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OF016</td>\n",
       "      <td>SUL_T_3860</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OF005</td>\n",
       "      <td>SUL_T_290</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OF013</td>\n",
       "      <td>SUL_T_233</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OF018</td>\n",
       "      <td>SUL_T_4108</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OF003</td>\n",
       "      <td>SUL_T_257</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OF007</td>\n",
       "      <td>SUL_T_777</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OF011</td>\n",
       "      <td>SUL_T_1976</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OF015</td>\n",
       "      <td>SUL_T_4162</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>OF017</td>\n",
       "      <td>SUL_T_4174</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>OF002</td>\n",
       "      <td>SUL_T_343</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>OF004</td>\n",
       "      <td>SUL_T_551</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>OF006</td>\n",
       "      <td>SUL_T_1619</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>OF008</td>\n",
       "      <td>SUL_T_1952</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OF010</td>\n",
       "      <td>SUL_T_2017</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>OF012</td>\n",
       "      <td>SUL_T_2465</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>OF014</td>\n",
       "      <td>SUL_T_4210</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>cing1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Cingulate Cortex</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0003027</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>OJ15</td>\n",
       "      <td>SUL_T_81</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OJ16</td>\n",
       "      <td>SUL_T_124</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>OJ22</td>\n",
       "      <td>SUL_T_2007</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OJ5</td>\n",
       "      <td>SUL_T_2465</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OJ2</td>\n",
       "      <td>SUL_T_2791</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OJ4</td>\n",
       "      <td>SUL_T_3799</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OJ17</td>\n",
       "      <td>SUL_T_3870</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OJ10</td>\n",
       "      <td>SUL_T_3998</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OJ24</td>\n",
       "      <td>SUL_T_4116</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OJ13</td>\n",
       "      <td>SUL_T_4235</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>OJ3</td>\n",
       "      <td>SUL_T_4424</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>OJ1</td>\n",
       "      <td>SUL_T_4462</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>OJ14</td>\n",
       "      <td>SUL_T_4515</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>OJ9</td>\n",
       "      <td>SUL_T_4530</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>OJ19</td>\n",
       "      <td>SUL_T_4695</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>OJ12</td>\n",
       "      <td>SUL_T_4815</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>OJ23</td>\n",
       "      <td>SUL_T_4839</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>OJ11</td>\n",
       "      <td>SUL_T_4857</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>OJ18</td>\n",
       "      <td>SUL_T_4934</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>OJ21</td>\n",
       "      <td>SUL_T_4938</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>OJ28</td>\n",
       "      <td>SUL_T_5133</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>OJ27</td>\n",
       "      <td>SUL_T_5227</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>OJ7</td>\n",
       "      <td>SUL_T_5259</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>OJ25</td>\n",
       "      <td>SUL_T_5359</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>OJ6</td>\n",
       "      <td>SUL_T_5385</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>OJ20</td>\n",
       "      <td>SUL_T_5498</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002037</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>OJ26</td>\n",
       "      <td>SUL_T_5547</td>\n",
       "      <td>NA</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sn1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Substantia Nigra</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NA</td>\n",
       "      <td>UBERON:0002038</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45 rows √ó 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id  subject_id source_sample_id replicate replicate_count  \\\n",
       "0      OF001   SUL_T_362               NA      Rep1               1   \n",
       "1      OF009   SUL_T_308               NA      Rep1               1   \n",
       "2      OF016  SUL_T_3860               NA      Rep1               1   \n",
       "3      OF005   SUL_T_290               NA      Rep1               1   \n",
       "4      OF013   SUL_T_233               NA      Rep1               1   \n",
       "5      OF018  SUL_T_4108               NA      Rep1               1   \n",
       "6      OF003   SUL_T_257               NA      Rep1               1   \n",
       "7      OF007   SUL_T_777               NA      Rep1               1   \n",
       "8      OF011  SUL_T_1976               NA      Rep1               1   \n",
       "9      OF015  SUL_T_4162               NA      Rep1               1   \n",
       "10     OF017  SUL_T_4174               NA      Rep1               1   \n",
       "11     OF002   SUL_T_343               NA      Rep1               1   \n",
       "12     OF004   SUL_T_551               NA      Rep1               1   \n",
       "13     OF006  SUL_T_1619               NA      Rep1               1   \n",
       "14     OF008  SUL_T_1952               NA      Rep1               1   \n",
       "15     OF010  SUL_T_2017               NA      Rep1               1   \n",
       "16     OF012  SUL_T_2465               NA      Rep1               1   \n",
       "17     OF014  SUL_T_4210               NA      Rep1               1   \n",
       "18      OJ15    SUL_T_81               NA      Rep1               1   \n",
       "19      OJ16   SUL_T_124               NA      Rep1               1   \n",
       "20      OJ22  SUL_T_2007               NA      Rep1               1   \n",
       "21       OJ5  SUL_T_2465               NA      Rep1               1   \n",
       "22       OJ2  SUL_T_2791               NA      Rep1               1   \n",
       "23       OJ4  SUL_T_3799               NA      Rep1               1   \n",
       "24      OJ17  SUL_T_3870               NA      Rep1               1   \n",
       "25      OJ10  SUL_T_3998               NA      Rep1               1   \n",
       "27      OJ24  SUL_T_4116               NA      Rep1               1   \n",
       "28      OJ13  SUL_T_4235               NA      Rep1               1   \n",
       "29       OJ3  SUL_T_4424               NA      Rep1               1   \n",
       "30       OJ1  SUL_T_4462               NA      Rep1               1   \n",
       "31      OJ14  SUL_T_4515               NA      Rep1               1   \n",
       "32       OJ9  SUL_T_4530               NA      Rep1               1   \n",
       "33      OJ19  SUL_T_4695               NA      Rep1               1   \n",
       "34      OJ12  SUL_T_4815               NA      Rep1               1   \n",
       "35      OJ23  SUL_T_4839               NA      Rep1               1   \n",
       "36      OJ11  SUL_T_4857               NA      Rep1               1   \n",
       "37      OJ18  SUL_T_4934               NA      Rep1               1   \n",
       "38      OJ21  SUL_T_4938               NA      Rep1               1   \n",
       "39      OJ28  SUL_T_5133               NA      Rep1               1   \n",
       "40      OJ27  SUL_T_5227               NA      Rep1               1   \n",
       "41       OJ7  SUL_T_5259               NA      Rep1               1   \n",
       "42      OJ25  SUL_T_5359               NA      Rep1               1   \n",
       "43       OJ6  SUL_T_5385               NA      Rep1               1   \n",
       "44      OJ20  SUL_T_5498               NA      Rep1               1   \n",
       "45      OJ26  SUL_T_5547               NA      Rep1               1   \n",
       "\n",
       "   repeated_sample  batch tissue      brain_region hemisphere  ...  \\\n",
       "0                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "1                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "2                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "3                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "4                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "5                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "6                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "7                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "8                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "9                0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "10               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "11               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "12               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "13               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "14               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "15               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "16               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "17               0  cing1  Brain  Cingulate Cortex         NA  ...   \n",
       "18               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "19               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "20               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "21               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "22               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "23               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "24               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "25               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "27               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "28               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "29               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "30               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "31               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "32               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "33               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "34               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "35               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "36               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "37               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "38               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "39               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "40               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "41               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "42               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "43               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "44               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "45               0    sn1  Brain  Substantia Nigra         NA  ...   \n",
       "\n",
       "   self_reported_ethnicity_ontology_term_id disease_ontology_term_id  \\\n",
       "0                                   Unknown                       NA   \n",
       "1                                   Unknown                       NA   \n",
       "2                                   Unknown                       NA   \n",
       "3                                   Unknown                       NA   \n",
       "4                                   Unknown                       NA   \n",
       "5                                   Unknown                       NA   \n",
       "6                                   Unknown                       NA   \n",
       "7                                   Unknown                       NA   \n",
       "8                                   Unknown                       NA   \n",
       "9                                   Unknown                       NA   \n",
       "10                                  Unknown                       NA   \n",
       "11                                  Unknown                       NA   \n",
       "12                                  Unknown                       NA   \n",
       "13                                  Unknown                       NA   \n",
       "14                                  Unknown                       NA   \n",
       "15                                  Unknown                       NA   \n",
       "16                                  Unknown                       NA   \n",
       "17                                  Unknown                       NA   \n",
       "18                                  Unknown                       NA   \n",
       "19                                  Unknown                       NA   \n",
       "20                                  Unknown                       NA   \n",
       "21                                  Unknown                       NA   \n",
       "22                                  Unknown                       NA   \n",
       "23                                  Unknown                       NA   \n",
       "24                                  Unknown                       NA   \n",
       "25                                  Unknown                       NA   \n",
       "27                                  Unknown                       NA   \n",
       "28                                  Unknown                       NA   \n",
       "29                                  Unknown                       NA   \n",
       "30                                  Unknown                       NA   \n",
       "31                                  Unknown                       NA   \n",
       "32                                  Unknown                       NA   \n",
       "33                                  Unknown                       NA   \n",
       "34                                  Unknown                       NA   \n",
       "35                                  Unknown                       NA   \n",
       "36                                  Unknown                       NA   \n",
       "37                                  Unknown                       NA   \n",
       "38                                  Unknown                       NA   \n",
       "39                                  Unknown                       NA   \n",
       "40                                  Unknown                       NA   \n",
       "41                                  Unknown                       NA   \n",
       "42                                  Unknown                       NA   \n",
       "43                                  Unknown                       NA   \n",
       "44                                  Unknown                       NA   \n",
       "45                                  Unknown                       NA   \n",
       "\n",
       "   tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id  \\\n",
       "0           UBERON:0003027                         NA                     NA   \n",
       "1           UBERON:0003027                         NA                     NA   \n",
       "2           UBERON:0003027                         NA                     NA   \n",
       "3           UBERON:0003027                         NA                     NA   \n",
       "4           UBERON:0003027                         NA                     NA   \n",
       "5           UBERON:0003027                         NA                     NA   \n",
       "6           UBERON:0003027                         NA                     NA   \n",
       "7           UBERON:0003027                         NA                     NA   \n",
       "8           UBERON:0003027                         NA                     NA   \n",
       "9           UBERON:0003027                         NA                     NA   \n",
       "10          UBERON:0003027                         NA                     NA   \n",
       "11          UBERON:0003027                         NA                     NA   \n",
       "12          UBERON:0003027                         NA                     NA   \n",
       "13          UBERON:0003027                         NA                     NA   \n",
       "14          UBERON:0003027                         NA                     NA   \n",
       "15          UBERON:0003027                         NA                     NA   \n",
       "16          UBERON:0003027                         NA                     NA   \n",
       "17          UBERON:0003027                         NA                     NA   \n",
       "18          UBERON:0002037                         NA                     NA   \n",
       "19          UBERON:0002038                         NA                     NA   \n",
       "20          UBERON:0002037                         NA                     NA   \n",
       "21          UBERON:0002038                         NA                     NA   \n",
       "22          UBERON:0002037                         NA                     NA   \n",
       "23          UBERON:0002038                         NA                     NA   \n",
       "24          UBERON:0002037                         NA                     NA   \n",
       "25          UBERON:0002038                         NA                     NA   \n",
       "27          UBERON:0002038                         NA                     NA   \n",
       "28          UBERON:0002037                         NA                     NA   \n",
       "29          UBERON:0002038                         NA                     NA   \n",
       "30          UBERON:0002037                         NA                     NA   \n",
       "31          UBERON:0002038                         NA                     NA   \n",
       "32          UBERON:0002037                         NA                     NA   \n",
       "33          UBERON:0002038                         NA                     NA   \n",
       "34          UBERON:0002037                         NA                     NA   \n",
       "35          UBERON:0002038                         NA                     NA   \n",
       "36          UBERON:0002037                         NA                     NA   \n",
       "37          UBERON:0002038                         NA                     NA   \n",
       "38          UBERON:0002037                         NA                     NA   \n",
       "39          UBERON:0002038                         NA                     NA   \n",
       "40          UBERON:0002037                         NA                     NA   \n",
       "41          UBERON:0002038                         NA                     NA   \n",
       "42          UBERON:0002037                         NA                     NA   \n",
       "43          UBERON:0002038                         NA                     NA   \n",
       "44          UBERON:0002037                         NA                     NA   \n",
       "45          UBERON:0002038                         NA                     NA   \n",
       "\n",
       "   suspension_type DV200 pm_PH donor_id alternate_sample_id  \n",
       "0          Nucleus    NA    NA       NA                  NA  \n",
       "1          Nucleus    NA    NA       NA                  NA  \n",
       "2          Nucleus    NA    NA       NA                  NA  \n",
       "3          Nucleus    NA    NA       NA                  NA  \n",
       "4          Nucleus    NA    NA       NA                  NA  \n",
       "5          Nucleus    NA    NA       NA                  NA  \n",
       "6          Nucleus    NA    NA       NA                  NA  \n",
       "7          Nucleus    NA    NA       NA                  NA  \n",
       "8          Nucleus    NA    NA       NA                  NA  \n",
       "9          Nucleus    NA    NA       NA                  NA  \n",
       "10         Nucleus    NA    NA       NA                  NA  \n",
       "11         Nucleus    NA    NA       NA                  NA  \n",
       "12         Nucleus    NA    NA       NA                  NA  \n",
       "13         Nucleus    NA    NA       NA                  NA  \n",
       "14         Nucleus    NA    NA       NA                  NA  \n",
       "15         Nucleus    NA    NA       NA                  NA  \n",
       "16         Nucleus    NA    NA       NA                  NA  \n",
       "17         Nucleus    NA    NA       NA                  NA  \n",
       "18         Nucleus    NA    NA       NA                  NA  \n",
       "19         Nucleus    NA    NA       NA                  NA  \n",
       "20         Nucleus    NA    NA       NA                  NA  \n",
       "21         Nucleus    NA    NA       NA                  NA  \n",
       "22         Nucleus    NA    NA       NA                  NA  \n",
       "23         Nucleus    NA    NA       NA                  NA  \n",
       "24         Nucleus    NA    NA       NA                  NA  \n",
       "25         Nucleus    NA    NA       NA                  NA  \n",
       "27         Nucleus    NA    NA       NA                  NA  \n",
       "28         Nucleus    NA    NA       NA                  NA  \n",
       "29         Nucleus    NA    NA       NA                  NA  \n",
       "30         Nucleus    NA    NA       NA                  NA  \n",
       "31         Nucleus    NA    NA       NA                  NA  \n",
       "32         Nucleus    NA    NA       NA                  NA  \n",
       "33         Nucleus    NA    NA       NA                  NA  \n",
       "34         Nucleus    NA    NA       NA                  NA  \n",
       "35         Nucleus    NA    NA       NA                  NA  \n",
       "36         Nucleus    NA    NA       NA                  NA  \n",
       "37         Nucleus    NA    NA       NA                  NA  \n",
       "38         Nucleus    NA    NA       NA                  NA  \n",
       "39         Nucleus    NA    NA       NA                  NA  \n",
       "40         Nucleus    NA    NA       NA                  NA  \n",
       "41         Nucleus    NA    NA       NA                  NA  \n",
       "42         Nucleus    NA    NA       NA                  NA  \n",
       "43         Nucleus    NA    NA       NA                  NA  \n",
       "44         Nucleus    NA    NA       NA                  NA  \n",
       "45         Nucleus    NA    NA       NA                  NA  \n",
       "\n",
       "[45 rows x 34 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SUL_T_4530': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4105': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_5259': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_5385': 'Control, Primary age-related tauopathy (PART)',\n",
       " 'SUL_T_2465': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_3799': 'Other neurological disorder',\n",
       " 'SUL_T_4424': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_5133': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_5227': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_5547': 'Control, Primary age-related tauopathy (PART)',\n",
       " 'SUL_T_5359': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_4116': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_4839': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_2007': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4938': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_5498': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_2791': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_4695': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4934': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_3870': 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)',\n",
       " 'SUL_T_124': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_81': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4515': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4235': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_4815': \"Parkinson's disease\",\n",
       " 'SUL_T_4857': \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       " 'SUL_T_3998': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_4462': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_4108': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4174': 'Control, no misfolded protein or significant vascular pathology',\n",
       " 'SUL_T_3860': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4162': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_4210': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_233': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_1976': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_2017': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_308': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_1952': 'Control, Low level AD neuropathological change',\n",
       " 'SUL_T_777': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_1619': 'Control, no misfolded protein or significant vascular pathology',\n",
       " 'SUL_T_290': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_551': 'Control, no misfolded protein or significant vascular pathology',\n",
       " 'SUL_T_257': \"Parkinson's disease with dementia\",\n",
       " 'SUL_T_343': 'Control, no misfolded protein or significant vascular pathology',\n",
       " 'SUL_T_362': \"Parkinson's disease with dementia\"}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(zip(CLINPATH['subject_id'], CLINPATH['path_autopsy_dx_main']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                     Parkinson's disease with dementia\n",
       "1     Alzheimer's disease (intermediate level neurop...\n",
       "2                     Parkinson's disease with dementia\n",
       "3         Control, Primary age-related tauopathy (PART)\n",
       "4        Control, Low level AD neuropathological change\n",
       "5                           Other neurological disorder\n",
       "6                     Parkinson's disease with dementia\n",
       "7     Alzheimer's disease (intermediate level neurop...\n",
       "8     Alzheimer's disease (intermediate level neurop...\n",
       "9         Control, Primary age-related tauopathy (PART)\n",
       "10    Alzheimer's disease (intermediate level neurop...\n",
       "11    Alzheimer's disease (intermediate level neurop...\n",
       "12                    Parkinson's disease with dementia\n",
       "13                    Parkinson's disease with dementia\n",
       "14                    Parkinson's disease with dementia\n",
       "15                    Parkinson's disease with dementia\n",
       "16       Control, Low level AD neuropathological change\n",
       "17                    Parkinson's disease with dementia\n",
       "18                    Parkinson's disease with dementia\n",
       "19    Control, Limbic predominant age-related TDP43 ...\n",
       "20                    Parkinson's disease with dementia\n",
       "21                    Parkinson's disease with dementia\n",
       "22                    Parkinson's disease with dementia\n",
       "23       Control, Low level AD neuropathological change\n",
       "24                                  Parkinson's disease\n",
       "25    Alzheimer's disease (intermediate level neurop...\n",
       "26       Control, Low level AD neuropathological change\n",
       "27       Control, Low level AD neuropathological change\n",
       "28                    Parkinson's disease with dementia\n",
       "29    Control, no misfolded protein or significant v...\n",
       "30                    Parkinson's disease with dementia\n",
       "31                    Parkinson's disease with dementia\n",
       "32       Control, Low level AD neuropathological change\n",
       "33                    Parkinson's disease with dementia\n",
       "34       Control, Low level AD neuropathological change\n",
       "35                    Parkinson's disease with dementia\n",
       "36       Control, Low level AD neuropathological change\n",
       "37                    Parkinson's disease with dementia\n",
       "38       Control, Low level AD neuropathological change\n",
       "39                    Parkinson's disease with dementia\n",
       "40    Control, no misfolded protein or significant v...\n",
       "41                    Parkinson's disease with dementia\n",
       "42    Control, no misfolded protein or significant v...\n",
       "43                    Parkinson's disease with dementia\n",
       "44    Control, no misfolded protein or significant v...\n",
       "45                    Parkinson's disease with dementia\n",
       "Name: primary_diagnosis, dtype: object"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['primary_diagnosis'] = SUBJECT['subject_id'].map(dict(zip(CLINPATH['subject_id'], CLINPATH['path_autopsy_dx_main'])))\n",
    "SUBJECT['primary_diagnosis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# via chatGPT\n",
    "diagnosis_mapping = {\n",
    "    \"Parkinson's disease with dementia\": \"Dementia with Lewy bodies\",\n",
    "    \"Alzheimer's disease (intermediate level neuropathological change)\": \"Alzheimer's disease\",\n",
    "    \"Control, Primary age-related tauopathy (PART)\": \"Other neurological disorder\",\n",
    "    \"Control, Low level AD neuropathological change\": \"Prodromal non-motor PD\",\n",
    "    \"Control, Limbic predominant age-related TDP43 proteinopathy (LATE)\": \"Frontotemporal dementia\",\n",
    "    \"Parkinson's disease\": \"Idiopathic PD\",\n",
    "    \"Control, no misfolded protein or significant vascular pathology\": \"No PD nor other neurological disorder\"\n",
    "}\n",
    "\n",
    "\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace(diagnosis_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "v2_path = metadata_path / \"v2\"\n",
    "STUDY.to_csv(v2_path / f\"STUDY.csv\", index=False)\n",
    "SAMPLE.to_csv(v2_path / f\"SAMPLE.csv\", index=False)\n",
    "SUBJECT.to_csv(v2_path / f\"SUBJECT.csv\", index=False)\n",
    "PROTOCOL.to_csv(v2_path / f\"PROTOCOL.csv\", index=False)\n",
    "CLINPATH.to_csv(v2_path / f\"CLINPATH.csv\", index=False)\n",
    "DATA.to_csv(v2_path / f\"DATA.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "üö®‚ö†Ô∏è‚ùó **8 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- ASAP_grant_id: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "üö®‚ö†Ô∏è‚ùó **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **17 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- race: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 23/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 46/46 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*age_at_collection*_:  invalid values üí©'77', '89+ ', '79', '88', '83', '78', '87', '75', '86', '85', '68', '73', '80', '72', '71', '50', '74', '81', '70', '66', '62'\n",
      "    - valid ‚û°Ô∏è float or NULL ('NA')\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **28 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_second_dx: 44/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 30/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 1/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 19/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 46/46 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **14 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_sample_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- hemisphere: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- region_level_2: 18/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 27/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- input_cell_count: 27/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_sample_id: 45/45 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values üí©'915.1258 ng', '1411.8474 ng', '1608.6002 ng', '704.4034 ng', '1444.6272 ng', '896.3852 ng', '2392.286 ng', '714.5692 ng', '923.9646 ng', '866.4158 ng', '1420.8972 ng', '2169.106 ng', '1069.3564 ng', '1192.589 ng', '1008.9174 ng', '983.9328 ng', '1266.1952 ng', '945.216 ng', 'NA'\n",
      "    - valid ‚û°Ô∏è int or NULL ('NA')\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "üö®‚ö†Ô∏è‚ùó **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- file_MD5: 63/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 711/711 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Extra field in DATA: src**\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "tables = CDE['Table'].unique()\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    df = read_meta_table(v2_path / f\"{table}.csv\")\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    dfs[table] = full_table\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['STUDY', 'PROTOCOL', 'SUBJECT', 'CLINPATH', 'SAMPLE', 'DATA'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### save extras as auxillary tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved DATA_auxiliary.csv\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "tables = CDE['Table'].unique()\n",
    "\n",
    "dfs = {}\n",
    "for table in tables:\n",
    "    df = read_meta_table(v2_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v2_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v2_path / f\"{table}.csv\", index=False)\n",
    "    dfs[table] = df_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make tables conform to CDE and save extra columns as \"auxiliary\"\n",
    "v2_path = metadata_path / \"v2\"\n",
    "\n",
    "for table in tables:\n",
    "    df = dfs[table]\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(v2_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v2_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "validate v2 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "üö®‚ö†Ô∏è‚ùó **8 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- ASAP_grant_id: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCHID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "üö®‚ö†Ô∏è‚ùó **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **17 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- GP2_id: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- race: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- ethnicity: 23/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- family_history: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_onset: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- age_at_diagnosis: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- first_motor_symptom: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_dementia_mci: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- hx_melanoma: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- education_level: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- smoking_years: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cognitive_status: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- time_from_baseline: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 46/46 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*age_at_collection*_:  invalid values üí©'77', '89+ ', '79', '88', '83', '78', '87', '75', '86', '85', '68', '73', '80', '72', '71', '50', '74', '81', '70', '66', '62'\n",
      "    - valid ‚û°Ô∏è float or NULL ('NA')\n",
      "\n",
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **28 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- path_autopsy_second_dx: 44/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 30/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- cause_death: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_cause_death_1: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_braak_nft: 1/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_cerad: 19/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- path_thal: 46/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- known_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 46/46 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 46/46 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **14 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_sample_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- hemisphere: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- region_level_2: 18/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- RIN: 27/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- source_RIN: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- input_cell_count: 27/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- disease_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_ontology_term_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- DV200: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_sample_id: 45/45 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values üí©'915.1258 ng', '1411.8474 ng', '1608.6002 ng', '704.4034 ng', '1444.6272 ng', '896.3852 ng', '2392.286 ng', '714.5692 ng', '923.9646 ng', '866.4158 ng', '1420.8972 ng', '2169.106 ng', '1069.3564 ng', '1192.589 ng', '1008.9174 ng', '983.9328 ng', '1266.1952 ng', '945.216 ng', 'NA'\n",
      "    - valid ‚û°Ô∏è int or NULL ('NA')\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "üö®‚ö†Ô∏è‚ùó **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- file_MD5: 63/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- annotation: 711/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- configuration_file: 711/711 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv2\n",
    "v2_tables = dfs\n",
    "for table,df in v2_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_4530', 'SUL_T_4105', 'SUL_T_5259', 'SUL_T_5385',\n",
       "       'SUL_T_2465', 'SUL_T_3799', 'SUL_T_4424', 'SUL_T_5133',\n",
       "       'SUL_T_5227', 'SUL_T_5547', 'SUL_T_5359', 'SUL_T_4116',\n",
       "       'SUL_T_4839', 'SUL_T_2007', 'SUL_T_4938', 'SUL_T_5498',\n",
       "       'SUL_T_2791', 'SUL_T_4695', 'SUL_T_4934', 'SUL_T_3870',\n",
       "       'SUL_T_124', 'SUL_T_81', 'SUL_T_4515', 'SUL_T_4235', 'SUL_T_4815',\n",
       "       'SUL_T_4857', 'SUL_T_3998', 'SUL_T_4462', 'SUL_T_4108',\n",
       "       'SUL_T_4174', 'SUL_T_3860', 'SUL_T_4162', 'SUL_T_4210',\n",
       "       'SUL_T_233', 'SUL_T_1976', 'SUL_T_2017', 'SUL_T_308', 'SUL_T_1952',\n",
       "       'SUL_T_777', 'SUL_T_1619', 'SUL_T_290', 'SUL_T_551', 'SUL_T_257',\n",
       "       'SUL_T_343', 'SUL_T_362'], dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2_tables[\"SUBJECT\"][['subject_id','primary_diagnosis'] ]\n",
    "    # primary_diagnosis = v3_tables[\"SUBJECT\"]['primary_diagnosis'].str.lower().str.replace(\" \", \"_\")\n",
    "    # diagnosis_mapper = dict(zip(subject_id, primary_diagnosis))\n",
    "v2_tables[\"SUBJECT\"]['subject_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SUL_T_362', 'SUL_T_308', 'SUL_T_3860', 'SUL_T_290', 'SUL_T_233',\n",
       "       'SUL_T_4108', 'SUL_T_257', 'SUL_T_777', 'SUL_T_1976', 'SUL_T_4162',\n",
       "       'SUL_T_4174', 'SUL_T_343', 'SUL_T_551', 'SUL_T_1619', 'SUL_T_1952',\n",
       "       'SUL_T_2017', 'SUL_T_2465', 'SUL_T_4210', 'SUL_T_81', 'SUL_T_124',\n",
       "       'SUL_T_2007', 'SUL_T_2791', 'SUL_T_3799', 'SUL_T_3870',\n",
       "       'SUL_T_3998', 'SUL_T_4116', 'SUL_T_4235', 'SUL_T_4424',\n",
       "       'SUL_T_4462', 'SUL_T_4515', 'SUL_T_4530', 'SUL_T_4695',\n",
       "       'SUL_T_4815', 'SUL_T_4839', 'SUL_T_4857', 'SUL_T_4934',\n",
       "       'SUL_T_4938', 'SUL_T_5133', 'SUL_T_5227', 'SUL_T_5259',\n",
       "       'SUL_T_5359', 'SUL_T_5385', 'SUL_T_5498', 'SUL_T_5547'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2_tables[\"SAMPLE\"]['subject_id'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## v2->v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\""
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "f\"{v3_meta_tables}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_of_brain_samples as int\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "recoding input_cell_count as int\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n"
     ]
    }
   ],
   "source": [
    "v3_path = metadata_path / \"v3\"\n",
    "\n",
    "v3_tables, aux_tables = v2_to_v3_PMDBS(v2_path, v3_path, CDEv2, CDEv3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "üö®‚ö†Ô∏è‚ùó **8 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- ASAP_grant_id: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- PI_ORCID: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PI_google_scholar_id: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- preprocessing_references: 1/1 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "üö®‚ö†Ô∏è‚ùó **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- github_url: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 28/46 empty rows (REQUIRED)\n",
      "\n",
      "\t- primary_diagnosis_text: 46/46 empty rows (OPTIONAL)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*age_at_collection*_:  invalid values üí©'77', '89+ ', '79', '88', '83', '78', '87', '75', '86', '85', '68', '73', '80', '72', '71', '50', '74', '81', '70', '66', '62'\n",
      "    - valid ‚û°Ô∏è float or NULL ('NA')\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- source_sample_id: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- organism: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- assay_type: 45/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- time: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- alternate_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- assay_ontology_term_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 45/45 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "üö®‚ö†Ô∏è‚ùó **4 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- file_MD5: 63/711 empty rows (REQUIRED)\n",
      "\n",
      "\t- header: 711/711 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 711/711 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 711/711 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "üö®‚ö†Ô∏è‚ùó **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- hemisphere: 45/45 empty rows (OPTIONAL)\n",
      "\n",
      "\t- region_level_2: 18/45 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "üö®‚ö†Ô∏è‚ùó **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- RIN: 27/45 empty rows (REQUIRED)\n",
      "\n",
      "\t- input_cell_count: 27/45 empty rows (REQUIRED)\n",
      "üö®‚ö†Ô∏è‚ùó **1 Fields with invalid entries:**\n",
      "- _*input_cell_count*_:  invalid values üí©'915.1258 ng', '1411.8474 ng', '1608.6002 ng', '704.4034 ng', '1444.6272 ng', '896.3852 ng', '2392.286 ng', '714.5692 ng', '923.9646 ng', '866.4158 ng', '1420.8972 ng', '2169.106 ng', '1069.3564 ng', '1192.589 ng', '1008.9174 ng', '983.9328 ng', '1266.1952 ng', '945.216 ng', 'NA'\n",
      "    - valid ‚û°Ô∏è int or NULL ('NA')\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 24/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 46/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 31/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_nft: 1/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 21/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 48/48 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_ad_level: 48/48 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "üö®‚ö†Ô∏è‚ùó **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 7/7 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 7/7 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-SULZER</td>\n",
       "      <td>Al-Dalahmah</td>\n",
       "      <td>The landscape of T-cell Immunity in Parkinson ...</td>\n",
       "      <td>sn_rnaseq</td>\n",
       "      <td>OJK_2022_ASAP_SEQ</td>\n",
       "      <td>Assessing the role of T cells in Parkinson's d...</td>\n",
       "      <td>Osama Al-Dalahmah</td>\n",
       "      <td>oa2298@cumc.columbia.edu</td>\n",
       "      <td>Kelly A Jakubiak, Adithya Kannan, Jaeseung Lee...</td>\n",
       "      <td>Kelly Jakubiak</td>\n",
       "      <td>...</td>\n",
       "      <td>44</td>\n",
       "      <td>Cingulate Cortex and Substantia nigra</td>\n",
       "      <td>Postmortem PD and control</td>\n",
       "      <td>v1</td>\n",
       "      <td>['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>v3.0_20241028</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name  \\\n",
       "0    TEAM-SULZER   Al-Dalahmah   \n",
       "\n",
       "                                        project_name team_dataset_id  \\\n",
       "0  The landscape of T-cell Immunity in Parkinson ...       sn_rnaseq   \n",
       "\n",
       "     project_dataset                                project_description  \\\n",
       "0  OJK_2022_ASAP_SEQ  Assessing the role of T cells in Parkinson's d...   \n",
       "\n",
       "        PI_full_name                  PI_email  \\\n",
       "0  Osama Al-Dalahmah  oa2298@cumc.columbia.edu   \n",
       "\n",
       "                                   contributor_names  submitter_name  ...  \\\n",
       "0  Kelly A Jakubiak, Adithya Kannan, Jaeseung Lee...  Kelly Jakubiak  ...   \n",
       "\n",
       "  number_samples                           sample_types  \\\n",
       "0             44  Cingulate Cortex and Substantia nigra   \n",
       "\n",
       "            types_of_samples DUA_version  \\\n",
       "0  Postmortem PD and control          v1   \n",
       "\n",
       "                                     metadata_tables PI_ORCID  \\\n",
       "0  ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...       NA   \n",
       "\n",
       "  PI_google_scholar_id preprocessing_references metadata_version_date  \\\n",
       "0                   NA                       NA         v3.0_20241028   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = v3_tables['STUDY']\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sulzer\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/sulzer-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-sulzer@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/**/*.gz\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "\n",
    "# make sure to get ALL the fastq files in the bucket\n",
    "prefix = \"**/*.gz\"\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def check_md5_sums()\n",
    "\n",
    "v3_path = metadata_path / \"v3\"\n",
    "data_df = read_meta_table(v3_path / \"DATA.csv\")\n",
    "\n",
    "# checksum = v3_tables['DATA'][['file_name','file_MD5','file_type']].copy()\n",
    "\n",
    "checksum = data_df[['file_name','file_MD5','file_type']].copy()\n",
    "checksum = checksum[checksum.file_type == \"fastq\"]\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "checksum['check1'] = checksum['file_MD5']\n",
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "#empty means success!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## prep metadata in raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/upload\n",
    "- 2. copy metadata/upload to dataset upload (upload subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata\n",
      "gsutil command succeeded: gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/CLINPATH.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/DATA.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/PROTOCOL.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/SAMPLE.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/STUDY.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/SUBJECT.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/v2-processed/\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/v2/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "metadata_subdir = \"metadata\"\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/CLINPATH.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/CLINPATH.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/DATA.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/DATA.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/PROTOCOL.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/PROTOCOL.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/SAMPLE.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/SAMPLE.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/STUDY.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/STUDY.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/SUBJECT.csv gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/SUBJECT.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv -r gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/v2-processed gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/v2-processed\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv -r gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/v2 gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/v2\n",
      "gsutil command succeeded: \n"
     ]
    }
   ],
   "source": [
    "metadata_subdir2 = \"metadata/upload\"\n",
    "bucket = current_files[0].split(\"/\")[2]\n",
    "\n",
    "for file in current_files:\n",
    "    if file == \"\":\n",
    "        continue\n",
    "    file_nm = Path(file).name\n",
    "    is_dir = not file_nm.__contains__(\".\")\n",
    "\n",
    "    source = f\"gs://{bucket}/{metadata_subdir}/{file_nm}\"\n",
    "\n",
    "    destination = f\"gs://{bucket}/{metadata_subdir2}/{file_nm}\"\n",
    "    gsutil_mv(source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload\n",
      "gsutil command succeeded: gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/CLINPATH.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/DATA.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/PROTOCOL.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/SAMPLE.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/STUDY.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/SUBJECT.csv\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/v2-processed/\n",
      "gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload/v2/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# verify\n",
    "metadata_subdir = \"metadata/upload\"\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archive the uploaded metadata locally\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload',\n",
       " '/Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_source = f\"gs://{bucket}/{metadata_subdir}\"\n",
    "destination = f\"{metadata_path}\"\n",
    "file_source, destination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons cp -r gs://asap-raw-team-sulzer-pmdbs-sn-rnaseq/metadata/upload /Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata\n",
      "gsutil command succeeded: \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "is_dir = True\n",
    "gsutil_cp(file_source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------\n",
    "## Create metadata package\n",
    "\n",
    "This will copy the final updated to v3.0 metadata to `asap-could-processing-resources`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(PosixPath('/Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata'),\n",
       " PosixPath('/Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/sulzer-pmdbs-sn-rnaseq'))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_source = metadata_path\n",
    "\n",
    "\n",
    "source = \"pmdbs\"\n",
    "archive_root = Path.home() / \"Projects/ASAP/asap-crn-metadata/datasets\"\n",
    "dataset_path = archive_root / f\"{team}-{source}-{dataset_name}\"\n",
    "# bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "metadata_source, dataset_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata/v2 to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/sulzer-pmdbs-sn-rnaseq/metadata/v2\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata/v3 to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/sulzer-pmdbs-sn-rnaseq/metadata/v3\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata/og to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/sulzer-pmdbs-sn-rnaseq/metadata/og\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/sulzer/sn-rnaseq/metadata/upload to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/sulzer-pmdbs-sn-rnaseq/metadata/upload\n"
     ]
    }
   ],
   "source": [
    "fnms = create_metadata_package(metadata_source, dataset_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "generate ASAP IDs + transfering back to raw data bucket via `asap-crn-metadata` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## transfer metadata to raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/archive\n",
    "- 2. copy package to metadata/ . i.e. /og/*.csv, /v??/*.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
