{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata compilation\n",
    "\n",
    "# Team Wood. ASAP CRN Metadata scrubbing\n",
    "\n",
    "confirm with Team Wood bulkRNAseq metadata\n",
    "\n",
    "29 Oct 2024\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os, sys\n",
    "\n",
    "sys.path.append(os.path.abspath((os.path.join(os.getcwd(), 'src/crn_utils'))))\n",
    "\n",
    "from util import read_CDE, NULL, prep_table, read_meta_table, create_metadata_package\n",
    "from validate import validate_table, ReportCollector\n",
    "from update_schema import v1_to_v2, v2_to_v3_PMDBS, intervention_typer\n",
    "from checksums import extract_md5_from_details2, get_md5_hashes \n",
    "from bucket_util import authenticate_with_service_account, gsutil_ls, gsutil_cp, gsutil_mv \n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "root_path = Path.home() / (\"Projects/ASAP/data/teams\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDEs\n",
    "load the relavent CDEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metadata_version: ASAP_CDE_v1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v2.1\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v2.1\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v2.1.csv\n",
      "read local file\n",
      "metadata_version: ASAP_CDE_v3.0\n",
      "https://docs.google.com/spreadsheets/d/1c0z5KvRELdT2AtQAH2Dus8kwAyyLrR0CROhKOjpU4Vc/gviz/tq?tqx=out:csv&sheet=v3.0\n",
      "/Users/ergonyc/Projects/ASAP/crn-utils/resource/CDE/ASAP_CDE_v3.0.csv\n",
      "read local file\n"
     ]
    }
   ],
   "source": [
    "schema_version = \"v1\"\n",
    "schema_path = Path.home() / \"Projects/ASAP/crn-utils/resource/CDE\"\n",
    "CDEv1 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v2.1\"\n",
    "CDEv2 = read_CDE(schema_version, local_path=schema_path)\n",
    "schema_version = \"v3.0\"\n",
    "CDEv3 = read_CDE(schema_version, local_path=schema_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load original tables \n",
    "These were submitted v3.0 (actually  v3.0.0 beta)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Team Wood bulk-rnaseq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "team = \"wood\"\n",
    "dataset_name = \"bulk-rnaseq\"\n",
    "\n",
    "metadata_path = root_path / f\"{team}/{dataset_name}/metadata/\"\n",
    "v3_path = metadata_path / \"v3\"\n",
    "og_path = metadata_path / \"og\"\n",
    "\n",
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "og_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "in_tables = [table_name for table_name in v3_meta_tables if f\"{table_name}.csv\" in os.listdir(og_path)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tables = {}\n",
    "for table_name in in_tables:\n",
    "    df = read_meta_table(f\"{og_path}/{table_name}.csv\")\n",
    "    og_tables[table_name] = df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq'])"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tables.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rationalize the team_dataset_id\n",
    "og_tables['STUDY']['team_dataset_id'] = dataset_name.replace(\"-\", \"_\").replace(\" \", \"_\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>project_name</th>\n",
       "      <th>team_dataset_id</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>...</th>\n",
       "      <th>number_samples</th>\n",
       "      <th>sample_types</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_tables</th>\n",
       "      <th>PI_ORCID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>preprocessing_references</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>alternate_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEAM-WOOD</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Parkinson's aggregate mapping</td>\n",
       "      <td>bulk_rnaseq</td>\n",
       "      <td>Wood_pmdbs_bulk-rnaseq</td>\n",
       "      <td>Oligomer mapping and functional genomic charac...</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne ; Christina Toomey ; J...</td>\n",
       "      <td>Jonathan Brenton</td>\n",
       "      <td>...</td>\n",
       "      <td>234</td>\n",
       "      <td>Substantia Nigra, Caudate, Putamen, Parahippoc...</td>\n",
       "      <td>(Braak 3-4) PD and control post-mortem brains</td>\n",
       "      <td>NA</td>\n",
       "      <td>STUDY, PROTOCOL, SUBJECT, SAMPLE, DATA, CLINPA...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>https://github.com/Jbrenton191/RNAseq_splicing...</td>\n",
       "      <td>Version 3, 10/2024</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASAP_team_name ASAP_lab_name                   project_name team_dataset_id  \\\n",
       "0      TEAM-WOOD     Ryten Lab  Parkinson's aggregate mapping     bulk_rnaseq   \n",
       "\n",
       "          project_dataset                                project_description  \\\n",
       "0  Wood_pmdbs_bulk-rnaseq  Oligomer mapping and functional genomic charac...   \n",
       "\n",
       "  PI_full_name              PI_email  \\\n",
       "0   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                   contributor_names    submitter_name  ...  \\\n",
       "0  Aine Fairbrother-Browne ; Christina Toomey ; J...  Jonathan Brenton  ...   \n",
       "\n",
       "  number_samples                                       sample_types  \\\n",
       "0            234  Substantia Nigra, Caudate, Putamen, Parahippoc...   \n",
       "\n",
       "                                types_of_samples DUA_version  \\\n",
       "0  (Braak 3-4) PD and control post-mortem brains          NA   \n",
       "\n",
       "                                     metadata_tables             PI_ORCID  \\\n",
       "0  STUDY, PROTOCOL, SUBJECT, SAMPLE, DATA, CLINPA...  0000-0001-9520-6957   \n",
       "\n",
       "                                PI_google_scholar_id  \\\n",
       "0  https://scholar.google.co.uk/citations?user=lt...   \n",
       "\n",
       "                            preprocessing_references metadata_version_date  \\\n",
       "0  https://github.com/Jbrenton191/RNAseq_splicing...    Version 3, 10/2024   \n",
       "\n",
       "  alternate_dataset_id  \n",
       "0                   NA  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tables['STUDY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- cell_extraction_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 234/234 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in SAMPLE: time, development_stage_ontology_term_id, sex_ontology_term_id, self_reported_ethnicity_ontology_term_id, disease_ontology_term_id, tissue_ontology_term_id, assay_ontology_term_id, donor_id, pm_PH, cell_type_ontology_term_id, source_RIN, DV200, suspension_type**\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in DATA: header, annotation, configuration_file**\n",
      "No empty entries (NULL) found .\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in CLINPATH: AMPPD_id, GP2_id, ethnicity, last_diagnosis, first_motor_symptom, hx_melanoma, education_level, smoking_status, smoking_years, APOE_e4_status, cognitive_status, time_from_baseline, PD_pathogenic_mutation**\n",
      "🚨⚠️❗ **30 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- family_history: 200/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 157/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 215/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 163/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 23/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 115/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 187/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 43/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 220/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 87/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 30/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 170/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 195/234 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'B0', 'B1'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3', 'NA'\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "No empty entries (NULL) found .\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*region_level_1*_:  invalid values 💩'Basal Gaglia'\n",
      "    - valid ➡️ 'Frontal lobe', 'Temporal lobe', 'Parietal lobe', 'Occipital lobe', 'Cingulate gyrus', 'Striatum', 'Thalamus', 'Midbrain', 'Pons', 'Medulla', 'Cerebellum', 'Intestine', 'Unknown', 'NA'\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **Missing Optional Fields in CONDITION: intervention_aux_table**\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 4/4 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- input_cell_count: 234/234 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "dfs = {}\n",
    "for table,df in og_tables.items():\n",
    "\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    dfs[table] = full_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = dfs['STUDY']\n",
    "PROTOCOL = dfs['PROTOCOL']\n",
    "SUBJECT = dfs['SUBJECT']\n",
    "SAMPLE = dfs['SAMPLE']\n",
    "DATA = dfs['DATA']\n",
    "CLINPATH = dfs['CLINPATH']\n",
    "PMDBS = dfs['PMDBS']\n",
    "ASSAY_RNAseq = dfs['ASSAY_RNAseq']\n",
    "CONDITION = dfs['CONDITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hemiparkinson/hemiatrophy syndrome</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            condition_id intervention_name intervention_id  \\\n",
       "0                          idiopathic_pd      Case-Control            Case   \n",
       "1  no_pd_nor_other_neurological_disorder      Case-Control         Control   \n",
       "2     Hemiparkinson/hemiatrophy syndrome      Case-Control            Case   \n",
       "3            other_neurological_disorder      Case-Control           Other   \n",
       "\n",
       "  protocol_id intervention_aux_table  \n",
       "0          NA                     NA  \n",
       "1          NA                     NA  \n",
       "2          NA                     NA  \n",
       "3          NA                     NA  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CONDITION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DA...\n",
       "Name: metadata_tables, dtype: object"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY['metadata_tables'] = f\"{v3_meta_tables}\"\n",
    "STUDY['metadata_tables']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_version = \"v3.0\"\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\"\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# SUBJECT[\"primary_diagnosis\"] = SUBJECT[\"primary_diagnosis\"].replace({\"NA\":\"Hemiparkinson/hemiatrophy syndrome\"})\n",
    "# SUBJECT[\"primary_diagnosis\"].unique()\n",
    "SUBJECT_ = SUBJECT.drop_duplicates(subset=[\"subject_id\"]).reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hemiparkinson_hemiatrophy_syndrome</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Field                           condition_id intervention_name  \\\n",
       "0      no_pd_nor_other_neurological_disorder      Case-Control   \n",
       "1                              idiopathic_pd      Case-Control   \n",
       "2         hemiparkinson_hemiatrophy_syndrome      Case-Control   \n",
       "\n",
       "Field intervention_id protocol_id intervention_aux_table  \n",
       "0             Control          NA                     NA  \n",
       "1                Case          NA                     NA  \n",
       "2                Case          NA                     NA  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT_ = SUBJECT.drop_duplicates(subset=[\"subject_id\"]).reset_index(drop=True)\n",
    "\n",
    "# SAMPLE \n",
    "SAMPLE['condition_id'] = SAMPLE['condition_id'].str.lower().str.replace(\" \", \"_\").str.replace(\"/\", \"_\")\n",
    "\n",
    "# CONDITION\n",
    "# construct this table.  needs to be checked by hand\n",
    "diagnosis_mapper = dict(zip(SUBJECT['subject_id'], SUBJECT['primary_diagnosis']))\n",
    "\n",
    "v3_tables[\"SAMPLE\"]['condition_id'] = v3_tables[\"SAMPLE\"]['subject_id'].map(diagnosis_mapper)\n",
    "\n",
    "CONDITIONv3 = pd.DataFrame(columns=CDEv3[CDEv3['Table'] == \"CONDITION\"]['Field'])\n",
    "# \n",
    "CONDITIONv3['condition_id'] = SUBJECT['primary_diagnosis'].unique()\n",
    "CONDITIONv3['intervention_name'] = \"Case-Control\"\n",
    "CONDITIONv3['intervention_id'] = CONDITIONv3['condition_id'].apply(intervention_typer)\n",
    "CONDITIONv3['condition_id'] = CONDITIONv3['condition_id'].str.lower().str.replace(\" \", \"_\").str.replace(\"/\", \"_\")\n",
    "CONDITIONv3 = CONDITIONv3.fillna(NULL)\n",
    "CONDITIONv3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 234/234 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table = \"SUBJECT\"\n",
    "schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(dfs[table].copy(), table, schema, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sex_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- tissue_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- assay_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 234/234 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table = \"SAMPLE\"\n",
    "schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(dfs[table].copy(), table, schema, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 200/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 157/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 215/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 163/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 23/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 115/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 187/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 43/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 220/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 87/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 30/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 170/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 195/234 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'B0', 'B1'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3', 'NA'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table = \"CLINPATH\"\n",
    "schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(dfs[table].copy(), table, schema, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sex_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- tissue_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- assay_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 234/234 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table = \"SAMPLE\"\n",
    "schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "report = ReportCollector(destination=\"NA\")\n",
    "full_table, report = validate_table(dfs[table].copy(), table, schema, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>family_history</th>\n",
       "      <th>age_at_onset</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>hx_dementia_mci</th>\n",
       "      <th>path_autopsy_dx_main</th>\n",
       "      <th>path_autopsy_second_dx</th>\n",
       "      <th>...</th>\n",
       "      <th>last_diagnosis</th>\n",
       "      <th>first_motor_symptom</th>\n",
       "      <th>hx_melanoma</th>\n",
       "      <th>education_level</th>\n",
       "      <th>smoking_status</th>\n",
       "      <th>smoking_years</th>\n",
       "      <th>APOE_e4_status</th>\n",
       "      <th>cognitive_status</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>PD_pathogenic_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C073</td>\n",
       "      <td>C073</td>\n",
       "      <td>29.0</td>\n",
       "      <td>71</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C073</td>\n",
       "      <td>C073</td>\n",
       "      <td>29.0</td>\n",
       "      <td>71</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C073</td>\n",
       "      <td>C073</td>\n",
       "      <td>29.0</td>\n",
       "      <td>71</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C073</td>\n",
       "      <td>C073</td>\n",
       "      <td>29.0</td>\n",
       "      <td>71</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C073</td>\n",
       "      <td>C073</td>\n",
       "      <td>29.0</td>\n",
       "      <td>71</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>PDC085</td>\n",
       "      <td>PDC085</td>\n",
       "      <td>20.0</td>\n",
       "      <td>82</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>PDC085</td>\n",
       "      <td>PDC085</td>\n",
       "      <td>20.0</td>\n",
       "      <td>82</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>PDC085</td>\n",
       "      <td>PDC085</td>\n",
       "      <td>20.0</td>\n",
       "      <td>82</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>PDC085</td>\n",
       "      <td>PDC085</td>\n",
       "      <td>20.0</td>\n",
       "      <td>82</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>PDC085</td>\n",
       "      <td>PDC085</td>\n",
       "      <td>20.0</td>\n",
       "      <td>82</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>Control, Low level AD neuropathological change</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>234 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    subject_id source_subject_id  duration_pmi age_at_death family_history  \\\n",
       "0         C073              C073          29.0           71             NA   \n",
       "1         C073              C073          29.0           71             NA   \n",
       "2         C073              C073          29.0           71             NA   \n",
       "3         C073              C073          29.0           71             NA   \n",
       "4         C073              C073          29.0           71             NA   \n",
       "..         ...               ...           ...          ...            ...   \n",
       "229     PDC085            PDC085          20.0           82             NA   \n",
       "230     PDC085            PDC085          20.0           82             NA   \n",
       "231     PDC085            PDC085          20.0           82             NA   \n",
       "232     PDC085            PDC085          20.0           82             NA   \n",
       "233     PDC085            PDC085          20.0           82             NA   \n",
       "\n",
       "    age_at_onset age_at_diagnosis hx_dementia_mci  \\\n",
       "0             NA               NA              NA   \n",
       "1             NA               NA              NA   \n",
       "2             NA               NA              NA   \n",
       "3             NA               NA              NA   \n",
       "4             NA               NA              NA   \n",
       "..           ...              ...             ...   \n",
       "229           NA               NA              No   \n",
       "230           NA               NA              No   \n",
       "231           NA               NA              No   \n",
       "232           NA               NA              No   \n",
       "233           NA               NA              No   \n",
       "\n",
       "                               path_autopsy_dx_main path_autopsy_second_dx  \\\n",
       "0    Control, Low level AD neuropathological change                     NA   \n",
       "1    Control, Low level AD neuropathological change                     NA   \n",
       "2    Control, Low level AD neuropathological change                     NA   \n",
       "3    Control, Low level AD neuropathological change                     NA   \n",
       "4    Control, Low level AD neuropathological change                     NA   \n",
       "..                                              ...                    ...   \n",
       "229  Control, Low level AD neuropathological change                     NA   \n",
       "230  Control, Low level AD neuropathological change                     NA   \n",
       "231  Control, Low level AD neuropathological change                     NA   \n",
       "232  Control, Low level AD neuropathological change                     NA   \n",
       "233  Control, Low level AD neuropathological change                     NA   \n",
       "\n",
       "     ... last_diagnosis first_motor_symptom hx_melanoma education_level  \\\n",
       "0    ...             NA                  NA          NA              NA   \n",
       "1    ...             NA                  NA          NA              NA   \n",
       "2    ...             NA                  NA          NA              NA   \n",
       "3    ...             NA                  NA          NA              NA   \n",
       "4    ...             NA                  NA          NA              NA   \n",
       "..   ...            ...                 ...         ...             ...   \n",
       "229  ...             NA                  NA          NA              NA   \n",
       "230  ...             NA                  NA          NA              NA   \n",
       "231  ...             NA                  NA          NA              NA   \n",
       "232  ...             NA                  NA          NA              NA   \n",
       "233  ...             NA                  NA          NA              NA   \n",
       "\n",
       "    smoking_status smoking_years APOE_e4_status cognitive_status  \\\n",
       "0               NA            NA             NA               NA   \n",
       "1               NA            NA             NA               NA   \n",
       "2               NA            NA             NA               NA   \n",
       "3               NA            NA             NA               NA   \n",
       "4               NA            NA             NA               NA   \n",
       "..             ...           ...            ...              ...   \n",
       "229             NA            NA             NA               NA   \n",
       "230             NA            NA             NA               NA   \n",
       "231             NA            NA             NA               NA   \n",
       "232             NA            NA             NA               NA   \n",
       "233             NA            NA             NA               NA   \n",
       "\n",
       "    time_from_baseline PD_pathogenic_mutation  \n",
       "0                   NA                     NA  \n",
       "1                   NA                     NA  \n",
       "2                   NA                     NA  \n",
       "3                   NA                     NA  \n",
       "4                   NA                     NA  \n",
       "..                 ...                    ...  \n",
       "229                 NA                     NA  \n",
       "230                 NA                     NA  \n",
       "231                 NA                     NA  \n",
       "232                 NA                     NA  \n",
       "233                 NA                     NA  \n",
       "\n",
       "[234 rows x 52 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_id</th>\n",
       "      <th>intervention_name</th>\n",
       "      <th>intervention_id</th>\n",
       "      <th>protocol_id</th>\n",
       "      <th>intervention_aux_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>idiopathic_pd</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_pd_nor_other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hemiparkinson/hemiatrophy syndrome</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Case</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>other_neurological_disorder</td>\n",
       "      <td>Case-Control</td>\n",
       "      <td>Other</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            condition_id intervention_name intervention_id  \\\n",
       "0                          idiopathic_pd      Case-Control            Case   \n",
       "1  no_pd_nor_other_neurological_disorder      Case-Control         Control   \n",
       "2     Hemiparkinson/hemiatrophy syndrome      Case-Control            Case   \n",
       "3            other_neurological_disorder      Case-Control           Other   \n",
       "\n",
       "  protocol_id intervention_aux_table  \n",
       "0          NA                     NA  \n",
       "1          NA                     NA  \n",
       "2          NA                     NA  \n",
       "3          NA                     NA  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CONDITION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now save the tables\n",
    "v3_tables = {\n",
    "    \"STUDY\": STUDY,\n",
    "    \"PROTOCOL\": PROTOCOL,\n",
    "    \"SUBJECT\": SUBJECT,\n",
    "    \"SAMPLE\": SAMPLE,\n",
    "    \"DATA\": DATA,\n",
    "    \"CLINPATH\": CLINPATH,\n",
    "    \"PMDBS\": PMDBS,\n",
    "    \"CONDITION\": CONDITIONv3,\n",
    "    \"ASSAY_RNAseq\": ASSAY_RNAseq\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### save extras as auxillary tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for table,df in v3_tables.items():\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "    valid_fields = schema['Field'].unique()\n",
    "    df_out = df[valid_fields]\n",
    "    aux_fields = set(df.columns) - set(valid_fields)\n",
    "    if aux_fields:\n",
    "        df_aux = df[list(aux_fields)]\n",
    "        df_aux.to_csv(og_path / f\"{table}_auxiliary.csv\", index=False)\n",
    "        print(f\"Saved {table}_auxiliary.csv\")\n",
    "    df_out.to_csv(v3_path / f\"{table}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### validate v3 tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "recoding number_samples as int\n",
      "All required fields are present in *STUDY* table.\n",
      "🚨⚠️❗ **5 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- other_funding_source: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- publication_PMID: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- DUA_version: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- alternate_dataset_id: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *PROTOCOL* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- cell_extraction_summary: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- protocols_io_DOI: 1/1 empty rows (REQUIRED)\n",
      "\n",
      "\t- other_reference: 1/1 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- race: 234/234 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **13 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- time: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- development_stage_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sex_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- self_reported_ethnicity_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- disease_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- tissue_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- assay_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- donor_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- pm_PH: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cell_type_ontology_term_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- source_RIN: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- DV200: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- suspension_type: 234/234 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding replicate_count as int\n",
      "recoding repeated_sample as int\n",
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **3 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- header: 468/468 empty rows (OPTIONAL)\n",
      "\n",
      "\t- annotation: 468/468 empty rows (OPTIONAL)\n",
      "\n",
      "\t- configuration_file: 468/468 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding age_at_onset as int\n",
      "recoding age_at_diagnosis as int\n",
      "recoding first_motor_symptom as int\n",
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **43 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- AMPPD_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- GP2_id: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- ethnicity: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- family_history: 200/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- last_diagnosis: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_onset: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- age_at_diagnosis: 113/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- first_motor_symptom: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_dementia_mci: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- hx_melanoma: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- education_level: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- smoking_years: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- APOE_e4_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cognitive_status: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- time_from_baseline: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_second_dx: 157/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_third_dx: 215/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fourth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_fifth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_sixth_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_seventh_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_autopsy_eight_dx: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_year_death: 163/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- cause_death: 23/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_1: 115/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- other_cause_death_2: 187/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- brain_weight: 43/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_braak_asyn: 24/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_cerad: 220/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_thal: 87/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- known_pathogenic_mutation: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- PD_pathogenic_mutation: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_mckeith: 30/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- sn_neuronal_loss: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_infarcs: 170/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_ri: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_a: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_b: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- path_nia_aa_c: 102/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- TDP43: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- arteriolosclerosis_severity_scale: 234/234 empty rows (OPTIONAL)\n",
      "\n",
      "\t- amyloid_angiopathy_severity_scale: 195/234 empty rows (OPTIONAL)\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'B0', 'B1'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3', 'NA'\n",
      "\n",
      "All required fields are present in *PMDBS* table.\n",
      "No empty entries (NULL) found .\n",
      "🚨⚠️❗ **1 Fields with invalid entries:**\n",
      "- _*region_level_1*_:  invalid values 💩'Basal Gaglia'\n",
      "    - valid ➡️ 'Frontal lobe', 'Temporal lobe', 'Parietal lobe', 'Occipital lobe', 'Cingulate gyrus', 'Striatum', 'Thalamus', 'Midbrain', 'Pons', 'Medulla', 'Cerebellum', 'Intestine', 'Unknown', 'NA'\n",
      "\n",
      "All required fields are present in *CONDITION* table.\n",
      "🚨⚠️❗ **2 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- protocol_id: 4/4 empty rows (REQUIRED)\n",
      "\n",
      "\t- intervention_aux_table: 4/4 empty rows (OPTIONAL)\n",
      "No invalid entries found in Enum fields.\n",
      "\n",
      "recoding input_cell_count as int\n",
      "recoding sequencing_length as int\n",
      "All required fields are present in *ASSAY_RNAseq* table.\n",
      "🚨⚠️❗ **1 Fields with empty (NULL) values:**\n",
      "\n",
      "\t- input_cell_count: 234/234 empty rows (REQUIRED)\n",
      "No invalid entries found in Enum fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CDE = CDEv3\n",
    "tables = CDE['Table'].unique()\n",
    "v3_meta_tables = ['STUDY', 'PROTOCOL', 'SUBJECT', 'SAMPLE', 'DATA', 'CLINPATH', 'PMDBS', 'CONDITION', 'ASSAY_RNAseq']\n",
    "\n",
    "v3_tables = {}\n",
    "for table in v3_meta_tables:\n",
    "    df = read_meta_table(v3_path / f\"{table}.csv\")\n",
    "    schema = CDE[CDE['Table'] == table]\n",
    "\n",
    "    report = ReportCollector(destination=\"NA\")\n",
    "    full_table, report = validate_table(df.copy(), table, schema, report)\n",
    "    report.print_log()\n",
    "    v3_tables[table] = full_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-------------------------\n",
    "## check md5s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wood\n",
      "CompletedProcess(args='gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/wood-credentials.json', returncode=0, stdout='', stderr='Activated service account credentials for: [raw-admin-wood@dnastack-asap-parkinsons.iam.gserviceaccount.com]\\n')\n",
      "gsutil -u dnastack-asap-parkinsons hash -h gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/**/*.gz\n"
     ]
    }
   ],
   "source": [
    "print(team)\n",
    "\n",
    "source = \"pmdbs\"\n",
    "\n",
    "bucket = f\"asap-raw-team-{team}-{source}-{dataset_name}\"\n",
    "bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "\n",
    "\n",
    "key_file_path = Path.home() / f\"Projects/ASAP/{team}-credentials.json\"\n",
    "\n",
    "res = authenticate_with_service_account(key_file_path)\n",
    "print(res)\n",
    "\n",
    "# make sure to get ALL the fastq files in the bucket\n",
    "prefix = \"Wood_bulkrnaseq_pm_hs/**/*.gz\" #\"**/*.gz\" # Wood_bulkrnaseq_pm_hs/\n",
    "bucket_files_md5 = get_md5_hashes( bucket, prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_99523/1352623229.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_99523/1352623229.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check1'] = checksum['file_MD5']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def check_md5_sums()\n",
    "\n",
    "\n",
    "checksum = v3_tables['DATA'][['file_name','file_MD5']]\n",
    "checksum['check2'] = checksum['file_name'].map(bucket_files_md5)\n",
    "checksum['check1'] = checksum['file_MD5']\n",
    "checksum[checksum.check1 != checksum.check2].file_name.to_list()\n",
    "#empty means success!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## prep metadata in raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/upload\n",
    "- 2. copy metadata/upload to dataset upload (upload subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata\n",
      "gsutil command succeeded: gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/ASSAY_RNAseq.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/CLINPATH.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/CONDITION.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/DATA.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/PMDBS.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/PROTOCOL.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/SAMPLE.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/STUDY.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/SUBJECT.csv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "metadata_subdir = \"Wood_bulkrnaseq_pm_hs/metadata\"\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/ASSAY_RNAseq.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/ASSAY_RNAseq.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/CLINPATH.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/CLINPATH.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/CONDITION.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/CONDITION.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/DATA.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/DATA.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/PMDBS.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/PMDBS.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/PROTOCOL.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/PROTOCOL.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/SAMPLE.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/SAMPLE.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/STUDY.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/STUDY.csv\n",
      "gsutil command succeeded: \n",
      "gsutil -u dnastack-asap-parkinsons mv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/SUBJECT.csv gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/SUBJECT.csv\n",
      "gsutil command succeeded: \n"
     ]
    }
   ],
   "source": [
    "metadata_subdir2 = \"Wood_bulkrnaseq_pm_hs/metadata/upload\"\n",
    "bucket = current_files[0].split(\"/\")[2]\n",
    "\n",
    "for file in current_files:\n",
    "    if file == \"\":\n",
    "        continue\n",
    "    file_nm = Path(file).name\n",
    "    is_dir = not file_nm.__contains__(\".\")\n",
    "\n",
    "    source = f\"gs://{bucket}/{metadata_subdir}/{file_nm}\"\n",
    "\n",
    "    destination = f\"gs://{bucket}/{metadata_subdir2}/{file_nm}\"\n",
    "    gsutil_mv(source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons ls gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload\n",
      "gsutil command succeeded: gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/ASSAY_RNAseq.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/CLINPATH.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/CONDITION.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/DATA.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/PMDBS.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/PROTOCOL.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/SAMPLE.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/STUDY.csv\n",
      "gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload/SUBJECT.csv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "metadata_subdir = \"Wood_bulkrnaseq_pm_hs/metadata/upload\"\n",
    "current_files = gsutil_ls(bucket,metadata_subdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Archive the uploaded metadata locally\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload',\n",
       " '/Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_source = f\"gs://{bucket}/{metadata_subdir}\"\n",
    "destination = f\"{metadata_path}\"\n",
    "file_source,destination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gsutil -u dnastack-asap-parkinsons cp -r gs://asap-raw-data-team-wood/Wood_bulkrnaseq_pm_hs/metadata/upload /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata\n",
      "gsutil command succeeded: \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "is_dir = True\n",
    "gsutil_cp(file_source, destination, is_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------\n",
    "## Create metadata package\n",
    "\n",
    "This will copy the final updated to v3.0 metadata to `asap-could-processing-resources`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(PosixPath('/Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata'),\n",
       " PosixPath('/Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/wood-pmdbs-bulk-rnaseq'))"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_source = metadata_path\n",
    "\n",
    "\n",
    "source = \"pmdbs\"\n",
    "archive_root = Path.home() / \"Projects/ASAP/asap-crn-metadata/datasets\"\n",
    "dataset_path = archive_root / f\"{team}-{source}-{dataset_name}\"\n",
    "# bucket = f\"asap-raw-data-team-{team}\" # for now old locations\n",
    "metadata_source, dataset_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping empty folder /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata/v1\n",
      "Skipping empty folder /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata/v2\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata/v3 to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/wood-pmdbs-bulk-rnaseq/metadata/v3\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata/og to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/wood-pmdbs-bulk-rnaseq/metadata/og\n",
      "Copied /Users/ergonyc/Projects/ASAP/data/teams/wood/bulk-rnaseq/metadata/upload to /Users/ergonyc/Projects/ASAP/asap-crn-metadata/datasets/wood-pmdbs-bulk-rnaseq/metadata/upload\n"
     ]
    }
   ],
   "source": [
    "fnms = create_metadata_package(metadata_source, dataset_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "generate ASAP IDs + transfering back to raw data bucket via `asap-crn-metadata` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "## transfer metadata to raw data bucket\n",
    "\n",
    "steps:\n",
    "- 1. archive whats there.  i.e. move to metadata/archive\n",
    "- 2. copy package to metadata/ . i.e. /og/*.csv, /v??/*.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nb_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
